<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Statistical Modeling – Modern Statistics for Modern Biology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./imgs/msmb-favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-af5ec82acda093b5ee751184164e9432.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d63985d606c7f07261ad561a6cfde940.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="msmb.css">
<meta property="og:title" content="2&nbsp; Statistical Modeling – Modern Statistics for Modern Biology">
<meta property="og:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta property="og:image" content="https://www.huber.embl.de/msmb/imgs/StatDiagram.png">
<meta property="og:site_name" content="Modern Statistics for Modern Biology">
<meta name="twitter:title" content="2&nbsp; Statistical Modeling – Modern Statistics for Modern Biology">
<meta name="twitter:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta name="twitter:image" content="https://www.huber.embl.de/msmb/imgs/StatDiagram.png">
<meta name="twitter:card" content="summary">
<script type="text/javascript">
  var _paq = _paq || [];
  
  // Call disableCookies before calling trackPageView 
  _paq.push(['disableCookies']);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u='//tr-denbi.embl.de/heimdall/';
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '26']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>

</head>

<body class="floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Modern Statistics for Modern Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-chapters" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Chapters</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-chapters">    
        <li>
    <a class="dropdown-item" href="./index.html">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./00-chap.html">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./01-chap.html">
 <span class="dropdown-text">Generative Models for Discrete Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-chap.html">
 <span class="dropdown-text">Statistical Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03-chap.html">
 <span class="dropdown-text">Data visualization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04-chap.html">
 <span class="dropdown-text">Mixture Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./05-chap.html">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06-chap.html">
 <span class="dropdown-text">Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./07-chap.html">
 <span class="dropdown-text">Multivariate Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./08-chap.html">
 <span class="dropdown-text">High-Throughput Count Data &amp; Generalized Linear Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./09-chap.html">
 <span class="dropdown-text">Multivariate methods for heterogeneous data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./10-chap.html">
 <span class="dropdown-text">Networks and Trees</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./11-chap.html">
 <span class="dropdown-text">Image data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./12-chap.html">
 <span class="dropdown-text">Supervised Learning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./13-chap.html">
 <span class="dropdown-text">Design of High Throughput Experiments and their Analyses</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./14-chap.html">
 <span class="dropdown-text">Statistical Concordance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./15-chap.html">
 <span class="dropdown-text">Acknowledgements</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./16-chap.html">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Sections</h2>
   
  <ul class="collapse">
  <li><a href="#goals-for-this-chapter" id="toc-goals-for-this-chapter" class="nav-link active" data-scroll-target="#goals-for-this-chapter"><span class="header-section-number">2.1</span> Goals for this chapter</a></li>
  <li><a href="#the-difference-between-statistical-and-probabilistic-models" id="toc-the-difference-between-statistical-and-probabilistic-models" class="nav-link" data-scroll-target="#the-difference-between-statistical-and-probabilistic-models"><span class="header-section-number">2.2</span> The difference between statistical and probabilistic models</a></li>
  <li><a href="#sec:stat:simplex" id="toc-sec:stat:simplex" class="nav-link" data-scroll-target="#sec\:stat\:simplex"><span class="header-section-number">2.3</span> A simple example of statistical modeling</a></li>
  <li><a href="#FreqBayes:sec:BinDistML" id="toc-FreqBayes:sec:BinDistML" class="nav-link" data-scroll-target="#FreqBayes\:sec\:BinDistML"><span class="header-section-number">2.4</span> Binomial distributions and maximum likelihood</a></li>
  <li><a href="#more-boxesmultinomial-data" id="toc-more-boxesmultinomial-data" class="nav-link" data-scroll-target="#more-boxesmultinomial-data"><span class="header-section-number">2.5</span> More boxes:multinomial data</a></li>
  <li><a href="#the-chi2-distribution" id="toc-the-chi2-distribution" class="nav-link" data-scroll-target="#the-chi2-distribution"><span class="header-section-number">2.6</span> The <span class="math inline">\(\chi^2\)</span> distribution</a></li>
  <li><a href="#sec-models-Chargaff" id="toc-sec-models-Chargaff" class="nav-link" data-scroll-target="#sec-models-Chargaff"><span class="header-section-number">2.7</span> Chargaff’s Rule</a></li>
  <li><a href="#sec-models-modelingdependencies" id="toc-sec-models-modelingdependencies" class="nav-link" data-scroll-target="#sec-models-modelingdependencies"><span class="header-section-number">2.8</span> Modeling sequential dependencies: Markov chains</a></li>
  <li><a href="#bayesian-thinking" id="toc-bayesian-thinking" class="nav-link" data-scroll-target="#bayesian-thinking"><span class="header-section-number">2.9</span> Bayesian Thinking</a></li>
  <li><a href="#example-occurrence-of-a-nucleotide-pattern-in-a-genome" id="toc-example-occurrence-of-a-nucleotide-pattern-in-a-genome" class="nav-link" data-scroll-target="#example-occurrence-of-a-nucleotide-pattern-in-a-genome"><span class="header-section-number">2.10</span> Example: occurrence of a nucleotide pattern in a genome</a></li>
  <li><a href="#summary-of-this-chapter" id="toc-summary-of-this-chapter" class="nav-link" data-scroll-target="#summary-of-this-chapter"><span class="header-section-number">2.11</span> Summary of this chapter</a></li>
  <li><a href="#further-reading" id="toc-further-reading" class="nav-link" data-scroll-target="#further-reading"><span class="header-section-number">2.12</span> Further reading</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">2.13</span> Exercises</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-models" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Statistical Modeling</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/StatDiagram.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="imgs/StatDiagram.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a></p>
</figure>
</div>
</div></div></div>
<p>In the previous chapter, the knowledge of both the generative model and the values of the parameters provided us with probabilities we could use for decision making – for instance, whether we had really found an epitope. In many real situations, neither the generative model nor the parameters are known, and we will need to estimate them using the data we have collected. Statistical modeling works from the data <em>upwards</em> to a model that <em>might</em> plausibly explain the data<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. This upward-reasoning step is called statistical <strong>inference</strong>. This chapter will show us some of the distributions and estimation mechanisms that serve as building blocks for inference. Although the examples in this chapter are all parametric (i.e., the statistical models only have a small number of unknown parameters), the principles we discuss will generalize.</p>
<div class="no-row-height column-margin column-container"><div id="fn1"><p><sup>1</sup>&nbsp;Even if we have found a model that perfectly explains all our current data, it could always be that reality is more complex. A new set of data lets us conclude that another model is needed, and may include the current model as a special case or approximation.</p></div></div><div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="In a statistical setting, we start with the data X and use them to estimate the parameters. These estimates are denoted by Greek letters with what we call hats on them, as in \widehat{\theta}."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="In a statistical setting, we start with the data X and use them to estimate the parameters. These estimates are denoted by Greek letters with what we call hats on them, as in \widehat{\theta}."></a></p>
</figure>
</div>
<figcaption>In a statistical setting, we start with the data <span class="math inline">\(X\)</span> and use them to <em>estimate</em> the parameters. These estimates are denoted by Greek letters with what we call hats on them, as in <span class="math inline">\(\widehat{\theta}\)</span>.</figcaption>
</figure>
</div>
</div></div></div>
<section id="goals-for-this-chapter" class="level2 page-columns page-full" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="goals-for-this-chapter"><span class="header-section-number">2.1</span> Goals for this chapter</h2>
<p>In this chapter we will:</p>
<ul>
<li><p>See that there is a difference between two subjects that are often confused: “Probability” and “Statistics”.</p></li>
<li><p>Fit data to probability distributions using histograms and other visualization tricks.</p></li>
<li><p>Have a first encounter with an estimating procedure known as <strong>maximum likelihood</strong> through a simulation experiment.</p></li>
<li><p>Make inferences from data for which we have prior information. For this we will use the Bayesian paradigm which will involve new distributions with specially tailored properties. We will use simulations and see how Bayesian estimation differs from simple application of maximum likelihood.</p></li>
<li><p>Use statistical models and estimation to evaluate dependencies in binomial and multinomial distributions.</p></li>
<li><p>Analyse some historically interesting genomic data assembled into tables.</p></li>
<li><p>Make Markov chain models for <strong>dependent</strong> data.</p></li>
<li><p>Do a few concrete applications counting motifs in whole genomes and manipulate special Bioconductor classes dedicated to genomic data.</p></li>
</ul>
<!-- -->
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/Parameters.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="imgs/Parameters.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a></p>
</figure>
</div>
</div></div></div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><strong>Examples of parameters</strong>: the single parameter <span class="math inline">\(\lambda\)</span> defines a Poisson distribution. The letter <span class="math inline">\(\mu\)</span> is often used for the mean of the normal. More generally, we use the Greek letter <span class="math inline">\(\theta\)</span> to designate a generic tuple of parameters necessary to specify a probability model. For instance, in the case of the binomial distribution, <span class="math inline">\(\theta=(n,p)\)</span> comprises two numbers, a positive integer and a real number between 0 and 1.</span></div></div>
<section id="parameters-are-the-key." class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="parameters-are-the-key.">Parameters are the key.</h4>
<p>We saw in <a href="01-chap.html" class="quarto-xref"><span>Chapter 1</span></a> that the knowledge of all the parameter values in the epitope example enabled us to use our probability model and test a null hypothesis based on the data we had at hand. We will see different approaches to statistical modeling through some real examples and computer simulations, but let’s start by making a distinction between two situations depending on how much information is available.</p>
</section>
</section>
<section id="the-difference-between-statistical-and-probabilistic-models" class="level2 page-columns page-full" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="the-difference-between-statistical-and-probabilistic-models"><span class="header-section-number">2.2</span> The difference between statistical and probabilistic models</h2>
<p>A probabilistic analysis is possible when we know a good generative model for the randomness in the data, <em>and</em> we know its parameters’ actual values.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-probabilitydiag" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-probabilitydiag-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/ProbaDiagram.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;2.1: The probabilistic model we obtained in sec-generative. The data are represented as x in green. If we know the true value of \theta, then we can compute the probability of observing x for all possible instances of x, in particular, for an x that we observed."><img src="imgs/ProbaDiagram.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-probabilitydiag-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.1: The probabilistic model we obtained in <a href="01-chap.html" class="quarto-xref"><span>Chapter 1</span></a>. The data are represented as <span class="math inline">\(x\)</span> in green. If we know the true value of <span class="math inline">\(\theta\)</span>, then we can compute the probability of observing <span class="math inline">\(x\)</span> for all possible instances of <span class="math inline">\(x\)</span>, in particular, for an <span class="math inline">\(x\)</span> that we observed.
</figcaption>
</figure>
</div>
</div></div></div>
<p>In the epitope example, knowing that false positives occur as Bernoulli(0.01) per position, the number of patient samples assayed and the length of the protein, meant that there were <em>no unknown parameters</em>.</p>
<p>In such a case, we can use mathematical <strong>deduction</strong> to compute the probability of an event as schematized in <a href="#fig-probabilitydiag" class="quarto-xref">Figure&nbsp;<span>2.1</span></a>. In the epitope examples, we used the Poisson probability as our <strong>null model</strong> with the given parameter <span class="math inline">\(\lambda=0.5\)</span>. We were able to conclude through mathematical deduction that the chances of seeing a maximum value of 7 or larger was around <span class="math inline">\(10^{-4}\)</span> and thus that in fact the observed data were highly unlikely under that model (or “null hypothesis”).</p>
<p>Now suppose that we know the number of patients and the length of the proteins (these are given by the experimental design) but not the distribution itself and the false positive rate. Once we observe data, we need to go <em>up</em> from the data to estimate both a probability model <span class="math inline">\(F\)</span> (Poisson, normal, binomial) and eventually the missing parameter(s) for that model. This is the type of statistical <strong>inference</strong> we will explain in this chapter.</p>
</section>
<section id="sec:stat:simplex" class="level2 page-columns page-full" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec:stat:simplex"><span class="header-section-number">2.3</span> A simple example of statistical modeling</h2>
<section id="start-with-the-data" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="start-with-the-data">Start with the data</h4>
<p>There are two parts to the modeling procedure. First we need a reasonable probability <em>distribution</em> to model the data generation process. As we saw in <a href="01-chap.html" class="quarto-xref"><span>Chapter 1</span></a>, discrete count data may be modeled by simple probability distributions such as binomial, multinomial or Poisson distributions. The normal distribution, or bell shaped curve, is often a good model for continuous measurements. Distributions can also be more complicated mixtures of these elementary ones (more on this in <a href="04-chap.html" class="quarto-xref"><span>Chapter 4</span></a>).</p>
<p>Let’s revisit the epitope example from the previous chapter, starting without the tricky outlier.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/e100.RData"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>e99 <span class="ot">=</span> e100[<span class="sc">-</span><span class="fu">which.max</span>(e100)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="goodness-of-fit-visual-evaluation" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="goodness-of-fit-visual-evaluation">Goodness-of-fit : visual evaluation</h4>
<p>Our first step is to find a fit from candidate distributions; this requires consulting graphical and quantitative goodness-of-fit plots. For discrete data, we can plot a barplot of frequencies (for continuous data, we would look at the histogram) as in <a href="#fig-twopoisson" class="quarto-xref">Figure&nbsp;<span>2.2</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(e99), <span class="at">space =</span> <span class="fl">0.8</span>, <span class="at">col =</span> <span class="st">"chartreuse4"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-twopoisson" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-twopoisson-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-twopoisson-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;2.2: The observed distribution of the epitope data without the outlier."><img src="02-chap_files/figure-html/fig-twopoisson-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-twopoisson-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.2: The observed distribution of the epitope data without the outlier.
</figcaption>
</figure>
</div>
</div></div></div>
<p>However, it is hard to decide which theoretical distribution fits the data best without using a comparison. One visual <strong>goodness-of-fit</strong> diagram is known as the <strong>rootogram</strong> <span class="citation" data-cites="Tukey:1988">(<a href="16-chap.html#ref-Tukey:1988" role="doc-biblioref">Cleveland 1988</a>)</span>; it hangs the bars with the observed counts from the theoretical red points. If the counts correspond exactly to their theoretical values, the bottom of the boxes will align exactly with the horizontal axis.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"vcd"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>gf1 <span class="ot">=</span> <span class="fu">goodfit</span>( e99, <span class="st">"poisson"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rootogram</span>(gf1, <span class="at">xlab =</span> <span class="st">""</span>, <span class="at">rect_gp =</span> <span class="fu">gpar</span>(<span class="at">fill =</span> <span class="st">"chartreuse4"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-stat-rooto" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stat-rooto-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-stat-rooto-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;2.3: Rootogram showing the square root of the theoretical values as red dots and the square root of the observed frequencies as drop down rectangles. (We’ll see a bit below how the goodfit function decided which \lambda to use.)"><img src="02-chap_files/figure-html/fig-stat-rooto-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stat-rooto-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.3: Rootogram showing the square root of the theoretical values as red dots and the square root of the observed frequencies as drop down rectangles. (We’ll see a bit below how the <code>goodfit</code> function decided which <span class="math inline">\(\lambda\)</span> to use.)
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-rootogrampoisson" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.1
</div>
</div>
<div class="callout-body-container callout-body">
<p>To calibrate what such a plot looks like with a known Poisson variable, use <code>rpois</code> with <span class="math inline">\(\lambda\)</span> = 0.05 to generate 100 Poisson distributed numbers and draw their rootogram.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>simp <span class="ot">=</span> <span class="fu">rpois</span>(<span class="dv">100</span>, <span class="at">lambda =</span> <span class="fl">0.05</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>gf2 <span class="ot">=</span> <span class="fu">goodfit</span>(simp, <span class="st">"poisson"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rootogram</span>(gf2, <span class="at">xlab =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>We see that the rootogram for <code>e99</code> seems to fit the Poisson model reasonably well. But remember, to make this happen we removed the outlier. The Poisson is completely determined by one parameter, often called the Poisson mean <span class="math inline">\(\lambda\)</span>. In most cases where we can guess the data follows a Poisson distribution, we will need to estimate the Poisson parameter from the data.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="The parameter is called the Poisson mean because it is the mean of the theoretical distribution and, as it turns out, is estimated by the sample mean. This overloading of the word is confusing to everyone."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="The parameter is called the Poisson mean because it is the mean of the theoretical distribution and, as it turns out, is estimated by the sample mean. This overloading of the word is confusing to everyone."></a></p>
</figure>
</div>
<figcaption>The parameter is called the Poisson mean because it is the mean of the theoretical distribution <em>and</em>, as it turns out, is estimated by the sample mean. This overloading of the word is confusing to everyone.</figcaption>
</figure>
</div>
</div></div></div>
<p>The most common way of estimating <span class="math inline">\(\lambda\)</span> is to choose the value <span class="math inline">\(\hat{\lambda}\)</span> that makes the observed data the most likely. This is called the <strong>maximum likelihood estimator</strong> <span class="citation" data-cites="Rice:2007">(<a href="16-chap.html#ref-Rice:2007" role="doc-biblioref">Rice 2006, chap. 8</a>, Section 5)</span>, often abbreviated <strong>MLE</strong>. We will illustrate this rather paradoxical idea in the next section.</p>
<p>Although we above took out the extreme observation before taking a guess at the probability distribution, we are going to return to the data with it for the rest of our analysis. In practice we would not know whether there is an outlier, and which data point(s) it is / they are. The effect of leaving it in is to make our estimate of the mean higher. In turns this would make it more likely that we’d observe a value of 7 under the null model, resulting in a larger p-value. So, if the resulting p-value is small even with the outlier included, we are assured that our analysis is up to something real. We call such a tactic being <strong>conservative</strong>: we err on the side of caution, of not detecting something.</p>
</section>
<section id="estimating-the-parameter-of-the-poisson-distribution" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="estimating-the-parameter-of-the-poisson-distribution">Estimating the parameter of the Poisson distribution</h4>
<p>What value for the Poisson mean makes the data the most probable? In a first step, we tally the outcomes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(e100)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>e100
 0  1  2  7 
58 34  7  1 </code></pre>
</div>
</div>
<p>Then we are going to try out different values for the Poisson mean and see which one gives the best fit to our data. If the mean <span class="math inline">\(\lambda\)</span> of the Poisson distribution were 3, the counts would look something like this:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rpois</span>(<span class="dv">100</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 0  1  2  3  4  5  6  7 
 4 12 23 24 14 16  4  3 </code></pre>
</div>
</div>
<p>which has many more 2’s and 3’s than we see in our data. So we see that <span class="math inline">\(\lambda=3\)</span> is unlikely to have produced our data, as the counts do not match up so well.</p>
<div id="wrn-models-simrpois" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.2
</div>
</div>
<div class="callout-body-container callout-body">
<p>Repeat this simulation with different values of <span class="math inline">\(\lambda\)</span>. Can you find one that gives counts close to the observed ones just by trial and error?</p>
</div>
</div>
<p>So we could try out many possible values and proceed by brute force. However, we’ll do something more elegant and use a little mathematics to see which value maximizes the probability of observing our data. Let’s calculate the probability of seeing the data if the value of the Poisson parameter is <span class="math inline">\(m\)</span>. Since we suppose the data derive from independent draws, this probability is simply the product of individual probabilities:</p>
<!-- TODO -->
<p><span class="math display">\[\begin{equation*}
\begin{aligned}
P(58 \times 0,
  34 \times 1,
  7 \times 2,
  \text{one }7 \;|\; \text{data are Poisson}(m)) =
P(0)^{58}\times
P(1)^{34}\times
P(2)^{7}\times
P(7)^{1}.\end{aligned}
\end{equation*}\]</span></p>
<p>For <span class="math inline">\(m=3\)</span> we can compute this<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn2"><p><sup>2</sup>&nbsp;Note how we here use R’s vectorization: the call to <code>dpois</code> returns four values, corresponding to the four different numbers. We then take these to the powers of 58, 34, 7 and 1, respectively, using the <code>^</code> operator, resulting again in four values. Finally, we collapse them into one number, the product, with the <code>prod</code> function.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prod</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span>), <span class="at">lambda =</span> <span class="dv">3</span>) <span class="sc">^</span> (<span class="fu">c</span>(<span class="dv">58</span>, <span class="dv">34</span>, <span class="dv">7</span>, <span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.392143e-110</code></pre>
</div>
</div>
<div id="wrn-models-dpois" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.3
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compute the probability as above for <span class="math inline">\(m=0,1,2\)</span>. Does <span class="math inline">\(m\)</span> have to be integer? Try computing the probability for <span class="math inline">\(m=0.4\)</span> for example.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prod</span>(<span class="fu">dpois</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span>), <span class="at">lambda =</span> <span class="fl">0.4</span>) <span class="sc">^</span> (<span class="fu">c</span>(<span class="dv">58</span>, <span class="dv">34</span>, <span class="dv">7</span>, <span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8.5483e-46</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="page-columns page-full"><p>This probability is the <strong>likelihood function</strong> of <span class="math inline">\(\lambda\)</span>, given the data, and we write it </p><div class="no-row-height column-margin column-container"><span class="margin-aside">Here <span class="math inline">\(L\)</span> stands for likelihood and <span class="math inline">\(f(k)=e^{-\lambda} \,\lambda^k\,/\,k!\)</span>, the Poisson probability we saw earlier.</span></div></div>
<p><span class="math display">\[
L\left(\lambda,\,x=(k_1,k_2,k_3,...)\right)=\prod_{i=1}^{100}f(k_i)
\]</span></p>
<p>Instead of working with multiplications of a hundred small numbers, it is convenient<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> to take the logarithm. Since the logarithm is strictly increasing, if there is a point where the logarithm achieves its maximum within an interval it will also be the maximum for the probability.</p>
<div class="no-row-height column-margin column-container"><div id="fn3"><p><sup>3</sup>&nbsp;That’s usually true both for pencil and paper and for computer calculations.</p></div><div id="fn4"><p><sup>4</sup>&nbsp;Here we again use R’s vector syntax that allows us to write the computation without an explicit loop over the data points. Compared to the code above, here we call <code>dpois</code> on each of the 100 data points, rather than tabulating <code>data</code> with the <code>table</code> function before calling <code>dpois</code> only on the distinct values. This is a simple example for alternative solutions whose results are equivalent, but may differ in how easy it is to read the code or how long it takes to execute.</p></div></div><p>Let’s start with a computational illustration. We compute the likelihood for many different values of the Poisson parameter. To do this we need to write a small function that computes the probability of the data for different values<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>loglikelihood  <span class="ot">=</span>  <span class="cf">function</span>(lambda, <span class="at">data =</span> e100) {</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sum</span>(<span class="fu">log</span>(<span class="fu">dpois</span>(data, lambda)))</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can compute the likelihood for a whole series of <code>lambda</code> values from 0.05 to 0.95 (<a href="#fig-poislikel-1" class="quarto-xref">Figure&nbsp;<span>2.4</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>lambdas <span class="ot">=</span> <span class="fu">seq</span>(<span class="fl">0.05</span>, <span class="fl">0.95</span>, <span class="at">length =</span> <span class="dv">100</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>loglik <span class="ot">=</span> <span class="fu">vapply</span>(lambdas, loglikelihood, <span class="fu">numeric</span>(<span class="dv">1</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lambdas, loglik, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">lwd =</span> <span class="dv">2</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">expression</span>(lambda))</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>m0 <span class="ot">=</span> <span class="fu">mean</span>(e100)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> m0, <span class="at">col =</span> <span class="st">"blue"</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="fu">loglikelihood</span>(m0), <span class="at">col =</span> <span class="st">"purple"</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>m0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.55</code></pre>
</div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-poislikel-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-poislikel-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-poislikel-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;2.4: The red curve is the log-likelihood function. The vertical line shows the value of m (the mean) and the horizontal line the log-likelihood of m. It looks like m maximizes the likelihood."><img src="02-chap_files/figure-html/fig-poislikel-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-poislikel-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.4: The red curve is the log-likelihood function. The vertical line shows the value of <code>m</code> (the mean) and the horizontal line the log-likelihood of <code>m</code>. It looks like <code>m</code> maximizes the likelihood.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-sapply" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.4
</div>
</div>
<div class="callout-body-container callout-body">
<p>What does the <code>vapply</code> function do in the above code? Hint: check its manual page.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><code>vapply</code> takes its first argument, the vector <code>lambdas</code> in this case, and iteratively applies the function <code>loglikelihood</code> (its second argument) to each of the vector elements. As a result, it returns a vector of the results. The function also needs a third argument, <code>numeric(1)</code> in this case, that specifies what type of value each individual call to <code>loglikelihood</code> is supposed to return: a single number. (In general, it could happen that the function sometimes returns something else, say, a character string, or two numbers; in that case it would not be possible to assemble the overall results into a coherent vector, and <code>vapply</code> would complain.)</p>
</div>
</div>
</div>
<p>In fact there is a shortcut: the function <code>goodfit</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>gf  <span class="ot">=</span>  <span class="fu">goodfit</span>(e100, <span class="st">"poisson"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "observed" "count"    "fitted"   "type"     "method"   "df"       "par"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>gf<span class="sc">$</span>par</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$lambda
[1] 0.55</code></pre>
</div>
</div>
<p>The output of <code>goodfit</code> is a composite object called a list. One of its components is called <code>par</code> and contains the value(s) of the fitted parameter(s) for the distribution studied. In this case it’s only one number, the estimate of <span class="math inline">\(\lambda\)</span>.</p>
<div id="wrn-models-goodfit" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.5
</div>
</div>
<div class="callout-body-container callout-body">
<p>What are the other components of the output from the <code>goodfit</code> function?</p>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compare the value of <code>m</code> to the value that we used previously for <span class="math inline">\(\lambda\)</span>, 0.5. Redo the modeling that we did in <a href="01-chap.html" class="quarto-xref"><span>Chapter 1</span></a> with <code>m</code> instead of 0.5.</p>
</div>
</div>
</section>
<section id="classical-statistics-for-classical-data" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="classical-statistics-for-classical-data"><span class="header-section-number">2.3.1</span> Classical statistics for classical data</h3>
<p>Here is a formal proof of our computational finding that the sample mean maximizes the (log-)likelihood.</p>
<p><span class="anchored-eq" data-anchor-id="eq-xzcwgrsaholu" id="eq-xzcwgrsaholu"><span class="math display">\[
\begin{align}
\log L(\lambda, x)
&amp;= \sum_{i=1}^{100} - \lambda + k_i\log\lambda - \log(k_i!) \\
&amp;= -100\lambda + \log\lambda\left(\sum_{i=1}^{100}k_i\right) + \text{const.}
\end{align}
\tag{2.1}\]</span></span></p>
<p>We use the catch-all “const.” for terms that do not depend on <span class="math inline">\(\lambda\)</span> (although they do depend on <span class="math inline">\(x\)</span>, i.e., on the <span class="math inline">\(k_i\)</span>). To find the <span class="math inline">\(\lambda\)</span> that maximizes this, we compute the derivative in <span class="math inline">\(\lambda\)</span> and set it to zero.</p>
<p><span class="anchored-eq" data-anchor-id="eq-zxjuykwlpbsf" id="eq-zxjuykwlpbsf"><span class="math display">\[
\begin{align}
\frac{d}{d\lambda}\log L
        &amp;= -100 + \frac{1}{\lambda} \sum_{i=1}^{100}k_i \stackrel{?}{=}0 \\
\lambda &amp;= \frac{1}{100} \sum_{i=1}^{100}k_i = \bar{k}
\end{align}
\tag{2.2}\]</span></span></p>
<p>You have just seen the first steps of a <em>statistical approach</em>, starting `from the ground up’ (from the data) to infer the model parameter(s): this is statistical <em>estimation</em> of a parameter from data. Another important component will be choosing which family of distributions we use to model our data; that part is done by evaluating the <em>goodness of fit</em>. We will encounter this later.</p>
<p>In the classical <em>statistical testing</em> framework, we consider one single model, that we call the <em>null model</em>, for the data. The null model formulates an “uninteresting” baseline, such as that all observations come from the same random distribution regardless of which group or treatment they are from. We then test whether there is something more interesting going on by computing the probability that the data are compatible with that model. Often, this is the best we can do, since we do not know in sufficient detail what the “interesting”, non-null or alternative model should be. In other situations, we have two competing models that we can compare, as we will see later.</p>
<div id="wrn-models-valuemodelingwithknown" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.6
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the value of modeling with a known distribution? For instance, why is it interesting to know a variable has a Poisson distribution ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Models are concise but expressive representations of the data generating process. For the Poisson for instance, knowing one number allows us to know everything about the distribution, including, as we saw earlier, the probabilities of extreme or rare events.</p>
</div>
</div>
</div>
<p>Another useful direction is <strong>regression</strong>. We may be interested in knowing how our count-based response variable (e.g., the result of counting sequencing reads) depends on a continuous covariate, say, temperature or nutrient concentration. You may already have encountered linear regression, where our model is that the response variable <span class="math inline">\(y\)</span> depends on the covariate <span class="math inline">\(x\)</span> via the equation <span class="math inline">\(y = ax+b + e\)</span>, with parameters <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> (that we need to estimate), and with residuals <span class="math inline">\(e\)</span> whose probability model is a normal distribution (whose variance we usually also need to estimate). For count data the same type of regression model is possible, although the probability distribution for the residuals then needs to be non-normal. In that case we use the <strong>generalized linear models</strong> framework. We will see examples when studying RNA-Seq in <a href="08-chap.html" class="quarto-xref"><span>Chapter 8</span></a> and another type of next generation sequencing data, 16S rRNA data, in <a href="09-chap.html" class="quarto-xref"><span>Chapter 9</span></a>.</p>
<p>Knowing that our probability model involves a Poisson, binomial, multinomial distribution or another parametric family will enable us to have quick answers to questions about the parameters of the model and compute quantities such as p-values and confidence intervals.</p>
</section>
</section>
<section id="FreqBayes:sec:BinDistML" class="level2 page-columns page-full" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="FreqBayes:sec:BinDistML"><span class="header-section-number">2.4</span> Binomial distributions and maximum likelihood</h2>
<p>In a binomial distribution there are two parameters: the number of trials <span class="math inline">\(n\)</span>, which is typically known, and the probability <span class="math inline">\(p\)</span> of seeing a 1 in a trial. This probability is often unknown.</p>
<section id="an-example" class="level3 page-columns page-full" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="an-example"><span class="header-section-number">2.4.1</span> An example</h3>
<p>Suppose we take a sample of <span class="math inline">\(n=120\)</span> males and test them for red-green colorblindness. We can code the data as 0 if the subject is not colorblind and 1 if he is. We summarize the data by the table:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cb
  0   1 
110  10 </code></pre>
</div>
</div>
<div id="wrn-models-mlp" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.7
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which value of <span class="math inline">\(p\)</span> is the most likely given these data?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><span class="math inline">\(\hat{p}=\frac{1}{12}\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(cb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.08333333</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="However, be careful: sometimes, maximum likelihood estimates are harder to guess and to compute, as well as being much less intuitive (see Exercise&nbsp;imp-models-mlmax)."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="However, be careful: sometimes, maximum likelihood estimates are harder to guess and to compute, as well as being much less intuitive (see Exercise&nbsp;imp-models-mlmax)."></a></p>
</figure>
</div>
<figcaption>However, be careful: sometimes, maximum likelihood estimates are harder to guess and to compute, as well as being much less intuitive (see <a href="#imp-models-mlmax" class="quarto-xref">Exercise&nbsp;<span>2.2</span></a>).</figcaption>
</figure>
</div>
</div></div></div>
<p>In this special case, your intuition may give you the estimate <span class="math inline">\(\hat{p}=\frac{1}{12}\)</span>, which turns out to be the maximum likelihood estimate. We put a hat over the letter to remind us that this is not (necessarily) the underlying true value, but an estimate we make from the data.</p>
<p>As before in the case of the Poisson, if we compute the likelihood for many possible <span class="math inline">\(p\)</span>, we can plot it and see where its maximum falls (<a href="#fig-likely1-1" class="quarto-xref">Figure&nbsp;<span>2.5</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>probs  <span class="ot">=</span>  <span class="fu">seq</span>(<span class="dv">0</span>, <span class="fl">0.3</span>, <span class="at">by =</span> <span class="fl">0.005</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>likelihood <span class="ot">=</span> <span class="fu">dbinom</span>(<span class="fu">sum</span>(cb), <span class="at">prob =</span> probs, <span class="at">size =</span> <span class="fu">length</span>(cb))</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(probs, likelihood, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">xlab =</span> <span class="st">"probability of success"</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">ylab =</span> <span class="st">"likelihood"</span>, <span class="at">cex=</span><span class="fl">0.6</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>probs[<span class="fu">which.max</span>(likelihood)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.085</code></pre>
</div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-likely1-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-likely1-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-likely1-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="Figure&nbsp;2.5: Plot of the likelihood as a function of the probabilities. The likelihood is a function on [0, 1]. Here we have zoomed into the range of [0, 0.3], as the likelihood is practically zero for larger values of p."><img src="02-chap_files/figure-html/fig-likely1-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-likely1-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.5: Plot of the likelihood as a function of the probabilities. The likelihood is a function on <span class="math inline">\([0, 1]\)</span>. Here we have zoomed into the range of <span class="math inline">\([0, 0.3]\)</span>, as the likelihood is practically zero for larger values of <span class="math inline">\(p\)</span>.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Note: 0.085 is not exactly the value we expected <span class="math inline">\((\frac{1}{12})\)</span>, and that is because the set of values that we tried (in <code>probs</code>) did not include the exact value of <span class="math inline">\(\frac{1}{12}\simeq 0.0833\)</span>, so we obtained the next best one. We could use numeric optimisation methods to overcome that.</p>
</section>
<section id="FreqBayes:subsec:likelihoodbinomial" class="level3 page-columns page-full" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="FreqBayes:subsec:likelihoodbinomial"><span class="header-section-number">2.4.2</span> Likelihood for the binomial distribution</h3>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="One can come up with different criteria than maximum likelihood, which lead to other estimators. They all carry hats. We’ll see other examples in sec-mixtures."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="One can come up with different criteria than maximum likelihood, which lead to other estimators. They all carry hats. We’ll see other examples in sec-mixtures."></a></p>
</figure>
</div>
<figcaption>One can come up with different criteria than maximum likelihood, which lead to other estimators. They all carry hats. We’ll see other examples in <a href="04-chap.html" class="quarto-xref"><span>Chapter 4</span></a>.</figcaption>
</figure>
</div>
</div></div></div>
<p>The likelihood and the probability are the same mathematical function, only interpreted in different ways – in one case, the function tells us how probable it is to see a particular set of values of the data, given the parameters; in the other case, we consider the data as given, and ask for the parameter value(s) that likely generated these data. Suppose <span class="math inline">\(n=300\)</span>, and we observe <span class="math inline">\(y=40\)</span> successes. Then, for the binomial distribution:</p>
<p><span class="anchored-eq" data-anchor-id="eq-stat" id="eq-stat-dbinom"><span class="math display">\[
f(p\,|\,n,y) = f(y\,|\,n,p)={n \choose y} \, p^y \, (1-p)^{(n-y)}.
\tag{2.3}\]</span></span></p>
<p>Again, it is more convenient to work with the logarithm of the likelihood,</p>
<p><span class="anchored-eq" data-anchor-id="eq-stat" id="eq-stat-loglik"><span class="math display">\[
\log f(p |y) = \log {n \choose y} + y\log(p) + (n-y)\log(1-p).
\tag{2.4}\]</span></span></p>
<p>Here’s a function we can use to calculate it<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>,</p>
<div class="no-row-height column-margin column-container"><div id="fn5"><p><sup>5</sup>&nbsp;In practice, we would try to avoid explicitly computing <code>choose(n, y)</code>, since it can be a very large number that tests the limits of our computer’s floating point arithmetic (for <code>n=300</code> and <code>y=40</code>, it is around 9.8e+49). One could approximate the term using Stirling’s formula, or indeed ignore it, as it is only an additive offset independent of <span class="math inline">\(p\)</span> that does not impact the maximization.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>loglikelihood <span class="ot">=</span> <span class="cf">function</span>(p, <span class="at">n =</span> <span class="dv">300</span>, <span class="at">y =</span> <span class="dv">40</span>) {</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">log</span>(<span class="fu">choose</span>(n, y)) <span class="sc">+</span> y <span class="sc">*</span> <span class="fu">log</span>(p) <span class="sc">+</span> (n <span class="sc">-</span> y) <span class="sc">*</span> <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">-</span> p)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>which we plot for the range of <span class="math inline">\(p\)</span> from 0 to 1 (<a href="#fig-loglikelihood-1" class="quarto-xref">Figure&nbsp;<span>2.6</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>p_seq <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.001</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p_seq, <span class="fu">loglikelihood</span>(p_seq), <span class="at">xlab =</span> <span class="st">"p"</span>, <span class="at">ylab =</span> <span class="st">"log f(p|y)"</span>, <span class="at">type =</span> <span class="st">"l"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-loglikelihood-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-loglikelihood-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-loglikelihood-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;2.6: Plot of the log likelihood function for n=300 and y=40."><img src="02-chap_files/figure-html/fig-loglikelihood-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="240"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-loglikelihood-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.6: Plot of the log likelihood function for <span class="math inline">\(n=300\)</span> and <span class="math inline">\(y=40\)</span>.
</figcaption>
</figure>
</div>
</div></div></div>
<p>The maximum lies at 40/300 = 0.1333… , consistent with intuition, but we see that other values of <span class="math inline">\(p\)</span> are almost equally likely, as the function is quite flat around the maximum. We will see in a later section how Bayesian methods enable us to work with a range of values for <span class="math inline">\(p\)</span> instead of just picking a single maximum.</p>
</section>
</section>
<section id="more-boxesmultinomial-data" class="level2 page-columns page-full" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="more-boxesmultinomial-data"><span class="header-section-number">2.5</span> More boxes:multinomial data</h2>
<section id="dna-count-modeling-base-pairs" class="level3" data-number="2.5.1">
<h3 data-number="2.5.1" class="anchored" data-anchor-id="dna-count-modeling-base-pairs"><span class="header-section-number">2.5.1</span> DNA count modeling: base pairs</h3>
<p>There are four basic molecules of DNA: A - adenine, C - cytosine, G - guanine, T - thymine. The nucleotides are classified into 2 groups: purines (A and G) and pyrimidines (C and T). The binomial would work as a model for the purine/pyrimidine groupings but not if we want to use A, C, G, T; for that we need the multinomial model from <a href="01-chap.html#sec-generative-multinomial" class="quarto-xref"><span>Section 1.4</span></a>. Let’s look at noticeable patterns that occur in these frequencies.</p>
</section>
<section id="Sec:nucleotideBias" class="level3 page-columns page-full" data-number="2.5.2">
<h3 data-number="2.5.2" class="anchored" data-anchor-id="Sec:nucleotideBias"><span class="header-section-number">2.5.2</span> Nucleotide bias</h3>
<p>This section combines estimation and testing by simulation in a real example. Data from one strand of DNA for the genes of <em>Staphylococcus aureus</em> bacterium are available in a <em>fasta</em> file <code>staphsequence.ffn.txt</code>, which we can read with a function from the Bioconductor package <strong><a href="https://bioconductor.org/packages/Biostrings/">Biostrings</a></strong>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Biostrings"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>staph <span class="ot">=</span> <span class="fu">readDNAStringSet</span>(<span class="st">"../data/staphsequence.ffn.txt"</span>, <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the first gene:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>staph[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 1:
    width seq                                               names               
[1]  1362 ATGTCGGAAAAAGAAATTTGGGA...AAAAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">letterFrequency</span>(staph[[<span class="dv">1</span>]], <span class="at">letters =</span> <span class="st">"ACGT"</span>, <span class="at">OR =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  A   C   G   T 
522 219 229 392 </code></pre>
</div>
</div>
<div id="wrn-models-squarebracket" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.8
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why did we use double square brackets in the second line?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The double square brackets <code>[[i]]</code> extract the sequence of the <code>i</code>-th gene as a <em>DNAString</em>, as opposed to the pair of single brackets <code>[i]</code>, which return a <em>DNAStringSet</em> with just a single <em>DNAString</em> in it. If you look at the length of <code>staph[1]</code>, it is 1, whereas <code>staph[[1]]</code> has length 1362.</p>
</div>
</div>
</div>
<div id="wrn-models-nucleotidesfair" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.9
</div>
</div>
<div class="callout-body-container callout-body">
<p>Following a similar procedure as in <a href="01-chap.html#imp-generative-genomefrequency" class="quarto-xref">Exercise&nbsp;<span>1.8</span></a>, test whether the nucleotides are equally distributed across the four nucleotides for this first gene.</p>
</div>
</div>
<p>Due to their different physical properties, evolutionary selection can act on the nucleotide frequencies. So we can ask whether, say, the first ten genes from these data come from the same multinomial. We do not have a prior reference, we just want to decide whether the nucleotides occur in the same proportions in the first 10 genes. If not, this would provide us with evidence for varying selective pressure on these ten genes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>letterFrq <span class="ot">=</span> <span class="fu">vapply</span>(staph, letterFrequency, <span class="at">FUN.VALUE =</span> <span class="fu">numeric</span>(<span class="dv">4</span>),</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">letters =</span> <span class="st">"ACGT"</span>, <span class="at">OR =</span> <span class="dv">0</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(letterFrq) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"gene"</span>, <span class="fu">seq</span>(<span class="at">along =</span> staph))</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>tab10 <span class="ot">=</span> letterFrq[, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>computeProportions <span class="ot">=</span> <span class="cf">function</span>(x) { x<span class="sc">/</span><span class="fu">sum</span>(x) }</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>prop10 <span class="ot">=</span> <span class="fu">apply</span>(tab10, <span class="dv">2</span>, computeProportions)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(prop10, <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
A  0.38  0.36  0.35  0.37  0.35  0.33  0.33  0.34  0.38   0.27
C  0.16  0.16  0.13  0.15  0.15  0.15  0.16  0.16  0.14   0.16
G  0.17  0.17  0.23  0.19  0.22  0.22  0.20  0.21  0.20   0.20
T  0.29  0.31  0.30  0.29  0.27  0.30  0.30  0.29  0.28   0.36</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>p0 <span class="ot">=</span> <span class="fu">rowMeans</span>(prop10)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>p0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        A         C         G         T 
0.3470531 0.1518313 0.2011442 0.2999714 </code></pre>
</div>
</div>
<p>So let’s suppose <code>p0</code> is the vector of multinomial probabilities for all the ten genes and use a Monte Carlo simulation to test whether the departures between the observed letter frequencies and expected values under this supposition are within a plausible range.</p>
<p>We compute the expected counts by taking the <code>outer</code> product of the vector of probabilities p0 with the sums of nucleotide counts from each of the 10 columns, <code>cs</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>cs <span class="ot">=</span> <span class="fu">colSums</span>(tab10)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>cs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9 gene10 
  1362   1134    246   1113   1932   2661    831   1515   1287    696 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>expectedtab10 <span class="ot">=</span> <span class="fu">outer</span>(p0, cs, <span class="at">FUN =</span> <span class="st">"*"</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(expectedtab10)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene1 gene2 gene3 gene4 gene5 gene6 gene7 gene8 gene9 gene10
A   473   394    85   386   671   924   288   526   447    242
C   207   172    37   169   293   404   126   230   195    106
G   274   228    49   224   389   535   167   305   259    140
T   409   340    74   334   580   798   249   454   386    209</code></pre>
</div>
</div>
<p>We can now create a random table with the correct column sums using the <code>rmultinom</code> function. This table is generated according to the null hypothesis that the true proportions are given by <code>p0</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>randomtab10 <span class="ot">=</span> <span class="fu">sapply</span>(cs, <span class="cf">function</span>(s) { <span class="fu">rmultinom</span>(<span class="dv">1</span>, s, p0) } )</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">colSums</span>(randomtab10) <span class="sc">==</span> cs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Now we repeat this B = 1000 times. For each table we compute our test statistic from <a href="01-chap.html#sec-generative-SimulatingForPower" class="quarto-xref"><span>Section 1.4.1</span></a> in <a href="01-chap.html" class="quarto-xref"><span>Chapter 1</span></a> (the function <code>stat</code>) and store the results in the vector <code>simulstat</code>. Together, these values constitute our null distribution, as they were generated under the null hypothesis that <code>p0</code> is the vector of multinomial proportions for each of the 10 genes.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>stat <span class="ot">=</span> <span class="cf">function</span>(obsvd, exptd) {</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">sum</span>((obsvd <span class="sc">-</span> exptd)<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> exptd)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>B <span class="ot">=</span> <span class="dv">1000</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>simulstat <span class="ot">=</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  randomtab10 <span class="ot">=</span> <span class="fu">sapply</span>(cs, <span class="cf">function</span>(s) { <span class="fu">rmultinom</span>(<span class="dv">1</span>, s, p0) })</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat</span>(randomtab10, expectedtab10)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>S1 <span class="ot">=</span> <span class="fu">stat</span>(tab10, expectedtab10)</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(simulstat <span class="sc">&gt;=</span> S1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(simulstat, <span class="at">col =</span> <span class="st">"lavender"</span>, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">75</span>, <span class="at">length.out=</span><span class="dv">50</span>))</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> S1, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">quantile</span>(simulstat, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.95</span>, <span class="fl">0.99</span>)),</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"darkgreen"</span>, <span class="st">"blue"</span>), <span class="at">lty =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-quant12-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-quant12-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-quant12-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Figure&nbsp;2.7: Histogram of simulstat. The value of S1 is marked by the vertical red line, those of the 0.95 and 0.99 quantiles (see next section) by the dotted lines."><img src="02-chap_files/figure-html/fig-quant12-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-quant12-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.7: Histogram of <code>simulstat</code>. The value of <code>S1</code> is marked by the vertical red line, those of the 0.95 and 0.99 quantiles (see next section) by the dotted lines.
</figcaption>
</figure>
</div>
</div></div></div>
<p>The histogram is shown in <a href="#fig-quant12-1" class="quarto-xref">Figure&nbsp;<span>2.7</span></a>. We see that the probability of seeing a value as large as <code>S1</code>=70.1 is very small under the <em>null model</em>. It happened 0 times in our 1000 simulations that a value as big as <code>S1</code> occurred. Thus the ten genes do not seem to come from the same multinomial model.</p>
</section>
</section>
<section id="the-chi2-distribution" class="level2 page-columns page-full" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="the-chi2-distribution"><span class="header-section-number">2.6</span> The <span class="math inline">\(\chi^2\)</span> distribution</h2>
<p>In fact, we could have used statistical theory to come to the same conclusion without running these simulations. The theoretical distribution of the <code>simulstat</code> statistic is called the <span class="math inline">\(\chi^2\)</span> (chi-squared) distribution<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> with parameter 30 (<span class="math inline">\(=10\times(4-1)\)</span>). We can use this for computing the probability of having a value as large as <code>S1</code> <span class="math inline">\(=\)</span> 70.1. As we just saw above, small probabilities are difficult to compute by Monte Carlo: the granularity of the computation is <span class="math inline">\(1/B\)</span>, so we cannot estimate any probabilities smaller than that, and in fact the uncertainty of the estimate is larger. So if any theory is applicable, that tends to be useful. We can check how well theory and simulation match up in our case using another visual goodness-of-fit tool: the <strong>quantile-quantile</strong> (<strong>QQ</strong>) plot. When comparing two distributions, whether from two different samples or from one sample versus a theoretical model, just looking at histograms is not informative enough. We use a method based on the quantiles of each of the distributions.</p>
<div class="no-row-height column-margin column-container"><div id="fn6"><p><sup>6</sup>&nbsp;Strictly speaking, the distribution of <code>simulstat</code> is approximately described by a <span class="math inline">\(\chi^2\)</span> distribution; the approximation is particularly good if the counts in the table are large.</p></div></div><section id="intermezzo-quantiles-and-the-quantile-quantile-plot" class="level3 page-columns page-full" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="intermezzo-quantiles-and-the-quantile-quantile-plot"><span class="header-section-number">2.6.1</span> Intermezzo: quantiles and the quantile-quantile plot</h3>
<p>In the previous chapter, we ordered the 100 sample values <span class="math inline">\(x_{(1)},x_{(2)},...,x_{(100)}\)</span>. Say we want the 22nd percentile. We can take any value between the 22nd and the 23rd value, i.e., any value that fulfills <span class="math inline">\(x_{(22)} \leq c_{0.22} &lt; x_{(23)}\)</span> is acceptable as a 0.22 <strong>quantile</strong> (<span class="math inline">\(c_{0.22}\)</span>). In other words, <span class="math inline">\(c_{0.22}\)</span> is defined by</p>
<p><span class="math display">\[
\frac{\# x_i's \leq c_{0.22}}{n} = 0.22.
\]</span></p>
<p>In <a href="03-chap.html#sec-graphics-ecdf" class="quarto-xref"><span>Section 3.6.7</span></a>, we’ll introduce the <strong>empirical cumulative distribution</strong> function (<strong>ECDF</strong>) <span class="math inline">\(\hat{F}\)</span>, and we’ll see that our definition of <span class="math inline">\(c_{0.22}\)</span> can also be written as <span class="math inline">\(\hat{F}_n(c_{0.22}) = 0.22\)</span>. In <a href="#fig-quant12-1" class="quarto-xref">Figure&nbsp;<span>2.7</span></a>, our histogram of the distribution of <code>simulstat</code>, the quantiles <span class="math inline">\(c_{0.95}\)</span> and <span class="math inline">\(c_{0.99}\)</span> are also shown.</p>
<div id="wrn-models-simulstat" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.10
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="a">
<li><p>Compare the <code>simulstat</code> values and 1000 randomly generated <span class="math inline">\(\chi^2_{30}\)</span> random numbers by displaying them in histograms with 50 bins each.</p></li>
<li><p>Compute the quantiles of the <code>simulstat</code> values and compare them to those of the <span class="math inline">\(\chi_{30}^2\)</span> distribution. Hint:</p></li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>qs <span class="ot">=</span> <span class="fu">ppoints</span>(<span class="dv">100</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(simulstat, qs)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(<span class="fu">qchisq</span>(qs, <span class="at">df =</span> <span class="dv">30</span>), qs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="A name collision occurs here. Statisticians call the summary statistic we just computed as simulstat (sum of squares of weighted differences), the chi-squared or \chi^2 statistic. The theoretical distribution \chi^2_\nu is a distribution in its own right, with a parameter \nu called the degrees of freedom. When reading about the chi-squared or \chi^2, you will need to pay attention to the context to see which meaning is appropriate."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="A name collision occurs here. Statisticians call the summary statistic we just computed as simulstat (sum of squares of weighted differences), the chi-squared or \chi^2 statistic. The theoretical distribution \chi^2_\nu is a distribution in its own right, with a parameter \nu called the degrees of freedom. When reading about the chi-squared or \chi^2, you will need to pay attention to the context to see which meaning is appropriate."></a></p>
</figure>
</div>
<figcaption>A name collision occurs here. Statisticians call the summary statistic we just computed as <code>simulstat</code> (sum of squares of weighted differences), the <strong>chi-squared</strong> or <span class="math inline">\(\chi^2\)</span> <em>statistic</em>. The theoretical <em>distribution</em> <span class="math inline">\(\chi^2_\nu\)</span> is a distribution in its own right, with a parameter <span class="math inline">\(\nu\)</span> called the degrees of freedom. When reading about the chi-squared or <span class="math inline">\(\chi^2\)</span>, you will need to pay attention to the context to see which meaning is appropriate.</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-median" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.11
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you know another name for the 0.5 quantile?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The median.</p>
</div>
</div>
</div>
<div id="wrn-models-quantile" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.12
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the above definition, we were a little vague on how the quantile is defined in general, i.e., not just for 0.22. How is the quantile computed for any number between 0 and 1, including ones that are not multiples of <span class="math inline">\(1/n\)</span>?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Check the manual page of the <code>quantile</code> function and its argument named <code>type</code>.</p>
</div>
</div>
</div>
<p>Now that we have an idea what quantiles are, we can do the quantile-quantile plot. We plot the quantiles of the <code>simulstat</code> values, which we simulated under the null hypothesis, against the theoretical null distribution <span class="math inline">\(\chi^2_{30}\)</span> (<a href="#fig-qqplot3-1" class="quarto-xref">Figure&nbsp;<span>2.8</span></a>):</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqplot</span>(<span class="fu">qchisq</span>(<span class="fu">ppoints</span>(B), <span class="at">df =</span> <span class="dv">30</span>), simulstat, <span class="at">main =</span> <span class="st">""</span>,</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="fu">expression</span>(chi[nu<span class="sc">==</span><span class="dv">30</span>]<span class="sc">^</span><span class="dv">2</span>), <span class="at">asp =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">pch =</span> <span class="dv">16</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-qqplot3-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qqplot3-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-qqplot3-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15" title="Figure&nbsp;2.8: Our simulated statistic’s distribution compared to \chi_{30}^2 using a quantile-quantile (QQ) plot, which shows the theoretical quantiles for the \chi^2_{30} distribution on the horizontal axis and the sampled ones on the vertical axis."><img src="02-chap_files/figure-html/fig-qqplot3-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="272"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qqplot3-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.8: Our simulated statistic’s distribution compared to <span class="math inline">\(\chi_{30}^2\)</span> using a quantile-quantile (QQ) plot, which shows the theoretical <strong>quantiles</strong> for the <span class="math inline">\(\chi^2_{30}\)</span> distribution on the horizontal axis and the sampled ones on the vertical axis.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Having convinced ourselves that <code>simulstat</code> is well described by a <span class="math inline">\(\chi^2_{30}\)</span> distribution, we can use that to compute our p-value, i.e., the probability that under the null hypothesis (counts are distributed as multinomial with probabilities <span class="math inline">\(p_{\text{A}} = 0.35\)</span>, <span class="math inline">\(p_{\text{C}} = 0.15\)</span>, <span class="math inline">\(p_{\text{G}} = 0.2\)</span>, <span class="math inline">\(p_{\text{T}} = 0.3\)</span>) we observe a value as high as <code>S1</code>=70.1:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">pchisq</span>(S1, <span class="at">df =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.74342e-05</code></pre>
</div>
</div>
<p>With such a small p-value, the null hypothesis seems improbable. Note how this computation did not require the 1000 simulations and was faster.</p>
</section>
</section>
<section id="sec-models-Chargaff" class="level2 page-columns page-full" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="sec-models-Chargaff"><span class="header-section-number">2.7</span> Chargaff’s Rule</h2>
<p>The most important pattern in the nucleotide frequencies was discovered by Chargaff <span class="citation" data-cites="Chargaff">(<a href="16-chap.html#ref-Chargaff" role="doc-biblioref">Elson and Chargaff 1952</a>)</span>.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/ChargaffColdSpring_web.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="imgs/ChargaffColdSpring_web.jpg" class="img-fluid quarto-figure quarto-figure-center figure-img"></a></p>
</figure>
</div>
</div></div></div>
<p>Long before DNA sequencing was available, using the weight of the molecules, he asked whether the nucleotides occurred at equal frequencies. He called this the tetranucleotide hypothesis. We would translate that into asking whether <span class="math inline">\(p_{\text{A}} = p_{\text{C}} = p_{\text{G}} = p_{\text{T}}\)</span>.</p>
<p>Unfortunately, Chargaff only published the <em>percentages</em> of the mass present in different organisms for each of the nucleotides, not the measurements themselves.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/ChargaffTable.RData"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>ChargaffTable</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  A    T    C    G
Human-Thymus   30.9 29.4 19.9 19.8
Mycobac.Tuber  15.1 14.6 34.9 35.4
Chicken-Eryth. 28.8 29.2 20.5 21.5
Sheep-liver    29.3 29.3 20.5 20.7
Sea Urchin     32.8 32.1 17.7 17.3
Wheat          27.3 27.1 22.7 22.8
Yeast          31.3 32.9 18.7 17.1
E.coli         24.7 23.6 26.0 25.7</code></pre>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="cell-output-display page-columns page-full">
<div id="fig-ChargaffBars" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-ChargaffBars-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-ChargaffBars-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17" title="Figure&nbsp;2.9: Barplots for the different rows in ChargaffTable. Can you spot the pattern?"><img src="02-chap_files/figure-html/fig-ChargaffBars-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="960"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-ChargaffBars-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.9: Barplots for the different rows in <code>ChargaffTable</code>. Can you spot the pattern?
</figcaption>
</figure>
</div>
</div>
</div>
<div id="wrn-models-chargaff" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.13
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>Do these data seem to come from equally likely multinomial categories?</p></li>
<li><p>Can you suggest an alternative pattern? puted from simulations u</p></li>
<li><p>Can you do a quantitative analysis of the pattern, perhaps inspired by the simulations above?</p></li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Chargaff <em>saw</em> the answer to this question and postulated a pattern called <em>base pairing</em>, which ensured a perfect match of the amount of adenine (A) in the DNA of an organism to the amount of thymine (T). Similarly, whatever the amount of guanine (G), the amount of cytosine (C) would be the same. This is now called Chargaff’s rule. On the other hand, the amount of C/G in an organism could be quite different from that of A/T, with no obvious pattern across organisms. Based on Chargaff’s rule, we might define a statistic</p>
<p><span class="math display">\[
(p_{\text{C}} - p_{\text{G}})^2 + (p_{\text{A}} - p_{\text{T}})^2,
\]</span></p>
<p>summed over all rows of the table. We are going to look at a comparison between the data and what would occur if the nucleotides were ‘exchangeable’, in the sense that the probabilities observed in each row were in no particular order, so that there were no special relationship between the proportions of As and Ts, or between those of Cs and Gs.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>statChf <span class="ot">=</span> <span class="cf">function</span>(x){</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sum</span>((x[, <span class="st">"C"</span>] <span class="sc">-</span> x[, <span class="st">"G"</span>])<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> (x[, <span class="st">"A"</span>] <span class="sc">-</span> x[, <span class="st">"T"</span>])<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>chfstat <span class="ot">=</span> <span class="fu">statChf</span>(ChargaffTable)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>permstat <span class="ot">=</span> <span class="fu">replicate</span>(<span class="dv">100000</span>, {</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>     permuted <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">apply</span>(ChargaffTable, <span class="dv">1</span>, sample))</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">colnames</span>(permuted) <span class="ot">=</span> <span class="fu">colnames</span>(ChargaffTable)</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>     <span class="fu">statChf</span>(permuted)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>pChf <span class="ot">=</span> <span class="fu">mean</span>(permstat <span class="sc">&lt;=</span> chfstat)</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>pChf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.00014</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(permstat, <span class="at">breaks =</span> <span class="dv">100</span>, <span class="at">main =</span> <span class="st">""</span>, <span class="at">col =</span> <span class="st">"lavender"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> chfstat, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-permstatChf-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-permstatChf-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-permstatChf-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18" title="Figure&nbsp;2.10: Histogram of our statistic statChf computed from simulations using per-row permutations of the columns. The value it yields for the observed data is shown by the red line."><img src="02-chap_files/figure-html/fig-permstatChf-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-permstatChf-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.10: Histogram of our statistic <code>statChf</code> computed from simulations using per-row permutations of the columns. The value it yields for the observed data is shown by the red line.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The histogram in <a href="#fig-permstatChf-1" class="quarto-xref">Figure&nbsp;<span>2.10</span></a> shows that it is quite rare to have a value as small as the observed 11.1, where the red line is drawn. The probability of observing a value as small or smaller is <code>pChf</code>=1.4^{-4}. Thus the data strongly support Chargaff’s insight.</p>
</div>
</div>
</div>
<div id="wrn-models-onesided" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.14
</div>
</div>
<div class="callout-body-container callout-body">
<p>When computing <code>pChf</code>, we only looked at the values in the null distribution smaller than the observed value. Why did we do this in a one-sided way here?</p>
</div>
</div>
<section id="two-categorical-variables" class="level3" data-number="2.7.1">
<h3 data-number="2.7.1" class="anchored" data-anchor-id="two-categorical-variables"><span class="header-section-number">2.7.1</span> Two categorical variables</h3>
<p>Up to now, we have visited cases where the data are taken from a sample that can be classified into different boxes: the binomial for Yes/No binary boxes and the multinomial distribution for categorical variables such as A, C, G, T or different genotypes such aa, aA, AA. However it might be that we measure two (or more) categorical variables on a set of subjects, for instance eye color and hair color. We can then cross-tabulate the counts for every combination of eye and hair color. We obtain a table of counts called a <strong>contingency table</strong>. This concept is very useful for many biological data types.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>HairEyeColor[,, <span class="st">"Female"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Eye
Hair    Brown Blue Hazel Green
  Black    36    9     5     2
  Brown    66   34    29    14
  Red      16    7     7     7
  Blond     4   64     5     8</code></pre>
</div>
</div>
<div id="wrn-models-haireyecolor" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.15
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore the <code>HairEyeColor</code> object in R. What data type, shape and dimensions does it have?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>It is a numeric array with three dimensions:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(HairEyeColor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 'table' num [1:4, 1:4, 1:2] 32 53 10 3 11 50 10 30 10 25 ...
 - attr(*, "dimnames")=List of 3
  ..$ Hair: chr [1:4] "Black" "Brown" "Red" "Blond"
  ..$ Eye : chr [1:4] "Brown" "Blue" "Hazel" "Green"
  ..$ Sex : chr [1:2] "Male" "Female"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="do">## ?HairEyeColor</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<section id="color-blindness-and-sex" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="color-blindness-and-sex">Color blindness and sex</h4>
<p>Deuteranopia is a form of red-green color blindness due to the fact that medium wavelength sensitive cones (green) are missing. A deuteranope can only distinguish 2 to 3 different hues, whereas somebody with normal vision sees 7 different hues. A survey for this type of color blindness in human subjects produced a two-way table crossing color blindness and sex.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/Deuteranopia.RData"</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>Deuteranopia</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Men Women
Deute      19     2
NonDeute 1981  1998</code></pre>
</div>
</div>
<p>How do we test whether there is a relationship between sex and the occurrence of color blindness? We postulate the null model with two independent binomials: one for sex and one for color blindness. Under this model we can estimate all the cells’ multinomial probabilities, and we can compare the observed counts to the expected ones. This is done through the <code>chisq.test</code> function in R.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(Deuteranopia)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  Deuteranopia
X-squared = 12.255, df = 1, p-value = 0.0004641</code></pre>
</div>
</div>
<p>The small p value tells us that we should expect to see such a table with only a very small probability under the null model – i.e., if the fractions of deuteranopic color blind among women and men were the same.</p>
<p>We’ll see another test for this type of data called Fisher’s exact test (also known as the hypergeometric test) in <a href="10-chap.html#sec-graphs-GSEA" class="quarto-xref"><span>Section 10.3.2</span></a>. This test is widely used for testing the over-representations of certain types of genes in a list of significantly expressed ones.</p>
</section>
</section>
<section id="a-special-multinomial-hardy-weinberg-equilibrium" class="level3 page-columns page-full" data-number="2.7.2">
<h3 data-number="2.7.2" class="anchored" data-anchor-id="a-special-multinomial-hardy-weinberg-equilibrium"><span class="header-section-number">2.7.2</span> A special multinomial: Hardy-Weinberg equilibrium</h3>
<p>Here we highlight the use of a multinomial with three possible levels created by combining two alleles M and N. Suppose that the overall frequency of allele M in the population is <span class="math inline">\(p\)</span>, so that of N is <span class="math inline">\(q = 1-p\)</span>. The Hardy-Weinberg model looks at the relationship between <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> if there is independence of the frequency of both alleles in a genotype, the so-called <strong>Hardy-Weinberg equilibrium</strong> (HWE). This would be the case if there is random mating in a large population with equal distribution of the alleles among sexes. The probabilities of the three genotypes are then as follows:</p>
<p><span class="anchored-eq" data-anchor-id="eq-HWE" id="eq-HWE"><span class="math display">\[
p_{\text{MM}}=p^2,\quad p_{\text{NN}}=q^2,\quad p_{\text{MN}}=2pq
\tag{2.5}\]</span></span></p>
<p>We only observe the frequencies <span class="anchored-eq" data-anchor-id="eq-HWE" class="math inline">\((n_{\text{MM}},\,n_{\text{MN}},\,n_{\text{NN}})\)</span> for the genotypes MM, MN, NN and the total number <span class="math inline">\(S=n_{\text{MM}}+ n_{\text{MN}}+n_{\text{NN}}\)</span>. We can write the likelihood, i.e., the probability of the observed data when the probabilities of the categories are given by <a href="#eq-HWE" class="quarto-xref">Equation&nbsp;<span>2.5</span></a>, using the multinomial formula</p>
<p><span class="math display">\[
P(n_{\text{MM}},\,n_{\text{MN}},\,n_{\text{NN}}\;|\;p) =
{S \choose n_{\text{MM}},n_{\text{MN}},n_{\text{NN}}}
(p^2)^{n_{\text{MM}}} \,\times\,
(2pq)^{n_{\text{MN}}} \,\times\,
(q^2)^{n_{\text{NN}}},
\]</span></p>
<p>and the log-likelihood under HWE</p>
<p><span class="math display">\[
L(p)=n_{\text{MM}}\log(p^2)+n_{\text{MN}} \log(2pq)+n_{\text{NN}}\log(q^2).
\]</span></p>
<p>The value of <span class="math inline">\(p\)</span> that maximizes the log-likelihood is</p>
<p><span class="math display">\[
p = \frac{n_{\text{MM}} + n_{\text{MN}}/2}{S}.
\]</span></p>
<p>See <span class="citation" data-cites="Rice:2007">(<a href="16-chap.html#ref-Rice:2007" role="doc-biblioref">Rice 2006, chap. 8</a>, Section 5)</span> for the proof. Given the data <span class="math inline">\((n_{\text{MM}},\,n_{\text{MN}},\,n_{\text{NN}})\)</span>, the log-likelihood <span class="math inline">\(L\)</span> is a function of only one parameter, <span class="math inline">\(p\)</span>. <a href="#fig-HardyWeinberg-1" class="quarto-xref">Figure&nbsp;<span>2.11</span></a> shows this log-likelihood function for different values of <span class="math inline">\(p\)</span> for the 216th row of the Mourant data<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>, computed in the following code.</p>
<div class="no-row-height column-margin column-container"><div id="fn7"><p><sup>7</sup>&nbsp;This is genotype frequency data of blood group alleles from <span class="citation" data-cites="Mourant1976">Mourant, Kopec, and Domaniewska-Sobczak (<a href="16-chap.html#ref-Mourant1976" role="doc-biblioref">1976</a>)</span> available through the R package <strong><a href="https://cran.r-project.org/web/packages/HardyWeinberg/">HardyWeinberg</a></strong>.</p></div></div><div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"HardyWeinberg"</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"Mourant"</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>Mourant[<span class="dv">214</span><span class="sc">:</span><span class="dv">216</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Population    Country Total  MM  MN  NN
214    Oceania Micronesia   962 228 436 298
215    Oceania Micronesia   678  36 229 413
216    Oceania     Tahiti   580 188 296  96</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>nMM <span class="ot">=</span> Mourant<span class="sc">$</span>MM[<span class="dv">216</span>]</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>nMN <span class="ot">=</span> Mourant<span class="sc">$</span>MN[<span class="dv">216</span>]</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>nNN <span class="ot">=</span> Mourant<span class="sc">$</span>NN[<span class="dv">216</span>]</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>loglik <span class="ot">=</span> <span class="cf">function</span>(p, <span class="at">q =</span> <span class="dv">1</span> <span class="sc">-</span> p) {</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="dv">2</span> <span class="sc">*</span> nMM <span class="sc">*</span> <span class="fu">log</span>(p) <span class="sc">+</span> nMN <span class="sc">*</span> <span class="fu">log</span>(<span class="dv">2</span><span class="sc">*</span>p<span class="sc">*</span>q) <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> nNN <span class="sc">*</span> <span class="fu">log</span>(q)</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>xv <span class="ot">=</span> <span class="fu">seq</span>(<span class="fl">0.01</span>, <span class="fl">0.99</span>, <span class="at">by =</span> <span class="fl">0.01</span>)</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>yv <span class="ot">=</span> <span class="fu">loglik</span>(xv)</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> xv, <span class="at">y =</span> yv, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lwd =</span> <span class="dv">2</span>,</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"p"</span>, <span class="at">ylab =</span> <span class="st">"log-likelihood"</span>)</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>imax <span class="ot">=</span> <span class="fu">which.max</span>(yv)</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> xv[imax], <span class="at">h =</span> yv[imax], <span class="at">lwd =</span> <span class="fl">1.5</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> yv[imax], <span class="at">lwd =</span> <span class="fl">1.5</span>, <span class="at">col =</span> <span class="st">"purple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-HardyWeinberg-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-HardyWeinberg-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-HardyWeinberg-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19" title="Figure&nbsp;2.11: Plot of the log-likelihood for the Tahiti data."><img src="02-chap_files/figure-html/fig-HardyWeinberg-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-HardyWeinberg-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.11: Plot of the log-likelihood for the Tahiti data.
</figcaption>
</figure>
</div>
</div></div></div>
<p>The maximum likelihood estimate for the probabilities in the multinomial is also obtained by using the observed frequencies as in the binomial case, however the estimates have to account for the relationships between the three probabilities. We can compute <span class="math inline">\(\hat{p}_{\text{MM}}\)</span>, <span class="math inline">\(\hat{p}_{\text{MN}}\)</span> and <span class="math inline">\(\hat{p}_{\text{NN}}\)</span> using the <code>af</code> function from the <strong><a href="https://cran.r-project.org/web/packages/HardyWeinberg/">HardyWeinberg</a></strong> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>phat  <span class="ot">=</span>  <span class="fu">af</span>(<span class="fu">c</span>(nMM, nMN, nNN))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>phat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        A 
0.5793103 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>pMM   <span class="ot">=</span>  phat<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>qhat  <span class="ot">=</span>  <span class="dv">1</span> <span class="sc">-</span> phat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The expected values under Hardy-Weinberg equilibrium are then</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>pHW <span class="ot">=</span> <span class="fu">c</span>(<span class="at">MM =</span> phat<span class="sc">^</span><span class="dv">2</span>, <span class="at">MN =</span> <span class="dv">2</span><span class="sc">*</span>phat<span class="sc">*</span>qhat, <span class="at">NN =</span> qhat<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">c</span>(nMM, nMN, nNN)) <span class="sc">*</span> pHW</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    MM.A     MN.A     NN.A 
194.6483 282.7034 102.6483 </code></pre>
</div>
</div>
<p>which we can compare to the observed values above. We can see that they are quite close to the observed values. We could further test whether the observed values allow us to reject the Hardy-Weinberg model, either by doing a simulation or a <span class="math inline">\(\chi^2\)</span> test as above. A visual evaluation of the goodness-of-fit of Hardy-Weinberg was designed by de Finetti <span class="citation" data-cites="definetti26 Cannings1968">(<a href="16-chap.html#ref-definetti26" role="doc-biblioref">Finetti 1926</a>; <a href="16-chap.html#ref-Cannings1968" role="doc-biblioref">Cannings and Edwards 1968</a>)</span>. It places every sample at a point whose coordinates are given by the proportions of each of the different alleles.</p>
<section id="visual-comparison-to-the-hardy-weinberg-equilibrium" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="visual-comparison-to-the-hardy-weinberg-equilibrium">Visual comparison to the Hardy-Weinberg equilibrium</h4>
<p>We use the <code>HWTernaryPlot</code> function to display the data and compare it to Hardy-Weinberg equilibrium graphically.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>pops <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">69</span>, <span class="dv">128</span>, <span class="dv">148</span>, <span class="dv">192</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>genotypeFrequencies <span class="ot">=</span> <span class="fu">as.matrix</span>(Mourant[, <span class="fu">c</span>(<span class="st">"MM"</span>, <span class="st">"MN"</span>, <span class="st">"NN"</span>)])</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">HWTernaryPlot</span>(genotypeFrequencies[pops, ],</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">markerlab =</span> Mourant<span class="sc">$</span>Country[pops],</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">alpha =</span> <span class="fl">0.0001</span>, <span class="at">curvecols =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="fu">rep</span>(<span class="st">"purple"</span>, <span class="dv">4</span>)),</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">mcex =</span> <span class="fl">0.75</span>, <span class="at">vertex.cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-HWtern" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-HWtern-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-HWtern-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20" title="Figure&nbsp;2.12: This de Finetti plot shows the points as barycenters of the three genotypes using the frequencies as weights on each of the corners of the triangle. The Hardy-Weinberg model is the red curve, the acceptance region is between the two purple lines. We see that the US is the furthest from being in HW equilibrium."><img src="02-chap_files/figure-html/fig-HWtern-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-HWtern-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.12: This <strong>de Finetti plot</strong> shows the points as barycenters of the three genotypes using the frequencies as weights on each of the corners of the triangle. The Hardy-Weinberg model is the red curve, the acceptance region is between the two purple lines. We see that the US is the furthest from being in HW equilibrium.
</figcaption>
</figure>
</div>
</div>
</div>
<div id="wrn-models-HWChisq" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.16
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make the ternary plot as in the code above, then add the other data points to it, what do you notice? You could back up your discussion using the <code>HWChisq</code> function.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HWTernaryPlot</span>(genotypeFrequencies[<span class="sc">-</span>pops, ], </span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">newframe =</span> <span class="cn">FALSE</span>, <span class="at">alpha =</span> <span class="fl">0.0001</span>, <span class="at">cex =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div id="wrn-models-totfreq" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.17
</div>
</div>
<div class="callout-body-container callout-body">
<p>Divide all total frequencies by 50, keeping the same proportions for each of the genotypes, and recreate the ternary plot.</p>
<ul>
<li><p>What happens to the points ?</p></li>
<li><p>What happens to the confidence regions and why?</p></li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-30-contents" aria-controls="callout-30" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-30" class="callout-30-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>newgf <span class="ot">=</span> <span class="fu">round</span>(genotypeFrequencies <span class="sc">/</span> <span class="dv">50</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HWTernaryPlot</span>(newgf[pops, ],</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">markerlab =</span> Mourant<span class="sc">$</span>Country[pops],</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">curvecols =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="fu">rep</span>(<span class="st">"purple"</span>, <span class="dv">4</span>)),</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="fl">0.0001</span>, <span class="at">mcex =</span> <span class="fl">0.75</span>, <span class="at">vertex.cex =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="concatenating-several-multinomials-sequence-motifs-and-logos" class="level3 page-columns page-full" data-number="2.7.3">
<h3 data-number="2.7.3" class="anchored" data-anchor-id="concatenating-several-multinomials-sequence-motifs-and-logos"><span class="header-section-number">2.7.3</span> Concatenating several multinomials: sequence motifs and logos</h3>
<p>The <a href="http://www.sciencegateway.org/resources/kozak.htm">Kozak Motif</a> is a sequence that occurs close to the start codon <strong>ATG</strong> of a coding region. The start codon itself always has a fixed spelling but in positions 5 to the left of it, there is a nucleotide pattern in which the letters are quite far from being equally likely.</p>
<p>We summarize this by giving the <strong>position weight matrix</strong> (PWM) or <strong>position-specific scoring matrix</strong> (PSSM), which provides the multinomial probabilities at every position. This is encoded graphically by the <strong>sequence logo</strong> (<a href="#fig-seqlogo-1" class="quarto-xref">Figure&nbsp;<span>2.13</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb75" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"seqLogo"</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/kozak.RData"</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>kozak</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre data-cap-location="margin"><code>  [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
A 0.33 0.25  0.4 0.15 0.20    1    0    0 0.05
C 0.12 0.25  0.1 0.40 0.40    0    0    0 0.05
G 0.33 0.25  0.4 0.20 0.25    0    0    1 0.90
T 0.22 0.25  0.1 0.25 0.15    0    1    0 0.00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>pwm <span class="ot">=</span> <span class="fu">makePWM</span>(kozak)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">seqLogo</span>(pwm, <span class="at">ic.scale =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-seqlogo-1" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-seqlogo-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-seqlogo-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21" title="Figure&nbsp;2.13: Here is a diagram called a sequence logo for the position dependent multinomial used to model the Kozak motif. It codifies the amount of variation in each of the positions on a log scale. The large letters represent positions where there is no uncertainty about which nucleotide occurs."><img src="02-chap_files/figure-html/fig-seqlogo-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-seqlogo-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.13: Here is a diagram called a sequence logo for the position dependent multinomial used to model the Kozak motif. It codifies the amount of variation in each of the positions on a log scale. The large letters represent positions where there is no uncertainty about which nucleotide occurs.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Over the last sections, we’ve seen how the different “boxes” in the multinomial distributions we have encountered very rarely have equal probabilities. In other words, the parameters <span class="math inline">\(p_1, p_2, ...\)</span> are often different, depending on what is being modeled. Examples of multinomials with unequal frequencies include the twenty different amino acids, blood types and hair color.</p>
<p>If there are multiple categorical variables, we have seen that they are rarely independent (sex and colorblindness, hair and eye color, …). We will see later in <a href="09-chap.html" class="quarto-xref"><span>Chapter 9</span></a> that we can explore the patterns in these dependencies by using multivariate decompositions of the contingency tables. Here, we’ll look at an important special case of dependencies between categorical variables: those that occur along a sequence (or “chain”) of categorical variables, e.g., over time or along a biopolymer.</p>
</section>
</section>
<section id="sec-models-modelingdependencies" class="level2 page-columns page-full" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="sec-models-modelingdependencies"><span class="header-section-number">2.8</span> Modeling sequential dependencies: Markov chains</h2>
<p>If we want to predict tomorrow’s weather, a reasonably good guess is that it will most likely be the same as today’s weather, in addition we may state the probabilities for various kinds of possible changes^[The same reasoning can also be applied in reverse: we could “predict” yesterday’s weather from today’s.. This method for weather forecasting is an example for the Markov assumption: the prediction for tomorrow only depends on the state of things today, but not on yesterday or three weeks ago (all information we could potentially use is already contained in today’s weather). The weather example also highlights that such an assumption need not necessarily be exactly true, but it should be a good enough assumption. It is fairly straightforward to extend this assumption to dependencies on the previous <span class="math inline">\(k\)</span> days, where <span class="math inline">\(k\)</span> is a finite and hopefully not too large number. The essence of the Markov assumption is that the process has a finite “memory”, so that predictions only need to look back for a finite amount of time.</p>
<p>Instead of temporal sequences, we can also apply this to biological sequences. In DNA, we may see specific succession of patterns so that pairs of nucleotides, called digrams, say, [CG, CA, CC] and [CT] are not equally frequent. For instance, in parts of the genome we see more frequent instances of [CA] than we would expect under independence:</p>
<p><span class="math display">\[
P(\mathtt{CA}) \neq P(\mathtt{C}) \, P(\mathtt{A}).
\]</span></p>
<p>We model this dependency in the sequence as a <strong>Markov chain</strong>:</p>
<p><span class="math display">\[
P(\mathtt{CA})
= P(\mathtt{NCA})
= P(\mathtt{NNCA})
= P(...\mathtt{CA})
= P(\mathtt{C}) \, P(\mathtt{A|C}),
\]</span></p>
<p>where N stands for any nucleotide, and <span class="math inline">\(P(\mathtt{A|C})\)</span> stands for “the probability of <span class="math inline">\(\mathtt{A}\)</span>, given that the preceding base is a <span class="math inline">\(\mathtt{C}\)</span>”. <a href="#fig-statsfourstateMC" class="quarto-xref">Figure&nbsp;<span>2.14</span></a> shows a schematic representation of such transitions on a graph.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-statsfourstateMC" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-statsfourstateMC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-statsfourstateMC-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22" title="Figure&nbsp;2.14: Visualisation of a 4-state Markov chain. The probability of each possible digram (e.,g., CA) is given by the weight of the edge between the corresponding nodes. So for instance, the probability of CA is given by the edge C$ o$ A. We’ll see in sec-images how to use R packages to draw these type of network graphs."><img src="02-chap_files/figure-html/fig-statsfourstateMC-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-statsfourstateMC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.14: Visualisation of a 4-state Markov chain. The probability of each possible digram (e.,g., CA) is given by the weight of the edge between the corresponding nodes. So for instance, the probability of CA is given by the edge C$ o$ A. We’ll see in <a href="11-chap.html" class="quarto-xref"><span>Chapter 11</span></a> how to use R packages to draw these type of network graphs.
</figcaption>
</figure>
</div>
</div></div></div>
</section>
<section id="bayesian-thinking" class="level2 page-columns page-full" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="bayesian-thinking"><span class="header-section-number">2.9</span> Bayesian Thinking</h2>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-FreqBayes-turtles" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-FreqBayes-turtles-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/turtlesalltheway.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23" title="Figure&nbsp;2.15: Turtles all the way down. Bayesian modeling of the uncertainty of the parameter of a distribution is done by using a random variable whose distribution may depend on parameters whose uncertainty can be modeled as a random variable; these are called hierarchical models."><img src="imgs/turtlesalltheway.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-FreqBayes-turtles-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.15: Turtles all the way down. Bayesian modeling of the uncertainty of the parameter of a distribution is done by using a random variable whose distribution may depend on parameters whose uncertainty can be modeled as a random variable; these are called hierarchical models.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Up to now we have followed a classical approach, where the parameters of our models and the distributions they use, i.e., the probabilities of the possible different outcomes, represent long term frequencies. The parameters are—at least conceptually—definite, knowable and fixed. We may not know them, so we estimate them from the data at hand. However, such an approach does not take into account any information that we might already have, and that might inform us on the parameters or make certain parameter values or their combinations more likely than others—even <em>before</em> we see any of the current set of data. For that we need a different approach, in which we use probabilistic models (i.e., distributions) to express our prior knowledge<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a> about the parameters, and use the current data to <em>update</em> such knowledge, for instance by shifting those distributions or making them more narrow. Such an approach is provided by the Bayesian paradigm (<a href="#fig-FreqBayes-turtles" class="quarto-xref">Figure&nbsp;<span>2.15</span></a>).</p>
<div class="no-row-height column-margin column-container"><div id="fn8"><p><sup>8</sup>&nbsp;Some like to say “our belief(s)”.</p></div></div><p>The Bayesian paradigm is a practical approach where <em>prior</em> and <em>posterior</em> distributions to model our knowledge <em>before</em> and <em>after</em> collecting some data and making an observation. It can be iterated ad infinitum: the posterior after one round of data generation can be used as the prior for the next round. Thus, it is also particularly useful for integrating or combining information from different sources.</p>
<p>The same idea can also be applied to hypothesis testing, where we want to use data to decide whether we believe that a certain statement—which we might call the hypothesis <span class="math inline">\(H\)</span>—is true. Here, our “parameter” is the probability that <span class="math inline">\(H\)</span> is true, and we can formalize our prior knowledge in the form of a <strong>prior</strong> probability, written <span class="math inline">\(P(H)\)</span><a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>. After we see the data, we have the <strong>posterior</strong> probability. We write it as <span class="math inline">\(P(H\,|\,D)\)</span>, the probability of <span class="math inline">\(H\)</span> given that we saw <span class="math inline">\(D\)</span>. This may be higher or lower than <span class="math inline">\(P(H)\)</span>, depending on what the data <span class="math inline">\(D\)</span> were.</p>
<div class="no-row-height column-margin column-container"><div id="fn9"><p><sup>9</sup>&nbsp;For a so-called frequentist, such a probability does not exist. Their viewpoint is that, although the truth is unknown, in reality the hypothesis is either true or false; there is no meaning in calling it, say, “70% true”.</p></div></div><section id="sec-models-haplotypefrequencies" class="level3 page-columns page-full" data-number="2.9.1">
<h3 data-number="2.9.1" class="anchored" data-anchor-id="sec-models-haplotypefrequencies"><span class="header-section-number">2.9.1</span> Example: haplotype frequencies</h3>
<p>To keep the mathematical formalism to a minimum, we start with an example from forensics, using combined signatures (haplotypes) from the Y chromosome.</p>
<p>A <em>haplotype</em> is a collection of alleles (DNA sequence variants) that are spatially adjacent on a chromosome, are usually inherited together (recombination tends not to disconnect them), and thus are genetically linked. In this case we are looking at linked variants on the Y chromosome.</p>
<p>First we’ll look at the motivation behind haplotype frequency analyses, then we’ll revisit the idea of likelihood. After this, we’ll explain how we can think of unknown parameters as being random numbers themselves, modeling their uncertainty with a prior distribution. Then we will see how to incorporate new data observed into the probability distributions and compute posterior confidence statements about the parameters.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="cell-output-display page-columns page-full">
<div id="fig-STRDefinition" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-STRDefinition-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/STRDefinition.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24" title="Figure&nbsp;2.16: A short tandem repeat (STR) in DNA occurs when a pattern of two or more nucleotides is repeated, and the repeated sequences are directly adjacent to each other. An STR is also known as a microsatellite. The pattern can range in length from 2 to 13 nucleotides, and the number of repeats is highly variable across individuals. STR numbers can be used as genetic signatures."><img src="imgs/STRDefinition.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-STRDefinition-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.16: A short tandem repeat (STR) in DNA occurs when a pattern of two or more nucleotides is repeated, and the repeated sequences are directly adjacent to each other. An STR is also known as a microsatellite. The pattern can range in length from 2 to 13 nucleotides, and the number of repeats is highly variable across individuals. STR numbers can be used as genetic signatures.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-YSTRPositions" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-YSTRPositions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/YSTRPositions.jpg" class="lightbox" data-gallery="quarto-lightbox-gallery-25" title="Figure&nbsp;2.17: Location of short tandem repeats (STR) on the human Y chromosome. Source: https://strbase.nist.gov/ystrpos1.htm"><img src="imgs/YSTRPositions.jpg" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-YSTRPositions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.17: Location of short tandem repeats (STR) on the human Y chromosome. Source: <a href="https://strbase.nist.gov/ystrpos1.htm">https://strbase.nist.gov/ystrpos1.htm</a>
</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-USY-STR" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-USY-STR-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/USY-STR.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26" title="Figure&nbsp;2.18: Y STR haplotype lookup from a database used by the FBI."><img src="imgs/USY-STR.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-USY-STR-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.18: Y STR haplotype lookup from a database used by the FBI.
</figcaption>
</figure>
</div>
</div></div></div>
<p>We’re interested in the frequencies of particular Y-haplotypes that consist of a set of different short tandem repeats (STR). The combination of STR numbers at the specific locations used for DNA forensics are labeled by the number of repeats at the specific positions. Here is a short excerpt of such an STR haplotype table:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>haplo6 <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/haplotype6.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>haplo6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Individual DYS19 DXYS156Y DYS389m DYS389n DYS389p
1         H1    14       12       4      12       3
2         H3    15       13       4      13       3
3         H4    15       11       5      11       3
4         H5    17       13       4      11       3
5         H7    13       12       5      12       3
6         H8    16       11       5      12       3</code></pre>
</div>
</div>
<div class="page-columns page-full"><p>The table says that the haplotype H1 has 14 repeats at position <code>DYS19</code>, 12 repeats at position <code>DXYS156Y</code>, etc.  Suppose we want to find the underlying proportion <span class="math inline">\(p\)</span> of a particular haplotype in a population of interest, by haplotyping <span class="math inline">\(n=300\)</span> men; and suppose we found H1 in <span class="math inline">\(y=40\)</span> of them. We are going to use the binomial distribution <span class="math inline">\(B(n,p)\)</span> to model this, with <span class="math inline">\(p\)</span> unknown.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">The haplotypes created through the use of these Y-STR profiles are shared between men in the same patriarchal lineages. Thus, it is possible that two different men share the same profile.</span></div></div>
</section>
<section id="subsection:simulationbinomial" class="level3 page-columns page-full" data-number="2.9.2">
<h3 data-number="2.9.2" class="anchored" data-anchor-id="subsection:simulationbinomial"><span class="header-section-number">2.9.2</span> Simulation study of the Bayesian paradigm for the binomial distribution</h3>
<p>Instead of assuming that our parameter <span class="math inline">\(p\)</span> has one single value (e.g., the maximum likelihood estimate 40/300), the Bayesian approach allows us to see it as a draw from a statistical distribution. The distribution expresses our belief about the possible values of the parameter <span class="math inline">\(p\)</span>. In principle, we can use any distribution that we like whose possible values are permissible for <span class="math inline">\(p\)</span>. As here we are looking at a parameter that expresses a proportion or a probability, and which takes its values between 0 and 1, it is convenient to use the <em>Beta distribution</em>. Its density formula is written</p>
<p><span class="anchored-eq" data-anchor-id="eq-stat" id="eq-stat-densitybeta"><span class="math display">\[
f_{\alpha,\beta}(x) = \frac{x^{\alpha-1}\,(1-x)^{\beta-1}}{\text{B}(\alpha,\beta)}\quad\text{where}\quad
\text{B}(\alpha,\beta)=\frac{\Gamma(\alpha)\Gamma(\beta)}{\Gamma(\alpha+\beta)}.
\tag{2.6}\]</span></span></p>
<p>We can see in <a href="#fig-histobeta2" class="quarto-xref">Figure&nbsp;<span>2.19</span></a> how this function depends on two parameters <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span>, which makes it a very flexible family of distributions (it can “fit” a lot different situations). And it has a nice mathematical property: if we start with a prior belief on <span class="math inline">\(p\)</span> that is Beta-shaped, observe a dataset of <span class="math inline">\(n\)</span> binomial trials, then update our belief, the posterior distribution on <span class="math inline">\(p\)</span> will also have a Beta distribution, albeit with updated parameters. This is a mathematical fact. We will not prove it here, however we demonstrate it by simulation.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-histobeta2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-histobeta2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-histobeta2-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-27" title="Figure&nbsp;2.19: Beta distributions with \alpha=10,20,50 and \beta=30,60,150. We can use these as a prior for probability of success in a binomial experiment. These three distributions have the same mean (\frac{\alpha}{\alpha +\beta}), but different concentrations around the mean."><img src="02-chap_files/figure-html/fig-histobeta2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-histobeta2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.19: Beta distributions with <span class="math inline">\(\alpha=10,20,50\)</span> and <span class="math inline">\(\beta=30,60,150\)</span>. We can use these as a <strong>prior</strong> for probability of success in a binomial experiment. These three distributions have the same mean (<span class="math inline">\(\frac{\alpha}{\alpha +\beta}\)</span>), but different concentrations around the mean.
</figcaption>
</figure>
</div>
</div></div></div>
</section>
<section id="the-distribution-of-y" class="level3 page-columns page-full" data-number="2.9.3">
<h3 data-number="2.9.3" class="anchored" data-anchor-id="the-distribution-of-y"><span class="header-section-number">2.9.3</span> The distribution of <span class="math inline">\(Y\)</span></h3>
<p>For a given choice of <span class="anchored-eq" data-anchor-id="eq-stat" class="math inline">\(p\)</span>, we know what the distribution of <span class="math inline">\(Y\)</span> is, by virtue of <a href="#eq-stat-dbinom" class="quarto-xref">Equation&nbsp;<span>2.3</span></a>. But what is the distribution of <span class="math inline">\(Y\)</span> if <span class="math inline">\(p\)</span> itself also varies according to some distribution? We call this the <strong>marginal distribution</strong> of <span class="math inline">\(Y\)</span>. Let’s simulate that. First we generate a random sample <code>rp</code> of 100000 <span class="math inline">\(p\)</span>s. For each of them, we then generate a random sample of <span class="math inline">\(Y\)</span>, shown in <a href="#fig-histmarginal-1" class="quarto-xref">Figure&nbsp;<span>2.20</span></a>. In the code below, for the sake of demonstration we use the parameters 50 and 350 for the prior. Such a prior is already quite informative (“peaked”) and may, e.g., reflect beliefs we have based on previous studies. In <a href="#wrn-models-softerprior" class="quarto-xref">Question&nbsp;<span>2.20</span></a> you have the opportunity to try out a “softer” (less informative) prior. We again use <code>vapply</code> to apply a function, the unnamed (anonymous) function of <code>x</code>, across all elements of <code>rp</code> to obtain as a result another vector <code>y</code> of the same length.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>rp <span class="ot">=</span> <span class="fu">rbeta</span>(<span class="dv">100000</span>, <span class="dv">50</span>, <span class="dv">350</span>)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">=</span> <span class="fu">vapply</span>(rp, </span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>           <span class="cf">function</span>(x) <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">prob =</span> x, <span class="at">size =</span> <span class="dv">300</span>), </span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>           <span class="fu">integer</span>(<span class="dv">1</span>))</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(y, <span class="at">breaks =</span> <span class="dv">50</span>, <span class="at">col =</span> <span class="st">"orange"</span>, <span class="at">main =</span> <span class="st">""</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-histmarginal-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-histmarginal-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-histmarginal-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-28" title="Figure&nbsp;2.20: Marginal Distribution of Y."><img src="02-chap_files/figure-html/fig-histmarginal-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-histmarginal-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.20: Marginal Distribution of <span class="math inline">\(Y\)</span>.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-vectorize" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.18
</div>
</div>
<div class="callout-body-container callout-body">
<p>Verify that we could have gotten the same result as in the above code chunk by using R’s vectorisation capabilities and writing <code>rbinom(length(rp), rp, size = 300)</code>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-32-contents" aria-controls="callout-32" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-32" class="callout-32-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0xbebe</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>y1 <span class="ot">=</span> <span class="fu">vapply</span>(rp, </span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>            <span class="cf">function</span>(x) <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">prob =</span> x, <span class="at">size =</span> <span class="dv">300</span>), </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>            <span class="fu">integer</span>(<span class="dv">1</span>))</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0xbebe</span>)</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>y2 <span class="ot">=</span> <span class="fu">rbinom</span>(<span class="fu">length</span>(rp), rp, <span class="at">size =</span> <span class="dv">300</span>)</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(y1, y2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="histogram-of-all-the-ps-such-that-y40-the-posterior-distribution" class="level3 page-columns page-full" data-number="2.9.4">
<h3 data-number="2.9.4" class="anchored" data-anchor-id="histogram-of-all-the-ps-such-that-y40-the-posterior-distribution"><span class="header-section-number">2.9.4</span> Histogram of all the <span class="math inline">\(p\)</span>s such that <span class="math inline">\(Y=40\)</span>: the posterior distribution</h3>
<p>So now let’s compute the posterior distribution of <span class="math inline">\(p\)</span> by conditioning on those outcomes where <span class="math inline">\(Y\)</span> was 40. We compare it to the theoretical posterior, <code>densPostTheory</code>, of which more below. The results are shown in <a href="#fig-densityposterior-1" class="quarto-xref">Figure&nbsp;<span>2.21</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>pPostEmp <span class="ot">=</span> rp[ y <span class="sc">==</span> <span class="dv">40</span> ]</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pPostEmp, <span class="at">breaks =</span> <span class="dv">40</span>, <span class="at">col =</span> <span class="st">"chartreuse4"</span>, <span class="at">main =</span> <span class="st">""</span>,</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">probability =</span> <span class="cn">TRUE</span>, <span class="at">xlab =</span> <span class="st">"posterior p"</span>)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>p_seq <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.001</span>)</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>densPostTheory <span class="ot">=</span> <span class="fu">dbeta</span>(p_seq, <span class="dv">50</span> <span class="sc">+</span> <span class="dv">40</span>, <span class="dv">350</span> <span class="sc">+</span> <span class="dv">260</span>)</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(p_seq, densPostTheory, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-densityposterior-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-densityposterior-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-densityposterior-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-29" title="Figure&nbsp;2.21: Only choosing the values of the distribution with Y=40 gives the posterior distribution of p. The histogram (green) shows the simulated values for the posterior distribution, the line the density of a Beta distribution with the theoretical parameters."><img src="02-chap_files/figure-html/fig-densityposterior-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-densityposterior-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.21: Only choosing the values of the distribution with <span class="math inline">\(Y=40\)</span> gives the posterior distribution of <span class="math inline">\(p\)</span>. The histogram (green) shows the simulated values for the posterior distribution, the line the density of a Beta distribution with the theoretical parameters.
</figcaption>
</figure>
</div>
</div></div></div>
<p>We can also check the means of both distributions computed above and see that they are close to 4 significant digits.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(pPostEmp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.128726</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>dp <span class="ot">=</span> p_seq[<span class="dv">2</span>] <span class="sc">-</span> p_seq[<span class="dv">1</span>]</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(p_seq <span class="sc">*</span> densPostTheory <span class="sc">*</span> dp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1285714</code></pre>
</div>
</div>
<p>To approximate the mean of the theoretical density <code>densPostTheory</code>, we have above literally computed the integral</p>
<p><span class="math display">\[
\int_0^1 p \, f(p) \, dp
\]</span></p>
<p>using numerical integration, i.e., the <code>sum</code> over the integral. This is not always convenient (or feasible), in particular if our model involves not just a single, scalar parameter <span class="math inline">\(p\)</span>, but has many parameters, so that we are dealing with a high-dimensional parameter vector and a high-dimensional integral. If the integral cannot be computed analytically, we can use <strong>Monte Carlo integration</strong>. You already saw a very simple instance of Monte Carlo integration in the code above, where we sampled the posterior with <code>pPostEmp</code> and performed integration to compute the posterior mean by calling R’s <code>mean</code> function. In this case, an alternative Monte Carlo algorithm is to generate posterior samples using the <code>rbeta</code> function directly with the right parameters.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>pPostMC <span class="ot">=</span> <span class="fu">rbeta</span>(<span class="at">n =</span> <span class="dv">100000</span>, <span class="dv">90</span>, <span class="dv">610</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(pPostMC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1285718</code></pre>
</div>
</div>
<p>We can check the concordance between the Monte Carlo samples <code>pPostMC</code> and <code>pPostEmp</code>, generated in slightly different ways, using a <strong>quantile-quantile plot</strong> (<strong>QQ-plot</strong>, <a href="#fig-qqplotbeta-1" class="quarto-xref">Figure&nbsp;<span>2.22</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqplot</span>(pPostMC, pPostEmp, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">asp =</span> <span class="dv">1</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-qqplotbeta-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-qqplotbeta-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-qqplotbeta-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-30" title="Figure&nbsp;2.22: Quantile-quantile (QQ) plot of our Monte Carlo sample pPostMC from the theoretical distribution and our simulation sample pPostEmp. We could also similarly compare either of these two distributions to the theoretical distribution function pbeta(., 90, 610). If the curve lies on the line y=x, this indicates a good agreement. There are some random differences at the tails."><img src="02-chap_files/figure-html/fig-qqplotbeta-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-qqplotbeta-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.22: Quantile-quantile (QQ) plot of our Monte Carlo sample <code>pPostMC</code> from the theoretical distribution and our simulation sample <code>pPostEmp</code>. We could also similarly compare either of these two distributions to the theoretical distribution function <code>pbeta(., 90, 610)</code>. If the curve lies on the line <span class="math inline">\(y=x\)</span>, this indicates a good agreement. There are some random differences at the tails.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-MC" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.19
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the difference between the simulation that results in <code>pPostEmp</code> and the Monte Carlo simulation that leads to <code>pPostMC</code>?</p>
</div>
</div>
</section>
<section id="the-posterior-distribution-is-also-a-beta" class="level3 page-columns page-full" data-number="2.9.5">
<h3 data-number="2.9.5" class="anchored" data-anchor-id="the-posterior-distribution-is-also-a-beta"><span class="header-section-number">2.9.5</span> The posterior distribution is also a Beta</h3>
<p>Now we have seen that the posterior distribution is also a Beta. In our case its parameters <span class="math inline">\(\alpha=90\)</span> and <span class="math inline">\(\beta=610\)</span> were obtained by summing the prior parameters <span class="math inline">\(\alpha=50\)</span>, <span class="math inline">\(\beta=350\)</span> with the observed successes <span class="math inline">\(y=40\)</span> and the observed failures <span class="math inline">\(n-y=260\)</span>, thus obtaining the posterior</p>
<p><span class="math display">\[
\text{Beta}(90,\, 610)=\text{Beta}(\alpha+y,\text{Beta}+(n-y)).
\]</span></p>
<p>We can use it to give the best<a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a> estimate we can for <span class="math inline">\(p\)</span> with its uncertainty given by the posterior distribution.</p>
<div class="no-row-height column-margin column-container"><div id="fn10"><p><sup>10</sup>&nbsp;We could take the value that maximizes the posterior distribution as our best estimate, this is called the <strong>MAP</strong> estimate, and in this case it would be <span class="math inline">\(\frac{\alpha-1}{\alpha+\beta-2}=\frac{89}{698}\doteq 0.1275\)</span>.</p></div></div></section>
<section id="suppose-we-had-a-second-series-of-data" class="level3" data-number="2.9.6">
<h3 data-number="2.9.6" class="anchored" data-anchor-id="suppose-we-had-a-second-series-of-data"><span class="header-section-number">2.9.6</span> Suppose we had a second series of data</h3>
<p>After seeing our previous data, we now have a new prior, <span class="math inline">\(\text{Beta}(90, 610)\)</span>. Suppose we collect a new set of data with <span class="math inline">\(n=150\)</span> observations and <span class="math inline">\(y=25\)</span> successes, thus 125 failures. Now what would we take to be our best guess at <span class="math inline">\(p\)</span>?</p>
<p>Using the same reasoning as before, the new posterior will be <span class="math inline">\(\text{Beta}(90+25=115,\, 610+125=735)\)</span>. The mean of this distribution is <span class="math inline">\(\frac{115}{115+735}=\frac{115}{850}\simeq 0.135\)</span>, thus one estimate of <span class="math inline">\(p\)</span> would be 0.135. The <strong>maximum a posteriori</strong> (MAP) estimate would be the mode of <span class="math inline">\(\text{Beta}(115, 735)\)</span>, ie <span class="math inline">\(\frac{114}{848}\simeq 0.134\)</span>. Let’s check this numerically.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>densPost2 <span class="ot">=</span> <span class="fu">dbeta</span>(p_seq, <span class="dv">115</span>, <span class="dv">735</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>mcPost2   <span class="ot">=</span> <span class="fu">rbeta</span>(<span class="fl">1e6</span>, <span class="dv">115</span>, <span class="dv">735</span>)</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(p_seq <span class="sc">*</span> densPost2 <span class="sc">*</span> dp)   <span class="co"># mean, by numeric integration</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1352941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(mcPost2)                 <span class="co"># mean by MC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1352655</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>p_seq[<span class="fu">which.max</span>(densPost2)]   <span class="co"># MAP estimate</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.134</code></pre>
</div>
</div>
<div id="wrn-models-softerprior" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.20
</div>
</div>
<div class="callout-body-container callout-body">
<p>Redo all the computations replacing our original prior with a softer prior (less peaked), meaning that we use less prior information. For instance, try Beta(1,1), which is the uniform distribution. How much does this change the final result?</p>
</div>
</div>
<p>As a general rule, the prior rarely changes the posterior distribution substantially except if it is very peaked. This would be the case if, at the outset, we were already rather sure of what to expect. Another case when the prior has an influence is if there is very little data.</p>
<p>The best situation to be in is to have enough data to swamp the prior so that its choice doesn’t have much impact on the final result.</p>
</section>
<section id="confidence-statements-for-the-proportion-parameter" class="level3 page-columns page-full" data-number="2.9.7">
<h3 data-number="2.9.7" class="anchored" data-anchor-id="confidence-statements-for-the-proportion-parameter"><span class="header-section-number">2.9.7</span> Confidence Statements for the proportion parameter</h3>
<p>Now it is time to conclude about what the proportion <span class="math inline">\(p\)</span> actually is, given the data. One summary is a posterior credibility interval, which is a Bayesian analog of the confidence interval. We can take the 2.5 and 97.5-th percentiles of the posterior distribution: <span class="math inline">\(P(q_{2.5\%} \leq p \leq q_{97.5\%})=0.95\)</span>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(mcPost2, <span class="fu">c</span>(<span class="fl">0.025</span>, <span class="fl">0.975</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     2.5%     97.5% 
0.1131080 0.1590221 </code></pre>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-DESeq2-Prediction-Interval" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-DESeq2-Prediction-Interval-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/DESeq2-Prediction-Interval.png" class="lightbox" data-gallery="quarto-lightbox-gallery-31" title="Figure&nbsp;2.23: An example from @LoveDESeq2 shows plots of the likelihoods (solid lines, scaled to integrate to 1) and the posteriors (dashed lines) for the green and purple genes and of the prior (solid black line): due to the higher dispersion of the purple gene, its likelihood is wider and less peaked (indicating less information), and the prior has more influence on its posterior than for the green gene. The stronger curvature of the green posterior at its maximum translates to a smaller reported standard error for the MAP logarithmic fold change (LFC) estimate (horizontal error bar)."><img src="imgs/DESeq2-Prediction-Interval.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-DESeq2-Prediction-Interval-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.23: An example from <span class="citation" data-cites="LoveDESeq2">Love, Huber, and Anders (<a href="16-chap.html#ref-LoveDESeq2" role="doc-biblioref">2014</a>)</span> shows plots of the likelihoods (solid lines, scaled to integrate to 1) and the posteriors (dashed lines) for the green and purple genes and of the prior (solid black line): due to the higher dispersion of the purple gene, its likelihood is wider and less peaked (indicating less information), and the prior has more influence on its posterior than for the green gene. The stronger curvature of the green posterior at its maximum translates to a smaller reported standard error for the MAP logarithmic fold change (LFC) estimate (horizontal error bar).
</figcaption>
</figure>
</div>
</div></div></div>
</section>
</section>
<section id="example-occurrence-of-a-nucleotide-pattern-in-a-genome" class="level2 page-columns page-full" data-number="2.10">
<h2 data-number="2.10" class="anchored" data-anchor-id="example-occurrence-of-a-nucleotide-pattern-in-a-genome"><span class="header-section-number">2.10</span> Example: occurrence of a nucleotide pattern in a genome</h2>
<p>The examples we have seen up to now have concentrated on distributions of discrete counts and categorical data. Let’s look at an example of distributions of distances, which are quasi-continuous. This case study of the distributions of the distances between instances of a specific motif in genome sequences will also allow us to explore specific genomic sequence manipulations in Bioconductor.</p>
<div class="page-columns page-full"><p>The <strong><a href="https://bioconductor.org/packages/Biostrings/">Biostrings</a></strong> package provides tools for working with sequence data. The essential data structures, or <em>classes</em> as they are known in R, are <em>DNAString</em> and <em>DNAStringSet</em>. These enable us to work with one or multiple DNA sequences efficiently .</p><div class="no-row-height column-margin column-container"><span class="margin-aside">The <strong><a href="https://bioconductor.org/packages/Biostrings/">Biostrings</a></strong> package also contains additional classes for representing amino acid sequences, and more general, biology-inspired sequences.</span></div></div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Biostrings"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="wrn-models-Biostrings" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.21
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore some of the useful data and functions provided in the Biostrings package by exploring the tutorial vignette.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-36-contents" aria-controls="callout-36" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-36" class="callout-36-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The first line prints genetic code information, the second one returns IUPAC nucleotide ambiguity codes. The third line lists all the vignettes available in the <strong><a href="https://bioconductor.org/packages/Biostrings/">Biostrings</a></strong> package, the fourth display one particular vignette.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>GENETIC_CODE</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>IUPAC_CODE_MAP</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="fu">vignette</span>(<span class="at">package =</span> <span class="st">"Biostrings"</span>)</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vignette</span>(<span class="st">"BiostringsQuickOverview"</span>, <span class="at">package =</span> <span class="st">"Biostrings"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>This last command will open a list in your browser window from which you can access the documentation<a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a>. The <strong><a href="https://bioconductor.org/packages/BSgenome/">BSgenome</a></strong> package provides access to many genomes, and you can access the names of the data packages that contain the whole genome sequences by typing</p>
<div class="no-row-height column-margin column-container"><div id="fn11"><p><sup>11</sup>&nbsp;Vignettes are manuals for the packages complete with examples and case studies.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BSgenome"</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>ag <span class="ot">=</span> <span class="fu">available.genomes</span>()</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(ag)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 113</code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>ag[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BSgenome.Alyrata.JGI.v1"              
[2] "BSgenome.Amellifera.BeeBase.assembly4"</code></pre>
</div>
</div>
<p>We are going to explore the occurrence of the <code>AGGAGGT</code> motif<a href="#fn12" class="footnote-ref" id="fnref12" role="doc-noteref"><sup>12</sup></a> in the genome of E.coli. We use the genome sequence of one particular strain, <strong>Escherichia coli</strong> str. K12 substr.DH10B<a href="#fn13" class="footnote-ref" id="fnref13" role="doc-noteref"><sup>13</sup></a>, whose NCBI accession number is NC_010473.</p>
<div class="no-row-height column-margin column-container"><div id="fn12"><p><sup>12</sup>&nbsp;This is the <a href="https://en.wikipedia.org/wiki/Shine-Dalgarno_sequence">Shine-Dalgarno</a> motif which helps initiate protein synthesis in bacteria.</p></div><div id="fn13"><p><sup>13</sup>&nbsp;It is known as the laboratory workhorse, often used in experiments.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BSgenome.Ecoli.NCBI.20080805"</span>)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>Ecoli</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>shineDalgarno <span class="ot">=</span> <span class="st">"AGGAGGT"</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>ecoli <span class="ot">=</span> Ecoli<span class="sc">$</span>NC_010473</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can count the pattern’s occurrence in windows of width 50000 using the <code>countPattern</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>window <span class="ot">=</span> <span class="dv">50000</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>starts <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">length</span>(ecoli) <span class="sc">-</span> window, <span class="at">by =</span> window)</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>ends   <span class="ot">=</span> starts <span class="sc">+</span> window <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>numMatches <span class="ot">=</span> <span class="fu">vapply</span>(<span class="fu">seq_along</span>(starts), <span class="cf">function</span>(i) {</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">countPattern</span>(shineDalgarno, ecoli[starts[i]<span class="sc">:</span>ends[i]],</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">max.mismatch =</span> <span class="dv">0</span>)</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>  }, <span class="fu">numeric</span>(<span class="dv">1</span>))</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(numMatches)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>numMatches
 0  1  2  3  4 
48 32  8  3  2 </code></pre>
</div>
</div>
<div id="wrn-models-countpatt" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.22
</div>
</div>
<div class="callout-body-container callout-body">
<p>What distribution might this table fit ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-38-contents" aria-controls="callout-38" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-38" class="callout-38-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The Poisson is a good candidate, as a quantitative and graphical evaluation (see <a href="#fig-poissonness" class="quarto-xref">Figure&nbsp;<span>2.24</span></a>) for these data shows.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"vcd"</span>)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">=</span> <span class="fu">goodfit</span>(numMatches, <span class="st">"poisson"</span>)</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(gf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Goodness-of-fit test for poisson distribution

                      X^2 df  P(&gt; X^2)
Likelihood Ratio 4.134932  3 0.2472577</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">distplot</span>(numMatches, <span class="at">type =</span> <span class="st">"poisson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-poissonness" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-poissonness-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-poissonness-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-32" title="Figure&nbsp;2.24: Evaluation of a Poisson model for motif counts along the sequence Ecoli$NC_010473."><img src="02-chap_files/figure-html/fig-poissonness-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-poissonness-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.24: Evaluation of a Poisson model for motif counts along the sequence <code>Ecoli$NC_010473</code>.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>We can inspect the matches using the <code>matchPattern</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>sdMatches <span class="ot">=</span> <span class="fu">matchPattern</span>(shineDalgarno, ecoli, <span class="at">max.mismatch =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can type <code>sdMatches</code> in the R command line to obtain a summary of this object. It contains the locations of all 65 pattern matches, represented as a set of so-called <em>views</em> on the original sequence. Now what are the distances between them?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>betweenmotifs <span class="ot">=</span> <span class="fu">gaps</span>(sdMatches)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So these are in fact the 66 complementary regions. Now let’s find a model for the distribution of the gap sizes between motifs. If the motifs occur at random locations, we expect the gap lengths to follow an exponential distribution<a href="#fn14" class="footnote-ref" id="fnref14" role="doc-noteref"><sup>14</sup></a>. The code below (whose output is shown in <a href="#fig-expplotdata-1" class="quarto-xref">Figure&nbsp;<span>2.25</span></a>) assesses this assumption. If the exponential distribution is a good fit, the points should lie roughly on a straight line. The exponential distribution has one parameter, the rate, and the line with slope corresponding to an estimate from the data is also shown.</p>
<div class="no-row-height column-margin column-container"><div id="fn14"><p><sup>14</sup>&nbsp;How could we guess that the exponential is the right fit here? Whenever we have independent, random Bernoulli occurrences along a sequence, the gap lengths are exponential. You may be familiar with radioactive decay, where the waiting times between emissions are also exponentially distributed. It is a good idea if you are not familiar with this distribution to look up more details in the <a href="http://en.wikipedia.org/wiki/Exponential_distribution">Wikipedia</a>.</p></div></div><div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Renext"</span>)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a><span class="fu">expplot</span>(<span class="fu">width</span>(betweenmotifs), <span class="at">rate =</span> <span class="dv">1</span><span class="sc">/</span><span class="fu">mean</span>(<span class="fu">width</span>(betweenmotifs)),</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels =</span> <span class="st">"fit"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-expplotdata-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-expplotdata-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-expplotdata-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-33" title="Figure&nbsp;2.25: Evaluation of fit to the exponential distribution (black line) of the gaps between the motifs."><img src="02-chap_files/figure-html/fig-expplotdata-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="280"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-expplotdata-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.25: Evaluation of fit to the exponential distribution (black line) of the gaps between the motifs.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-models-exponential" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.23
</div>
</div>
<div class="callout-body-container callout-body">
<p>There appears to be a slight deviation from the fitted line in <a href="#fig-expplotdata-1" class="quarto-xref">Figure&nbsp;<span>2.25</span></a> at the right tail of the distribution, i.e., for the largest values. What could be the reason?</p>
</div>
</div>
<section id="modeling-in-the-case-of-dependencies" class="level3 page-columns page-full" data-number="2.10.1">
<h3 data-number="2.10.1" class="anchored" data-anchor-id="modeling-in-the-case-of-dependencies"><span class="header-section-number">2.10.1</span> Modeling in the case of dependencies</h3>
<p>As we saw in <a href="#sec-models-modelingdependencies" class="quarto-xref"><span>Section 2.8</span></a>, nucleotide sequences are often dependent: the probability of seing a certain nucleotide at a given position tends to depend on the surrounding sequence. Here we are going to put into practice dependency modeling using a <strong>Markov chain</strong>. We are going to look at regions of chromosome 8 of the human genome and try to discover differences between regions called CpG<a href="#fn15" class="footnote-ref" id="fnref15" role="doc-noteref"><sup>15</sup></a> islands and the rest.</p>
<div class="no-row-height column-margin column-container"><div id="fn15"><p><sup>15</sup>&nbsp;CpG stands for 5’-C-phosphate-G-3’; this means that a C is connected to a G through a phosphate along the strand (this is unrelated to C-G base-pairing of <a href="#sec-models-Chargaff" class="quarto-xref"><span>Section 2.7</span></a>). The cytosines in the CpG dinucleotide can be methylated, changing the levels of gene expression. This type of gene regulation is part of <strong>epigenetics</strong>. Some more information is on Wikipedia: <a href="https://en.wikipedia.org/wiki/CpG_site">CpG site</a> and <a href="https://en.wikipedia.org/wiki/Epigenetics">epigenetics</a>.</p></div></div><p>We use data (generated by <span class="citation" data-cites="Irizarry2009">Irizarry, Wu, and Feinberg (<a href="16-chap.html#ref-Irizarry2009" role="doc-biblioref">2009</a>)</span>) that tell us where the start and end points of the islands are in the genome and look at the frequencies of nucleotides and of the digrams ‘CG’, ‘CT’, ‘CA’, ‘CC’. So we can ask whether there are dependencies between the nucleotide occurrences and if so, how to model them.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>chr8  <span class="ot">=</span>  Hsapiens<span class="sc">$</span>chr8</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>CpGtab <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/model-based-cpg-islands-hg19.txt"</span>,</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(CpGtab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 65699</code></pre>
</div>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(CpGtab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    chr  start    end length CpGcount GCcontent pctGC obsExp
1 chr10  93098  93818    721       32       403 0.559  0.572
2 chr10  94002  94165    164       12        97 0.591  0.841
3 chr10  94527  95302    776       65       538 0.693  0.702
4 chr10 119652 120193    542       53       369 0.681  0.866
5 chr10 122133 122621    489       51       339 0.693  0.880
6 chr10 180265 180720    456       32       256 0.561  0.893</code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>irCpG <span class="ot">=</span> <span class="fu">with</span>(dplyr<span class="sc">::</span><span class="fu">filter</span>(CpGtab, chr <span class="sc">==</span> <span class="st">"chr8"</span>),</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>         <span class="fu">IRanges</span>(<span class="at">start =</span> start, <span class="at">end =</span> end))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-34" title="We use the :: operator to call the filter function specifically from the dplyr package—and not from any other packages that may happen to be loaded and defining functions of the same name. This precaution is particularly advisable in the case of the filter function, since this name is used by quite a few other packages. You can think of the normal (without ::) way of calling R functions like calling people by their first (given) names; whereas the fully qualified version with :: corresponds to calling someone by their full name. At least within the reach of the CRAN and Bioconductor repositories, such fully qualified names are guaranteed to be unique."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="We use the :: operator to call the filter function specifically from the dplyr package—and not from any other packages that may happen to be loaded and defining functions of the same name. This precaution is particularly advisable in the case of the filter function, since this name is used by quite a few other packages. You can think of the normal (without ::) way of calling R functions like calling people by their first (given) names; whereas the fully qualified version with :: corresponds to calling someone by their full name. At least within the reach of the CRAN and Bioconductor repositories, such fully qualified names are guaranteed to be unique."></a></p>
</figure>
</div>
<figcaption>We use the <code>::</code> operator to call the <code>filter</code> function specifically from the <code>dplyr</code> package—and not from any other packages that may happen to be loaded and defining functions of the same name. This precaution is particularly advisable in the case of the <code>filter</code> function, since this name is used by quite a few other packages. You can think of the normal (without <code>::</code>) way of calling R functions like calling people by their first (given) names; whereas the fully qualified version with <code>::</code> corresponds to calling someone by their full name. At least within the reach of the CRAN and Bioconductor repositories, such fully qualified names are guaranteed to be unique.</figcaption>
</figure>
</div>
</div></div></div>
<p>In the line above, we subset (<code>filter</code>) the data frame <code>CpGtab</code> to only chromosome 8, and then we create an <em>IRanges</em> object whose start and end positions are defined by the equally named columns of the data frame. In the <code>IRanges</code> function call (which constructs the object from its arguments), the first <code>start</code> is the argument name of the function, the second <code>start</code> refers to the column in the data frame obtained as an output from <code>filter</code>; and similarly for <code>end</code>. <em>IRanges</em> is a general container for mathematical intervals. We create the biological context<a href="#fn16" class="footnote-ref" id="fnref16" role="doc-noteref"><sup>16</sup></a> with the next line.</p>
<div class="no-row-height column-margin column-container"><div id="fn16"><p><sup>16</sup>&nbsp;The “I” in <em>IRanges</em> stands for “interval”; the “G” in <em>GRanges</em> for “genomic”.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>grCpG <span class="ot">=</span> <span class="fu">GRanges</span>(<span class="at">ranges =</span> irCpG, <span class="at">seqnames =</span> <span class="st">"chr8"</span>, <span class="at">strand =</span> <span class="st">"+"</span>)</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(grCpG) <span class="ot">=</span> <span class="st">"hg19"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s visualize; see the output in <a href="#fig-freqandbayes-ideo" class="quarto-xref">Figure&nbsp;<span>2.26</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Gviz"</span>)</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>ideo <span class="ot">=</span> <span class="fu">IdeogramTrack</span>(<span class="at">genome =</span> <span class="st">"hg19"</span>, <span class="at">chromosome =</span> <span class="st">"chr8"</span>)</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTracks</span>(</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="fu">GenomeAxisTrack</span>(),</span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">AnnotationTrack</span>(grCpG, <span class="at">name =</span> <span class="st">"CpG"</span>), ideo),</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">from =</span> <span class="dv">2200000</span>, <span class="at">to =</span> <span class="dv">5800000</span>,</span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape =</span> <span class="st">"box"</span>, <span class="at">fill =</span> <span class="st">"#006400"</span>, <span class="at">stacking =</span> <span class="st">"dense"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-freqandbayes-ideo" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-freqandbayes-ideo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-freqandbayes-ideo-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-35" title="Figure&nbsp;2.26: Gviz plot of CpG locations in a selected region of chromosome 8."><img src="02-chap_files/figure-html/fig-freqandbayes-ideo-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-freqandbayes-ideo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.26: <strong><a href="https://bioconductor.org/packages/Gviz/">Gviz</a></strong> plot of CpG locations in a selected region of chromosome 8.
</figcaption>
</figure>
</div>
</div></div></div>
<p>We now define so-called views on the chromosome sequence that correspond to the CpG islands, <code>irCpG</code>, and to the regions in between (<code>gaps(irCpG)</code>). The resulting objects <code>CGIview</code> and <code>NonCGIview</code> only contain the coordinates, not the sequences themselves (these stay in the big object <code>Hsapiens$chr8</code>), so they are fairly lightweight in terms of storage.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>CGIview    <span class="ot">=</span> <span class="fu">Views</span>(<span class="fu">unmasked</span>(Hsapiens<span class="sc">$</span>chr8), irCpG)</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>NonCGIview <span class="ot">=</span> <span class="fu">Views</span>(<span class="fu">unmasked</span>(Hsapiens<span class="sc">$</span>chr8), <span class="fu">gaps</span>(irCpG))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We compute transition counts in CpG islands and non-islands using the data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>seqCGI      <span class="ot">=</span> <span class="fu">as</span>(CGIview, <span class="st">"DNAStringSet"</span>)</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>seqNonCGI   <span class="ot">=</span> <span class="fu">as</span>(NonCGIview, <span class="st">"DNAStringSet"</span>)</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>dinucCpG    <span class="ot">=</span> <span class="fu">sapply</span>(seqCGI, dinucleotideFrequency)</span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>dinucNonCpG <span class="ot">=</span> <span class="fu">sapply</span>(seqNonCGI, dinucleotideFrequency)</span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>dinucNonCpG[, <span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT 
389 351 400 436 498 560 112 603 359 336 403 336 330 527 519 485 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>NonICounts <span class="ot">=</span> <span class="fu">rowSums</span>(dinucNonCpG)</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>IslCounts  <span class="ot">=</span> <span class="fu">rowSums</span>(dinucCpG)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For a four state Markov chain as we have, we define the transition matrix as a matrix where the rows are the <code>from</code> state and the columns are the <code>to</code> state.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>TI  <span class="ot">=</span> <span class="fu">matrix</span>( IslCounts, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>TnI <span class="ot">=</span> <span class="fu">matrix</span>(NonICounts, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dimnames</span>(TI) <span class="ot">=</span> <span class="fu">dimnames</span>(TnI) <span class="ot">=</span></span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>), <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use the counts of numbers of transitions of each type to compute frequencies and put them into two matrices.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-36" title="The transition probabilities are probabilities so the rows need to sum to 1."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="The transition probabilities are probabilities so the rows need to sum to 1."></a></p>
</figure>
</div>
<figcaption>The transition probabilities are probabilities so the rows need to sum to 1.</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>MI <span class="ot">=</span> TI <span class="sc">/</span><span class="fu">rowSums</span>(TI)</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>MI</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           A         C         G         T
A 0.20457773 0.2652333 0.3897678 0.1404212
C 0.20128250 0.3442381 0.2371595 0.2173200
G 0.18657245 0.3145299 0.3450223 0.1538754
T 0.09802105 0.3352314 0.3598984 0.2068492</code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>MN <span class="ot">=</span> TnI <span class="sc">/</span> <span class="fu">rowSums</span>(TnI)</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>MN</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          A         C          G         T
A 0.3351380 0.1680007 0.23080886 0.2660524
C 0.3641054 0.2464366 0.04177094 0.3476871
G 0.2976696 0.2029017 0.24655406 0.2528746
T 0.2265813 0.1972407 0.24117528 0.3350027</code></pre>
</div>
</div>
<div id="wrn-models-transrows" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.24
</div>
</div>
<div class="callout-body-container callout-body">
<p>Are the transitions different in the different rows? This would mean that, for instance, <span class="math inline">\(P(\mathtt{A}\,|\,\mathtt{C}) \neq P(\mathtt{A}\,|\,\mathtt{T})\)</span>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-41-contents" aria-controls="callout-41" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-41" class="callout-41-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The transitions are different. For instance, the transitions from C to A and T to A for in the islands (MI) transition matrix seem very different (0.201 versus 0.098).</p>
</div>
</div>
</div>
<div id="wrn-models-relfreq" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.25
</div>
</div>
<div class="callout-body-container callout-body">
<p>Are the relative frequencies of the different nucleotides different in CpG islands compared to elsewhere?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-43-contents" aria-controls="callout-43" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-43" class="callout-43-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>freqIsl <span class="ot">=</span> <span class="fu">alphabetFrequency</span>(seqCGI, <span class="at">baseOnly =</span> <span class="cn">TRUE</span>, <span class="at">collapse =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>freqIsl <span class="sc">/</span> <span class="fu">sum</span>(freqIsl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        A         C         G         T 
0.1781693 0.3201109 0.3206298 0.1810901 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>freqNon <span class="ot">=</span> <span class="fu">alphabetFrequency</span>(seqNonCGI, <span class="at">baseOnly =</span> <span class="cn">TRUE</span>, <span class="at">collapse =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>freqNon <span class="sc">/</span> <span class="fu">sum</span>(freqNon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        A         C         G         T 
0.3008292 0.1993832 0.1993737 0.3004139 </code></pre>
</div>
</div>
<p>This shows an inverse pattern: in the CpG islands, C and G have frequencies around 0.32, whereas in the non-CpG islands, we have A and T that have frequencies around 0.30.</p>
</div>
</div>
</div>
<div id="wrn-models-iscp" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;2.26
</div>
</div>
<div class="callout-body-container callout-body">
<p>How can we use these differences to decide whether a given sequence comes from a CpG island?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-45-contents" aria-controls="callout-45" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-45" class="callout-45-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use a <span class="math inline">\(\chi^2\)</span> statistic to compare the frequencies between the observed and <code>freqIsl</code> and <code>freqNon</code> frequencies. For shorter sequences, this may not be sensitive enough, and a more sensitive approach is given below.</p>
</div>
</div>
</div>
<p>Given a sequence for which we do not know whether it is in a CpG island or not, we can ask what is the probability it belongs to a CpG island compared to somewhere else. We compute a score based on what is called the odds ratio. Let’s do an example: suppose our sequence <span class="math inline">\(x\)</span> is <code>ACGTTATACTACG</code>, and we want to decide whether it comes from a CpG island or not.</p>
<p>If we model the sequence as a first order Markov chain we can write, supposing that the sequence comes from a CpG island:</p>
<p><span class="math display">\[
\begin{align}
P_{\text{i}}(x = \mathtt{ACGTTATACTACG}) = \;
&amp;P_{\text{i}}(\mathtt{A}) \, P_{\text{i}}(\mathtt{AC})\, P_{\text{i}}(\mathtt{CG})\,   
P_{\text{i}}(\mathtt{GT})\, P_{\text{i}}(\mathtt{TT}) \times \\
&amp;P_{\text{i}}(\mathtt{TA})\, P_{\text{i}}(\mathtt{AT})\, P_{\text{i}}(\mathtt{TA})\,
P_{\text{i}}(\mathtt{AC})\, P_{\text{i}}(\mathtt{CG}).
\end{align}
\]</span></p>
<p>We are going to compare this probability to the probability for non-islands. As we saw above, these probabilities tend to be quite different. We will take their ratio and see if it is larger or smaller than 1. These probabilities are going to be products of many small terms and become very small. We can work around this by taking logarithms.</p>
<p><span class="anchored-eq" data-anchor-id="eq-stajcmvdowrq" id="eq-stajcmvdowrq"><span class="math display">\[
\begin{align}
\log&amp;\frac{P(x\,|\, \text{island})}{P(x\,|\,\text{non-island})}=\\
\log&amp;\left(
\frac{P_{\text{i}}(\mathtt{A})\, P_{\text{i}}(\mathtt{A}\rightarrow \mathtt{C})\,
P_{\text{i}}(\mathtt{C}\rightarrow \mathtt{G})\,
P_{\text{i}}(\mathtt{G}\rightarrow \mathtt{T})\, P_{\text{i}}(\mathtt{T}\rightarrow \mathtt{T})\, P_{\text{i}}(\mathtt{T}\rightarrow \mathtt{A})}
{P_{\text{n}}(\mathtt{A})\, P_{\text{n}}(\mathtt{A}\rightarrow \mathtt{C})\, P_{\text{n}}(\mathtt{C}\rightarrow \mathtt{G})\,
P_{\text{n}}(\mathtt{G}\rightarrow \mathtt{T})\, P_{\text{n}}( \mathtt{T}\rightarrow  \mathtt{T})\, P_{\text{n}}( \mathtt{T}\rightarrow \mathtt{A})} \right. \times\\
&amp;\left.
\frac{P_{\text{i}}(\mathtt{A}\rightarrow \mathtt{T})\, P_{\text{i}}(\mathtt{T}\rightarrow \mathtt{A})\, P_{\text{i}}(\mathtt{A}\rightarrow \mathtt{C})\,
P_{\text{i}}(\mathtt{C}\rightarrow \mathtt{G})}
{P_{\text{n}}(\mathtt{A}\rightarrow \mathtt{T})\, P_{\text{n}}(\mathtt{T}\rightarrow \mathtt{A})\, P_{\text{n}}(\mathtt{A}\rightarrow \mathtt{C})\, P_{\text{n}}(\mathtt{C}\rightarrow \mathtt{G})} \right)
\end{align}
\tag{2.7}\]</span></span></p>
<p>This is the <strong>log-likelihood ratio</strong> score. To speed up the calculation, we compute the log-ratios <span class="math inline">\(\log(P_{\text{i}}(\mathtt{A})/P_{\text{n}}(\mathtt{A})),..., \log(P_{\text{i}}(\mathtt{T}\rightarrow \mathtt{A})/P_{\text{n}}(\mathtt{T}\rightarrow \mathtt{A}))\)</span> once and for all and then sum up the relevant ones to obtain our score.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/book_icon.png" class="lightbox" data-gallery="quarto-lightbox-gallery-37" title="Worked out examples and many useful details can be found in @DEKM."><img src="imgs/book_icon.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="126" alt="Worked out examples and many useful details can be found in Durbin et al. (1998)."></a></p>
</figure>
</div>
<figcaption>Worked out examples and many useful details can be found in <span class="citation" data-cites="DEKM">Durbin et al. (<a href="16-chap.html#ref-DEKM" role="doc-biblioref">1998</a>)</span>.</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">=</span> <span class="fu">log</span>((freqIsl<span class="sc">/</span><span class="fu">sum</span>(freqIsl)) <span class="sc">/</span> (freqNon<span class="sc">/</span><span class="fu">sum</span>(freqNon)))</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>beta  <span class="ot">=</span> <span class="fu">log</span>(MI <span class="sc">/</span> MN)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="st">"ACGTTATACTACG"</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>scorefun <span class="ot">=</span> <span class="cf">function</span>(x) {</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  s <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(x, <span class="st">""</span>))</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>  score <span class="ot">=</span> alpha[s[<span class="dv">1</span>]]</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(s) <span class="sc">&gt;=</span> <span class="dv">2</span>)</span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(s))</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>      score <span class="ot">=</span> score <span class="sc">+</span> beta[s[j<span class="dv">-1</span>], s[j]]</span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a>  score</span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a><span class="fu">scorefun</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         A 
-0.2824623 </code></pre>
</div>
</div>
<p>In the code below, we pick sequences of length <code>len = 100</code> out of the 2855 sequences in the <code>seqCGI</code> object, and then out of the 2854 sequences in the <code>seqNonCGI</code> object (each of them is a <em>DNAStringSet</em>). In the first three lines of the <code>generateRandomScores</code> function, we drop sequences that contain any letters other than A, C, T, G; such as “.” (a character used for undefined nucleotides). Among the remaining sequences, we sample with probabilities proportional to their length minus <code>len</code> and then pick subsequences of length <code>len</code> out of them. The start points of the subsequences are sampled uniformly, with the constraint that the subsequences have to fit in.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>generateRandomScores <span class="ot">=</span> <span class="cf">function</span>(s, <span class="at">len =</span> <span class="dv">100</span>, <span class="at">B =</span> <span class="dv">1000</span>) {</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>  alphFreq <span class="ot">=</span> <span class="fu">alphabetFrequency</span>(s)</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>  isGoodSeq <span class="ot">=</span> <span class="fu">rowSums</span>(alphFreq[, <span class="dv">5</span><span class="sc">:</span><span class="fu">ncol</span>(alphFreq)]) <span class="sc">==</span> <span class="dv">0</span></span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>  s <span class="ot">=</span> s[isGoodSeq]</span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a>  slen <span class="ot">=</span> <span class="fu">sapply</span>(s, length)</span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>  prob <span class="ot">=</span> <span class="fu">pmax</span>(slen <span class="sc">-</span> len, <span class="dv">0</span>)</span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a>  prob <span class="ot">=</span> prob <span class="sc">/</span> <span class="fu">sum</span>(prob)</span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a>  idx  <span class="ot">=</span> <span class="fu">sample</span>(<span class="fu">length</span>(s), B, <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="at">prob =</span> prob)</span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>  ssmp <span class="ot">=</span> s[idx]</span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a>  start <span class="ot">=</span> <span class="fu">sapply</span>(ssmp, <span class="cf">function</span>(x) <span class="fu">sample</span>(<span class="fu">length</span>(x) <span class="sc">-</span> len, <span class="dv">1</span>))</span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a>  scores <span class="ot">=</span> <span class="fu">sapply</span>(<span class="fu">seq_len</span>(B), <span class="cf">function</span>(i)</span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scorefun</span>(<span class="fu">as.character</span>(ssmp[[i]][start[i]<span class="sc">+</span>(<span class="dv">1</span><span class="sc">:</span>len)]))</span>
<span id="cb136-13"><a href="#cb136-13" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb136-14"><a href="#cb136-14" aria-hidden="true" tabindex="-1"></a>  scores <span class="sc">/</span> len</span>
<span id="cb136-15"><a href="#cb136-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb136-16"><a href="#cb136-16" aria-hidden="true" tabindex="-1"></a>scoresCGI    <span class="ot">=</span> <span class="fu">generateRandomScores</span>(seqCGI)</span>
<span id="cb136-17"><a href="#cb136-17" aria-hidden="true" tabindex="-1"></a>scoresNonCGI <span class="ot">=</span> <span class="fu">generateRandomScores</span>(seqNonCGI)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>rgs <span class="ot">=</span> <span class="fu">range</span>(<span class="fu">c</span>(scoresCGI, scoresNonCGI))</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>br <span class="ot">=</span> <span class="fu">seq</span>(rgs[<span class="dv">1</span>], rgs[<span class="dv">2</span>], <span class="at">length.out =</span> <span class="dv">50</span>)</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>h1 <span class="ot">=</span> <span class="fu">hist</span>(scoresCGI,    <span class="at">breaks =</span> br, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>h2 <span class="ot">=</span> <span class="fu">hist</span>(scoresNonCGI, <span class="at">breaks =</span> br, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(h1, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">120</span>))</span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(h2, <span class="at">col =</span> <span class="fu">rgb</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">4</span>), <span class="at">add =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ScoreMixture-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ScoreMixture-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-ScoreMixture-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-38" title="Figure&nbsp;2.27: Island and non-island scores as generated by the function generateRandomScores. This is the first instance of a mixture we encounter. We will revisit them in sec-mixtures."><img src="02-chap_files/figure-html/fig-ScoreMixture-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ScoreMixture-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.27: Island and non-island scores as generated by the function <code>generateRandomScores</code>. This is the first instance of a <strong>mixture</strong> we encounter. We will revisit them in <a href="04-chap.html" class="quarto-xref"><span>Chapter 4</span></a>.
</figcaption>
</figure>
</div>
</div></div></div>
<p>We can consider these our <em>training data</em>: from data for which we know the types, we can see whether our score is useful for discriminating – see <a href="#fig-ScoreMixture-1" class="quarto-xref">Figure&nbsp;<span>2.27</span></a>.</p>
</section>
</section>
<section id="summary-of-this-chapter" class="level2" data-number="2.11">
<h2 data-number="2.11" class="anchored" data-anchor-id="summary-of-this-chapter"><span class="header-section-number">2.11</span> Summary of this chapter</h2>
<p>In this chapter we experienced the basic yoga of statistics: how to go from the data back to the possible generating distributions and how to estimate the parameters that define these distributions.</p>
<p><strong>Statistical models</strong> We showed some specific statistical models for experiments with categorical outcomes (binomial and multinomial).</p>
<p><strong>Goodness of fit</strong> We used different visualizations and showed how to run simulation experiments to test whether our data could be fit by a fair four box multinomial model. We encountered the chi-square statistic and saw how to compare simulation and theory using a qq-plot.</p>
<p><strong>Estimation</strong> We explained maximum likelihood and Bayesian estimation procedures. These approaches were illustrated on examples involving nucleotide pattern discovery and haplotype estimations.</p>
<p><strong>Prior and posterior distributions</strong> When assessing data of a type that has been been previously studied, such as haplotypes, it can be beneficial to compute the posterior distribution of the data. This enables us to incorporate uncertainty in the decision making, by way of a simple computation. The choice of the prior has little effect on the result as long as there is sufficient data.</p>
<p><strong>CpG islands and Markov chains</strong> We saw how dependencies along DNA sequences can be modeled by Markov chain transitions. We used this to build scores based on likelihood ratios that enable us to see whether long DNA sequences come from CpG islands or not. When we made the histogram of scores, we saw in <a href="#fig-ScoreMixture-1" class="quarto-xref">Figure&nbsp;<span>2.27</span></a> a noticeable feature: it seemed to be made of two pieces. This <strong>bimodality</strong> was our first encounter with mixtures, they are the subject of <a href="04-chap.html" class="quarto-xref"><span>Chapter 4</span></a>.</p>
<p>This is the first instance of building a model on some training data: sequences which we knew were in CpG islands, that we could use later to classify new data. We will develop a much more complete way of doing this in <a href="12-chap.html" class="quarto-xref"><span>Chapter 12</span></a>.</p>
</section>
<section id="further-reading" class="level2" data-number="2.12">
<h2 data-number="2.12" class="anchored" data-anchor-id="further-reading"><span class="header-section-number">2.12</span> Further reading</h2>
<p>One of the best introductory statistics books available is <span class="citation" data-cites="Freedman:1997">Freedman, Pisani, and Purves (<a href="16-chap.html#ref-Freedman:1997" role="doc-biblioref">1997</a>)</span>. It uses box models to explain the important concepts. If you have never taken a statistics class, or you feel you need a refresher, we highly recommend it. Many introductory statistics classes do not cover statistics for discrete data in any depth. The subject is an important part of what we need for biological applications. A book-long introduction to these types of analyses can be found in <span class="citation" data-cites="Agresti:2007">(<a href="16-chap.html#ref-Agresti:2007" role="doc-biblioref">Agresti 2007</a>)</span>.</p>
<p>Here we gave examples of simple unstructured multinomials. However, sometimes the categories (or boxes) of a multinomial have specific structure. For instance, the 64 possible codons code for 20 amino acids and the stop codons (61+3). So we can see the amino acids themselves as a multinomial with 20 degrees of freedom. Within each amino acid there are multinomials with differing numbers of categories (Proline has four: <code>CCA, CCG, CCC, CCT</code>, see <a href="#imp-models-codonusage" class="quarto-xref">Exercise&nbsp;<span>2.3</span></a>). Some multivariate methods have been specifically designed to decompose the variability between codon usage within the differently abundant amino-acids <span class="citation" data-cites="Grantham1981 Perriere2002">(<a href="16-chap.html#ref-Grantham1981" role="doc-biblioref">Grantham et al. 1981</a>; <a href="16-chap.html#ref-Perriere2002" role="doc-biblioref">Perrière and Thioulouse 2002</a>)</span>, and this enables discovery of latent gene transfer and translational selection. We will cover the specific methods used in those papers when we delve into the multivariate exploration of categorical data in <a href="09-chap.html" class="quarto-xref"><span>Chapter 9</span></a>.</p>
<p>There are many examples of successful uses of the Bayesian paradigm to quantify uncertainties. In recent years the computation of the posterior distribution has been revolutionized by special types of Monte Carlo that use either a Markov chain or random walk or Hamiltonian dynamics. These methods provide approximations that converge to the correct posterior distribution after quite a few iterations. For examples and much more see <span class="citation" data-cites="Casella2009 Marin2007 McElreath2015">(<a href="16-chap.html#ref-Casella2009" role="doc-biblioref">Robert and Casella 2009</a>; <a href="16-chap.html#ref-Marin2007" role="doc-biblioref">Marin and Robert 2007</a>; <a href="16-chap.html#ref-McElreath2015" role="doc-biblioref">McElreath 2015</a>)</span>.</p>
</section>
<section id="exercises" class="level2 page-columns page-full" data-number="2.13">
<h2 data-number="2.13" class="anchored" data-anchor-id="exercises"><span class="header-section-number">2.13</span> Exercises</h2>
<div id="imp-models-qqplot" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.1
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate 1,000 random 0/1 variables that model mutations occurring along a 1,000 long gene sequence. These occur independently at a rate of <span class="math inline">\(10^{-4}\)</span> each. Then sum the 1,000 positions to count how many mutations in sequences of length 1,000.</p>
<p>Find the correct distribution for these mutation sums using a goodness of fit test and make a plot to visualize the quality of the fit.</p>
</div>
</div>
<div id="imp-models-mlmax" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.2
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make a function that generates <span class="math inline">\(n\)</span> random uniform numbers between <span class="math inline">\(0\)</span> and <span class="math inline">\(7\)</span> and returns their maximum. Execute the function for <span class="math inline">\(n=25\)</span>. Repeat this procedure <span class="math inline">\(B=100\)</span> times. Plot the distribution of these maxima.<br>
What is the maximum likelihood estimate of the maximum of a sample of size 25 (call it <span class="math inline">\(\hat{\theta}\)</span>)?<br>
Can you find a theoretical justification and the true maximum <span class="math inline">\(\theta\)</span>?</p>
</div>
</div>
<div id="imp-models-codonusage" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.3
</div>
</div>
<div class="callout-body-container callout-body">
<p>A sequence of three nucleotides (a <strong>codon</strong>) taken in a coding region of a gene can be transcribed into one of 20 possible amino acids. There are <span class="math inline">\(4^3=64\)</span> possible codon sequences, but only 20 amino acids. We say the <strong>genetic code</strong> is redundant: there are several ways to <em>spell</em> each amino acid.</p>
<p>The multiplicity (the number of codons that code for the same amino acid) varies from 2 to 6. The different codon-spellings of each amino acid do not occur with equal probabilities. Let’s look at the data for the standard laboratory strain of tuberculosis (H37Rv):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>mtb <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/M_tuberculosis.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mtb, <span class="at">n =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  AmAcid Codon Number PerThous
1    Gly   GGG  25874    19.25
2    Gly   GGA  13306     9.90
3    Gly   GGT  25320    18.84
4    Gly   GGC  68310    50.82</code></pre>
</div>
</div>
<p>The codons for the amino acid proline are of the form <span class="math inline">\(CC*\)</span>, and they occur with the following frequencies in Mycobacterium turberculosis:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>pro  <span class="ot">=</span>  mtb[ mtb<span class="sc">$</span>AmAcid <span class="sc">==</span> <span class="st">"Pro"</span>, <span class="st">"Number"</span>]</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>pro<span class="sc">/</span><span class="fu">sum</span>(pro)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.54302025 0.10532985 0.05859765 0.29305225</code></pre>
</div>
</div>
<ol type="a">
<li><p>Explore the data <code>mtb</code> using <code>table</code> to tabulate the <code>AmAcid</code> and <code>Codon</code> variables.</p></li>
<li><p>How was the <code>PerThous</code> variable created?</p></li>
<li><p>Write an R function that you can apply to the table to find which of the amino acids shows the strongest <strong>codon bias</strong>, i.e., the strongest departure from uniform distribution among its possible spellings.</p></li>
</ol>
</div>
</div>
<div class="no-row-height column-margin column-container"><span class="margin-aside callout-margin-content"><span class="math inline">\(*\)</span> stands for any of the 4 letters, using the computer notation for a regular expression.</span></div><div id="imp-models-displaygc" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.4
</div>
</div>
<div class="callout-body-container callout-body">
<p>Display GC content in a running window along the sequence of <em>Staphylococcus Aureus</em>. Read in a <em>fasta</em> file sequence from a file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>staph <span class="ot">=</span> <span class="fu">readDNAStringSet</span>(<span class="st">"../data/staphsequence.ffn.txt"</span>, <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol type="a">
<li><p>Look at the complete <code>staph</code> object and then display the first three sequences in the set.</p></li>
<li><p>Find the GC content along the sequence in sliding windows of width 100.</p></li>
<li><p>Display the result of b).</p></li>
<li><p>How could we visualize the overall trends of these proportions along the sequence?</p></li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-50-contents" aria-controls="callout-50" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-50" class="callout-50-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="a">
<li>The data is displayed using:</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>staph[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 3:
    width seq                                               names               
[1]  1362 ATGTCGGAAAAAGAAATTTGGGA...AAAAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...
[2]  1134 ATGATGGAATTCACTATTAAAAG...TTTTACCAATCAGAACTTACTAA lcl|NC_002952.2_c...
[3]   246 GTGATTATTTTGGTTCAAGAAGT...TCATTCATCAAGGTGAACAATGA lcl|NC_002952.2_c...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>staph</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 2650:
       width seq                                            names               
   [1]  1362 ATGTCGGAAAAAGAAATTTGGG...AAAGAAATAAGAAATGTATAA lcl|NC_002952.2_c...
   [2]  1134 ATGATGGAATTCACTATTAAAA...TTACCAATCAGAACTTACTAA lcl|NC_002952.2_c...
   [3]   246 GTGATTATTTTGGTTCAAGAAG...ATTCATCAAGGTGAACAATGA lcl|NC_002952.2_c...
   [4]  1113 ATGAAGTTAAATACACTCCAAT...CAAGGTGAAATTATAAAGTAA lcl|NC_002952.2_c...
   [5]  1932 GTGACTGCATTGTCAGATGTAA...TATGCAAACTTAGACTTCTAA lcl|NC_002952.2_c...
   ...   ... ...
[2646]   720 ATGACTGTAGAATGGTTAGCAG...ACTCCTTTACTTGAAAAATAA lcl|NC_002952.2_c...
[2647]  1878 GTGGTTCAAGAATATGATGTAA...CTCCAAAGGGTGAGTGACTAA lcl|NC_002952.2_c...
[2648]  1380 ATGGATTTAGATACAATTACGA...CAATTCTGCTTAGGTAAATAG lcl|NC_002952.2_c...
[2649]   348 TTGGAAAAAGCTTACCGAATTA...TTTAATAAAAAGATTAAGTAA lcl|NC_002952.2_c...
[2650]   138 ATGGTAAAACGTACTTATCAAC...CGTAAAGTTTTATCTGCATAA lcl|NC_002952.2_c...</code></pre>
</div>
</div>
<ol start="2" type="a">
<li>We can compute the frequencies using the function <code>letterFrequency</code>.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">letterFrequency</span>(staph[[<span class="dv">1</span>]], <span class="at">letters =</span> <span class="st">"ACGT"</span>, <span class="at">OR =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  A   C   G   T 
522 219 229 392 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>GCstaph <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">ID =</span> <span class="fu">names</span>(staph),</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">GC =</span> <span class="fu">rowSums</span>(<span class="fu">alphabetFrequency</span>(staph)[, <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>] <span class="sc">/</span> <span class="fu">width</span>(staph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="a">
<li>Plotting can be done as follows, here exemplarily for sequence 364 (<a href="#fig-SlidingGC-1" class="quarto-xref">Figure&nbsp;<span>2.28</span></a>):</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>window <span class="ot">=</span> <span class="dv">100</span></span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>gc <span class="ot">=</span> <span class="fu">rowSums</span>( <span class="fu">letterFrequencyInSlidingView</span>(staph[[<span class="dv">364</span>]], window,</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">c</span>(<span class="st">"G"</span>,<span class="st">"C"</span>)))<span class="sc">/</span>window</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="at">along =</span> gc), <span class="at">y =</span> gc, <span class="at">type =</span> <span class="st">"l"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-SlidingGC-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-SlidingGC-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-SlidingGC-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-39" title="Figure&nbsp;2.28: GC content along sequence 364 of the Staphylococcus Aureus genome."><img src="02-chap_files/figure-html/fig-SlidingGC-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-SlidingGC-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.28: GC content along sequence 364 of the <em>Staphylococcus Aureus</em> genome.
</figcaption>
</figure>
</div>
</div>
</div>
<ol start="4" type="a">
<li>We can look at the overall trends by smoothing the data using the function <code>lowess</code> along a window.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="at">along =</span> gc), <span class="at">y =</span> gc, <span class="at">type =</span> <span class="st">"l"</span>)</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">lowess</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="at">along =</span> gc), <span class="at">y =</span> gc, <span class="at">f =</span> <span class="fl">0.2</span>), <span class="at">col =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-SmoothSlidingGC-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-SmoothSlidingGC-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-SmoothSlidingGC-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-40" title="Figure&nbsp;2.29: Similar to Figure&nbsp;fig-SlidingGC-1, with smoothing."><img src="02-chap_files/figure-html/fig-SmoothSlidingGC-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-SmoothSlidingGC-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.29: Similar to <a href="#fig-SlidingGC-1" class="quarto-xref">Figure&nbsp;<span>2.28</span></a>, with smoothing.
</figcaption>
</figure>
</div>
</div>
</div>
<p>We will see later an appropriate way of deciding whether the window has an abnormally high GC content by using the idea that as we move along the sequences, we are always in one of several possible <strong>states</strong>. However, we don’t directly observe the state, just the sequence. Such models are called <strong>hidden (state) Markov models</strong>, or HMM for short (see <a href="http://en.wikipedia.org/wiki/Hidden_Markov_model">Wikipedia</a>). The <em>Markov</em> in the name of these models is for how they model dependencies between neighboring positions, the <em>hidden</em> part indicates that the state is not directly observed, that is, hidden.</p>
</div>
</div>
</div>
<div id="imp-models-diffbetas" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.5
</div>
</div>
<div class="callout-body-container callout-body">
<p>Redo a figure similar to <a href="#fig-histobeta2" class="quarto-xref">Figure&nbsp;<span>2.19</span></a>, but include two other distributions: the uniform (which is Beta(1,1)) and Beta(<span class="math inline">\(\frac{1}{2},\frac{1}{2}\)</span>). What do you notice?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-52-contents" aria-controls="callout-52" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-52" class="callout-52-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>dfbetas <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">p =</span> <span class="fu">rep</span>(p_seq, <span class="dv">5</span>),</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">dbeta =</span> <span class="fu">c</span>(<span class="fu">dbeta</span>(p_seq, <span class="fl">0.5</span>, <span class="fl">0.5</span>), </span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>            <span class="fu">dbeta</span>(p_seq,   <span class="dv">1</span>,   <span class="dv">1</span>), </span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>            <span class="fu">dbeta</span>(p_seq,  <span class="dv">10</span>,  <span class="dv">30</span>),</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">dbeta</span>(p_seq,  <span class="dv">20</span>,  <span class="dv">60</span>), </span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">dbeta</span>(p_seq,  <span class="dv">50</span>, <span class="dv">150</span>)),</span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Beta(0.5,0.5)"</span>, <span class="st">"U(0,1)=Beta(1,1)"</span>, </span>
<span id="cb152-9"><a href="#cb152-9" aria-hidden="true" tabindex="-1"></a>               <span class="st">"Beta(10,30)"</span>, <span class="st">"Beta(20,60)"</span>, </span>
<span id="cb152-10"><a href="#cb152-10" aria-hidden="true" tabindex="-1"></a>               <span class="st">"Beta(50,150)"</span>), <span class="at">each =</span> <span class="fu">length</span>(p_seq)))</span>
<span id="cb152-11"><a href="#cb152-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dfbetas) <span class="sc">+</span></span>
<span id="cb152-12"><a href="#cb152-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> dbeta, <span class="at">colour =</span> pars)) <span class="sc">+</span></span>
<span id="cb152-13"><a href="#cb152-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb152-14"><a href="#cb152-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="fu">aes</span>(<span class="at">xintercept =</span> <span class="fl">0.25</span>), <span class="at">colour =</span> <span class="st">"#990000"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-histobeta4-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-histobeta4-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="02-chap_files/figure-html/fig-histobeta4-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-41" title="Figure&nbsp;2.30: Beta densities for different parameter choices."><img src="02-chap_files/figure-html/fig-histobeta4-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-histobeta4-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.30: Beta densities for different parameter choices.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Whereas the Beta distributions with parameters larger than one are unimodal, the Beta(0.5,0.5) distribution is bimodal and the Beta(1,1) is flat and has no mode.</p>
</div>
</div>
</div>
<div id="imp-models-drawyourown" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;2.6
</div>
</div>
<div class="callout-body-container callout-body">
<p>Choose your own prior for the parameters of the Beta distribution. You can do this by sketching it here: <a href="https://jhubiostatistics.shinyapps.io/drawyourprior" class="uri">https://jhubiostatistics.shinyapps.io/drawyourprior</a>. Once you have set up a prior, re-analyse the data from <a href="#sec-models-haplotypefrequencies" class="quarto-xref"><span>Section 2.9.1</span></a>, where we saw <span class="math inline">\(Y = 40\)</span> successes out of <span class="math inline">\(n=300\)</span> trials. Compare your posterior distribution to the one we obtained in that section using a QQ-plot.</p>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-Agresti:2007" class="csl-entry" role="listitem">
Agresti, Alan. 2007. <em>An Introduction to Categorical Data Analysis</em>. John Wiley.
</div>
<div id="ref-Cannings1968" class="csl-entry" role="listitem">
Cannings, Chris, and Anthony WF Edwards. 1968. <span>“Natural Selection and the <span class="nocase">de Finetti</span> Diagram.”</span> <em>Annals of Human Genetics</em> 31 (4): 421–28.
</div>
<div id="ref-Tukey:1988" class="csl-entry" role="listitem">
Cleveland, William S. 1988. <em>The Collected Works of John w. Tukey: Graphics 1965-1985</em>. Vol. 5. CRC Press.
</div>
<div id="ref-DEKM" class="csl-entry" role="listitem">
Durbin, Richard, Sean Eddy, Anders Krogh, and Graeme Mitchison. 1998. <em>Biological Sequence Analysis</em>. Cambridge University Press.
</div>
<div id="ref-Chargaff" class="csl-entry" role="listitem">
Elson, D, and E Chargaff. 1952. <span>“On the Desoxyribonucleic Acid Content of Sea Urchin Gametes.”</span> <em>Experientia</em> 8 (4): 143–45.
</div>
<div id="ref-definetti26" class="csl-entry" role="listitem">
Finetti, Bruno de. 1926. <span>“Considerazioni Matematiche Sull’ereditarieta Mendeliana.”</span> <em>Metron</em> 6: 3–41.
</div>
<div id="ref-Freedman:1997" class="csl-entry" role="listitem">
Freedman, David, Robert Pisani, and Roger Purves. 1997. <em>Statistics</em>. New York, NY: WW Norton.
</div>
<div id="ref-Grantham1981" class="csl-entry" role="listitem">
Grantham, Richard, Christian Gautier, Manolo Gouy, M Jacobzone, and R Mercier. 1981. <span>“Codon Catalog Usage Is a Genome Strategy Modulated for Gene Expressivity.”</span> <em>Nucleic Acids Research</em> 9 (1): 213–13.
</div>
<div id="ref-Irizarry2009" class="csl-entry" role="listitem">
Irizarry, Rafael A, Hao Wu, and Andrew P Feinberg. 2009. <span>“A Species-Generalized Probabilistic Model-Based Definition of CpG Islands.”</span> <em>Mammalian Genome</em> 20 (9-10): 674–80.
</div>
<div id="ref-LoveDESeq2" class="csl-entry" role="listitem">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Gnome Biology</em> 15 (12): 1–21.
</div>
<div id="ref-Marin2007" class="csl-entry" role="listitem">
Marin, Jean-Michel, and Christian Robert. 2007. <em>Bayesian Core: A Practical Approach to Computational <span>B</span>ayesian Statistics</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-McElreath2015" class="csl-entry" role="listitem">
McElreath, Richard. 2015. <em>Statistical Rethinking: A <span>B</span>ayesian Course with Examples in <span class="nocase">R and Stan</span></em>. Chapman; Hall/CRC.
</div>
<div id="ref-Mourant1976" class="csl-entry" role="listitem">
Mourant, AE, Ada Kopec, and K Domaniewska-Sobczak. 1976. <span>“The Distribution of the Human Blood Groups 2nd Edition.”</span> Oxford University Press London.
</div>
<div id="ref-Perriere2002" class="csl-entry" role="listitem">
Perrière, Guy, and Jean Thioulouse. 2002. <span>“Use and Misuse of Correspondence Analysis in Codon Usage Studies.”</span> <em>Nucleic Acids Research</em> 30 (20): 4548–55.
</div>
<div id="ref-Rice:2007" class="csl-entry" role="listitem">
Rice, John. 2006. <em>Mathematical Statistics and Data Analysis</em>. Cengage Learning.
</div>
<div id="ref-Casella2009" class="csl-entry" role="listitem">
Robert, Christian, and George Casella. 2009. <em>Introducing <span>M</span>onte <span>C</span>arlo Methods with <span>R</span></em>. Springer Science &amp; Business Media.
</div>
</div>
</section>


<p>Page built at 01:33 on 2025-09-01 using R version 4.5.1 (2025-06-13)</p></main> <!-- /main --></p><p class="build-date">
Support for maintaining the online version of this book is provided by <a href="http://www.denbi.de">de.NBI</a>
  <img src="imgs/denbi.png" style="vertical-align: text-top;height: 16px"><br>
For website support please contact msmith@embl.de
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
  const anchorJS_eq = new window.AnchorJS();
  anchorJS_eq.options = {
    placement: 'left',
    class: 'eq-link',
    icon: icon
  };
  anchorJS_eq.add('.anchored-eq');
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/www\.huber\.embl\.de\/msmb");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>
