![](imgs/xkcd-plotting.png)

ë°ì´í„° ì‹œê°í™”ì—ëŠ” ì ì–´ë„ ë‘ ê°€ì§€ ìœ í˜•ì´ ìˆìŠµë‹ˆë‹¤. ì²« ë²ˆì§¸ëŠ” ê³¼í•™ìê°€ ë°ì´í„°ë¥¼ íƒìƒ‰í•˜ê³  ë³µì¡í•œ í”„ë¡œì„¸ìŠ¤ì— ëŒ€í•œ ë°œê²¬ì„ í•  ìˆ˜ ìˆë„ë¡ í•©ë‹ˆë‹¤. ë‹¤ë¥¸ ìœ í˜•ì˜ ì‹œê°í™”ëŠ” ë‹¤ë¥¸ ì‚¬ëŒë“¤ì—ê²Œ ë³´ì—¬ì£¼ê³  ê¶ê·¹ì ìœ¼ë¡œ ì¶œíŒë¬¼ì— í¬í•¨í•  ìˆ˜ ìˆëŠ” ìœ ìµí•˜ê³  ëª…í™•í•˜ë©° ì‹œê°ì ìœ¼ë¡œ ë§¤ë ¥ì ì¸ ê²°ê³¼ì˜ ì‚½í™”ë¥¼ ì œê³µí•©ë‹ˆë‹¤.

ì´ ë‘ ê°€ì§€ ìœ í˜•ì˜ ì‹œê°í™”ëŠ” ëª¨ë‘ Rë¡œ ë§Œë“¤ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì‹¤ì œë¡œ Rì€ ì—¬ëŸ¬ ê·¸ë˜í”½ ì‹œìŠ¤í…œì„ ì œê³µí•©ë‹ˆë‹¤. ì´ëŠ” Rì´ í™•ì¥ ê°€ëŠ¥í•˜ê¸° ë•Œë¬¸ì´ë©°, ìˆ˜ë…„ì— ê±¸ì¹œ R ê·¸ë˜í”½ì˜ ë°œì „ì€ ê¸°ì¡´ í•¨ìˆ˜ë¥¼ ëŒ€ì²´í•˜ëŠ” ê²ƒì´ ì•„ë‹ˆë¼ ì£¼ë¡œ íŒ¨í‚¤ì§€ë¥¼ ì¶”ê°€í•¨ìœ¼ë¡œì¨ ì´ë£¨ì–´ì¡Œê¸° ë•Œë¬¸ì…ë‹ˆë‹¤. ê° ê·¸ë˜í”½ ì‹œìŠ¤í…œì—ëŠ” ì¥ì ê³¼ í•œê³„ê°€ ìˆìŠµë‹ˆë‹¤. ì´ ì¥ì—ì„œëŠ” ê·¸ ì¤‘ ë‘ ê°€ì§€ë¥¼ ì•Œì•„ë³´ê² ìŠµë‹ˆë‹¤. ë¨¼ì € ê¸°ë³¸ R í”Œë¡œíŒ… í•¨ìˆ˜ë¥¼ ê°„ëµíˆ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤1. ê·¸ í›„ì— **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ë¡œ ì „í™˜í•  ê²ƒì…ë‹ˆë‹¤.

1 ì´ í•¨ìˆ˜ë“¤ì€ ëª¨ë“  ê¸°ë³¸ R ì„¤ì¹˜ì™€ í•¨ê»˜ ì œê³µë˜ëŠ” **[graphics](https://cran.r-project.org/web/packages/graphics/)** íŒ¨í‚¤ì§€ì— ìˆìŠµë‹ˆë‹¤.

[![](imgs/Automatisches_Zeichengeraet_ZUSE_Z64_ubt_web.jpg)](imgs/Automatisches_Zeichengeraet_ZUSE_Z64_ubt_web.jpg
"ê·¸ë¦¼ 3.1: ZUSE í”Œë¡œí„° Z64 (1961ë…„ ë°œí‘œ). ì¶œì²˜: https://en.wikipedia.org/wiki/Plotter.")

ê·¸ë¦¼ 3.1: ZUSE í”Œë¡œí„° Z64 (1961ë…„ ë°œí‘œ). ì¶œì²˜: <https://en.wikipedia.org/wiki/Plotter>.

ê¸°ë³¸ R ê·¸ë˜í”½ì€ ì—­ì‚¬ì ìœ¼ë¡œ ë¨¼ì € ë‚˜ì™”ìŠµë‹ˆë‹¤. ê°„ë‹¨í•˜ê³  ì ˆì°¨ì ì´ë©° ê°œë…ì ìœ¼ë¡œ ìº”ë²„ìŠ¤ì— ê·¸ë¦¬ëŠ” ê²ƒì—ì„œ ë™ê¸°ë¥¼ ì–»ì—ˆìŠµë‹ˆë‹¤. ë‹¤ì–‘í•œ ìœ í˜•ì˜ í”Œë¡¯ì„ ìœ„í•œ íŠ¹ìˆ˜ í•¨ìˆ˜ê°€ ìˆìŠµë‹ˆë‹¤. ì´ í•¨ìˆ˜ë“¤ì€ í˜¸ì¶œí•˜ê¸° ì‰½ì§€ë§Œ, ë” ë³µì¡í•œ í”Œë¡¯ì„ ë§Œë“¤ê¸° ìœ„í•´ ê²°í•©í•˜ê±°ë‚˜ ì„œë¡œ êµí™˜í•˜ê³  ì‹¶ì„ ë•Œ, ì´ê²ƒì€ ê¸ˆë°© ì§€ì €ë¶„í•´ì§€ê±°ë‚˜ ì‹¬ì§€ì–´ ë¶ˆê°€ëŠ¥í•´ì§‘ë‹ˆë‹¤. ì‚¬ìš©ìëŠ” (ê°œë…ì ì¸) ìº”ë²„ìŠ¤ì— ì§ì ‘ í”Œë¡¯ì„ ê·¸ë¦½ë‹ˆë‹¤(ì´ ë‹¨ì–´ëŠ” ìµœì´ˆì˜ ê·¸ë˜í”½ ì¥ì¹˜ ì¤‘ ì¼ë¶€ë¡œ ê±°ìŠ¬ëŸ¬ ì˜¬ë¼ê°‘ë‹ˆë‹¤ - ê·¸ë¦¼ 3.1 ì°¸ì¡°). ì‚¬ìš©ìëŠ” ì—¬ë°±, ì¶• ë ˆì´ë¸”, ì œëª©, ë²”ë¡€, í•˜ìœ„ íŒ¨ë„ì— í• ë‹¹í•  ê³µê°„ê³¼ ê°™ì€ ê²°ì •ì„ ëª…ì‹œì ìœ¼ë¡œ ì²˜ë¦¬í•´ì•¼ í•©ë‹ˆë‹¤. ì¼ë‹¨ ë¬´ì–¸ê°€ê°€ 'í”Œë¡¯'ë˜ë©´ ì´ë™í•˜ê±°ë‚˜ ì§€ìš¸ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.

ë” ë†’ì€ ìˆ˜ì¤€ì˜ ì ‘ê·¼ ë°©ì‹ì´ ìˆìŠµë‹ˆë‹¤. _ê·¸ë˜í”½ì˜ ë¬¸ë²•_ì—ì„œëŠ” ê·¸ë˜í”½ì´ ëª¨ë“ˆì‹ ë…¼ë¦¬ ì¡°ê°ìœ¼ë¡œ êµ¬ì„±ë˜ë¯€ë¡œ, ì¸ê°„ ì–¸ì–´ì—ì„œ ë¬¸ì¥ì˜ ì¼ë¶€ë¥¼ ë°”ê¾¸ê³  ì œì™¸í•  ìˆ˜ ìˆëŠ” ê²ƒì²˜ëŸ¼ ì§ê´€ì ì´ê³  ì‰½ê²Œ í•´ë…í•  ìˆ˜ ìˆëŠ” ë°©ì‹ìœ¼ë¡œ ë°ì´í„°ì— ëŒ€í•œ ë‹¤ì–‘í•œ ì‹œê°í™” ìœ í˜•ì„ ì‰½ê²Œ ì‹œë„í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ìº”ë²„ìŠ¤ë‚˜ í”Œë¡œí„°ì˜ ê°œë…ì´ ì—†ìŠµë‹ˆë‹¤. ëŒ€ì‹  ì‚¬ìš©ìëŠ” **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì— R ê°ì²´ í˜•íƒœë¡œ ì›í•˜ëŠ” í”Œë¡¯ì— ëŒ€í•œ ë†’ì€ ìˆ˜ì¤€ì˜ ì„¤ëª…ì„ ì œê³µí•˜ê³ , ë Œë”ë§ ì—”ì§„ì€ ì¥ë©´ì„ ì „ì²´ì ìœ¼ë¡œ íŒŒì•…í•˜ì—¬ ê·¸ë˜í”½ì„ ë°°ì¹˜í•˜ê³  ì¶œë ¥ ì¥ì¹˜ì— ë Œë”ë§í•©ë‹ˆë‹¤.

## 3.1 ì´ ì¥ì˜ ëª©í‘œ

ì´ ì¥ì—ì„œëŠ” ë‹¤ìŒì„ ìˆ˜í–‰í•©ë‹ˆë‹¤:

  * ì‹œê°í™”ë¥¼ í†µí•´ ë°ì´í„° ì„¸íŠ¸ë¥¼ ë¹ ë¥´ê³  ìœ ì—°í•˜ê²Œ íƒìƒ‰í•˜ëŠ” ë°©ë²•ì„ ë°°ì›ë‹ˆë‹¤.

  * ê³¼í•™ ë°œí‘œ ë° ì¶œíŒë¬¼ì„ ìœ„í•œ ì•„ë¦„ë‹µê³  ì§ê´€ì ì¸ ê·¸ë¦¼ì„ ë§Œë“­ë‹ˆë‹¤.

  * ê¸°ë³¸ R í”Œë¡œíŒ…ì˜ ê¸°ë³¸ ì‚¬í•­ì„ ê²€í† í•©ë‹ˆë‹¤.

  * _ê·¸ë˜í”½ì˜ ë¬¸ë²•_ ê°œë… ë’¤ì— ìˆëŠ” ë…¼ë¦¬ë¥¼ ì´í•´í•©ë‹ˆë‹¤.

  * **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì˜ `ggplot` í•¨ìˆ˜ë¥¼ ì†Œê°œí•©ë‹ˆë‹¤.

  * 1, 2 ë˜ëŠ” 3~5ì°¨ì›ìœ¼ë¡œ ë°ì´í„°ë¥¼ í”Œë¡œíŒ…í•˜ëŠ” ë°©ë²•ì„ ë³´ê³  íŒ¨ì‹¯(facet)ì„ íƒìƒ‰í•©ë‹ˆë‹¤.

  * ë¶„ì ìƒë¬¼í•™ ë°ì´í„°(ë˜ëŠ” ë‹¤ë¥¸ ì„œì—´, ì˜ˆ: í©íƒ€ì´ë“œ)ì— ëŒ€í•œ 'ê²Œë†ˆ' í”Œë¡¯ì„ ë§Œë“­ë‹ˆë‹¤.

  * ëŒ€í™”í˜• ê·¸ë˜í”½ì˜ ëª‡ ê°€ì§€ ì˜µì…˜ì— ëŒ€í•´ ë…¼ì˜í•©ë‹ˆë‹¤.

## 3.2 ê¸°ë³¸ R í”Œë¡œíŒ…

ê°€ì¥ ê¸°ë³¸ì ì¸ í•¨ìˆ˜ëŠ” `plot`ì…ë‹ˆë‹¤. ì•„ë˜ ì½”ë“œ(ê·¸ë¦¼ 3.2ì— ì¶œë ¥ì´ í‘œì‹œë¨)ì—ì„œëŠ” íš¨ì†Œ ê²°í•© ë©´ì—­ í¡ì°© ë¶„ì„(ELISA)ì˜ ë°ì´í„°ë¥¼ í”Œë¡œíŒ…í•˜ëŠ” ë° ì‚¬ìš©ë©ë‹ˆë‹¤. ì´ ë¶„ì„ì€ DNAë¥¼ ë¶„í•´í•˜ëŠ” íš¨ì†Œì¸ ë°ì˜¥ì‹œë¦¬ë³´ë‰´í´ë ˆì•„ì œ(DNase)ì˜ í™œì„±ì„ ì •ëŸ‰í™”í•˜ëŠ” ë° ì‚¬ìš©ë˜ì—ˆìŠµë‹ˆë‹¤. ë°ì´í„°ëŠ” ê¸°ë³¸ Rê³¼ í•¨ê»˜ í¸ë¦¬í•˜ê²Œ ì œê³µë˜ëŠ” R ê°ì²´ `DNase`ì— ìˆ˜ì§‘ë©ë‹ˆë‹¤. `DNase` ê°ì²´ëŠ” `Run`(ë¶„ì„ ì‹¤í–‰), `conc`(ì‚¬ìš©ëœ ë‹¨ë°±ì§ˆ ë†ë„), `density`(ì¸¡ì •ëœ ê´‘í•™ ë°€ë„) ì—´ì´ ìˆëŠ” ë°ì´í„° í”„ë ˆì„ì…ë‹ˆë‹¤.

    
    
    head(DNase)__
    
    
      Run       conc density
    1   1 0.04882812   0.017
    2   1 0.04882812   0.018
    3   1 0.19531250   0.121
    4   1 0.19531250   0.124
    5   1 0.39062500   0.206
    6   1 0.39062500   0.215
    
    
    plot(DNase$conc, DNase$density)__

[![](03-chap_files/figure-html/fig-graphics-
basicplotting1-1.png)](03-chap_files/figure-html/fig-graphics-
basicplotting1-1.png "FigureÂ 3.2: Plot of concentration vs.Â density for an
ELISA assay of DNase.")

Figure 3.2: Plot of concentration vs. density for an ELISA assay of DNase.

This basic plot can be customized, for example by changing the plot symbol and
axis labels using the parameters `xlab`, `ylab` and `pch` (plot character), as
shown in Figure 3.3. Information about the variables is stored in the object
`DNase`, and we can access it with the `attr` function.

    
    
    plot(DNase$conc, DNase$density,
      ylab = attr(DNase, "labels")$y,
      xlab = paste(attr(DNase, "labels")$x, attr(DNase, "units")$x),
      pch = 3,
      col = "blue")__

[![](03-chap_files/figure-html/fig-graphics-
basicplotting2-1.png)](03-chap_files/figure-html/fig-graphics-
basicplotting2-1.png "FigureÂ 3.3: Same data as in FigureÂ fig-graphics-
basicplotting1 but with better axis labels and a different plot symbol.")

ê·¸ë¦¼ 3.3: ê·¸ë¦¼ 3.2ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ ë” ë‚˜ì€ ì¶• ë ˆì´ë¸”ê³¼ ë‹¤ë¥¸ í”Œë¡¯ ê¸°í˜¸ë¥¼ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.

__

ì§ˆë¬¸ 3.1

ë” ê¸´ ì„¤ëª…, ë¬¼ë¦¬ì  ë‹¨ìœ„, ì¶œì²˜ ì •ë³´ ë“±ê³¼ ê°™ì€ "ë©”íƒ€ë°ì´í„°"ë¡œ ë°ì´í„° í”„ë ˆì„ ì—´ì— ì£¼ì„ì„ ë‹¬ì•„ë‘ëŠ” ê²ƒì€ ìœ ìš©í•œ ê¸°ëŠ¥ì²˜ëŸ¼ ë³´ì…ë‹ˆë‹¤. `DNase` ê°ì²´ì²˜ëŸ¼ ì´ëŸ¬í•œ ì •ë³´ë¥¼ ì €ì¥í•˜ëŠ” ë°©ì‹ì´ R ìƒíƒœê³„ ì „ë°˜ì—ì„œ í‘œì¤€í™”ë˜ê±°ë‚˜ ì¼ë°˜ì ì¸ê°€ìš”? ì´ ì‘ì—…ì„ ìˆ˜í–‰í•˜ëŠ” ë‹¤ë¥¸ í‘œì¤€í™”ë˜ê±°ë‚˜ ì¼ë°˜ì ì¸ ë°©ë²•ì´ ìˆìŠµë‹ˆê¹Œ?

__

í•´ê²°ì±…

__

ì¼ë°˜ì ì¸ R _data.frame_ì´ë‚˜ tidyverse(_data_frame_, _tibble_)ì—ëŠ” ì´ë¥¼ ìœ„í•œ í›Œë¥­í•˜ê±°ë‚˜ ë„ë¦¬ ì‚¬ìš©ë˜ëŠ” ì¸í”„ë¼ê°€ ì—†ìŠµë‹ˆë‹¤. í•˜ì§€ë§Œ Bioconductor íŒ¨í‚¤ì§€ **[S4Vectors](https://bioconductor.org/packages/S4Vectors/)**ì˜ _DataFrame_ í´ë˜ìŠ¤ë¥¼ ì‚´í´ë³´ì„¸ìš”. ë¬´ì—‡ë³´ë‹¤ë„ ì´ê²ƒì€ _SummarizedExperiment_ì˜ í–‰ê³¼ ì—´ì— ì£¼ì„ì„ ë‹¤ëŠ” ë° ì‚¬ìš©ë©ë‹ˆë‹¤.

ì‚°ì ë„ ì™¸ì—ë„ ë‚´ì¥ í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ íˆìŠ¤í† ê·¸ë¨ ë° ìƒì ê·¸ë¦¼ì„ ë§Œë“¤ ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤(ê·¸ë¦¼ 3.4).

    
    
    hist(DNase$density, breaks=25, main = "")
    boxplot(density ~ Run, data = DNase)__

[![](03-chap_files/figure-html/fig-graphics-
basicplotting3-1.png)](03-chap_files/figure-html/fig-graphics-
basicplotting3-1.png "FigureÂ 3.4Â \(a\): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
basicplotting3-2.png)](03-chap_files/figure-html/fig-graphics-
basicplotting3-2.png "FigureÂ 3.4Â \(b\): ")

(b)

Figure 3.4: (a) Histogram of the density from the ELISA assay, and (b)
boxplots of these values stratified by the assay run. The boxes are ordered
along the axis in lexicographical order because the runs were stored as text
strings. We could use Râ€™s type conversion functions to achieve numerical
ordering.

ìƒì ê·¸ë¦¼(Boxplot)ì€ ì¢ì€ ê³µê°„ì— ì—¬ëŸ¬ ë¶„í¬ë¥¼ ë‚˜ë€íˆ ë³´ì—¬ì£¼ê¸°ì— í¸ë¦¬í•©ë‹ˆë‹¤. 3.6ì ˆì—ì„œ ì—¬ëŸ¬ ë‹¨ë³€ëŸ‰ ë¶„í¬ë¥¼ ê·¸ë¦¬ëŠ” ê²ƒì— ëŒ€í•´ ë” ìì„¸íˆ ë‹¤ë£° ê²ƒì…ë‹ˆë‹¤.

ê¸°ë³¸ R í”Œë¡œíŒ… í•¨ìˆ˜ëŠ” ë¹ ë¥¸ ëŒ€í™”í˜• ë°ì´í„° íƒìƒ‰ì— ìœ ìš©í•˜ì§€ë§Œ, ë” ì •êµí•œ ë””ìŠ¤í”Œë ˆì´ë¥¼ ë§Œë“¤ê³  ì‹¶ë‹¤ë©´ ê³§ ê·¸ í•œê³„ì— ë¶€ë”ªíˆê²Œ ë©ë‹ˆë‹¤. ìš°ë¦¬ëŠ” **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** íŒ¨í‚¤ì§€ì— êµ¬í˜„ëœ 'ê·¸ë˜í”½ì˜ ë¬¸ë²•(grammar of graphics)'ì´ë¼ëŠ” ì‹œê°í™” í”„ë ˆì„ì›Œí¬ë¥¼ ì‚¬ìš©í•  ê²ƒì…ë‹ˆë‹¤. ì´ í”„ë ˆì„ì›Œí¬ëŠ” ë…¼ë¦¬ì ì´ê³  ìš°ì•„í•œ ë°©ì‹ìœ¼ë¡œ ê³ í’ˆì§ˆ ê·¸ë˜í”½ì„ ë‹¨ê³„ë³„ë¡œ êµ¬ì„±í•  ìˆ˜ ìˆê²Œ í•´ì¤ë‹ˆë‹¤. ë¨¼ì € ì˜ˆì œ ë°ì´í„° ì„¸íŠ¸ë¥¼ ì†Œê°œí•˜ê³  ë¡œë“œí•´ ë³´ê² ìŠµë‹ˆë‹¤.

## 3.3 ì˜ˆì œ ë°ì´í„° ì„¸íŠ¸

[![](imgs/Yusukecells-2_web.jpg)](imgs/Yusukecells-2_web.jpg "ê·¸ë¦¼ 3.5: ì›ì‹œ ë‚´ë°°ì—½(primitive endoderm)ì˜ ë§ˆì»¤ì¸ Serpinh1(íŒŒë€ìƒ‰), Gata6(ë¹¨ê°„ìƒ‰), Nanog(ë…¹ìƒ‰)ìœ¼ë¡œ ì—¼ìƒ‰ëœ E3.5 ë§ˆìš°ìŠ¤ ë°°ë°˜í¬ì˜ ë‹¨ì¼ ì ˆí¸ ë©´ì—­í˜•ê´‘ ì´ë¯¸ì§€.")

ê·¸ë¦¼ 3.5: ì›ì‹œ ë‚´ë°°ì—½(primitive endoderm)ì˜ ë§ˆì»¤ì¸ Serpinh1(íŒŒë€ìƒ‰), Gata6(ë¹¨ê°„ìƒ‰), Nanog(ë…¹ìƒ‰)ìœ¼ë¡œ ì—¼ìƒ‰ëœ E3.5 ë§ˆìš°ìŠ¤ ë°°ë°˜í¬ì˜ ë‹¨ì¼ ì ˆí¸ ë©´ì—­í˜•ê´‘ ì´ë¯¸ì§€.

**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** ê¸°ëŠ¥ì„ ì œëŒ€ë¡œ í…ŒìŠ¤íŠ¸í•˜ë ¤ë©´, ì—¬ëŸ¬ ë‹¤ë¥¸ ê°ë„ì—ì„œ ìë¥´ê³  ë³¼ ìˆ˜ ìˆë„ë¡ ì¶©ë¶„íˆ í¬ê³  ë³µì¡í•œ ë°ì´í„° ì„¸íŠ¸ê°€ í•„ìš”í•©ë‹ˆë‹¤. ì´ˆê¸° ë°œë‹¬ì˜ ë‹¤ë¥¸ ì‹œì ì—ì„œ ì–»ì€ ì•½ 100ê°œì˜ ê°œë³„ ë§ˆìš°ìŠ¤ ë°°ì•„ ì„¸í¬ì˜ ì „ì‚¬ì²´ë¥¼ ë³´ê³ í•˜ëŠ” ìœ ì „ì ë°œí˜„ ë§ˆì´í¬ë¡œì–´ë ˆì´ ë°ì´í„° ì„¸íŠ¸ë¥¼ ì‚¬ìš©í•  ê²ƒì…ë‹ˆë‹¤. í¬ìœ ë¥˜ ë°°ì•„ëŠ” ë‹¨ì¼ ì„¸í¬ì¸ ìˆ˜ì •ë€ì—ì„œ ì‹œì‘ë©ë‹ˆë‹¤. ë™ê¸°í™”ëœ ì„¸í¬ ë¶„ì—´ì˜ ë¬¼ê²°ì„ í†µí•´, ë‚œìëŠ” ì²˜ìŒì—ëŠ” ëˆˆì— ë„ëŠ” ì°¨ì´ê°€ ì—†ëŠ” ì„¸í¬ ë©ì–´ë¦¬ë¡œ ì¦ì‹í•©ë‹ˆë‹¤. ê·¸ëŸ¬ë‚˜ ì–´ëŠ ì‹œì ì—ì„œ ì„¸í¬ëŠ” ë‹¤ë¥¸ ê³„í†µ(lineage)ì„ ì„ íƒí•©ë‹ˆë‹¤. ì ì  ë” ì„¸ë¶„í™”ë˜ë©´ì„œ ì™„ì „í•œ ìœ ê¸°ì²´ì— í•„ìš”í•œ ë‹¤ì–‘í•œ ì„¸í¬ ìœ í˜•ê³¼ ì¡°ì§ì´ ë°œìƒí•©ë‹ˆë‹¤. Ohnishi ë“±([2014](16-chap.html#ref-Ohnishi2014))ì´ ì„¤ëª…í•œ ì‹¤í—˜ì˜ ëª©í‘œëŠ” ë°°ì•„ì—ì„œ ì²« ë²ˆì§¸ ëŒ€ì¹­ì„± ê¹¨ì§ í˜„ìƒê³¼ ê´€ë ¨ëœ ìœ ì „ì ë°œí˜„ ë³€í™”ë¥¼ ì¡°ì‚¬í•˜ëŠ” ê²ƒì´ì—ˆìŠµë‹ˆë‹¤. ì§„í–‰í•˜ë©´ì„œ ë°ì´í„°ë¥¼ ë” ì„¤ëª…í•˜ê² ìŠµë‹ˆë‹¤. ìì„¸í•œ ë‚´ìš©ì€ ë…¼ë¬¸ê³¼ Bioconductor ë°ì´í„° íŒ¨í‚¤ì§€ **[Hiiragi2013](https://bioconductor.org/packages/Hiiragi2013/)**ì˜ ë¬¸ì„œì—ì„œ ì°¾ì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë¨¼ì € ë°ì´í„°ë¥¼ ë¡œë“œí•©ë‹ˆë‹¤.

![](imgs/devil.png)

ë°ì´í„° ê°ì²´ê°€ ë” ì„¤ëª…ì ì¸ ì´ë¦„ ëŒ€ì‹  ë‹¤ì†Œ ì¼ë°˜ì ì¸ ì´ë¦„ì¸ `x`ë¥¼ ê°–ëŠ” ê²ƒì€ ìœ ê°ì…ë‹ˆë‹¤. ì´ë¦„ ì¶©ëŒì„ í”¼í•˜ê¸° ìœ„í•´ ê°€ì¥ ì‹¤ìš©ì ì¸ í•´ê²°ì±…ì€ `esHiiragi = x; rm(list="x")`ì™€ ê°™ì€ ì½”ë“œë¥¼ ì‹¤í–‰í•˜ëŠ” ê²ƒì…ë‹ˆë‹¤.

    
    
    library("Hiiragi2013")__
    
    
    In chunk 'loadHiiragi': Warning: replacing previous import 'boot::logit' by 'gtools::logit' whenloading 'Hiiragi2013'
    
    
    In chunk 'loadHiiragi': Warning: replacing previous import 'boot::inv.logit' by 'gtools::inv.logit'when loading 'Hiiragi2013'
    
    
    data("x")
    dim(Biobase::exprs(x))__
    
    
    [1] 45101   101

R í”„ë¡¬í”„íŠ¸ì— `x`ë¥¼ ì…ë ¥í•˜ì—¬ _ExpressionSet_ ê°ì²´ `x`ì˜ ë” ìì„¸í•œ ìš”ì•½ì„ ì¶œë ¥í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë°ì´í„° í–‰ë ¬ì˜ 101ê°œ ì—´(**[Biobase](https://bioconductor.org/packages/Biobase/)** íŒ¨í‚¤ì§€ì˜ `exprs` í•¨ìˆ˜ë¥¼ í†µí•´ ìœ„ì—ì„œ ì ‘ê·¼)ì€ í‘œë³¸(ê°ê° ë‹¨ì¼ ì„¸í¬)ì— í•´ë‹¹í•˜ê³ , 45101ê°œ í–‰ì€ Affymetrix mouse4302 ì–´ë ˆì´ì— ì˜í•´ í”„ë¡œë¸Œëœ ìœ ì „ìì— í•´ë‹¹í•©ë‹ˆë‹¤. ë°ì´í„°ëŠ” RMA ë°©ë²•ì„ ì‚¬ìš©í•˜ì—¬ ì •ê·œí™”ë˜ì—ˆìŠµë‹ˆë‹¤([Irizarry et al. 2003](16-chap.html#ref-Irizarry:Biostat:2003)). ì›ì‹œ ë°ì´í„°ëŠ” íŒ¨í‚¤ì§€(ë°ì´í„° ê°ì²´ `a`)ì™€ EMBL-EBIì˜ ArrayExpress ë°ì´í„°ë² ì´ìŠ¤(ì ‘ê·¼ ì½”ë“œ E-MTAB-1681)ì—ì„œë„ ì‚¬ìš©í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

í‘œë³¸ì— ëŒ€í•´ ì–´ë–¤ ì •ë³´ê°€ ìˆëŠ”ì§€ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤2.

2 #CAB2D6 í‘œê¸°ë²•ì€ ìƒ‰ìƒì˜ RGB ì¢Œí‘œë¥¼ 16ì§„ìˆ˜ë¡œ í‘œí˜„í•œ ê²ƒì…ë‹ˆë‹¤. ì´ì— ëŒ€í•œ ìì„¸í•œ ë‚´ìš©ì€ 3.10.2ì ˆì„ ì°¸ì¡°í•˜ì„¸ìš”.

    
    
    head(pData(x), n = 2)__
    
    
            File.name Embryonic.day Total.number.of.cells lineage genotype
    1 E3.25  1_C32_IN         E3.25                    32               WT
    2 E3.25  2_C32_IN         E3.25                    32               WT
              ScanDate sampleGroup sampleColour
    1 E3.25 2011-03-16       E3.25      #CAB2D6
    2 E3.25 2011-03-16       E3.25      #CAB2D6

ì œê³µëœ ì •ë³´ëŠ” ì„¸í¬ì— ëŒ€í•œ ì •ë³´(ì¦‰, ì„¸í¬ë¥¼ ì–»ì€ ë°°ì•„ì˜ ë‚˜ì´, í¬ê¸°, ìœ ì „ìí˜•)ì™€ ê¸°ìˆ ì  ì •ë³´(ìŠ¤ìº” ë‚ ì§œ, ì›ì‹œ ë°ì´í„° íŒŒì¼ ì´ë¦„)ê°€ í˜¼í•©ëœ ê²ƒì…ë‹ˆë‹¤. ê´€ë¡€ì— ë”°ë¼, ë§ˆìš°ìŠ¤ ë°°ì•„ì˜ ë°œë‹¬ ì‹œê°„ì€ ì¼ ë‹¨ìœ„ë¡œ ì¸¡ì •ë˜ë©°, ì˜ˆë¥¼ ë“¤ì–´ `E3.5`ë¡œ ë³´ê³ ë©ë‹ˆë‹¤. ë˜í•œ, ë…¼ë¬¸ì—ì„œ ì €ìë“¤ì€ ë‚˜ì´, ìœ ì „ìí˜•, ê³„í†µì„ ê¸°ì¤€ìœ¼ë¡œ ì„¸í¬ë¥¼ 8ê°œì˜ ìƒë¬¼í•™ì  ê·¸ë£¹(`sampleGroup`)ìœ¼ë¡œ ë‚˜ëˆ„ê³ , ì´ ê·¸ë£¹ë“¤ì„ ë‚˜íƒ€ë‚´ëŠ” ìƒ‰ìƒ êµ¬ì„±í‘œ(`sampleColour`3)ë¥¼ ì •ì˜í–ˆìŠµë‹ˆë‹¤. ë‹¤ìŒ ì½”ë“œ(ì„¤ëª…ì€ ì•„ë˜ ì°¸ì¡°)ë¥¼ ì‚¬ìš©í•˜ì—¬ ê° ê·¸ë£¹ì— ëŒ€í•œ ìš”ì•½ ì •ë³´, ì¦‰ ì„¸í¬ ìˆ˜ì™€ ì„ í˜¸í•˜ëŠ” ìƒ‰ìƒì„ í¬í•¨í•˜ëŠ” ì‘ì€ ë°ì´í„° í”„ë ˆì„ `groups`ë¥¼ ì •ì˜í•©ë‹ˆë‹¤.

3 ë°ì´í„° ì„¸íŠ¸ì˜ ì´ ì‹ë³„ìëŠ” ì˜êµ­ì‹ ì² ìë²•ì„ ì‚¬ìš©í•©ë‹ˆë‹¤. ì´ ì±…ì˜ ë‹¤ë¥¸ ê³³ì—ì„œëŠ” ë¯¸êµ­ì‹ ì² ìë²•(color)ì„ ì‚¬ìš©í•©ë‹ˆë‹¤. **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** íŒ¨í‚¤ì§€ëŠ” ì¼ë°˜ì ìœ¼ë¡œ ë‘ ì² ìë²•ì„ ëª¨ë‘ í—ˆìš©í•©ë‹ˆë‹¤.

    
    
    library("dplyr")
    groups = group_by(pData(x), sampleGroup) |>
      summarise(n = n(), color = unique(sampleColour))
    groups __
    
    
    # A tibble: 8 Ã— 3
      sampleGroup         n color  
      <chr>           <int> <chr>  
    1 E3.25              36 #CAB2D6
    2 E3.25 (FGF4-KO)    17 #FDBF6F
    3 E3.5 (EPI)         11 #A6CEE3
    4 E3.5 (FGF4-KO)      8 #FF7F00
    5 E3.5 (PE)          11 #B2DF8A
    6 E4.5 (EPI)          4 #1F78B4
    7 E4.5 (FGF4-KO)     10 #E31A1C
    8 E4.5 (PE)           4 #33A02C

`FGF4-KO`ë¼ëŠ” ì´ë¦„ì´ í¬í•¨ëœ ê·¸ë£¹ì˜ ì„¸í¬ëŠ” ì„¸í¬ ë¶„í™”ì˜ ì¤‘ìš”í•œ ì¡°ì ˆìì¸ FGF4 ìœ ì „ìê°€ ë…¹ì•„ì›ƒëœ ë°°ì•„ì—ì„œ ìœ ë˜í–ˆìŠµë‹ˆë‹¤. E3.5ë¶€í„°, ì•¼ìƒí˜• ì„¸í¬(FGF4 ë…¹ì•„ì›ƒ ì—†ìŒ)ëŠ” ì²« ë²ˆì§¸ ëŒ€ì¹­ì„± ê¹¨ì§ í˜„ìƒì„ ê²ªê³ , ì „ë¶„í™”ëŠ¥ ì™¸ë°°ì—½(pluripotent epiblast, EPI)ê³¼ ì›ì‹œ ë‚´ë°°ì—½(primitive endoderm, PE)ì´ë¼ëŠ” ë‹¤ë¥¸ ì„¸í¬ ê³„í†µìœ¼ë¡œ ë¶„í™”ë©ë‹ˆë‹¤.

ìœ„ì˜ ì½”ë“œ ì²­í¬ëŠ” **[dplyr](https://cran.r-project.org/web/packages/dplyr/)** íŒ¨í‚¤ì§€ì˜ íŒŒì´í”„ ì—°ì‚°ì `|>`ì™€ `group_by`, `summarise` í•¨ìˆ˜ë¥¼ ì²˜ìŒ ì ‘í•˜ëŠ” ê²½ìš°ì´ë¯€ë¡œ, ì½”ë“œë¥¼ í’€ì–´ë³´ê² ìŠµë‹ˆë‹¤. ë¨¼ì € íŒŒì´í”„ `|>`4ì…ë‹ˆë‹¤. ì¼ë°˜ì ìœ¼ë¡œ íŒŒì´í”„ëŠ” ì¤‘ì²©ëœ í•¨ìˆ˜ í˜¸ì¶œì„ ì‚¬ëŒì´ ë” ì‰½ê²Œ ì½ì„ ìˆ˜ ìˆë„ë¡ ë§Œë“œëŠ” ë° ìœ ìš©í•©ë‹ˆë‹¤. ë‹¤ìŒ ë‘ ì¤„ì˜ ì½”ë“œëŠ” Rì—ì„œ ë™ì¼í•©ë‹ˆë‹¤.

4 `|>`ëŠ” 2021ë…„ì— ì¶œì‹œëœ ë²„ì „ 4.1ë¶€í„° ê¸°ë³¸ Rê³¼ í•¨ê»˜ ì œê³µë˜ëŠ” íŒŒì´í”„ ì—°ì‚°ìì…ë‹ˆë‹¤. **[magrittr](https://cran.r-project.org/web/packages/magrittr/)** íŒ¨í‚¤ì§€ëŠ” ì˜ë¯¸ëŠ” ë™ì¼í•˜ì§€ ì•Šì§€ë§Œ ìœ ì‚¬í•œ `%>%` ì—°ì‚°ìë¥¼ ì˜¤ë˜ ì „ë¶€í„° ì œê³µí•´ ì™”ìœ¼ë©°, `%<>%` ë° `%T>%`ì™€ ê°™ì€ ì—¬ëŸ¬ ë‹¤ë¥¸ íŒŒì´í•‘ ê´€ë ¨ ì—°ì‚°ìë„ ì œê³µí•©ë‹ˆë‹¤. ì´ ì±…ì˜ ëŒ€ë¶€ë¶„ì€ 2021ë…„ ì´ì „ì— ì‘ì„±ë˜ì—ˆìœ¼ë¯€ë¡œ **[magrittr](https://cran.r-project.org/web/packages/magrittr/)**ì˜ `%>%` ì—°ì‚°ìê°€ ì—¬ëŸ¬ ê³³ì—ì„œ ì‚¬ìš©ë©ë‹ˆë‹¤. ì—¬ê¸°ì— í‘œì‹œëœ ì½”ë“œì—ì„œì²˜ëŸ¼ ì±…ì„ ìœ ì§€ ê´€ë¦¬í•˜ëŠ” ë™ì•ˆ `%>%`ë¥¼ `|>`ë¡œ ê°€ë” ì—…ë°ì´íŠ¸í•˜ê³  ìˆìŠµë‹ˆë‹¤.

    
    
    f(x) |> g(y) |> h()
    h(g(f(x), y))__

ì´ê²ƒì€ "ë¨¼ì € `f(x)`ë¥¼ í‰ê°€í•œ ë‹¤ìŒ, ê·¸ ê²°ê³¼ë¥¼ `g` í•¨ìˆ˜ì˜ ì²« ë²ˆì§¸ ì¸ìˆ˜ë¡œ ì „ë‹¬í•˜ê³ , `y`ëŠ” `g`ì˜ ë‘ ë²ˆì§¸ ì¸ìˆ˜ë¡œ ì „ë‹¬í•©ë‹ˆë‹¤. ê·¸ëŸ° ë‹¤ìŒ `g`ì˜ ì¶œë ¥ì„ `h` í•¨ìˆ˜ì— ì „ë‹¬í•©ë‹ˆë‹¤."ë¼ëŠ” ì˜ë¯¸ì…ë‹ˆë‹¤. ì´ë¥¼ ë¬´í•œíˆ ë°˜ë³µí•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. íŠ¹íˆ ì¸ìˆ˜ `x`ì™€ `y`ê°€ ê·¸ ìì²´ë¡œ ë³µì¡í•œ í‘œí˜„ì‹ì´ê±°ë‚˜, ì—¬ëŸ¬ í•¨ìˆ˜ì˜ ì—°ì‡„ê°€ ê´€ë ¨ëœ ê²½ìš° ì²« ë²ˆì§¸ ë²„ì „ì´ ë” ì½ê¸° ì‰¬ìš´ ê²½í–¥ì´ ìˆìŠµë‹ˆë‹¤.

`group_by` í•¨ìˆ˜ëŠ” ëª¨ë“  í›„ì† ì‘ì—…ì´ ì „ì²´ ë°ì´í„° í”„ë ˆì„ì— í•œ ë²ˆì— ì ìš©ë˜ëŠ” ê²ƒì´ ì•„ë‹ˆë¼ `sampleGroup` ìš”ì¸ì— ì˜í•´ ì •ì˜ëœ ë¸”ë¡ì— ì ìš©ë˜ì–´ì•¼ í•œë‹¤ëŠ” ë©”ëª¨ë¡œ ë°ì´í„° í”„ë ˆì„ì„ "í‘œì‹œ"í•©ë‹ˆë‹¤. ë§ˆì§€ë§‰ìœ¼ë¡œ, `summarise`ëŠ” ìš”ì•½ í†µê³„ë¥¼ ê³„ì‚°í•©ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ `mean`, `sum` ë“±ì´ ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì´ ê²½ìš°, ìš°ë¦¬ëŠ” ë‹¨ìˆœíˆ ê° ë¸”ë¡ì˜ í–‰ ìˆ˜(`n()`)ì™€ ëŒ€í‘œ ìƒ‰ìƒì„ ê³„ì‚°í•©ë‹ˆë‹¤.

## 3.4 ggplot2

**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** is a package
by Hadley Wickham ([Wickham 2016](16-chap.html#ref-ggplot2)) that implements
the idea of **grammar of graphics** â€“ a concept created by Leland Wilkinson in
his eponymous book ([Wilkinson 2005](16-chap.html#ref-GrofGrbook)). We will
explore some of its functionality in this chapter, and you will see many
examples of how it can be used in the rest of this book. Comprehensive
documentation for the package can be found [on its
website](https://ggplot2.tidyverse.org). The online documentation includes
example use cases for each of the graphic types that are introduced in this
chapter (and many more) and is an invaluable resource when creating figures.

Letâ€™s start by loading the package and redoing the simple plot of Figure 3.2.

    
    
    library("ggplot2")
    ggplot(DNase, aes(x = conc, y = density)) + geom_point()__

[![](03-chap_files/figure-html/fig-graphics-
figredobasicplottingwithggplot-1.png)](03-chap_files/figure-html/fig-graphics-
figredobasicplottingwithggplot-1.png "FigureÂ 3.6: Our first ggplot2 figure,
similar to the base graphics FigureÂ fig-graphics-basicplotting1.")

Figure 3.6: Our first
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** figure,
similar to the base graphics Figure 3.2.

We just wrote our first â€œsentenceâ€ using the grammar of graphics. Letâ€™s
deconstruct this sentence. First, we specified the dataframe that contains the
data, `DNase`. The `aes` (this stands for **aesthetic**) argument states which
variables we want to see mapped to the \\(x\\)\- and \\(y\\)-axes,
respectively. Finally, we stated that we want the plot to use points (as
opposed to, say, lines or bars), by adding the result of calling the function
`geom_point`.

Now letâ€™s turn to the mouse single cell data and plot the number of samples
for each of the 8 groups using the `ggplot` function. The result is shown in
Figure 3.7.

    
    
    ggplot(groups, aes(x = sampleGroup, y = n)) +
      geom_bar(stat = "identity")__

[![](03-chap_files/figure-html/fig-graphics-
qplot1-1.png)](03-chap_files/figure-html/fig-graphics-qplot1-1.png
"FigureÂ 3.7: A barplot, produced with the ggplot function from the table of
group sizes in the mouse single cell data.")

Figure 3.7: A barplot, produced with the `ggplot` function from the table of
group sizes in the mouse single cell data.

With `geom_bar` we now told `ggplot` that we want each data item (each row of
`groups`) to be represented by a bar. Bars are one example of geometric object
(**geom** in the
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** packageâ€™s
parlance) that `ggplot` knows about. We have already seen another such object
in Figure 3.6: points, indicated by the `geom_point` function. We will
encounter many other geoms later. We used the `aes` to indicate that we want
the groups shown along the \\(x\\)-axis and the sizes along the \\(y\\)-axis.
Finally, we provided the argument `stat = "identity"` (in other words, do
nothing) to the `geom_bar` function, since otherwise it would try to compute a
histogram of the data (the default value of `stat` is `"count"`). `stat` is
short for _statistic_ , which is what we call any function of data. Besides
the identity and count statistic, there are others, such as smoothing,
averaging, binning, or other operations that reduce the data in some way.

These concepts â€“data, geometrical objects, statisticsâ€“ are some of the
ingredients of the grammar of graphics, just as nouns, verbs and adverbs are
ingredients of an English sentence.

__

Task

Flip the \\(x\\)\- and \\(y\\)-aesthetics to produce a horizontal barplot.

The plot in Figure 3.7 is not bad, but there are several potential
improvements. We can use color for the bars to help us quickly see which bar
corresponds to which group. This is particularly useful if we use the same
color scheme in several plots. To this end, letâ€™s define a named vector
`groupColor` that contains our desired colors for each possible value of
`sampleGroup`.5

5 The information is completely equivalent to that in the `sampleGroup` and
`color` columns of the dataframe `groups`; we are just adapting to the fact
that **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** expects
this information in the form of a named vector.

    
    
    groupColor = setNames(groups$color, groups$sampleGroup)__

Another thing that we need to fix in Figure 3.7 is the readability of the bar
labels. Right now they are running into each other â€” a common problem when you
have descriptive names.

    
    
    ggplot(groups, aes(x = sampleGroup, y = n, fill = sampleGroup)) +
      geom_bar(stat = "identity") +
      scale_fill_manual(values = groupColor, name = "Groups") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1))__

[![](03-chap_files/figure-html/fig-graphics-
qplot2-1.png)](03-chap_files/figure-html/fig-graphics-qplot2-1.png
"FigureÂ 3.8: Similar to FigureÂ fig-graphics-qplot1, but with colored bars and
better bar labels.")

Figure 3.8: Similar to Figure 3.7, but with colored bars and better bar
labels.

This is now already a longer and more complex sentence. Letâ€™s dissect it. We
added an argument, `fill` to the `aes` function that states that we want the
bars to be colored (filled) based on `sampleGroup` (which in this case co-
incidentally is also the value of the `x` argument, but that need not always
be so). Furthermore we added a call to the `scale_fill_manual` function, which
takes as its input a color map â€“ i.e., the mapping from the possible values of
a variable to the associated colors â€“ as a named vector. We also gave this
color map a title (note that in more complex plots, there can be several
different color maps involved). Had we omitted the call to
`scale_fill_manual`,
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** would have
used its choice of default colors. We also added a call to `theme` stating
that we want the \\(x\\)-axis labels rotated by 90 degrees and right-aligned
(`hjust`; the default would be to center).

### 3.4.1 ë°ì´í„° íë¦„

![](imgs/devil.png)

`ggplot` í•¨ìˆ˜ëŠ” ë°ì´í„°ë¥¼ ë°ì´í„° í”„ë ˆì„ í˜•ì‹ìœ¼ë¡œ ê¸°ëŒ€í•©ë‹ˆë‹¤. ë°ì´í„°ê°€ í–‰ë ¬ì´ë‚˜ ë³„ë„ì˜ ë²¡í„°, ë˜ëŠ” ë‹¤ë¥¸ ìœ í˜•ì˜ ê°ì²´ì— ìˆë‹¤ë©´ ë³€í™˜í•´ì•¼ í•©ë‹ˆë‹¤. **[dplyr](https://cran.r-project.org/web/packages/dplyr/)** ë° **[broom](https://cran.r-project.org/web/packages/broom/)** ë“±ì˜ íŒ¨í‚¤ì§€ê°€ ì´ ëª©ì ì„ ìœ„í•œ ê¸°ëŠ¥ì„ ì œê³µí•©ë‹ˆë‹¤. ì´ì— ëŒ€í•´ì„œëŠ” 13.10ì ˆì—ì„œ ë” ìì„¸íˆ ë‹¤ë£° ê²ƒì´ë©°, ì±… ì „ì²´ì—ì„œ ì´ëŸ¬í•œ ë³€í™˜ì˜ ì˜ˆì‹œë¥¼ ë³´ê²Œ ë  ê²ƒì…ë‹ˆë‹¤.

ì—¬ê¸°ì—ëŠ” ê¸°ë³¸ Rì˜ _data.frame_ë¿ë§Œ ì•„ë‹ˆë¼ tidyverseì˜ **[tibble](https://cran.r-project.org/web/packages/tibble/)** íŒ¨í‚¤ì§€ì— ìˆëŠ” _tibble_(ë° ë™ì˜ì–´ _data_frame_) í´ë˜ìŠ¤ë„ í¬í•¨ë©ë‹ˆë‹¤.

`ggplot` í˜¸ì¶œì˜ ê²°ê³¼ëŠ” _ggplot_ ê°ì²´ì…ë‹ˆë‹¤. ì•ì„œ ë³´ì•˜ë˜ ì½”ë“œ ì¡°ê°ì„ ë‹¤ì‹œ ìƒê¸°í•´ ë´…ì‹œë‹¤.

    
    
    gg = ggplot(DNase, aes(x = conc, y = density)) + geom_point()__

ì—¬ê¸°ì„œëŠ” `ggplot`ì˜ ì¶œë ¥ì„ ì½˜ì†”ë¡œ ì§ì ‘ ë³´ë‚´ ê·¸ë¦¼ 3.6ì„ "ì¶œë ¥"í•˜ëŠ” ëŒ€ì‹ , `gg` ê°ì²´ì— í• ë‹¹í–ˆìŠµë‹ˆë‹¤. ì´ëŠ” R ì½˜ì†”ì—ì„œ ì‘ì—…í•  ë•Œì™€ ì™„ì „íˆ ë™ì¼í•©ë‹ˆë‹¤. `1+1` ê°™ì€ í‘œí˜„ì‹ì„ ì…ë ¥í•˜ê³  "Enter"ë¥¼ ëˆ„ë¥´ë©´ ê²°ê³¼ê°€ ì¶œë ¥ë˜ì§€ë§Œ, `s = 1+1`ì²˜ëŸ¼ í• ë‹¹ í‘œí˜„ì‹ì¸ ê²½ìš° ë¶€ì‘ìš©(ì´ë¦„ `"s"`ê°€ ë©”ëª¨ë¦¬ì˜ ê°ì²´ì— ë°”ì¸ë”©ë¨)ë§Œ ë°œìƒí•˜ê³  ì•„ë¬´ê²ƒë„ ì¶œë ¥ë˜ì§€ ì•ŠëŠ” ê²ƒê³¼ ê°™ìŠµë‹ˆë‹¤. ë§ˆì°¬ê°€ì§€ë¡œ `source`ë¡œ í˜¸ì¶œëœ ìŠ¤í¬ë¦½íŠ¸ ë‚´ì—ì„œ í‘œí˜„ì‹ì´ í‰ê°€ë  ë•Œë„ ì¶œë ¥ë˜ì§€ ì•ŠìŠµë‹ˆë‹¤. ë”°ë¼ì„œ ìœ„ ì½”ë“œë„ `print` ë©”ì„œë“œê°€ í˜¸ì¶œë˜ì§€ ì•Šì•˜ê¸° ë•Œë¬¸ì— ê·¸ë˜í”½ ì¶œë ¥ì„ ìƒì„±í•˜ì§€ ì•ŠìŠµë‹ˆë‹¤. `gg`ë¥¼ ì¶œë ¥í•˜ë ¤ë©´ (ëŒ€í™”í˜• ì„¸ì…˜ì—ì„œ) ê·¸ ì´ë¦„ì„ ì…ë ¥í•˜ê±°ë‚˜ `print`ë¥¼ í˜¸ì¶œí•˜ë©´ ë©ë‹ˆë‹¤.

    
    
    gg
    print(gg)__

### 3.4.2 ê·¸ë¦¼ ì €ì¥í•˜ê¸°

**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì—ëŠ” `ggsave`ë¼ëŠ” ë‚´ì¥ í”Œë¡¯ ì €ì¥ í•¨ìˆ˜ê°€ ìˆìŠµë‹ˆë‹¤.

    
    
    ggplot2::ggsave("DNAse-histogram-demo.pdf", plot = gg)__

í”Œë¡¯ì„ ì €ì¥í•˜ëŠ” ë‘ ê°€ì§€ ì£¼ìš” ë°©ë²•ì€ ë²¡í„° ê·¸ë˜í”½ê³¼ ë˜ìŠ¤í„°(í”½ì…€) ê·¸ë˜í”½ì…ë‹ˆë‹¤. ë²¡í„° ê·¸ë˜í”½ì—ì„œ í”Œë¡¯ì€ ì , ì„ , ê³¡ì„ , ë„í˜• ë° íƒ€ì´í¬ê·¸ë˜í”¼ ë¬¸ìì™€ ê°™ì€ ì¼ë ¨ì˜ ê¸°í•˜í•™ì  ì›í˜•(primitives)ìœ¼ë¡œ ì €ì¥ë©ë‹ˆë‹¤. Rì—ì„œ í”Œë¡¯ì„ ë²¡í„° ê·¸ë˜í”½ í˜•ì‹ìœ¼ë¡œ ì €ì¥í•  ë•Œ ì„ í˜¸ë˜ëŠ” í˜•ì‹ì€ PDFì…ë‹ˆë‹¤. ë˜ìŠ¤í„° ê·¸ë˜í”½ì—ì„œ í”Œë¡¯ì€ ë„íŠ¸ ë§¤íŠ¸ë¦­ìŠ¤ ë°ì´í„° êµ¬ì¡°ì— ì €ì¥ë©ë‹ˆë‹¤. ë˜ìŠ¤í„° í˜•ì‹ì˜ ì£¼ìš” í•œê³„ëŠ” ì‚¬ìš© ê°€ëŠ¥í•œ í”½ì…€ ìˆ˜ì— ë”°ë¥¸ ì œí•œëœ í•´ìƒë„ì…ë‹ˆë‹¤. Rì—ì„œ ë˜ìŠ¤í„° ê·¸ë˜í”½ ì¶œë ¥ì— ê°€ì¥ í”íˆ ì‚¬ìš©ë˜ëŠ” ì¥ì¹˜ëŠ” `png`ì…ë‹ˆë‹¤. ì¼ë°˜ì ìœ¼ë¡œ í”Œë¡¯ì„ ë²¡í„° ê·¸ë˜í”½ í˜•ì‹ìœ¼ë¡œ ì €ì¥í•˜ëŠ” ê²ƒì´ ë°”ëŒì§í•©ë‹ˆë‹¤. ë‚˜ì¤‘ì— ì–¸ì œë“ ì§€ ë²¡í„° ê·¸ë˜í”½ íŒŒì¼ì„ ì›í•˜ëŠ” í•´ìƒë„ì˜ ë˜ìŠ¤í„° í˜•ì‹ìœ¼ë¡œ ë³€í™˜í•  ìˆ˜ ìˆëŠ” ë°˜ë©´, ê·¸ ë°˜ëŒ€ëŠ” ë§¤ìš° ì–´ë µê¸° ë•Œë¬¸ì…ë‹ˆë‹¤. ë°œí‘œë‚˜ ë…¼ë¬¸ì˜ ê·¸ë¦¼ì´ í”½ì…€í™” ì•„í‹°íŒ©íŠ¸ ë•Œë¬¸ì— ë³´ê¸° í‰í•´ì§€ëŠ” ê²ƒì„ ì›ì¹˜ ì•Šìœ¼ì‹¤ ê²ƒì…ë‹ˆë‹¤!

## 3.5 The grammar of graphics

The components of
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** â€™s grammar of
graphics are

  1. one or more datasets,

  2. one or more geometric objects that serve as the visual representations of the data, â€“ for instance, points, lines, rectangles, contours,

  3. descriptions of how the variables in the data are mapped to visual properties (aesthetics) of the geometric objects, and an associated scale (e. g., linear, logarithmic, rank),

  4. one or more coordinate systems,

  5. statistical summarization rules,

  6. a facet specification, i.e. the use of multiple similar subplots to look at subsets of the same data,

  7. optional parameters that affect the layout and rendering, such text size, font and alignment, legend positions.

In the examples above, Figures 3.7 and 3.8, the dataset was `groupsize`, the
variables were the numeric values as well as the names of `groupsize`, which
we mapped to the aesthetics \\(y\\)-axis and \\(x\\)-axis respectively, the
scale was linear on the \\(y\\) and rank-based on the \\(x\\)-axis (the bars
are ordered alphanumerically and each has the same width), and the geometric
object was the rectangular bar.

Items 4â€“7 in the above list are optional. If you donâ€™t specify them, then the
Cartesian is used as the coordinate system, the statistical summary is the
trivial one (i.e., the identity), and no facets or subplots are made (weâ€™ll
see examples later on, in Section 3.8). The first three items are required: a
valid **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**
â€œsentenceâ€ needs to contain at least one of each of them.

In fact, **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** â€™s
implementation of the grammar of graphics allows you to use the same type of
component multiple times, in what are called **layers** ([Wickham
2010](16-chap.html#ref-Wickham:LayeredGrammar)). For example, the code below
uses three types of geometric objects in the same plot, for the same data:
points, a line and a confidence band.

    
    
    dftx = data.frame(t(Biobase::exprs(x)), pData(x))
    ggplot( dftx, aes( x = X1426642_at, y = X1418765_at )) +
      geom_point( shape = 1 ) +
      geom_smooth( method = "loess" )__

[![](03-chap_files/figure-html/fig-graphics-
scp2layers1-1.png)](03-chap_files/figure-html/fig-graphics-scp2layers1-1.png
"FigureÂ 3.9: A scatterplot with three layers that show different statistics of
the same data: points \(geom_point\), a smooth regression line and a
confidence band \(the latter two from geom_smooth\).")

Figure 3.9: A scatterplot with three layers that show different statistics of
the same data: points (`geom_point`), a smooth regression line and a
confidence band (the latter two from `geom_smooth`).

Here we had to assemble a copy of the expression data (`Biobase::exprs(x)`)
and the sample annotation data (`pData(x)`) all together into the dataframe
`dftx` â€“ since this is the data format that
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** functions most
easily take as input (more on this in [Section 13.10](13-chap.html#sec-design-
datarep)).

We can further enhance the plot by using colors â€“ since each of the points in
Figure 3.9 corresponds to one sample, it makes sense to use the `sampleColour`
information in the object `x`.

    
    
    ggplot(dftx, aes(x = X1426642_at, y = X1418765_at))  +
      geom_point(aes(color = sampleGroup), shape = 19) +
      scale_color_manual(values = groupColor, guide = "none") +
      geom_smooth(method = "loess")__

[![](03-chap_files/figure-html/fig-graphics-
scp2layers2-1.png)](03-chap_files/figure-html/fig-graphics-scp2layers2-1.png
"FigureÂ 3.10: As FigureÂ fig-graphics-scp2layers1, but in addition with points
colored by the time point and cell lineage \(as defined in FigureÂ fig-
graphics-qplot2\). We can now see that the expression values of the gene Timd2
\(targeted by the probe 1418765_at, along the y-axis\) are consistently high
in the early time points, whereas its expression goes down in the EPI samples
at days 3.5 and 4.5. In the FGF4-KO, this decrease is delayed - at E3.5, its
expression is still high. Conversely, the gene Fn1 \(1426642_at, x-axis\) is
off in the early timepoints and then goes up at days 3.5 and 4.5. The PE
samples \(green\) show a high degree of cell-to-cell variability.")

Figure 3.10: As Figure 3.9, but in addition with points colored by the time
point and cell lineage (as defined in Figure 3.8). We can now see that the
expression values of the gene Timd2 (targeted by the probe 1418765_at, along
the y-axis) are consistently high in the early time points, whereas its
expression goes down in the EPI samples at days 3.5 and 4.5. In the FGF4-KO,
this decrease is delayed - at E3.5, its expression is still high. Conversely,
the gene Fn1 (1426642_at, x-axis) is off in the early timepoints and then goes
up at days 3.5 and 4.5. The PE samples (green) show a high degree of cell-to-
cell variability.

__

ì§ˆë¬¸ 3.2

ìœ„ì˜ ì½”ë“œì—ì„œ ìš°ë¦¬ëŠ” `geom_point` ë ˆì´ì–´ì— ëŒ€í•´ì„œë§Œ `color` ë¯¸í•™(`aes`)ì„ ì •ì˜í•œ ë°˜ë©´, `x`ì™€ `y` ë¯¸í•™ì€ ëª¨ë“  ë ˆì´ì–´ì— ëŒ€í•´ ì •ì˜í–ˆìŠµë‹ˆë‹¤. ë§Œì•½ ëª¨ë“  ë ˆì´ì–´ì— ëŒ€í•´ `color` ë¯¸í•™ì„ ì„¤ì •í•œë‹¤ë©´, ì¦‰ `ggplot`ì˜ ì¸ìˆ˜ ëª©ë¡ìœ¼ë¡œ ì˜®ê¸´ë‹¤ë©´ ì–´ë–»ê²Œ ë ê¹Œìš”?

__

ì§ˆë¬¸ 3.3

ê·¸ë¦¼ 3.9ì™€ 3.10ì—ì„œì²˜ëŸ¼ ì‚°ì ë„ ë°ì´í„°ë¥¼ íšŒê·€ì„ ê³¼ í•¨ê»˜ ì‹œê°í™”í•˜ëŠ” ê²ƒì´ í•­ìƒ ì˜ë¯¸ê°€ ìˆì„ê¹Œìš”?

ì—¬ë‹´ìœ¼ë¡œ, ì´ëŸ¬í•œ í”„ë¡œë¸Œ ì‹ë³„ìê°€ ì–´ë–¤ ìœ ì „ìë¥¼ íƒ€ê²ŸíŒ…í•˜ê³  ê·¸ ìœ ì „ìë“¤ì´ ì–´ë–¤ ì—­í• ì„ í•˜ëŠ”ì§€ ì•Œê³  ì‹¶ë‹¤ë©´ ë‹¤ìŒê³¼ ê°™ì´ í˜¸ì¶œí•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

![](imgs/devil.png)

ì—¬ê¸°ì„œëŠ” ì›ë˜ì˜ í”¼ì²˜ ì‹ë³„ì(ì˜ˆ: ì„ í–‰ "X"ê°€ ì—†ëŠ” "1426642_at")ë¥¼ ì‚¬ìš©í•´ì•¼ í•©ë‹ˆë‹¤. ì´ëŠ” ë§ˆì´í¬ë¡œì–´ë ˆì´ ì œì¡°ì—…ì²´, Bioconductor ì£¼ì„ íŒ¨í‚¤ì§€, ê·¸ë¦¬ê³  ê°ì²´ `x` ë‚´ë¶€ì—ì„œ ì‚¬ìš©ë˜ëŠ” í‘œê¸°ë²•ì…ë‹ˆë‹¤. ìœ„ì—ì„œ `dftx`ë¡œ ì‘ì—…í•  ë•Œ ì‚¬ìš©í•œ ì„ í–‰ "X"ëŠ” `data.frame` ìƒì„±ì í•¨ìˆ˜ì— ì˜í•´ `dftx` ìƒì„± ì¤‘ì— ì‚½ì…ëœ ê²ƒì¸ë°, ì´ëŠ” `check.names` ì¸ìˆ˜ê°€ ê¸°ë³¸ì ìœ¼ë¡œ `TRUE`ë¡œ ì„¤ì •ë˜ì–´ ìˆê¸° ë•Œë¬¸ì…ë‹ˆë‹¤. ë˜ëŠ” `check.names = FALSE`ë¡œ ì„¤ì •í•˜ì—¬ ì›ë˜ ì‹ë³„ì í‘œê¸°ë²•ì„ ìœ ì§€í•  ìˆ˜ë„ ìˆì—ˆì§€ë§Œ, ê·¸ëŸ´ ê²½ìš° ì½”ë“œ(ì˜ˆ: `aes()` í˜¸ì¶œ)ì—ì„œ Rì´ ì´ë¥¼ ì˜¬ë°”ë¥´ê²Œ í•´ì„í•  ìˆ˜ ìˆë„ë¡ ì‹ë³„ì ì£¼ìœ„ì— ì—­ë”°ì˜´í‘œ(backtick)ë¥¼ ì‚¬ìš©í•´ì•¼ í•©ë‹ˆë‹¤.

íŒ¨í‚¤ì§€ ì´ë¦„ì„ í¬í•¨í•œ ì „ì²´ ì´ë¦„(fully qualified name)ìœ¼ë¡œ `select` í•¨ìˆ˜ë¥¼ í˜¸ì¶œí•˜ê¸° ìœ„í•´ `::` ì—°ì‚°ìë¥¼ ì‚¬ìš©í•œ ê²ƒì— ì£¼ëª©í•˜ì„¸ìš”. ìš°ë¦¬ëŠ” ì´ë¯¸ [2ì¥](02-chap.html)ì—ì„œ ì´ë¥¼ ì ‘í–ˆìŠµë‹ˆë‹¤.

    
    
    library("mouse4302.db")__
    
    
    AnnotationDbi::select(mouse4302.db,
       keys = c("1426642_at", "1418765_at"), keytype = "PROBEID",
       columns = c("SYMBOL", "GENENAME"))__
    
    
         PROBEID SYMBOL                                            GENENAME
    1 1426642_at    Fn1                                       fibronectin 1
    2 1418765_at  Timd2 T cell immunoglobulin and mucin domain containing 2

`ggplot`ì„ ì‚¬ìš©í•  ë•Œ ì¢…ì¢… ë°ì´í„°, ë¯¸í•™, ê·¸ë¦¬ê³  ê¸°í•˜í•™ì  ê°ì²´ë§Œ ì§€ì •í•˜ë©´ ë©ë‹ˆë‹¤. ëŒ€ë¶€ë¶„ì˜ ê¸°í•˜í•™ì  ê°ì²´ëŠ” ë°ì´í„°ì— ì í•©í•œ ê¸°ë³¸ í†µê³„ ìš”ì•½ì„ ì•”ì‹œì ìœ¼ë¡œ í˜¸ì¶œí•©ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ `geom_smooth`ë¥¼ ì‚¬ìš©í•˜ë©´ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ëŠ” ê¸°ë³¸ì ìœ¼ë¡œ `stat = "smooth"`ë¥¼ ì‚¬ìš©í•˜ì—¬ ì„ ì„ í‘œì‹œí•©ë‹ˆë‹¤. `geom_histogram`ì„ ì‚¬ìš©í•˜ë©´ ë°ì´í„°ê°€ êµ¬ê°„í™”(binning)ë˜ê³  ê²°ê³¼ê°€ ë§‰ëŒ€ ê·¸ë¦¼ í˜•ì‹ìœ¼ë¡œ í‘œì‹œë©ë‹ˆë‹¤. ë‹¤ìŒì€ ì˜ˆì‹œì…ë‹ˆë‹¤.

    
    
    dfx = as.data.frame(Biobase::exprs(x))
    ggplot(dfx, aes(x = `20 E3.25`)) + geom_histogram(binwidth = 0.2)__

[![](03-chap_files/figure-html/fig-graphics-hists-1.png)](03-chap_files/figure-html/fig-graphics-hists-1.png "ê·¸ë¦¼ 3.11: E3.25ì¼ì°¨ì˜ íŠ¹ì • ìƒ˜í”Œ í•˜ë‚˜(ì„¸í¬ ë²ˆí˜¸ 20)ì— ëŒ€í•œ í”„ë¡œë¸Œ ê°•ë„ì˜ íˆìŠ¤í† ê·¸ë¨ì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.11: E3.25ì¼ì°¨ì˜ íŠ¹ì • ìƒ˜í”Œ í•˜ë‚˜(ì„¸í¬ ë²ˆí˜¸ 20)ì— ëŒ€í•œ í”„ë¡œë¸Œ ê°•ë„ì˜ íˆìŠ¤í† ê·¸ë¨.

__

ì§ˆë¬¸ 3.4

ê°ì²´ `dfx`ì™€ `dftx`ì˜ ì°¨ì´ì ì€ ë¬´ì—‡ì¸ê°€ìš”? ì™œ ë‘ ê°ì²´ë¥¼ ëª¨ë‘ ë§Œë“¤ì–´ì•¼ í–ˆì„ê¹Œìš”?

ìœ„ì˜ ë§‰ëŒ€ ê·¸ë¦¼ ì˜ˆì œë¡œ ëŒì•„ê°€ ë´…ì‹œë‹¤.

    
    
    pb = ggplot(groups, aes(x = sampleGroup, y = n))__

ì´ê²ƒì€ í”Œë¡¯ ê°ì²´ `pb`ë¥¼ ìƒì„±í•©ë‹ˆë‹¤. ì´ë¥¼ í‘œì‹œí•˜ë ¤ê³  í•˜ë©´ ê¸°í•˜í•™ì  ê°ì²´ë¥¼ ì§€ì •í•˜ì§€ ì•Šì•˜ê¸° ë•Œë¬¸ì— ë¹ˆ í”Œë¡¯ì´ ìƒì„±ë©ë‹ˆë‹¤. ì§€ê¸ˆê¹Œì§€ `pb` ê°ì²´ì— ìˆëŠ” ê²ƒì€ ë°ì´í„°ì™€ ë¯¸í•™ë¿ì…ë‹ˆë‹¤(ê·¸ë¦¼ 3.12).

    
    
    class(pb)__
    
    
    [1] "gg"     "ggplot"
    
    
    pb __

[![](03-chap_files/figure-html/fig-graphics-figbpempty-1.png)](03-chap_files/figure-html/fig-graphics-figbpempty-1.png "ê·¸ë¦¼ 3.12: pb: ê¸°í•˜í•™ì  ê°ì²´(geom)ê°€ ì—†ìœ¼ë©´ í”Œë¡¯ ì˜ì—­ì´ ë¹„ì–´ ìˆìŠµë‹ˆë‹¤. ê¸°ë³¸ ìŠ¤íƒ€ì¼ ë§¤ê°œë³€ìˆ˜ í•˜ì—ì„œëŠ” xì¶•ì˜ ëˆˆê¸ˆ ë ˆì´ë¸”ì„ ì½ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.12: `pb`: ê¸°í•˜í•™ì  ê°ì²´(`geom`)ê°€ ì—†ìœ¼ë©´ í”Œë¡¯ ì˜ì—­ì´ ë¹„ì–´ ìˆìŠµë‹ˆë‹¤. ê¸°ë³¸ ìŠ¤íƒ€ì¼ ë§¤ê°œë³€ìˆ˜ í•˜ì—ì„œëŠ” \(x\)ì¶•ì˜ ëˆˆê¸ˆ ë ˆì´ë¸”ì„ ì½ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.

ì´ì œ `+` ì—°ì‚°ìë¥¼ ì‚¬ìš©í•˜ì—¬ í”Œë¡¯ì˜ ë‹¤ë¥¸ êµ¬ì„± ìš”ì†Œë“¤ì„ ê°„ë‹¨íˆ ì¶”ê°€í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤(ê·¸ë¦¼ 3.13).

    
    
    pb = pb + geom_bar(stat = "identity")
    pb = pb + aes(fill = sampleGroup)
    pb = pb + theme(axis.text.x = element_text(angle = 90, hjust = 1))
    pb = pb + scale_fill_manual(values = groupColor, name = "Groups")
    pb __

[![](03-chap_files/figure-html/fig-graphics-bpgg3-1.png)](03-chap_files/figure-html/fig-graphics-bpgg3-1.png "ê·¸ë¦¼ 3.13: ê·¸ë˜í”½ ê°ì²´ bpì˜ ì™„ì „í•œ ëª¨ìŠµì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.13: ê·¸ë˜í”½ ê°ì²´ `bp`ì˜ ì™„ì „í•œ ëª¨ìŠµ.

ì–´ë–¤ ë°©ì‹ìœ¼ë¡œë“  ì´ë¯¸ ìƒì„±ëœ ê·¸ë˜í”½ ê°ì²´ë¥¼ ê°€ì ¸ì™€ì„œ ë” ì •ì œí•˜ëŠ” ì´ëŸ¬í•œ ë‹¨ê³„ë³„ êµ¬ì¶• ë°©ì‹ì€, ê·¸ë˜í”½ì„ ìƒì„±í•˜ëŠ” ë‹¨ì¼ í•¨ìˆ˜ í˜¸ì¶œì— ëª¨ë“  ì§€ì¹¨ì„ ë¯¸ë¦¬ ì œê³µí•˜ëŠ” ê²ƒë³´ë‹¤ ë” í¸ë¦¬í•˜ê³  ê´€ë¦¬í•˜ê¸° ì‰¬ìš¸ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

ë§¤ë²ˆ ì²˜ìŒë¶€í„° í”Œë¡¯ì„ ë‹¤ì‹œ ë§Œë“¤ í•„ìš” ì—†ì´, ë¶€ë¶„ì ìœ¼ë¡œ ì™„ì„±ëœ ê°ì²´ë¥¼ ì €ì¥í•œ ë‹¤ìŒ ë‹¤ë¥¸ ë°©ì‹ìœ¼ë¡œ ìˆ˜ì •í•¨ìœ¼ë¡œì¨ ë‹¤ì–‘í•œ ì‹œê°í™” ì•„ì´ë””ì–´ë¥¼ ë¹ ë¥´ê²Œ ì‹œë„í•´ ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´, ë§‰ëŒ€ ê·¸ë¦¼ì˜ ëŒ€ì•ˆì ì¸ ì‹œê°í™”ë¥¼ ë§Œë“¤ê¸° ìœ„í•´ í”Œë¡¯ì„ ê·¹ì¢Œí‘œê³„(polar coordinates)ë¡œ ì „í™˜í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

    
    
    pb.polar = pb + coord_polar() +
      theme(axis.text.x = element_text(angle = 0, hjust = 1),
            axis.text.y = element_blank(),
            axis.ticks = element_blank()) +
      xlab("") + ylab("")
    pb.polar __

[![](03-chap_files/figure-html/fig-graphics-bpgg7-1.png)](03-chap_files/figure-html/fig-graphics-bpgg7-1.png "ê·¸ë¦¼ 3.14: ê·¹ì¢Œí‘œê³„ì—ì„œì˜ ë§‰ëŒ€ ê·¸ë¦¼ì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.14: ê·¹ì¢Œí‘œê³„ì—ì„œì˜ ë§‰ëŒ€ ê·¸ë¦¼.

ìœ„ì—ì„œ ì´ì „ì— ì„¤ì •í•œ `theme` ë§¤ê°œë³€ìˆ˜ë¥¼ ë‹¨ìˆœíˆ ìƒˆ ê°’ìœ¼ë¡œ ì„¤ì •í•˜ì—¬ ë®ì–´ì“¸ ìˆ˜ ìˆìŒì— ì£¼ëª©í•˜ì„¸ìš”. ì›ë˜ ì„¤ì •í–ˆë˜ `pb`ë¥¼ ë‹¤ì‹œ ìƒì„±í•  í•„ìš”ê°€ ì—†ìŠµë‹ˆë‹¤.

## 3.6 1ì°¨ì› ë°ì´í„° ì‹œê°í™”

ìƒë¬¼í•™ì  ë°ì´í„° ë¶„ì„ì—ì„œ ì¼ë°˜ì ì¸ ì‘ì—…ì€ ë‹¨ë³€ëŸ‰ ì¸¡ì •ê°’(univariate measurements)ì˜ ì—¬ëŸ¬ ìƒ˜í”Œì„ ë¹„êµí•˜ëŠ” ê²ƒì…ë‹ˆë‹¤. ì´ ì ˆì—ì„œëŠ” ì´ëŸ¬í•œ ìƒ˜í”Œì„ ì‹œê°í™”í•˜ê³  ë¹„êµí•˜ëŠ” ëª‡ ê°€ì§€ ê°€ëŠ¥ì„±ì„ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤. ì˜ˆì‹œë¡œ Fgf4, Gata4, Gata6, Sox2ë¼ëŠ” ë„¤ ìœ ì „ìì˜ ê°•ë„ë¥¼ ì‚¬ìš©í•˜ê² ìŠµë‹ˆë‹¤6. ë§ˆì´í¬ë¡œì–´ë ˆì´ì—ì„œ ì´ë“¤ì€ ë‹¤ìŒê³¼ ê°™ì´ í‘œì‹œë©ë‹ˆë‹¤.

6 ì´ ìœ ì „ìë“¤ì— ëŒ€í•´ì„œëŠ” ([Ohnishi et al. 2014](16-chap.html#ref-Ohnishi2014))ì—ì„œ ë” ìì„¸íˆ ì½ì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

    
    
    selectedProbes = c( Fgf4 = "1420085_at", Gata4 = "1418863_at",
                       Gata6 = "1425463_at",  Sox2 = "1416967_at")__

ì´ í‘œí˜„ì—ì„œ ë°ì´í„°ë¥¼ ì¶”ì¶œí•˜ì—¬ ë°ì´í„° í”„ë ˆì„ìœ¼ë¡œ ë³€í™˜í•˜ê¸° ìœ„í•´ **[reshape2](https://cran.r-project.org/web/packages/reshape2/)** íŒ¨í‚¤ì§€ì˜ `melt` í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•©ë‹ˆë‹¤7.

7 ë‹¤ì–‘í•œ ë°ì´í„° í‘œí˜„ ë°©ì‹ì˜ ê°œë…ê³¼ ì›ë¦¬ì— ëŒ€í•´ì„œëŠ” [13.10ì ˆ](13-chap.html#sec-design-datarep)ì—ì„œ ë” ìì„¸íˆ ë‹¤ë£° ê²ƒì…ë‹ˆë‹¤.

    
    
    library("reshape2")
    genes = melt(Biobase::exprs(x)[selectedProbes, ],
                 varnames = c("probe", "sample"))__

For good measure, we also add a column that provides the gene symbol along
with the probe identifiers.

    
    
    genes$gene =
      names(selectedProbes)[match(genes$probe, selectedProbes)]
    head(genes)__
    
    
           probe  sample    value  gene
    1 1420085_at 1 E3.25 3.027715  Fgf4
    2 1418863_at 1 E3.25 4.843137 Gata4
    3 1425463_at 1 E3.25 5.500618 Gata6
    4 1416967_at 1 E3.25 1.731217  Sox2
    5 1420085_at 2 E3.25 9.293016  Fgf4
    6 1418863_at 2 E3.25 5.530016 Gata4

### 3.6.1 ë§‰ëŒ€ ê·¸ë¦¼(Barplots)

ìš°ë¦¬ ë°ì´í„° í”„ë ˆì„ `genes`ì™€ ê°™ì€ ë°ì´í„°ë¥¼ í‘œì‹œí•˜ëŠ” ì¼ë°˜ì ì¸ ë°©ë²•ì€ ë§‰ëŒ€ ê·¸ë¦¼ì…ë‹ˆë‹¤(ê·¸ë¦¼ 3.15).

    
    
    ggplot(genes, aes(x = gene, y = value)) +
      stat_summary(fun = mean, geom = "bar")__

[![](03-chap_files/figure-html/fig-graphics-
onedbp1-1.png)](03-chap_files/figure-html/fig-graphics-onedbp1-1.png
"FigureÂ 3.15: Barplots showing the means of the distributions of expression
measurements from four probes.")

Figure 3.15: Barplots showing the means of the distributions of expression
measurements from four probes.

ê·¸ë¦¼ 3.15ì—ì„œ ê° ë§‰ëŒ€ëŠ” í•´ë‹¹ ìœ ì „ìì˜ ê°’ì˜ í‰ê· ì„ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ì´ëŸ¬í•œ í”Œë¡¯ì€ ìƒë¬¼ ê³¼í•™ë¿ë§Œ ì•„ë‹ˆë¼ ëŒ€ì¤‘ ë§¤ì²´ì—ì„œë„ í”íˆ ì‚¬ìš©ë©ë‹ˆë‹¤. í•˜ì§€ë§Œ ë°ì´í„°ë¥¼ í‰ê· ì´ë¼ëŠ” ë‹¨ í•˜ë‚˜ì˜ ìˆ«ìë¡œ ìš”ì•½í•˜ë©´ ë§ì€ ì •ë³´ê°€ ì†ì‹¤ë©ë‹ˆë‹¤. ì°¨ì§€í•˜ëŠ” ê³µê°„ì— ë¹„í•´ ë§‰ëŒ€ ê·¸ë¦¼ì€ ë°ì´í„°ë¥¼ ì‹œê°í™”í•˜ëŠ” íš¨ìœ¨ì ì¸ ë°©ë²•ì´ ì•„ë‹™ë‹ˆë‹¤8.

8 ì‹¤ì œë¡œ ë°ì´í„° ë¶„í¬ê°€ ì‹¬í•˜ê²Œ ì¹˜ìš°ì¹˜ê±°ë‚˜(skewed), ë‹¤ë´‰ ë¶„í¬(multimodal)ì´ê±°ë‚˜, í° ì´ìƒì¹˜(outlier)ê°€ ìˆëŠ” ê²½ìš°ì²˜ëŸ¼ í‰ê· ì´ ì ì ˆí•œ ìš”ì•½ í†µê³„ëŸ‰ì´ ì•„ë‹ ë•ŒëŠ” ì´ëŸ¬í•œ ì¢…ë¥˜ì˜ ì‹œê°í™”ê°€ ì™„ì „íˆ ì˜¤í•´ë¥¼ ë¶ˆëŸ¬ì¼ìœ¼í‚¬ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

ê°€ë”ì€ ì˜¤ì°¨ ë§‰ëŒ€(error bars)ë¥¼ ì¶”ê°€í•˜ê³  ì‹¶ì„ ë•Œê°€ ìˆëŠ”ë°, **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì—ì„œ ì´ë¥¼ ìˆ˜í–‰í•˜ëŠ” í•œ ê°€ì§€ ë°©ë²•ì€ ë‹¤ìŒê³¼ ê°™ìŠµë‹ˆë‹¤.

    
    
    library("Hmisc")
    ggplot(genes, aes( x = gene, y = value, fill = gene)) +
      stat_summary(fun = mean, geom = "bar") +
      stat_summary(fun.data = mean_cl_normal, geom = "errorbar",
                   width = 0.25)__

[![](03-chap_files/figure-html/fig-graphics-
onedbp2-1.png)](03-chap_files/figure-html/fig-graphics-onedbp2-1.png
"FigureÂ 3.16: Barplots with error bars indicating standard error of the
mean.")

Figure 3.16: Barplots with error bars indicating standard error of the mean.

ì—¬ê¸°ì„œ ìš°ë¦¬ëŠ” ë‹¤ì‹œ ê³„ì¸µí™”ëœ ê·¸ë˜í”½(layered graphics)ì˜ ì›ì¹™ì„ ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë‘ ê°œì˜ ìš”ì•½ í•¨ìˆ˜ `mean`ê³¼ `mean_cl_normal`, ê·¸ë¦¬ê³  ì´ì™€ ì—°ê´€ëœ ë‘ ê°œì˜ ê¸°í•˜í•™ì  ê°ì²´ `bar`ì™€ `errorbar`ë¥¼ ì‚¬ìš©í•©ë‹ˆë‹¤. `mean_cl_normal` í•¨ìˆ˜ëŠ” **[Hmisc](https://cran.r-project.org/web/packages/Hmisc/)** íŒ¨í‚¤ì§€ì—ì„œ ì œê³µí•˜ë©°, í‰ê· ì˜ í‘œì¤€ ì˜¤ì°¨(ë˜ëŠ” ì‹ ë¢° í•œê³„, **c**onfidence **l** imits)ë¥¼ ê³„ì‚°í•©ë‹ˆë‹¤. ì´ëŠ” ê°„ë‹¨í•œ í•¨ìˆ˜ì´ë¯€ë¡œ, ì›í•œë‹¤ë©´ ê¸°ë³¸ R í‘œí˜„ì‹ì„ ì‚¬ìš©í•˜ì—¬ ì§ì ‘ ê³„ì‚°í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤. ë˜í•œ í”Œë¡¯ì„ ì‹œê°ì ìœ¼ë¡œ ë” ë³´ê¸° ì¢‹ê²Œ ë§Œë“¤ê¸° ìœ„í•´ ë§‰ëŒ€ì— ìƒ‰ìƒì„ ì…í˜”ìŠµë‹ˆë‹¤.

### 3.6.2 ìƒì ê·¸ë¦¼(Boxplots)

ìƒì ê·¸ë¦¼ì€ ë§‰ëŒ€ ê·¸ë¦¼ê³¼ ë¹„ìŠ·í•œ ê³µê°„ì„ ì°¨ì§€í•˜ë©´ì„œë„ í›¨ì”¬ ë” ë§ì€ ì •ë³´ë¥¼ ì œê³µí•©ë‹ˆë‹¤.

    
    
    p = ggplot(genes, aes( x = gene, y = value, fill = gene))
    p + geom_boxplot()__

[![](03-chap_files/figure-html/fig-graphics-
onedboxpl-1.png)](03-chap_files/figure-html/fig-graphics-onedboxpl-1.png
"FigureÂ 3.17: Boxplots.")

Figure 3.17: Boxplots.

ê·¸ë¦¼ 3.17ì—ì„œ ë‘ ìœ ì „ì(Gata4, Gata6)ëŠ” ë¶„í¬ê°€ ë¹„êµì  ì§‘ì¤‘ë˜ì–´ ìˆìœ¼ë©°, ë†’ì€ ê°’ ë°©í–¥ìœ¼ë¡œ ëª‡ ê°œì˜ ë°ì´í„° í¬ì¸íŠ¸ë§Œ ë²—ì–´ë‚˜ ìˆìŒì„ ì•Œ ìˆ˜ ìˆìŠµë‹ˆë‹¤. Fgf4ì˜ ê²½ìš° ë¶„í¬ê°€ ì˜¤ë¥¸ìª½ìœ¼ë¡œ ì¹˜ìš°ì³(right-skewed) ìˆìŠµë‹ˆë‹¤. ìƒì ì•ˆì˜ ê²€ì€ ìˆ˜í‰ì„ ìœ¼ë¡œ í‘œì‹œëœ ì¤‘ì•™ê°’ì´ ìƒìì˜ ì•„ë˜ìª½(ë˜ëŠ” ì™¼ìª½)ì— ë” ê°€ê¹ê¸° ë•Œë¬¸ì…ë‹ˆë‹¤. ë°˜ëŒ€ë¡œ Sox2ì˜ ë¶„í¬ëŠ” ì™¼ìª½ìœ¼ë¡œ ì¹˜ìš°ì³(left-skewed) ìˆìŠµë‹ˆë‹¤.

### 3.6.3 ì  ê·¸ë¦¼(Dot plots)ê³¼ ë²Œë–¼ ê·¸ë¦¼(Beeswarm plots)

ë°ì´í„° í¬ì¸íŠ¸ì˜ ìˆ˜ê°€ ë„ˆë¬´ ë§ì§€ ì•Šë‹¤ë©´, ì•ì„œ ë³¸ ìš”ì•½ ë°©ì‹ ëŒ€ì‹  ë°ì´í„° í¬ì¸íŠ¸ë¥¼ ì§ì ‘ ë³´ì—¬ì£¼ëŠ” ê²ƒì´ ê°€ëŠ¥í•˜ë©° ê¶Œì¥ë˜ëŠ” ë°©ë²•ì…ë‹ˆë‹¤. í•˜ì§€ë§Œ ë°ì´í„°ë¥¼ ì§ì ‘ í”Œë¡œíŒ…í•˜ë©´ ì ë“¤ì´ ì„œë¡œ ê²¹ì¹˜ëŠ”(overlapping) ê²½ìš°ê°€ ë§ì•„ ì‹œê°ì ìœ¼ë¡œ ì¢‹ì§€ ì•Šê±°ë‚˜ ë°ì´í„°ë¥¼ ê°€ë¦´ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì ë“¤ì´ ê²¹ì¹˜ì§€ ì•Šìœ¼ë©´ì„œ ì›ë˜ ìœ„ì¹˜ì— ìµœëŒ€í•œ ê°€ê¹ê²Œ ë°°ì¹˜ë˜ë„ë¡ ì‹œë„í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤([Wilkinson 1999](16-chap.html#ref-Wilkinson:DotPlots:1999)).

    
    
    p + geom_dotplot(binaxis = "y", binwidth = 1/6,
           stackdir = "center", stackratio = 0.75,
           aes(color = gene))
    library("ggbeeswarm")
    p + geom_beeswarm(aes(color = gene))__

[![](03-chap_files/figure-html/fig-graphics-
oneddot-1.png)](03-chap_files/figure-html/fig-graphics-oneddot-1.png
"FigureÂ 3.18Â \(a\): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
oneddot-2.png)](03-chap_files/figure-html/fig-graphics-oneddot-2.png
"FigureÂ 3.18Â \(b\): ")

(b)

Figure 3.18: (a) **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì˜ `geom_dotplot`ì„ ì‚¬ìš©í•´ ë§Œë“  ì  ê·¸ë¦¼. (b) **[ggbeeswarm](https://cran.r-project.org/web/packages/ggbeeswarm/)** íŒ¨í‚¤ì§€ì˜ `geom_beeswarm`ì„ ì‚¬ìš©í•´ ë§Œë“  ë²Œë–¼ ê·¸ë¦¼.

ê·¸ë¦¼ 3.18ì˜ (a) íŒ¨ë„ì— ìˆëŠ” í”Œë¡¯ì€ ì ë“¤ì˜ \(y\) ì¢Œí‘œë¥¼ êµ¬ê°„(bin)ìœ¼ë¡œ ì´ì‚°í™”í•˜ê³ (ìœ„ì—ì„œëŠ” êµ¬ê°„ í¬ê¸°ë¥¼ 1/6ë¡œ ì„ íƒí•¨), ê·¸ ë‹¤ìŒ ì„œë¡œ ì˜†ìœ¼ë¡œ ìŒ“ì•„ ì˜¬ë¦° ê²ƒì…ë‹ˆë‹¤.

**[ggbeeswarm](https://cran.r-project.org/web/packages/ggbeeswarm/)** íŒ¨í‚¤ì§€ì˜ `geom_beeswarm`ì„ ì‚¬ìš©í•˜ëŠ” ëŒ€ì•ˆë„ ìˆìŠµë‹ˆë‹¤. ì´ í”Œë¡¯ì€ ê·¸ë¦¼ 3.18ì˜ (b) íŒ¨ë„ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤. ë°°ì¹˜ ì•Œê³ ë¦¬ì¦˜ì€ ì ë“¤ ì‚¬ì´ì˜ ê²¹ì¹¨ì„ í”¼í•˜ëŠ” ê²ƒì„ ëª©í‘œë¡œ í•©ë‹ˆë‹¤. ì ì´ ê¸°ì¡´ ì ê³¼ ê²¹ì¹˜ê²Œ ë˜ë©´, ê²¹ì¹¨ì„ í”¼í•  ìˆ˜ ìˆì„ ë§Œí¼ë§Œ \(x\)ì¶• ë°©í–¥ìœ¼ë¡œ ì¡°ê¸ˆ ì´ë™ì‹œí‚µë‹ˆë‹¤. ì  ê·¸ë¦¼ì´ë‚˜ ë²Œë–¼ ê·¸ë¦¼ì„ ë³´ê¸° ì¢‹ê²Œ ë§Œë“¤ë ¤ë©´ ëŒ€ê°œ ìƒˆë¡œìš´ ë°ì´í„° ì„¸íŠ¸ë§ˆë‹¤ ë°°ì¹˜ ë§¤ê°œë³€ìˆ˜ë¥¼ ì•½ê°„ì”© ì¡°ì •í•´ì•¼ í•©ë‹ˆë‹¤.

### 3.6.4 ë°€ë„ ê·¸ë¦¼(Density plots)

ë™ì¼í•œ ë°ì´í„°ë¥¼ ë‚˜íƒ€ë‚´ëŠ” ë˜ ë‹¤ë¥¸ ë°©ë²•ì€ ë°€ë„ ê·¸ë¦¼ì…ë‹ˆë‹¤. ì—¬ê¸°ì„œëŠ” ë°ì´í„° í¬ì¸íŠ¸ë¥¼ í‰í™œí™”(smoothing)í•˜ì—¬ ê¸°ì €ì˜ ë°ì´í„° ìƒì„± ë°€ë„ë¥¼ ì¶”ì •í•˜ë ¤ê³  ì‹œë„í•©ë‹ˆë‹¤(ê·¸ë¦¼ 3.19).

    
    
    ggplot(genes, aes( x = value, color = gene)) + geom_density()__

[![](03-chap_files/figure-html/fig-graphics-
oneddens-1.png)](03-chap_files/figure-html/fig-graphics-oneddens-1.png
"FigureÂ 3.19: Density plots.")

Figure 3.19: ë°€ë„ ê·¸ë¦¼.

ê·¸ë¦¼ 3.17â€”3.19ì—ì„œ ë³¼ ìˆ˜ ìˆë“¯ì´, ìƒì ê·¸ë¦¼ì€ ë‹¨ë´‰(unimodal) ë°ì´í„°ì—ëŠ” ê½¤ ì˜ ì‘ë™í•˜ì§€ë§Œ ë°ì´í„° ë¶„í¬ê°€ ë‹¤ë´‰(multimodal)ì¼ ê²½ìš°ì—ëŠ” ì˜¤í•´ë¥¼ ë¶ˆëŸ¬ì¼ìœ¼í‚¬ ìˆ˜ ìˆìœ¼ë©°, ê¸´ ê¼¬ë¦¬(long-tailed) ë¶„í¬ì˜ íŠ¹ì„±ì„ í•­ìƒ ê³µì •í•˜ê²Œ ë³´ì—¬ì£¼ì§€ëŠ” ëª»í•©ë‹ˆë‹¤. ë°ì´í„° í¬ì¸íŠ¸ë‚˜ ê·¸ ë°€ë„ë¥¼ ì§ì ‘ ë³´ì—¬ì£¼ëŠ” ê·¸ë¦¼ 3.18â€”3.19ëŠ” ì´ëŸ¬í•œ íŠ¹ì§•ì„ ë” ì˜ ë³´ì—¬ì¤ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ Fgf4ì™€ Sox2ì˜ ë°ì´í„°ëŠ” ì´ë´‰(bimodal) ë¶„í¬ì„ì„ ì•Œ ìˆ˜ ìˆê³ , Gata4ì™€ Gata6ì€ ëŒ€ë¶€ë¶„ì˜ ê°’ì´ ê¸°ì¤€ì„  ê·¼ì²˜ì— ëª¨ì—¬ ìˆì§€ë§Œ ì¼ë¶€ ì„¸í¬ë“¤ì€ ë„“ì€ ê°’ ë²”ìœ„ì— ê±¸ì³ ë°œí˜„ ìˆ˜ì¤€ì´ ë†’ì•„ì ¸ ìˆìŒì„ ì•Œ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

ê·¸ëŸ¬ë‚˜ ë°€ë„ ì¶”ì •ì—ëŠ” ì—¬ëŸ¬ ë³µì¡í•œ ë¬¸ì œê°€ ìˆìŠµë‹ˆë‹¤. íŠ¹íˆ í‰í™œ ì°½(smoothing window)ì˜ í¬ê¸°ë¥¼ ì„ íƒí•´ì•¼ í•©ë‹ˆë‹¤. ë°ì´í„°ê°€ ë°€ì§‘ëœ ì˜ì—­ì˜ ë´‰ìš°ë¦¬ë¥¼ í¬ì°©í•  ë§Œí¼ ì‘ì€ ì°½ í¬ê¸°ë¥¼ ì„ íƒí•˜ë©´ ë‹¤ë¥¸ ì˜ì—­ì—ì„œëŠ” ë¶ˆì•ˆì •í•œ("êµ¬ë¶ˆêµ¬ë¶ˆí•œ") ì¶”ì •ì¹˜ê°€ ë‚˜ì˜¬ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë°˜ë©´ ì°½ì„ ë„ˆë¬´ í¬ê²Œ ì¡ìœ¼ë©´ ë¾°ì¡±í•œ ë´‰ìš°ë¦¬ì™€ ê°™ì€ ë°€ë„ì˜ ëšœë ·í•œ íŠ¹ì§•ë“¤ì´ ë­‰ê°œì§ˆ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë˜í•œ ë°€ë„ ì„ ì€ ë°€ë„ë¥¼ ì¶”ì •í•˜ëŠ” ë° ì–¼ë§ˆë‚˜ ë§ì€ ë°ì´í„°ê°€ ì‚¬ìš©ë˜ì—ˆëŠ”ì§€ì— ëŒ€í•œ ì •ë³´ë¥¼ ì „ë‹¬í•˜ì§€ ì•Šìœ¼ë¯€ë¡œ, ê·¸ë¦¼ 3.19ì™€ ê°™ì€ í”Œë¡¯ì€ ê³¡ì„ ë“¤ ê°„ì˜ ìƒ˜í”Œ í¬ê¸°ê°€ ë‹¤ë¥¼ ê²½ìš° íŠ¹íˆ ë¬¸ì œê°€ ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

### 3.6.5 ë°”ì´ì˜¬ë¦° ê·¸ë¦¼(Violin plots)

ê·¸ë¦¼ 3.19ì™€ ê°™ì€ ë°€ë„ ê·¸ë¦¼ì€ ë°€ë„ê°€ ì—¬ëŸ¬ ê°œì¼ ê²½ìš° ë³µì¡í•´ ë³´ì¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. í•œ ê°€ì§€ ì•„ì´ë””ì–´ëŠ” ìƒì ê·¸ë¦¼ì—ì„œ ì˜ê°ì„ ì–»ì€ ë ˆì´ì•„ì›ƒìœ¼ë¡œ ë°€ë„ë¥¼ ë°°ì¹˜í•˜ëŠ” ê²ƒì¸ë°, ì´ê²ƒì´ ë°”ë¡œ ë°”ì´ì˜¬ë¦° ê·¸ë¦¼ì…ë‹ˆë‹¤(ê·¸ë¦¼ 3.20). ì—¬ê¸°ì„œëŠ” ë°€ë„ ì¶”ì •ì¹˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ëŒ€ì¹­ì ì¸ ëª¨ì–‘ì„ ê·¸ë¦¬ëŠ”ë°, ì´ ëª¨ì–‘ì´ ë•Œë•Œë¡œ ì‹œê°ì ìœ¼ë¡œ ë°”ì´ì˜¬ë¦°ì„ ì—°ìƒì‹œí‚µë‹ˆë‹¤.

    
    
    p + geom_violin()__

[![](03-chap_files/figure-html/fig-graphics-
onedviolin-1.png)](03-chap_files/figure-html/fig-graphics-onedviolin-1.png
"FigureÂ 3.20: Violin plots.")

Figure 3.20: ë°”ì´ì˜¬ë¦° ê·¸ë¦¼.

### 3.6.6 ëŠ¥ì„  ê·¸ë¦¼(Ridgeline plots)

ë°€ë„ ê·¸ë¦¼ì˜ ë˜ ë‹¤ë¥¸ ë³€í˜•ì€ ëŠ¥ì„  ê·¸ë¦¼ì…ë‹ˆë‹¤(ê·¸ë¦¼ 3.21).

    
    
    library("ggridges")
    ggplot(genes, aes(x = value, y = gene, fill = gene)) + 
      geom_density_ridges()__

[![](03-chap_files/figure-html/fig-graphics-
onedridge4-1.png)](03-chap_files/figure-html/fig-graphics-onedridge4-1.png
"FigureÂ 3.21: Ridgeline plots.")

Figure 3.21: ëŠ¥ì„  ê·¸ë¦¼.

ì´ëŸ¬í•œ ë°©ì‹ì˜ í‘œì‹œëŠ” í‘œì‹œí•´ì•¼ í•  ë°€ë„ì˜ ìˆ˜ê°€ ìˆ˜ì‹­ ê°œì— ë‹¬í•  ë•Œ ê°€ì¥ ì í•©í•©ë‹ˆë‹¤.

    
    
    top42 = order(rowMeans(Biobase::exprs(x)), decreasing = TRUE)[1:42]
    g42 = melt(Biobase::exprs(x)[rev(top42), ], varnames = c("probe", "sample"))
    ggplot(g42, aes(x = value, y = probe)) __

[![](03-chap_files/figure-html/fig-graphics-
onedridge42-1.png)](03-chap_files/figure-html/fig-graphics-onedridge42-1.png
"FigureÂ 3.22: Like FigureÂ fig-graphics-onedridge4, with more genes.")

ê·¸ë¦¼ 3.22: ê·¸ë¦¼ 3.21ê³¼ ìœ ì‚¬í•˜ì§€ë§Œ ë” ë§ì€ ìœ ì „ìë¥¼ í¬í•¨í•˜ê³  ìˆìŠµë‹ˆë‹¤.

### 3.6.7 ECDF ê·¸ë¦¼

1ì°¨ì› í™•ë¥  ë³€ìˆ˜ \(X\)ì˜ ë¶„í¬ë¥¼ ì„¤ëª…í•˜ëŠ” ìˆ˜í•™ì ìœ¼ë¡œ ê°€ì¥ í¸ë¦¬í•œ ë°©ë²•ì€ **ëˆ„ì  ë¶„í¬ í•¨ìˆ˜(cumulative distribution function, CDF)**ì…ë‹ˆë‹¤. ì¦‰, ë‹¤ìŒê³¼ ê°™ì´ ì •ì˜ë˜ëŠ” í•¨ìˆ˜ì…ë‹ˆë‹¤.

\[ F(x) = P(X \le x), \tag{3.1}\]

ì—¬ê¸°ì„œ \(x\)ëŠ” ì‹¤ìˆ˜ì¶• ì „ì²´ì˜ ê°’ì„ ê°€ì§‘ë‹ˆë‹¤. \(X\)ì˜ ë°€ë„ëŠ” \(F\)ê°€ ì¡´ì¬í•  ê²½ìš° \(F\)ì˜ ë„í•¨ìˆ˜ì…ë‹ˆë‹¤9. ì‹ 3.1ì˜ í™•ë¥ ì— ëŒ€í•œ ìœ í•œ í‘œë³¸ ë²„ì „ì€ **ê²½í—˜ì  ëˆ„ì  ë¶„í¬ í•¨ìˆ˜(empirical cumulative distribution function, ECDF)**ë¼ê³  í•©ë‹ˆë‹¤.

9 ì •ì˜ì— ë”°ë¼ \(F\)ëŠ” \(x\)ê°€ ì‘ì•„ì§ˆìˆ˜ë¡ 0ì— ìˆ˜ë ´í•˜ê³ (\(x \to -\infty\)), \(x\)ê°€ ì»¤ì§ˆìˆ˜ë¡ 1ì— ìˆ˜ë ´í•©ë‹ˆë‹¤(\(x \to +\infty\)).

\[ F_{n}(x) = \frac{\text{\(x_i \le x\)ë¥¼ ë§Œì¡±í•˜ëŠ” \(i\)ì˜ ê°œìˆ˜}}{n} = \frac{1}{n}\sum_{i=1}^n ğŸ™({x \le x_i}), \tag{3.2}\]

ì—¬ê¸°ì„œ \(x_1, ..., x_n\)ì€ \(X\)ì—ì„œ ì¶”ì¶œí•œ \(n\)ê°œì˜ í‘œë³¸ì„ ë‚˜íƒ€ë‚´ê³ , \(ğŸ™\)ì€ ì¸ìˆ˜ì˜ í‘œí˜„ì‹ì´ ì°¸ì´ë©´ 1, ì•„ë‹ˆë©´ 0ì˜ ê°’ì„ ê°–ëŠ” ì§€ì‹œ í•¨ìˆ˜(indicator function)ì…ë‹ˆë‹¤. ì´ ì„¤ëª…ì´ ì¶”ìƒì ìœ¼ë¡œ ë“¤ë¦°ë‹¤ë©´ ë‹¤ìŒ ì˜ˆì œ(ê·¸ë¦¼ 3.23)ë¥¼ í†µí•´ ì¢€ ë” ì§ê´€ì ìœ¼ë¡œ ì´í•´í•  ìˆ˜ ìˆì„ ê²ƒì…ë‹ˆë‹¤.

    
    
    simdata = rnorm(70)
    tibble(index = seq(along = simdata),
              sx = sort(simdata)) %>%
    ggplot(aes(x = sx, y = index)) + geom_step()__

[![](03-chap_files/figure-html/fig-graphics-
ecdfexample-1.png)](03-chap_files/figure-html/fig-graphics-ecdfexample-1.png
"FigureÂ 3.23: Sorted values of simdata versus their index. This is the
empirical cumulative distribution function of simdata.")

Figure 3.23: Sorted values of `simdata` versus their index. This is the
empirical cumulative distribution function of `simdata`.

Plotting the sorted values against their ranks gives the essential features of
the ECDF (Figure 3.23). In practice, we do not need to do the sorting and the
other steps in the above code manually and will rather use the `stat_ecdf()`
geometric object. The ECDFs of our data are shown in Figure 3.24.

    
    
    ggplot(genes, aes( x = value, color = gene)) + stat_ecdf()__

[![](03-chap_files/figure-html/fig-graphics-
onedecdf-1.png)](03-chap_files/figure-html/fig-graphics-onedecdf-1.png
"FigureÂ 3.24: Empirical cumulative distribution functions \(ECDF\).")

Figure 3.24: Empirical cumulative distribution functions (ECDF).

ECDFëŠ” ëª‡ ê°€ì§€ ìœ ìš©í•œ ì†ì„±ì„ ê°€ì§€ê³  ìˆìŠµë‹ˆë‹¤:

  * ë¬´ì†ì‹¤(Lossless): ECDF \\(F_{n}(x)\\)ëŠ” ê°’ì˜ ìˆœì„œ(ì¤‘ìš”í•˜ì§€ ì•Šë‹¤ê³  ê°€ì •í•¨)ë¥¼ ì œì™¸í•˜ê³  ì›ë˜ í‘œë³¸ \\(x_1,...,x_n\\)ì— í¬í•¨ëœ ëª¨ë“  ì •ë³´ë¥¼ ìœ ì§€í•©ë‹ˆë‹¤.

  * \\(n\\)ì´ ì»¤ì§ì— ë”°ë¼ ECDF \\(F_{n}(x)\\)ëŠ” ì‹¤ì œ CDF \\(F(x)\\)ë¡œ ìˆ˜ë ´í•©ë‹ˆë‹¤. ì œí•œëœ í‘œë³¸ í¬ê¸° \\(n\\)ì˜ ê²½ìš°ì—ë„ ë‘ í•¨ìˆ˜ ì‚¬ì´ì˜ ì°¨ì´ëŠ” ì‘ì€ ê²½í–¥ì´ ìˆìŠµë‹ˆë‹¤. ì´ëŠ” ê²½í—˜ì  ë°€ë„(empirical density)ì˜ ê²½ìš°ì—ëŠ” í•´ë‹¹ë˜ì§€ ì•ŠëŠ”ë‹¤ëŠ” ì ì— ìœ ì˜í•˜ì„¸ìš”! í‰í™œí™”(smoothing)ê°€ ì—†ìœ¼ë©´ ìœ í•œ í‘œë³¸ì˜ ê²½í—˜ì  ë°€ë„ëŠ” ë””ë™ ë¸íƒ€ í•¨ìˆ˜(Dirac delta functions)ì˜ í•©ìœ¼ë¡œ ë‚˜íƒ€ë‚˜ë©°, ì´ëŠ” ì‹œê°í™”í•˜ê¸° ì–´ë µê³  ê¸°ì €ì˜ ë§¤ë„ëŸ¬ìš´ ì‹¤ì œ ë°€ë„ì™€ëŠ” ìƒë‹¹íˆ ë‹¤ë¦…ë‹ˆë‹¤. í‰í™œí™”ë¥¼ ì‚¬ìš©í•˜ë©´ ê·¸ ì°¨ì´ê°€ ëœ ë‘ë“œëŸ¬ì§ˆ ìˆ˜ ìˆì§€ë§Œ, ì•ì„œ ë…¼ì˜í•œ ê²ƒì²˜ëŸ¼ ì´ë¥¼ ì œì–´í•˜ê¸°ëŠ” ì–´ë µìŠµë‹ˆë‹¤.

[![](imgs/Lawrence-TCGA-Nature-2013-Fig1.png)](imgs/Lawrence-TCGA-
Nature-2013-Fig1.png "ê·¸ë¦¼ 3.25: Lawrence ë“±(2013)ì˜ ê·¸ë¦¼ 1ì˜ ì¼ë¶€ì…ë‹ˆë‹¤. ê° ì ì€ ì¢…ì–‘-ì •ìƒ ìŒì— ëŒ€ì‘í•˜ë©°, ìˆ˜ì§ ìœ„ì¹˜ëŠ” ì—‘ì†œ(exome) ë‚´ ì²´ì„¸í¬ ëŒì—°ë³€ì´ì˜ ì´ ë¹ˆë„ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ê²°ê³¼ ê³¡ì„ ì€ ë³¸ì§ˆì ìœ¼ë¡œ ECDF ê·¸ë¦¼ì´ë©°, ê°œë…ì ìœ¼ë¡œ ì´ í”Œë¡¯ì€ ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ ê·¸ë˜í”„ê°€ 90ë„ íšŒì „(ì¦‰, xì¶•ê³¼ yì¶•ì˜ ì—­í• ì´ ë°”ë€œ)ë˜ì–´ ìˆê³  ê°œë³„ ì¢…ì–‘ ìœ í˜•ì— ëŒ€í•œ ê³¡ì„ ì´ ì„œë¡œ ë” ì˜ ë¶„ë¦¬ë˜ë„ë¡ ìˆ˜í‰ìœ¼ë¡œ ì´ë™ë˜ì–´ ìˆë‹¤ëŠ” ì ì´ ë‹¤ë¦…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.25: Lawrence ë“± ([2013](16-chap.html#ref-Lawrence:Nature:2013))ì˜ ê·¸ë¦¼ 1ì˜ ì¼ë¶€. ê° ì ì€ ì¢…ì–‘-ì •ìƒ ìŒì— ëŒ€ì‘í•˜ë©°, ìˆ˜ì§ ìœ„ì¹˜ëŠ” ì—‘ì†œ(exome) ë‚´ ì²´ì„¸í¬ ëŒì—°ë³€ì´ì˜ ì´ ë¹ˆë„ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ê²°ê³¼ ê³¡ì„ ì€ ë³¸ì§ˆì ìœ¼ë¡œ ECDF ê·¸ë¦¼ì´ë©°, ê°œë…ì ìœ¼ë¡œ ì´ í”Œë¡¯ì€ ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ ê·¸ë˜í”„ê°€ 90ë„ íšŒì „(ì¦‰, \\(x\\)\-ì™€ \\(y\\)-ì¶•ì˜ ì—­í• ì´ ë°”ë€œ)ë˜ì–´ ìˆê³  ê°œë³„ ì¢…ì–‘ ìœ í˜•ì— ëŒ€í•œ ê³¡ì„ ì´ ì„œë¡œ ë” ì˜ ë¶„ë¦¬ë˜ë„ë¡ ìˆ˜í‰ìœ¼ë¡œ ì´ë™ë˜ì–´ ìˆë‹¤ëŠ” ì ì´ ë‹¤ë¦…ë‹ˆë‹¤.

__

íƒœìŠ¤í¬

**í‹°ë¸”(tibbles).** ìœ„ ì½”ë“œì—ì„œ ì²˜ìŒìœ¼ë¡œ `tibble`ì„ ì ‘í–ˆìŠµë‹ˆë‹¤. **[tibble](https://cran.r-project.org/web/packages/tibble/)** íŒ¨í‚¤ì§€ì˜ ë¹„ë„¤íŠ¸(vignette)ë¥¼ ì‚´í´ë³´ê³  ê·¸ ê¸°ëŠ¥ì„ í™•ì¸í•´ ë³´ì„¸ìš”.

### 3.6.8 ë°€ë„ì— ëŒ€í•œ ë³€í™˜ì˜ ì˜í–¥

íˆìŠ¤í† ê·¸ë¨ì´ë‚˜ ë°€ë„ ê·¸ë¦¼ì„ ë³´ê³  ì¼ë¶€ ê¸°ì €ì˜ ìƒë¬¼í•™ì  í˜„ìƒì˜ ì§€í‘œë¡œì„œ **ì´ë´‰ì„±(bimodality)**(ë˜ëŠ” ë‹¤ë´‰ì„±)ì˜ ì¦ê±°ë¥¼ ì¡°ì‚¬í•˜ê³  ì‹¶ì€ ìœ í˜¹ì´ ìƒê¸¸ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ê·¸ë ‡ê²Œ í•˜ê¸° ì „ì—, ë°€ë„ì˜ ìµœë¹ˆê°’(mode)ì˜ ìˆ˜ëŠ” **ì—°ì‡„ ë²•ì¹™(chain rule)**ì„ í†µí•´ ë°ì´í„°ê°€ ë³€í™˜ë˜ëŠ” ì²™ë„ì— ë”°ë¼ ë‹¬ë¼ì§„ë‹¤ëŠ” ì ì„ ê¸°ì–µí•˜ëŠ” ê²ƒì´ ì¤‘ìš”í•©ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´, Hiiragi ë°ì´í„° ì„¸íŠ¸ì˜ ì–´ë ˆì´ ì¤‘ í•˜ë‚˜ì—ì„œ ì–»ì€ ë°ì´í„°ë¥¼ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤(ê·¸ë¦¼ 3.26).

    
    
    ggplot(dfx, aes(x = `64 E4.5 (EPI)`)) + geom_histogram(bins = 100)
    ggplot(dfx, aes(x = 2 ^ `64 E4.5 (EPI)`)) + 
      geom_histogram(binwidth = 20) + xlim(0, 1500)__

[![](03-chap_files/figure-html/fig-graphics-
onedtrsf-1.png)](03-chap_files/figure-html/fig-graphics-onedtrsf-1.png
"ê·¸ë¦¼ 3.26 (a): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
onedtrsf-2.png)](03-chap_files/figure-html/fig-graphics-onedtrsf-2.png
"ê·¸ë¦¼ 3.26 (b): ")

(b)

ê·¸ë¦¼ 3.26: ë¡œê·¸ ë³€í™˜ ì—¬ë¶€ì— ë”°ë¥¸ ë™ì¼í•œ ë°ì´í„°ì˜ íˆìŠ¤í† ê·¸ë¨. (a) ë°ì´í„°ëŠ” ë°ì´í„° ê°ì²´ `x`ì— ì €ì¥ëœ ì²™ë„ë¡œ í‘œì‹œë˜ë©°, ì´ëŠ” ë§ˆì´í¬ë¡œì–´ë ˆì´ í˜•ê´‘ ê°•ë„ë¥¼ ë¡œê·¸(ë°‘ 2) ë³€í™˜í•œ ê²°ê³¼ì…ë‹ˆë‹¤([Irizarry et al. 2003](16-chap.html#ref-Irizarry:Biostat:2003)). (b) ë‹¤ì‹œ ì§€ìˆ˜í™”í•˜ì—¬ ì›ë˜ì˜ í˜•ê´‘ ì²™ë„ë¡œ ë˜ëŒë¦° í›„ì˜ ëª¨ìŠµì…ë‹ˆë‹¤. ê³µê°„ì„ ë” ì˜ í™œìš©í•˜ê¸° ìœ„í•´ \(x\)ì¶•ì˜ ë²”ìœ„ë¥¼ 1500ìœ¼ë¡œ ì œí•œí–ˆìŠµë‹ˆë‹¤.

__

ì§ˆë¬¸ 3.5

(ì‹¬í™”:) í™•ë¥  ë³€ìˆ˜ \\(X\\)ì™€, ë³€í™˜ëœ í™•ë¥  ë³€ìˆ˜ \\(Y = f(X)\\)ë¥¼ ì •ì˜í•˜ëŠ” ë¹„ì„ í˜• 1:1 ë³€í™˜ \\(f: x \mapsto y\\)ë¥¼ ìƒê°í•´ ë´…ì‹œë‹¤. \\(Y\\)ì˜ ë°€ë„ í•¨ìˆ˜ë¥¼ \\(p(y)\\)ë¼ê³  ê°€ì •í•  ë•Œ, \\(X\\)ì˜ ë°€ë„ëŠ” ë¬´ì—‡ì¼ê¹Œìš”? \\(X\\)ì˜ ìµœë¹ˆê°’(ë˜ëŠ” ìµœë¹ˆê°’ë“¤)ì€ \\(Y\\)ì˜ ìµœë¹ˆê°’(ë“¤)ê³¼ ì–´ë–¤ ê´€ë ¨ì´ ìˆì„ê¹Œìš”?  
íŒíŠ¸: í•¨ìˆ˜ \\(p\\)ì˜ ìµœë¹ˆê°’ì€ ê·¸ ë„í•¨ìˆ˜ \\(p'=dp/dx\\)ì˜ ê·¼(root)ì´ë¼ëŠ” ì ì— ìœ ì˜í•˜ì„¸ìš”. \\(x_0\\)ì´ \\(X\\)ì˜ ìµœë¹ˆê°’ì´ë¼ë©´ \\(y_0=f(x_0)\\)ì´ \\(Y\\)ì˜ ìµœë¹ˆê°’ì´ë¼ëŠ” ê²ƒì´ ì¼ë°˜ì ìœ¼ë¡œ ì„±ë¦½í• ê¹Œìš”?

__

í•´ê²°ì±…

__

ì—°ì‡„ ë²•ì¹™ì— ë”°ë¥´ë©´ \\(p(y)\, dy = p(f(x))\,f'(x)\,dx\\)ì´ë¯€ë¡œ, \\(X\\)ì˜ ë°€ë„ëŠ” \\(\tilde{p}(x) = p(f(x))\,f'(x)\\)ì…ë‹ˆë‹¤. \\(\tilde{p}\\)ì˜ ìµœë¹ˆê°’ì€ ê·¸ ë„í•¨ìˆ˜ \\(d\tilde{p}/dx\\)ì˜ ê·¼ì´ë©°, ì¦‰ ì´ë“¤ì€ \\(p'(f(x))\,f'^2(x) + p(f(x))f"(x) = 0\\)ì„ ë”°ë¦…ë‹ˆë‹¤. í•©ì˜ ë‘ ë²ˆì§¸ í•­ì€ \\(f\\)ê°€ ì•„í•€ ì„ í˜•(affine linear, \\(f"\equiv0\\))ì¸ ê²½ìš° ì‚¬ë¼ì§€ì§€ë§Œ, ì¼ë°˜ì ìœ¼ë¡œ ë‘ ë°€ë„ì˜ ê·¼ ì‚¬ì´ì—ëŠ”, ë”°ë¼ì„œ \\(X\\)ì™€ \\(Y\\)ì˜ ìµœë¹ˆê°’ ì‚¬ì´ì—ëŠ” ê°„ë‹¨í•œ ê´€ê³„ê°€ ì—†ìŠµë‹ˆë‹¤.

## 3.7 2ì°¨ì› ë°ì´í„° ì‹œê°í™”: ì‚°ì ë„(Scatterplots)

ì‚°ì ë„ëŠ” ì²˜ë¦¬-ë°˜ì‘ ë¹„êµ(ê·¸ë¦¼ 3.3ì—ì„œì²˜ëŸ¼), ë³€ìˆ˜ ê°„ì˜ ì—°ê´€ì„±(ê·¸ë¦¼ 3.10ì—ì„œì²˜ëŸ¼), ë˜ëŠ” ìŒì„ ì´ë£¬ ë°ì´í„°(ì˜ˆ: ì¹˜ë£Œ ì „í›„ ì—¬ëŸ¬ í™˜ìì˜ ì§ˆë³‘ ë°”ì´ì˜¤ë§ˆì»¤)ë¥¼ ì‹œê°í™”í•˜ëŠ” ë° ìœ ìš©í•©ë‹ˆë‹¤. ë‘ ë³€ìˆ˜ë¥¼ ë‚˜íƒ€ë‚´ê¸° ìœ„í•´ í”Œë¡œíŒ… ì¢…ì´ë‚˜ í™”ë©´ì˜ ë‘ ì°¨ì›ì„ ì‚¬ìš©í•©ë‹ˆë‹¤. ì•¼ìƒí˜•ê³¼ FGF4-KO ìƒ˜í”Œ ì‚¬ì´ì˜ ì°¨ë“± ë°œí˜„ì„ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤.

    
    
    scp = ggplot(dfx, aes(x = `59 E4.5 (PE)` ,
                          y = `92 E4.5 (FGF4-KO)`))
    scp + geom_point()__

[![](03-chap_files/figure-html/fig-graphics-
twodsp1-1.png)](03-chap_files/figure-html/fig-graphics-twodsp1-1.png
"ê·¸ë¦¼ 3.27: ë‘ ìƒ˜í”Œì— ëŒ€í•œ 45,101ê°œ ë°œí˜„ ì¸¡ì •ê°’ì˜ ì‚°ì ë„ì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.27: ë‘ ìƒ˜í”Œì— ëŒ€í•œ 45,101ê°œ ë°œí˜„ ì¸¡ì •ê°’ì˜ ì‚°ì ë„.

ë ˆì´ë¸” 59 E4.5 (PE)ì™€ 92 E4.5 (FGF4-KO)ëŠ” ìœ„ì—ì„œ ë§Œë“  ë°ì´í„° í”„ë ˆì„ `dfx`ì˜ ì—´ ì´ë¦„(ìƒ˜í”Œ ì´ë¦„)ì„ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ì´ë“¤ì€ íŠ¹ìˆ˜ ë¬¸ì(ê³µë°±, ê´„í˜¸, í•˜ì´í”ˆ)ë¥¼ í¬í•¨í•˜ê³  ìˆ«ìë¡œ ì‹œì‘í•˜ê¸° ë•Œë¬¸ì—, Rì´ êµ¬ë¬¸ì ìœ¼ë¡œ ì´í•´í•  ìˆ˜ ìˆë„ë¡ ì—­ë”°ì˜´í‘œ(backtick)ë¡œ ë¬¶ì–´ì¤„ í•„ìš”ê°€ ìˆìŠµë‹ˆë‹¤. í”Œë¡¯ì€ ê·¸ë¦¼ 3.27ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤. ë°€ì§‘ëœ ì  êµ¬ë¦„(point cloud)ì„ ì–»ê²Œ ë˜ëŠ”ë°, êµ¬ë¦„ì˜ ì™¸ê³½ ë¶€ë¶„ì€ í•´ì„ì„ ì‹œë„í•´ ë³¼ ìˆ˜ ìˆê² ì§€ë§Œ, í”Œë¡¯ì˜ ë°€ì§‘ëœ ì˜ì—­ ë‚´ì—ì„œ ë°ì´í„°ê°€ ì–´ë–»ê²Œ ë¶„í¬ë˜ì–´ ìˆëŠ”ì§€ëŠ” ì‹œê°ì ìœ¼ë¡œ ì•Œ ë°©ë²•ì´ ì—†ìŠµë‹ˆë‹¤.

ê²¹ì³ê·¸ë¦¬ê¸°(overplotting) ë¬¸ì œë¥¼ ê°œì„ í•˜ëŠ” í•œ ê°€ì§€ ì‰¬ìš´ ë°©ë²•ì€ `geom_point`ì˜ `alpha` ë§¤ê°œë³€ìˆ˜ë¥¼ ìˆ˜ì •í•˜ì—¬ ì ë“¤ì˜ íˆ¬ëª…ë„(ì•ŒíŒŒ ê°’)ë¥¼ ì¡°ì ˆí•˜ëŠ” ê²ƒì…ë‹ˆë‹¤(ê·¸ë¦¼ 3.28).

    
    
    scp  + geom_point(alpha = 0.1)__

[![](03-chap_files/figure-html/fig-graphics-
twodsp2-1.png)](03-chap_files/figure-html/fig-graphics-twodsp2-1.png
"ê·¸ë¦¼ 3.28: ê·¸ë¦¼ 3.27ê³¼ ìœ ì‚¬í•˜ì§€ë§Œ, ì¼ë¶€ ê²¹ì³ê·¸ë¦¬ê¸°ë¥¼ í•´ê²°í•˜ê¸° ìœ„í•´ ë°˜íˆ¬ëª…í•œ ì ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.28: ê·¸ë¦¼ 3.27ê³¼ ìœ ì‚¬í•˜ì§€ë§Œ, ì¼ë¶€ ê²¹ì³ê·¸ë¦¬ê¸°ë¥¼ í•´ê²°í•˜ê¸° ìœ„í•´ ë°˜íˆ¬ëª…í•œ ì ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.

ì´ê²ƒì€ ì´ë¯¸ ê·¸ë¦¼ 3.27ë³´ë‹¤ ë‚«ì§€ë§Œ, ë” ë°€ì§‘ëœ ì˜ì—­ì—ì„œëŠ” ë°˜íˆ¬ëª…í•œ ì ë“¤ì¡°ì°¨ íŠ¹ì§• ì—†ëŠ” ê²€ì€ ë©ì–´ë¦¬ë¡œ ê²¹ì³ì§€ëŠ” ë°˜ë©´, ë” ê³ ë¦½ëœ ë°”ê¹¥ìª½ ì ë“¤ì€ í¬ë¯¸í•´ì§‘ë‹ˆë‹¤. ëŒ€ì•ˆìœ¼ë¡œëŠ” 2D ë°€ë„ì˜ ë“±ê³ ì„  ê·¸ë¦¼(contour plot)ì´ ìˆìœ¼ë©°, ì´ëŠ” ê·¸ë¦¼ 3.29ì—ì„œì²˜ëŸ¼ í”Œë¡¯ì˜ ëª¨ë“  ì ì„ ë‹¤ ê·¸ë¦¬ì§€ ì•Šì•„ë„ ëœë‹¤ëŠ” ì¥ì ì´ ìˆìŠµë‹ˆë‹¤.

í•˜ì§€ë§Œ ê·¸ë¦¼ 3.29ë¥¼ ë³´ë©´ ì˜¤ë¥¸ìª½ í•˜ë‹¨ì˜ ì  êµ¬ë¦„(ìƒëŒ€ì ìœ¼ë¡œ ì ì€ ìˆ˜ì˜ ì ì„ í¬í•¨)ì´ ë” ì´ìƒ í‘œí˜„ë˜ì§€ ì•ŠëŠ” ê²ƒì„ ì•Œ ìˆ˜ ìˆìŠµë‹ˆë‹¤. `geom_density2d`ì˜ ëŒ€ì—­í­(bandwidth)ê³¼ êµ¬ê°„í™”(binning) ë§¤ê°œë³€ìˆ˜ë¥¼ ì¡°ì •í•˜ì—¬ ì´ ë¬¸ì œë¥¼ ì–´ëŠ ì •ë„ ê·¹ë³µí•  ìˆ˜ ìˆìŠµë‹ˆë‹¤(ê·¸ë¦¼ 3.30, ì™¼ìª½ íŒ¨ë„).

    
    
    scp + geom_density2d()__

[![](03-chap_files/figure-html/fig-graphics-
twodsp3-1.png)](03-chap_files/figure-html/fig-graphics-twodsp3-1.png
"ê·¸ë¦¼ 3.29: ê·¸ë¦¼ 3.27ê³¼ ìœ ì‚¬í•˜ì§€ë§Œ, 2D ë°€ë„ ì¶”ì •ì¹˜ì˜ ë“±ê³ ì„  ê·¸ë¦¼ìœ¼ë¡œ ë Œë”ë§ë˜ì—ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.29: ê·¸ë¦¼ 3.27ê³¼ ìœ ì‚¬í•˜ì§€ë§Œ, 2D ë°€ë„ ì¶”ì •ì¹˜ì˜ ë“±ê³ ì„  ê·¸ë¦¼ìœ¼ë¡œ ë Œë”ë§ë˜ì—ˆìŠµë‹ˆë‹¤.

    
    
    scp + geom_density2d(h = 0.5, bins = 60)
    library("RColorBrewer")
    colorscale = scale_fill_gradientn(
        colors = rev(brewer.pal(9, "YlGnBu")),
        values = c(0, exp(seq(-5, 0, length.out = 100))))
    
    scp + stat_density2d(h = 0.5, bins = 60,
              aes( fill = after_stat(level)), geom = "polygon") +
      colorscale + coord_fixed()__

[![](03-chap_files/figure-html/fig-graphics-
twodsp4-1.png)](03-chap_files/figure-html/fig-graphics-twodsp4-1.png
"ê·¸ë¦¼ 3.30 (a): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
twodsp4-2.png)](03-chap_files/figure-html/fig-graphics-twodsp4-2.png
"ê·¸ë¦¼ 3.30 (b): ")

(b)

ê·¸ë¦¼ 3.30: 2D ë°€ë„ ê·¸ë¦¼. (a) ê·¸ë¦¼ 3.29ì™€ ê°™ì§€ë§Œ ë” ì‘ì€ í‰í™œ ëŒ€ì—­í­ê³¼ ë” ì´˜ì´˜í•œ ë“±ê³ ì„  êµ¬ê°„ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤. (b) ìƒ‰ìƒ ì±„ìš°ê¸°ê°€ ì¶”ê°€ëœ ëª¨ìŠµì…ë‹ˆë‹¤.

ê·¸ë¦¼ 3.30ì˜ ì˜¤ë¥¸ìª½ íŒ¨ë„ì—ì„œì²˜ëŸ¼, `stat_density2d` í•¨ìˆ˜(`geom_density2d`ì˜ ë˜í¼ í•¨ìˆ˜)ë¥¼ ëª…ì‹œì ìœ¼ë¡œ í˜¸ì¶œí•˜ê³  ê¸°í•˜í•™ì  ê°ì²´ë¡œ _polygon_ì„ ì‚¬ìš©í•¨ìœ¼ë¡œì¨ ë“±ê³ ì„  ì‚¬ì´ì˜ ê° ê³µê°„ì„ ì ë“¤ì˜ ìƒëŒ€ì  ë°€ë„ë¡œ ì±„ìš¸ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

ìš°ë¦¬ëŠ” **[RColorBrewer](https://cran.r-project.org/web/packages/RColorBrewer/)** íŒ¨í‚¤ì§€ì˜ `brewer.pal` í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•´ ìƒ‰ìƒ ì²™ë„ë¥¼ ì •ì˜í–ˆê³ , `coord_fixed`ë¥¼ í˜¸ì¶œí•˜ì—¬ í”Œë¡¯ì˜ ì¢…íš¡ë¹„(aspect ratio)ë¥¼ ê³ ì •í–ˆìŠµë‹ˆë‹¤. ì´ëŠ” ë‘ ë³€ìˆ˜ì— ëŒ€í•´ ë°ì´í„° ë²”ìœ„ì—ì„œ \(x\) ë° \(y\) ì¢Œí‘œë¡œì˜ ë§¤í•‘ì´ ë™ì¼í•˜ê²Œ ì´ë£¨ì–´ì§€ë„ë¡ ë³´ì¥í•©ë‹ˆë‹¤. ì´ ë‘ ì´ìŠˆ ëª¨ë‘ ë” ê¹Šì´ ì‚´í´ë³¼ ê°€ì¹˜ê°€ ìˆìœ¼ë©°, 3.7.1ì ˆì—ì„œëŠ” í”Œë¡¯ì˜ ëª¨ì–‘ì— ëŒ€í•´, 3.9ì ˆì—ì„œëŠ” ìƒ‰ìƒì— ëŒ€í•´ ë” ì´ì•¼ê¸°í•˜ê² ìŠµë‹ˆë‹¤.

ê·¸ë¦¼ 3.30ì˜ ë°€ë„ ê¸°ë°˜ í”Œë¡œíŒ… ë°©ë²•ì€ ê·¸ë¦¼ 3.27 ë° 3.28ì˜ ê²¹ì³ì§„ ì  êµ¬ë¦„ë³´ë‹¤ ì‹œê°ì ìœ¼ë¡œ ë” ë§¤ë ¥ì ì´ê³  í•´ì„í•˜ê¸° ì‰½ì§€ë§Œ, í”Œë¡¯ì˜ í¬ì†Œí•œ ì˜ì—­ì— ìˆëŠ” ì´ìƒì¹˜ ì ë“¤ì— ëŒ€í•œ ì •ë³´ë¥¼ ë§ì´ ìƒê²Œ ë˜ë¯€ë¡œ ì£¼ì˜í•´ì„œ ì‚¬ìš©í•´ì•¼ í•©ë‹ˆë‹¤. í•œ ê°€ì§€ ë°©ë²•ì€ `geom_point`ë¥¼ ì‚¬ìš©í•˜ì—¬ ê·¸ëŸ¬í•œ ì ë“¤ì„ ë‹¤ì‹œ ì¶”ê°€í•˜ëŠ” ê²ƒì…ë‹ˆë‹¤.

í•˜ì§€ë§Œ í‰í™œí™”ì˜ í•œê³„ë¥¼ í”¼í•  ìˆ˜ ìˆëŠ” ì•„ë§ˆë„ ê°€ì¥ ì¢‹ì€ ëŒ€ì•ˆì€ ìœ¡ê°í˜• êµ¬ê°„í™”(hexagonal binning)ì¼ ê²ƒì…ë‹ˆë‹¤([Carr et al. 1987](16-chap.html#ref-Carr1987hexbin)).

    
    
    scp + geom_hex() + coord_fixed()
    scp + geom_hex(binwidth = c(0.2, 0.2)) + colorscale +
      coord_fixed()__

[![](03-chap_files/figure-html/fig-graphics-
twodsp6-1.png)](03-chap_files/figure-html/fig-graphics-twodsp6-1.png
"ê·¸ë¦¼ 3.31 (a): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
twodsp6-2.png)](03-chap_files/figure-html/fig-graphics-twodsp6-2.png
"ê·¸ë¦¼ 3.31 (b): ")

(b)

ê·¸ë¦¼ 3.31: ìœ¡ê°í˜• êµ¬ê°„í™”. (a) ê¸°ë³¸ ë§¤ê°œë³€ìˆ˜ ì‚¬ìš©. (b) ë” ë¯¸ì„¸í•œ êµ¬ê°„ í¬ê¸°ì™€ ì‚¬ìš©ì ì •ì˜ ìƒ‰ìƒ ì²™ë„ ì‚¬ìš©.

### 3.7.1 í”Œë¡¯ ëª¨ì–‘(Plot shapes)

í”Œë¡¯ì— ì ì ˆí•œ ëª¨ì–‘ì„ ì„ íƒí•˜ëŠ” ê²ƒì€ ì •ë³´ê°€ ì˜ ì „ë‹¬ë˜ë„ë¡ í•˜ëŠ” ë° ì¤‘ìš”í•©ë‹ˆë‹¤. ê¸°ë³¸ì ìœ¼ë¡œ ê·¸ë˜í”„ì˜ ë†’ì´ì™€ ë„ˆë¹„ ì‚¬ì´ì˜ ë¹„ìœ¨ì¸ ëª¨ì–‘ ë§¤ê°œë³€ìˆ˜(shape parameter)ëŠ” í˜„ì¬ í”Œë¡œíŒ… ì¥ì¹˜(plotting device)ì—ì„œ ì‚¬ìš© ê°€ëŠ¥í•œ ê³µê°„ì„ ê¸°ë°˜ìœ¼ë¡œ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì— ì˜í•´ ì„ íƒë©ë‹ˆë‹¤. ì¥ì¹˜ì˜ ë„ˆë¹„ì™€ ë†’ì´ëŠ” ëª…ì‹œì ìœ¼ë¡œ ì§€ì •í•˜ê±°ë‚˜ ê¸°ë³¸ ë§¤ê°œë³€ìˆ˜10ë¥¼ í†µí•´ Rì—ì„œ ì—´ë¦´ ë•Œ ì§€ì •ë©ë‹ˆë‹¤. ë”ìš±ì´ ê·¸ë˜í”„ì˜ í¬ê¸°ëŠ” ê·¸ë¦¼ 3.31ì˜ ìƒ‰ìƒ ì²™ë„ ë§‰ëŒ€ì™€ ê°™ì€ ì¶”ê°€ì ì¸ ì¥ì‹ ìš”ì†Œì˜ ìœ ë¬´ì— ë”°ë¼ì„œë„ ë‹¬ë¼ì§‘ë‹ˆë‹¤.

10 See for example the manual pages of the `pdf` and `png` functions.

There are two simple rules that you can apply for scatterplots:

  * If the variables on the two axes are measured in the same units, then make sure that the same mapping of data space to physical space is used â€“ i.e., use `coord_fixed`. In the scatterplots above, both axes are the logarithm to base 2 of expression level measurements; that is, a change by one unit has the same meaning on both axes (a doubling of the expression level). Another case is principal component analysis (PCA), where the \\(x\\)-axis typically represents component 1, and the \\(y\\)-axis component 2. Since the axes arise from an orthonormal rotation of input data space, we want to make sure their scales match. Since the variance of the data is (by definition) smaller along the second component than along the first component (or at most, equal), well-made PCA plots usually have a width thatâ€™s larger than the height.

  * If the variables on the two axes are measured in different units, then we can still relate them to each other by comparing their dimensions. The default in many plotting routines in R, including **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** , is to look at the range of the data and map it to the available plotting region. However, in particular when the data more or less follow a line, looking at the typical slope of the line can be useful. This is called **banking** ([William S. Cleveland, McGill, and McGill 1988](16-chap.html#ref-Cleveland:1988:Banking)).

To illustrate banking, letâ€™s use the classic sunspot data from Clevelandâ€™s
paper.

    
    
    library("ggthemes")
    sunsp = tibble(year   = time(sunspot.year),
                   number = as.numeric(sunspot.year))
    sp = ggplot(sunsp, aes(x = year, y = number)) + geom_line()
    sp
    ratio = with(sunsp, bank_slopes(year, number))
    sp + coord_fixed(ratio = ratio)__

[![](03-chap_files/figure-html/fig-graphics-
banking-1.png)](03-chap_files/figure-html/fig-graphics-banking-1.png
"FigureÂ 3.32Â \(a\): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-
banking-2.png)](03-chap_files/figure-html/fig-graphics-banking-2.png
"FigureÂ 3.32Â \(b\): ")

(b)

Figure 3.32: The sunspot data. In (a), the plot shape is roughly quadratic, a
frequent default choice. In (b), a technique called **banking** was used to
choose the plot shape. (Note: the placement of the tick labels is not great in
this plot and would benefit from customization.)

The resulting plot is shown in the upper panel of Figure 3.32. We can clearly
see long-term fluctuations in the amplitude of sunspot activity cycles, with
particularly low maximum activities in the early 1700s, early 1800s, and
around the turn of the 20\\(^\text{th}\\) century. But now lets try out
banking.

How does the algorithm work? It aims to make the slopes in the curve be around
one. In particular, `bank_slopes` computes the median absolute slope, and then
with the call to `coord_fixed` we set the aspect ratio of the plot such that
this quantity becomes 1. The result is shown in the lower panel of Figure
3.32. Quite counter-intuitively, even though the plot takes much smaller
space, we see more on it! In particular, we can see the saw-tooth shape of the
sunspot cycles, with sharp rises and more slow declines.

## 3.8 2ì°¨ì› ì´ìƒì˜ ì‹œê°í™”

ë•Œë¡œëŠ” ë‘ ê°œ ì´ìƒì˜ ë³€ìˆ˜ ê°„ì˜ ê´€ê³„ë¥¼ ë³´ì—¬ì£¼ê³  ì‹¶ì„ ë•Œê°€ ìˆìŠµë‹ˆë‹¤. ì¶”ê°€ì ì¸ ì°¨ì›ì„ í¬í•¨í•˜ê¸° ìœ„í•œ ë‹¹ì—°í•œ ì„ íƒì€ í”Œë¡¯ ê¸°í˜¸ì˜ ëª¨ì–‘ê³¼ ìƒ‰ìƒì…ë‹ˆë‹¤. `geom_point` ê¸°í•˜í•™ì  ê°ì²´ëŠ” (\(x\)ì™€ \(y\) ì™¸ì—) ë‹¤ìŒê³¼ ê°™ì€ ë¯¸í•™(aesthetics)ì„ ì œê³µí•©ë‹ˆë‹¤:

  * `fill`

  * `color`

  * `shape`

  * `size`

  * `alpha`

ì´ë“¤ì€ `geom_point` í•¨ìˆ˜ì˜ ë§¤ë‰´ì–¼ í˜ì´ì§€ì—ì„œ ì‚´í´ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. `fill`ê³¼ `color`ëŠ” ê°ì²´ì˜ ì±„ìš°ê¸° ë° í…Œë‘ë¦¬ ìƒ‰ìƒì„ ì˜ë¯¸í•˜ë©°, `alpha`ëŠ” íˆ¬ëª…ë„ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ì•ì„œ ê·¸ë¦¼ 3.28 ë“±ì—ì„œ ìš°ë¦¬ëŠ” ì ì˜ ë°€ë„ë¥¼ ë°˜ì˜í•˜ê³  ê²¹ì³ê·¸ë¦¬ê¸°(overplotting)ì— ì˜í•œ ê°€ë ¤ì§ íš¨ê³¼ë¥¼ í”¼í•˜ê¸° ìœ„í•´ ìƒ‰ìƒì´ë‚˜ íˆ¬ëª…ë„ë¥¼ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤. ë˜í•œ ì´ëŸ¬í•œ ì†ì„±ë“¤ì„ ì‚¬ìš©í•˜ì—¬ ë°ì´í„°ì˜ ë‹¤ë¥¸ ì°¨ì›ì„ ë³´ì—¬ì¤„ ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤. ì›ì¹™ì ìœ¼ë¡œëŠ” ìœ„ì— ë‚˜ì—´ëœ 5ê°€ì§€ ë¯¸í•™ì„ ëª¨ë‘ ë™ì‹œì— ì‚¬ìš©í•˜ì—¬ ìµœëŒ€ 7ì°¨ì›ì˜ ë°ì´í„°ë¥¼ ë³´ì—¬ì¤„ ìˆ˜ ìˆì§€ë§Œ, ê·¸ëŸ¬í•œ í”Œë¡¯ì€ í•´ë…í•˜ê¸°ê°€ ë§¤ìš° ì–´ë ¤ìš¸ ê²ƒì…ë‹ˆë‹¤. ëŒ€ê°œëŠ” ì´ ë¯¸í•™ë“¤ ì¤‘ í•œë‘ ê°€ì§€ë§Œ ì„ íƒí•˜ì—¬ ë°ì´í„°ì˜ ì¶”ê°€ì ì¸ í•œë‘ ì°¨ì›ì„ ë³´ì—¬ì£¼ëŠ” ì •ë„ë¡œ ì œí•œí•˜ëŠ” ê²ƒì´ ì¢‹ìŠµë‹ˆë‹¤.

### 3.8.1 íŒ¨ì‹¯(Faceting)

[![ë•Œë¡œëŠ” ì´ëŸ¬í•œ í”Œë¡¯ ë°°ì—´ì˜ ëª¨ì–‘ì„ ë¹—ëŒ€ì–´ íŠ¸ë ë¦¬ìŠ¤(trellis) ë˜ëŠ” ë˜í‹°ìŠ¤(lattice) ê·¸ë˜í”½ì´ë¼ê³ ë„ ë¶€ë¦…ë‹ˆë‹¤. íŒ¨ì‹¯ì„ êµ¬í˜„í•œ ì²« ë²ˆì§¸ ì£¼ìš” R íŒ¨í‚¤ì§€ëŠ” latticeì˜€ìŠµë‹ˆë‹¤. ì´ ì±…ì—ì„œëŠ” ggplot2ë¥¼ í†µí•´ ì œê³µë˜ëŠ” íŒ¨ì‹¯ ê¸°ëŠ¥ì„ ì‚¬ìš©í•  ê²ƒì…ë‹ˆë‹¤.](imgs/devil.png)](imgs/devil.png "ë•Œë¡œëŠ” ì´ëŸ¬í•œ í”Œë¡¯ ë°°ì—´ì˜ ëª¨ì–‘ì„ ë¹—ëŒ€ì–´ íŠ¸ë ë¦¬ìŠ¤(trellis) ë˜ëŠ” ë˜í‹°ìŠ¤(lattice) ê·¸ë˜í”½ì´ë¼ê³ ë„ ë¶€ë¦…ë‹ˆë‹¤. íŒ¨ì‹¯ì„ êµ¬í˜„í•œ ì²« ë²ˆì§¸ ì£¼ìš” R íŒ¨í‚¤ì§€ëŠ” latticeì˜€ìŠµë‹ˆë‹¤. ì´ ì±…ì—ì„œëŠ” ggplot2ë¥¼ í†µí•´ ì œê³µë˜ëŠ” íŒ¨ì‹¯ ê¸°ëŠ¥ì„ ì‚¬ìš©í•  ê²ƒì…ë‹ˆë‹¤.")

ë•Œë¡œëŠ” ì´ëŸ¬í•œ í”Œë¡¯ ë°°ì—´ì˜ ëª¨ì–‘ì„ ë¹—ëŒ€ì–´ _íŠ¸ë ë¦¬ìŠ¤(trellis)_ ë˜ëŠ” _ë˜í‹°ìŠ¤(lattice)_ ê·¸ë˜í”½ì´ë¼ê³ ë„ ë¶€ë¦…ë‹ˆë‹¤. íŒ¨ì‹¯ì„ êµ¬í˜„í•œ ì²« ë²ˆì§¸ ì£¼ìš” R íŒ¨í‚¤ì§€ëŠ” **[lattice](https://cran.r-project.org/web/packages/lattice/)**ì˜€ìŠµë‹ˆë‹¤. ì´ ì±…ì—ì„œëŠ” **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ë¥¼ í†µí•´ ì œê³µë˜ëŠ” íŒ¨ì‹¯ ê¸°ëŠ¥ì„ ì‚¬ìš©í•  ê²ƒì…ë‹ˆë‹¤.

ë°ì´í„°ì˜ ì¶”ê°€ì ì¸ ì°¨ì›ì„ ë³´ì—¬ì£¼ëŠ” ë˜ ë‹¤ë¥¸ ë°©ë²•ì€ í•˜ë‚˜(ë˜ëŠ” ê·¸ ì´ìƒ)ì˜ ë³€ìˆ˜ë¥¼ ê¸°ë°˜ìœ¼ë¡œ ë°ì´í„°ë¥¼ ë°˜ë³µì ìœ¼ë¡œ í•˜ìœ„ ì§‘í•©í™”(subsetting, ë˜ëŠ” "ìŠ¬ë¼ì´ì‹±")í•˜ì—¬ ì—¬ëŸ¬ ê°œì˜ í”Œë¡¯ì„ ë³´ì—¬ì¤Œìœ¼ë¡œì¨ ê° ë¶€ë¶„ì„ ë”°ë¡œ ì‹œê°í™”í•˜ëŠ” ê²ƒì…ë‹ˆë‹¤. ì´ë¥¼ **íŒ¨ì‹¯(faceting)**ì´ë¼ê³  í•˜ë©°, ì´ë¥¼ í†µí•´ ìµœëŒ€ 4~5ì°¨ì›ì˜ ë°ì´í„°ë¥¼ ì‹œê°í™”í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´, íŒ¨ì‹¯ ë³€ìˆ˜ì˜ ë²”ìœ„ì— ë”°ë¼ ë‹¤ë¥¸ ë³€ìˆ˜ë“¤ ê°„ì— ê´€ì°°ëœ íŒ¨í„´ì´ ë™ì¼í•œì§€ ì•„ë‹ˆë©´ ë‹¤ë¥¸ì§€ ì¡°ì‚¬í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì˜ˆì œë¥¼ ì‚´í´ë³´ê² ìŠµë‹ˆë‹¤:

    
    
    library("magrittr")
    dftx$lineage %<>% sub("^$", "no", .)
    dftx$lineage %<>% factor(levels = c("no", "EPI", "PE", "FGF4-KO"))
    
    ggplot(dftx, aes(x = X1426642_at, y = X1418765_at)) +
      geom_point() + facet_grid( . ~ lineage )__

[![](03-chap_files/figure-html/fig-graphics-
facet1-1.png)](03-chap_files/figure-html/fig-graphics-facet1-1.png
"ê·¸ë¦¼ 3.33: íŒ¨ì‹¯ì˜ ì˜ˆì‹œ: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ì´ì œ ë²”ì£¼í˜• ë³€ìˆ˜ lineageì— ì˜í•´ ë¶„í• ë˜ì—ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.33: **íŒ¨ì‹¯(faceting)**ì˜ ì˜ˆì‹œ: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ì´ì œ ë²”ì£¼í˜• ë³€ìˆ˜ `lineage`ì— ì˜í•´ ë¶„í• ë˜ì—ˆìŠµë‹ˆë‹¤.

ê²°ê³¼ëŠ” ê·¸ë¦¼ 3.33ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤. ìš°ë¦¬ëŠ” Rì˜ í¬ë®¬ëŸ¬(formula) ì–¸ì–´ë¥¼ ì‚¬ìš©í•˜ì—¬ ë°ì´í„°ë¥¼ ë¶„í• í•  ë³€ìˆ˜ë¥¼ ì§€ì •í–ˆê³ , ê°œë³„ íŒ¨ë„ì´ ì„œë¡œ ë‹¤ë¥¸ ì—´ì— ë°°ì¹˜ë˜ë„ë¡ í–ˆìŠµë‹ˆë‹¤: `facet_grid( . ~ lineage )`. ì‚¬ì‹¤ ë‹¤ìŒê³¼ ê°™ì´ ë‘ ê°œì˜ íŒ¨ì‹¯ ë³€ìˆ˜ë¥¼ ì§€ì •í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤. ê·¸ ê²°ê³¼ëŠ” ê·¸ë¦¼ 3.34ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤.

    
    
    ggplot(dftx,
      aes(x = X1426642_at, y = X1418765_at)) + geom_point() +
       facet_grid( Embryonic.day ~ lineage )__

[![](03-chap_files/figure-html/fig-graphics-
facet2-1.png)](03-chap_files/figure-html/fig-graphics-facet2-1.png
"ê·¸ë¦¼ 3.34: íŒ¨ì‹¯: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ë²”ì£¼í˜• ë³€ìˆ˜ Embryonic.day(í–‰)ì™€ lineage(ì—´)ì— ì˜í•´ ë¶„í• ë˜ì—ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.34: **íŒ¨ì‹¯(Faceting)**: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ë²”ì£¼í˜• ë³€ìˆ˜ `Embryonic.day`(í–‰)ì™€ `lineage`(ì—´)ì— ì˜í•´ ë¶„í• ë˜ì—ˆìŠµë‹ˆë‹¤.

ë˜ ë‹¤ë¥¸ ìœ ìš©í•œ í•¨ìˆ˜ëŠ” `facet_wrap`ì…ë‹ˆë‹¤. íŒ¨ì‹¯ ë³€ìˆ˜ì˜ ìˆ˜ì¤€(level)ì´ ë„ˆë¬´ ë§ì•„ ëª¨ë“  í”Œë¡¯ì„ í•œ í–‰ì´ë‚˜ í•œ ì—´ì— ë§ì¶”ê¸° ì–´ë ¤ìš¸ ë•Œ, ì´ í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ì§€ì •ëœ ìˆ˜ì˜ ì—´ì´ë‚˜ í–‰ìœ¼ë¡œ ë°°ì—´í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì§€ê¸ˆê¹Œì§€ëŠ” ë²”ì£¼í˜• ë³€ìˆ˜ì— ì˜í•œ íŒ¨ì‹¯ì„ ë³´ì•˜ì§€ë§Œ, ì—°ì†í˜• ë³€ìˆ˜ë¥¼ ìˆ˜ì¤€ë³„ë¡œ ì´ì‚°í™”í•˜ì—¬ ì‚¬ìš©í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤. ì´ë¥¼ ìœ„í•´ `cut` í•¨ìˆ˜ê°€ ìœ ìš©í•©ë‹ˆë‹¤.

    
    
    ggplot(mutate(dftx, Tdgf1 = cut(X1450989_at, breaks = 4)),
       aes(x = X1426642_at, y = X1418765_at)) + geom_point() +
       facet_wrap( ~ Tdgf1, ncol = 2 )__

[![](03-chap_files/figure-html/fig-graphics-
facet3-1.png)](03-chap_files/figure-html/fig-graphics-facet3-1.png
"ê·¸ë¦¼ 3.35: íŒ¨ì‹¯: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ì—°ì†í˜• ë³€ìˆ˜ X1450989_atì— ì˜í•´ ë¶„í• ë˜ê³  facet_wrapì— ì˜í•´ ë°°ì—´ë˜ì—ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.35: **íŒ¨ì‹¯(Faceting)**: ê·¸ë¦¼ 3.9ì™€ ë™ì¼í•œ ë°ì´í„°ì´ì§€ë§Œ, ì—°ì†í˜• ë³€ìˆ˜ `X1450989_at`ì— ì˜í•´ ë¶„í• ë˜ê³  `facet_wrap`ì— ì˜í•´ ë°°ì—´ë˜ì—ˆìŠµë‹ˆë‹¤.

ê·¸ë¦¼ 3.35ì—ì„œ ë„¤ íŒ¨ë„ì˜ ì  ê°œìˆ˜ê°€ ì„œë¡œ ë‹¤ë¥¸ ê²ƒì„ ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì´ëŠ” `cut`ì´ ì  ê°œìˆ˜ê°€ ì•„ë‹ˆë¼ êµ¬ê°„ì˜ ê¸¸ì´ë¥¼ ë™ì¼í•˜ê²Œ ë‚˜ëˆ„ê¸° ë•Œë¬¸ì…ë‹ˆë‹¤. ë§Œì•½ ì  ê°œìˆ˜ë¥¼ ë™ì¼í•˜ê²Œ ë§ì¶”ê³  ì‹¶ë‹¤ë©´ `cut`ê³¼ í•¨ê»˜ `quantile`ì„ ì‚¬ìš©í•˜ê±°ë‚˜, ë³€ìˆ˜ ê°’ì˜ ìˆœìœ„(rank)ë¥¼ ê¸°ì¤€ìœ¼ë¡œ ë‚˜ëˆŒ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

#### ì¶• ìŠ¤ì¼€ì¼(Axes scales)

ê·¸ë¦¼ 3.33â€”3.35ì—ì„œ ì¶•ì˜ ìŠ¤ì¼€ì¼ì€ ëª¨ë“  íŒ¨ë„ì—ì„œ ë™ì¼í•©ë‹ˆë‹¤. ëŒ€ì‹  `facet_grid`ì™€ `facet_wrap` í•¨ìˆ˜ì˜ `scales` ì¸ìˆ˜ë¥¼ ì„¤ì •í•˜ì—¬ ì´ë¥¼ ë³€ê²½í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì´ ì¸ìˆ˜ë¥¼ í†µí•´ ê° íŒ¨ë„ì˜ \(x\)ì¶•ê³¼ \(y\)ì¶•ì´ ë™ì¼í•œ ìŠ¤ì¼€ì¼ì„ ê°€ì§ˆì§€, ì•„ë‹ˆë©´ ê° íŒ¨ë„ì˜ ë°ì´í„° ë²”ìœ„ì— ë§ê²Œ ì¡°ì •ë ì§€ ì œì–´í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì—¬ê¸°ì—ëŠ” ìƒì¶© ê´€ê³„(trade-off)ê°€ ìˆìŠµë‹ˆë‹¤. ì¶• ìŠ¤ì¼€ì¼ì„ íŒ¨ë„ë³„ë¡œ ì¡°ì •í•˜ë©´ ì„¸ë¶€ ì‚¬í•­ì„ ë” ì˜ ë³¼ ìˆ˜ ìˆëŠ” ë°˜ë©´, ê·¸ë£¹ ê°„ íŒ¨ë„ ë¹„êµëŠ” ë” ì–´ë ¤ì›Œì§‘ë‹ˆë‹¤.

#### ì•”ì‹œì  íŒ¨ì‹¯(Implicit faceting)

`facet_grid`ë‚˜ `facet_wrap`ì„ ëª…ì‹œì ìœ¼ë¡œ í˜¸ì¶œí•˜ì§€ ì•Šê³ ë„ ë¯¸í•™(aesthetics)ì„ ì§€ì •í•˜ì—¬ í”Œë¡¯ì„ íŒ¨ì‹¯í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì•”ì‹œì  íŒ¨ì‹¯ì˜ ë§¤ìš° ê°„ë‹¨í•œ ì˜ˆëŠ” ê·¸ë¦¼ 3.15â€”3.18ì—ì„œì²˜ëŸ¼ ìš”ì¸(factor)ì„ \(x\)ì¶•ìœ¼ë¡œ ì‚¬ìš©í•˜ëŠ” ê²ƒì…ë‹ˆë‹¤.

### 3.8.2 ëŒ€í™”í˜• ê·¸ë˜í”½

ì§€ê¸ˆê¹Œì§€ ìƒì„±ëœ í”Œë¡¯ì€ ì •ì ì¸ ì´ë¯¸ì§€ì˜€ìŠµë‹ˆë‹¤. í”Œë¡¯ì„ ëŒ€í™”í˜•(interactive)ìœ¼ë¡œ ë§Œë“¤ë©´ ì—„ì²­ë‚œ ì–‘ì˜ ì •ë³´ì™€ í‘œí˜„ë ¥ì„ ë”í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì—¬ê¸°ì„œëŠ” ëŒ€í™”í˜• ì‹œê°í™”ë¥¼ ê¹Šì´ ìˆê²Œ ë‹¤ë£¨ì§€ëŠ” ì•Šê² ì§€ë§Œ, ëª‡ ê°€ì§€ ì¤‘ìš”í•œ ìë£Œë¥¼ ì†Œê°œí•˜ê² ìŠµë‹ˆë‹¤. ì´ëŠ” ì—­ë™ì ì¸ ë¶„ì•¼ì´ë¯€ë¡œ ë…ìë“¤ì€ ìµœì‹  ë™í–¥ì„ íŒŒì•…í•˜ê¸° ìœ„í•´ R ìƒíƒœê³„ë¥¼ íƒìƒ‰í•´ ë³´ì•„ì•¼ í•©ë‹ˆë‹¤.

#### Shiny

Posit(êµ¬ RStudio)ì˜ **[shiny](https://cran.r-project.org/web/packages/shiny/)**ëŠ” Rìš© ì›¹ ì• í”Œë¦¬ì¼€ì´ì…˜ í”„ë ˆì„ì›Œí¬ì…ë‹ˆë‹¤. ëŒ€í™”í˜• ìš”ì†Œê°€ í”Œë¡¯ì„ ìƒì„±í•˜ëŠ” R ì½”ë“œë¥¼ ì§ì ‘ í˜¸ì¶œí•˜ë¯€ë¡œ, ìŠ¬ë¼ì´ë”, ì„ íƒê¸° ë° ê¸°íƒ€ ì œì–´ ìš”ì†Œë¥¼ ì‚¬ìš©í•˜ì—¬ í‘œì‹œë˜ëŠ” í”Œë¡¯ì˜ ëª¨ë“  ì¸¡ë©´ì„ ë³€ê²½í•  ìˆ˜ ìˆëŠ” ëŒ€í™”í˜• ë””ìŠ¤í”Œë ˆì´ë¥¼ ì‰½ê²Œ ë§Œë“¤ ìˆ˜ ìˆìŠµë‹ˆë‹¤. í›Œë¥­í•œ ì˜ˆì‹œë“¤ì„ [Shiny ê°¤ëŸ¬ë¦¬](https://shiny.posit.co/r/gallery)ì—ì„œ í™•ì¸í•´ ë³´ì„¸ìš”.

**[Shiny](https://cran.r-project.org/web/packages/shiny/)** ê¸°ë°˜ ëŒ€í™”í˜• ì‹œê°í™”ë¥¼ ìœ„í•œ ê·¸ë˜í”½ ì—”ì§„ìœ¼ë¡œ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ë¥¼ ì‚¬ìš©í•  ìˆ˜ ìˆìœ¼ë©°, ì‚¬ì‹¤ ê¸°ë³¸ R ê·¸ë˜í”½ì´ë‚˜ ë‹¤ë¥¸ ê·¸ë˜í”½ íŒ¨í‚¤ì§€ë„ ì‚¬ìš©í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì—¬ê¸°ì„œ ì•½ê°„ ì–´ìƒ‰í•œ ì ì€ ëŒ€í™”í˜• ì˜µì…˜ì„ ì„¤ëª…í•˜ëŠ” ë° ì‚¬ìš©ë˜ëŠ” ì–¸ì–´ê°€ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ë¥¼ í†µí•œ ê·¸ë˜í”½ ìƒì„± ë° ê·¸ë˜í”½ì˜ ë¬¸ë²•ê³¼ ë¶„ë¦¬ë˜ì–´ ìˆë‹¤ëŠ” ê²ƒì…ë‹ˆë‹¤. **[ggvis](https://cran.r-project.org/web/packages/ggvis/)** íŒ¨í‚¤ì§€ëŠ” ì´ëŸ¬í•œ í•œê³„ë¥¼ ê·¹ë³µí•˜ëŠ” ê²ƒì„ ëª©í‘œë¡œ í•©ë‹ˆë‹¤.

#### ggvis

**[ggvis](https://cran.r-project.org/web/packages/ggvis/)**11ëŠ” **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì˜ ì¥ì ë“¤ì„ ëŒ€í™”í˜• ê·¸ë˜í”½ ì˜ì—­ìœ¼ë¡œ í™•ì¥í•˜ë ¤ëŠ” ì‹œë„ì…ë‹ˆë‹¤. **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ê°€ Rì˜ ì „í†µì ì¸ ê·¸ë˜í”½ ì¥ì¹˜(PDF, PNG ë“±)ì— ê·¸ë˜í”½ì„ ìƒì„±í•˜ëŠ” ê²ƒê³¼ ëŒ€ì¡°ì ìœ¼ë¡œ, **[ggvis](https://cran.r-project.org/web/packages/ggvis/)**ëŠ” [Vega](https://github.com/trifacta/vega)ë¼ëŠ” JavaScript ì¸í”„ë¼ë¥¼ ê¸°ë°˜ìœ¼ë¡œ í•˜ë©°, ê·¸ í”Œë¡¯ì€ ì›¹ ë¸Œë¼ìš°ì €ì—ì„œ ë³¼ ìˆ˜ ìˆë„ë¡ ì˜ë„ë˜ì—ˆìŠµë‹ˆë‹¤. **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì™€ ë§ˆì°¬ê°€ì§€ë¡œ **[ggvis](https://cran.r-project.org/web/packages/ggvis/)** íŒ¨í‚¤ì§€ë„ ê·¸ë˜í”½ì˜ ë¬¸ë²• ê°œë…ì—ì„œ ì˜ê°ì„ ë°›ì•˜ì§€ë§Œ, ë³„ê°œì˜ êµ¬ë¬¸ì„ ì‚¬ìš©í•©ë‹ˆë‹¤. ìƒí˜¸ ì‘ìš©ì— í•„ìš”í•œ ê³„ì‚°ì„ ìˆ˜í–‰í•˜ê¸° ìœ„í•´ Rì— ì—°ê²°í•˜ëŠ” **[Shiny](https://cran.r-project.org/web/packages/shiny/)**ì˜ ì¸í”„ë¼ë¥¼ í™œìš©í•©ë‹ˆë‹¤. ì €ìê°€ ë§í–ˆë“¯ì´12, ëª©í‘œëŠ” Rì˜ ìµœê³ (ì˜ˆ: ìƒìƒí•  ìˆ˜ ìˆëŠ” ëª¨ë“  ëª¨ë¸ë§ ê¸°ëŠ¥)ì™€ ì›¹ì˜ ìµœê³ (ëª¨ë“  ì‚¬ëŒì´ ì›¹ ë¸Œë¼ìš°ì €ë¥¼ ê°€ì§€ê³  ìˆìŒ)ë¥¼ ê²°í•©í•˜ëŠ” ê²ƒì…ë‹ˆë‹¤. ë°ì´í„° ì¡°ì‘ ë° ë³€í™˜ì€ Rì—ì„œ ìˆ˜í–‰ë˜ë©°, ê·¸ë˜í”½ì€ Vegaë¥¼ ì‚¬ìš©í•˜ì—¬ ì›¹ ë¸Œë¼ìš°ì €ì—ì„œ ë Œë”ë§ë©ë‹ˆë‹¤.

11 ì´ ê¸€ì„ ì“°ëŠ” ì‹œì (2017ë…„ ì—¬ë¦„)ì—ëŠ” **[ggvis](https://cran.r-project.org/web/packages/ggvis/)** ê°œë°œì˜ ì´ˆê¸° ì¶”ì§„ë ¥ì´ ìœ ì§€ë ì§€ ë¶ˆë¶„ëª…í•˜ë©°, í˜„ì¬ ê¸°ëŠ¥ê³¼ ì™„ì„±ë„ëŠ” ì•„ì§ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì— ë¯¸ì¹˜ì§€ ëª»í•©ë‹ˆë‹¤.

12 https://ggvis.rstudio.com

**[Shiny](https://cran.r-project.org/web/packages/shiny/)**ì™€ **[ggvis](https://cran.r-project.org/web/packages/ggvis/)**ì˜ ìƒí˜¸ ì‘ìš© ê²°ê³¼ë¡œ, ì‚¬ìš©ìê°€ ê·¸ë˜í”½ì„ ë³´ëŠ” ë™ì•ˆ ì‚¬ìš©ìì˜ ë™ì‘ì— ì‘ë‹µí•˜ê¸° ìœ„í•´ ê¸°ë³¸ ë°ì´í„°ì™€ ì½”ë“œë¥¼ ì‹¤í–‰í•˜ëŠ” R ì¸í„°í”„ë¦¬í„°ê°€ í•„ìš”í•©ë‹ˆë‹¤. ì´ R ì¸í„°í”„ë¦¬í„°ëŠ” ë¡œì»¬ ë¨¸ì‹ ì´ë‚˜ ì„œë²„ì— ìˆì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë‘ ê²½ìš° ëª¨ë‘ ë·°ì–´ ì• í”Œë¦¬ì¼€ì´ì…˜ì€ ì›¹ ë¸Œë¼ìš°ì €ì´ë©°, Rê³¼ì˜ ìƒí˜¸ ì‘ìš©ì€ ì›¹ í”„ë¡œí† ì½œ(HTTP ë˜ëŠ” HTTPS)ì„ í†µí•´ ì´ë£¨ì–´ì§‘ë‹ˆë‹¤. ë¬¼ë¡  ì´ëŠ” Rì— ì˜í•´ í•œ ë²ˆ ìƒì„±ë˜ì–´ ì‹¤í–‰ ì¤‘ì¸ R ì¸ìŠ¤í„´ìŠ¤ì— ì—°ê²°ë˜ì§€ ì•Šê³  PDF ë˜ëŠ” HTML ë·°ì–´ì—ì„œ ë³¼ ìˆ˜ ìˆëŠ” ìì²´ í¬í•¨ íŒŒì¼ì— ì €ì¥ëœ ê·¸ë˜í”½ê³¼ëŠ” ë‹¤ë¦…ë‹ˆë‹¤.

#### plotly

ëŒ€í™”í˜• ê·¸ë˜í”½ ìƒì„±ì„ ìœ„í•œ ë˜ ë‹¤ë¥¸ í›Œë¥­í•œ ì›¹ ê¸°ë°˜ ë„êµ¬ëŠ” **plotly**ì…ë‹ˆë‹¤. [ì˜¨ë¼ì¸](https://plot.ly/r)ì—ì„œ ëŒ€í™”í˜• ê·¸ë˜í”½ì˜ ëª‡ ê°€ì§€ ì˜ˆë¥¼ ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. Rì—ì„œ ìì‹ ë§Œì˜ ëŒ€í™”í˜• í”Œë¡¯ì„ ë§Œë“¤ë ¤ë©´ ë‹¤ìŒê³¼ ê°™ì€ ì½”ë“œë¥¼ ì‚¬ìš©í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤:

    
    
    library("plotly")
    plot_ly(economics, x = ~date, y = ~unemploy/pop)__

**[Shiny](https://cran.r-project.org/web/packages/shiny/)** ë° **[ggvis](https://cran.r-project.org/web/packages/ggvis/)**ì™€ ë§ˆì°¬ê°€ì§€ë¡œ, ê·¸ë˜í”½ì€ ì›¹ ë¸Œë¼ìš°ì €ì—ì„œ ë³¼ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ê·¸ëŸ¬ë‚˜ ì‹¤í–‰ ì¤‘ì¸ R ì„¸ì…˜ì´ ë°˜ë“œì‹œ í•„ìš”í•œ ê²ƒì€ ì•„ë‹™ë‹ˆë‹¤. ê·¸ë˜í”½ì€ ë¡œì§ì´ JavaScript ë˜ëŠ” D3.js ì‹œìŠ¤í…œìœ¼ë¡œ ì½”ë”©ëœ ìì²´ í¬í•¨ HTML ë¬¸ì„œë¡œ êµ¬ì„±ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

[![](imgs/chap3-rglvolcano.png)](imgs/chap3-rglvolcano.png "ê·¸ë¦¼ 3.36: ì˜¤í´ëœë“œ í™”ì‚° ì§€ëŒ€ì˜ ì•½ 50ê°œ í™”ì‚° ì¤‘ í•˜ë‚˜ì¸ Maunga Whau(Mt Eden)ì˜ ì§€í˜• ì •ë³´ì¸ volcano ë°ì´í„°ì˜ rgl ë Œë”ë§ì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.36: **[rgl](https://cran.r-project.org/web/packages/rgl/)**ì„ ì´ìš©í•œ `volcano` ë°ì´í„° ë Œë”ë§. ì˜¤í´ëœë“œ í™”ì‚° ì§€ëŒ€ì˜ ì•½ 50ê°œ í™”ì‚° ì¤‘ í•˜ë‚˜ì¸ Maunga Whau(Mt Eden)ì˜ ì§€í˜• ì •ë³´ì…ë‹ˆë‹¤.

#### rgl, webgl

3D ê°ì²´(ì˜ˆ: ê¸°í•˜í•™ì  êµ¬ì¡°)ë¥¼ ì‹œê°í™”í•˜ë ¤ë©´ **[rgl](https://cran.r-project.org/web/packages/rgl/)** íŒ¨í‚¤ì§€ê°€ ìˆìŠµë‹ˆë‹¤. ì´ íŒ¨í‚¤ì§€ëŠ” í™”ë©´ì˜ íŠ¹ìˆ˜ ê·¸ë˜í”½ ì°½ì´ë‚˜ ì›¹ ë¸Œë¼ìš°ì €ë¥¼ í†µí•´ ì¥ë©´ì„ íšŒì „ì‹œí‚¤ê±°ë‚˜ í™•ëŒ€/ì¶•ì†Œí•  ìˆ˜ ìˆëŠ” ëŒ€í™”í˜• ë·°ì–´ë¥¼ ìƒì„±í•©ë‹ˆë‹¤. ì•„ë˜ ì½”ë“œì—ì„œ ìƒì„±ëœ ì¥ë©´ì˜ ìŠ¤í¬ë¦°ìƒ·ì´ ê·¸ë¦¼ 3.36ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤. ì´ëŸ¬í•œ ìŠ¤í¬ë¦°ìƒ·ì€ `snapshot3d` í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ìƒì„±í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

    
    
    data("volcano")
    volcanodata = list(
      x = 10 * seq_len(nrow(volcano)),
      y = 10 * seq_len(ncol(volcano)),
      z = volcano,
      col = terrain.colors(500)[cut(volcano, breaks = 500)]
    )
    library("rgl")
    with(volcanodata, persp3d(x, y, z, color = col))__

ìœ„ì˜ ì½”ë“œì—ì„œ ê¸°ë³¸ R í•¨ìˆ˜ `cut`ì€ `volcano` ë°ì´í„°ì˜ ê°’ ë²”ìœ„ì—ì„œ 1ë¶€í„° 500 ì‚¬ì´ì˜ ì •ìˆ˜ë¡œì˜ ë§¤í•‘ì„ ê³„ì‚°í•˜ë©°13, ìš°ë¦¬ëŠ” ì´ ì •ìˆ˜ë“¤ì„ ì‚¬ìš©í•˜ì—¬ `terrain.colors(500)`ì—ì„œ ìƒ‰ìƒ ìŠ¤ì¼€ì¼ì„ ì¸ë±ì‹±í•©ë‹ˆë‹¤. ìì„¸í•œ ë‚´ìš©ì€ ì´ íŒ¨í‚¤ì§€ì˜ í›Œë¥­í•œ ë¹„ë„¤íŠ¸(vignettes)ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”.

13 ë” ì •í™•í•˜ê²ŒëŠ”, `cut`ì€ 500ê°œì˜ ìˆ˜ì¤€(level)ì„ ê°€ì§„ ìš”ì¸(factor)ì„ ë°˜í™˜í•˜ë©°, ìš°ë¦¬ëŠ” ì´ë¥¼ ì •ìˆ˜ë¡œ ê°•ì œ ë³€í™˜(coerce)í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

## 3.9 ìƒ‰ìƒ(Color)

í”Œë¡¯ì„ ë§Œë“¤ ë•Œ ì¤‘ìš”í•˜ê²Œ ê³ ë ¤í•´ì•¼ í•  ì‚¬í•­ ì¤‘ í•˜ë‚˜ëŠ” í”Œë¡¯ì— ì‚¬ìš©í•˜ëŠ” ìƒ‰ìƒì…ë‹ˆë‹¤. ëŒ€ë¶€ë¶„ì˜ R ì‚¬ìš©ìëŠ” ê·¸ë¦¼ 3.37ì— í‘œì‹œëœ ê²ƒê³¼ ê°™ì´ ê¸°ë³¸ R ê·¸ë˜í”½ì—ì„œ ì‚¬ìš©í•˜ëŠ” ë‚´ì¥ ìƒ‰ìƒ ì²´ê³„ì— ìµìˆ™í•  ê²ƒì…ë‹ˆë‹¤.

    
    
    pie(rep(1, 8), col=1:8)__

[![](03-chap_files/figure-html/fig-graphics-
simplecolorpie-1.png)](03-chap_files/figure-html/fig-graphics-
simplecolorpie-1.png "ê·¸ë¦¼ 3.37: ê¸°ë³¸ R ìƒ‰ìƒ íŒ”ë ˆíŠ¸ì˜ ì²˜ìŒ 8ê°€ì§€ ìƒ‰ìƒì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.37: ê¸°ë³¸ R ìƒ‰ìƒ íŒ”ë ˆíŠ¸ì˜ ì²˜ìŒ 8ê°€ì§€ ìƒ‰ìƒ.

ì´ ìƒ‰ìƒ íŒ”ë ˆíŠ¸ì˜ ê¸°ì›ì€ 1980ë…„ëŒ€ í•˜ë“œì›¨ì–´ë¡œ ê±°ìŠ¬ëŸ¬ ì˜¬ë¼ê°‘ë‹ˆë‹¤. ë‹¹ì‹œ ê·¸ë˜í”½ ì¹´ë“œëŠ” ìŒê·¹ì„ ê´€(CRT)ì˜ ì„¸ ê°€ì§€ ê¸°ë³¸ ìƒ‰ìƒ ì±„ë„ì¸ ë¹¨ê°•, ì´ˆë¡, íŒŒë‘(RGB)ì„ ê°ê° ì¼œê±°ë‚˜ ë„ëŠ” ë°©ì‹ìœ¼ë¡œ ìƒ‰ìƒì„ ì²˜ë¦¬í–ˆìŠµë‹ˆë‹¤. ì´ë¡œ ì¸í•´ RGB ìƒ‰ìƒ ì…ë°©ì²´ì˜ ì—¬ëŸ ê°œ ëª¨ì„œë¦¬ì—ì„œ \\(2^3=8\\)ê°€ì§€ ì¡°í•©ì´ ë§Œë“¤ì–´ì¡ŒìŠµë‹ˆë‹¤14.

14 ë”°ë¼ì„œ ê·¸ë¦¼ 3.37ì˜ 8ë²ˆì§¸ ìƒ‰ìƒì€ í°ìƒ‰ì´ì–´ì•¼ í•˜ì§€ë§Œ, ëŒ€ì‹  íšŒìƒ‰ìœ¼ë¡œ í‘œì‹œë©ë‹ˆë‹¤.

The first eight colors for a categorical variable in
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** are shown in
Figure 3.38:

    
    
    ggplot(tibble(u = factor(1:8), v = 1), 
           aes(x = "",  y = v, fill = u)) +
      geom_bar(stat = "identity", width = 1) + 
      coord_polar("y", start = 0) + theme_void()__

[![](03-chap_files/figure-html/fig-graphics-
ggplot2colorpie-1.png)](03-chap_files/figure-html/fig-graphics-
ggplot2colorpie-1.png "FigureÂ 3.38: The first eight colors in the ggplot2
color palette.")

Figure 3.38: The first eight colors in the
**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** color palette.

These defaults are appropriate for simple use cases, but often we will want to
make our own choices. In Section 3.7 we already saw the function
`scale_fill_gradientn`, with which we created the smooth-looking color
gradient used in Figures 3.30 and 3.31 by interpolating a set of color steps
provided by the function `brewer.pal` in the
**[RColorBrewer](https://cran.r-project.org/web/packages/RColorBrewer/)**
package. This package defines a set of purpose-designed color palettes. We can
see all of them at a glance with the function `display.brewer.all` (Figure
3.39).

    
    
    display.brewer.all()__

[![](03-chap_files/figure-html/fig-graphics-
RColorBrewer-1.png)](03-chap_files/figure-html/fig-graphics-RColorBrewer-1.png
"FigureÂ 3.39: RColorBrewer palettes.")

Figure 3.39: RColorBrewer palettes.

We can get information about the available color palettes from
`brewer.pal.info`.

    
    
    head(brewer.pal.info)__
    
    
         maxcolors category colorblind
    BrBG        11      div       TRUE
    PiYG        11      div       TRUE
    PRGn        11      div       TRUE
    PuOr        11      div       TRUE
    RdBu        11      div       TRUE
    RdGy        11      div      FALSE
    
    
    table(brewer.pal.info$category)__
    
    
     div qual  seq 
       9    8   18 

The palettes are divided into three categories:

  * qualitative: for categorical properties that have no intrinsic ordering. The `Paired` palette supports up to 6 categories that each fall into two subcategories - such as _before_ and _after_ , _with_ and _without_ treatment, etc.

  * sequential: for quantitative properties that go from _low_ to _high_

  * diverging: for quantitative properties for which there is a natural midpoint or neutral value, and whose value can deviate both up- and down; weâ€™ll see an example in Figure 3.41.

To obtain the colors from a particular palette we use the function
`brewer.pal`. Its first argument is the number of colors we want (which can be
less than the available maximum number in `brewer.pal.info`).

    
    
    brewer.pal(4, "RdYlGn")__
    
    
    [1] "#D7191C" "#FDAE61" "#A6D96A" "#1A9641"

If we want more than the available number of preset colors (for example so we
can plot a heatmap with continuous colors) we can interpolate using the
`colorRampPalette` function15.

15 `colorRampPalette` returns a function of one parameter, an integer. In the
code shown, we call that function with the argument `100`.

    
    
    mypalette  = colorRampPalette(
        c("darkorange3", "white","darkblue")
      )(100)
    head(mypalette)__
    
    
    [1] "#CD6600" "#CE6905" "#CF6C0A" "#D06F0F" "#D17214" "#D27519"
    
    
    image(matrix(1:100, nrow = 100, ncol = 10), col = mypalette,
            xaxt = "n", yaxt = "n", useRaster = TRUE)__

[![](03-chap_files/figure-html/fig-graphics-
colorRampPalette-1.png)](03-chap_files/figure-html/fig-graphics-
colorRampPalette-1.png "FigureÂ 3.40: A quasi-continuous color palette derived
by interpolating between the colors darkorange3, white and darkblue.")

Figure 3.40: A quasi-continuous color palette derived by interpolating between
the colors `darkorange3`, `white` and `darkblue`.

## 3.10 Heatmaps

Heatmaps are a powerful way of visualizing large, matrix-like datasets and
providing a quick overview of the patterns that might be in the data. There
are a number of heatmap drawing functions in R; one that is convenient and
produces a good-looking output is the function `pheatmap` from the eponymous
package16. In the code below, we first select the top 500 most variable genes
in the dataset `x` and define a function `rowCenter` that centers each gene
(row) by subtracting the mean across columns. By default, `pheatmap` uses the
_RdYlBu_ color palette from
**[RcolorBrewer](https://cran.r-project.org/web/packages/RcolorBrewer/)** in
conjuction with the `colorRampPalette` function to interpolate the 11 colors
into a smooth-looking palette (Figure 3.41).

16 A very versatile and modular alternative is the
**[ComplexHeatmap](https://bioconductor.org/packages/ComplexHeatmap/)**
package.

    
    
    library("pheatmap")
    topGenes = order(rowVars(Biobase::exprs(x)), decreasing = TRUE)[1:500]
    rowCenter = function(x) { x - rowMeans(x) }
    pheatmap(rowCenter(dfx[topGenes, ]), 
      show_rownames = FALSE, 
      show_colnames = FALSE, 
      breaks = seq(-5, +5, length = 101),
      annotation_col = pData(x)[, c("sampleGroup", "Embryonic.day", "ScanDate", "genotype") ],
      annotation_colors = list(
        sampleGroup = groupColor,
        genotype = c(`FGF4-KO` = "chocolate1", `WT` = "azure2"),
        Embryonic.day = setNames(brewer.pal(9, "Blues")[c(3, 6, 9)], c("E3.25", "E3.5", "E4.5")),
        ScanDate = setNames(brewer.pal(nlevels(x$ScanDate), "YlGn"), levels(x$ScanDate))
      )
    )__

[![](03-chap_files/figure-html/fig-graphics-
heatmap-1.png)](03-chap_files/figure-html/fig-graphics-heatmap-1.png
"FigureÂ 3.41: A heatmap of relative expression values, i.e., logarithmic fold
change compared to the average expression of that gene \(row\) across all
samples \(columns\). The color scale uses a diverging palette whose midpoint
is at 0.")

Figure 3.41: A heatmap of relative expression values, i.e., logarithmic fold
change compared to the average expression of that gene (row) across all
samples (columns). The color scale uses a diverging palette whose midpoint is
at 0.

Letâ€™s take a minute to deconstruct this rather massive-looking call to
`pheatmap`. The options `show_rownames` and `show_colnames` control whether
the row and column names are printed at the sides of the matrix. Because our
matrix is large in relation to the available plotting space, the labels would
not be readable, so we might as well suppress them. The `annotation_col`
argument takes a data frame that carries additional information about the
samples. The information is shown in the colored bars on top of the heatmap.
There is also a similar `annotation_row` argument, which we havenâ€™t used here,
to annotate the rows (genes) with colored bars at the side. The argument
`annotation_colors` is a list of named vectors by which we can override the
default choice of colors for the annotation bars. The `pheatmap` function has
many further options, and if you want to use it for your own data
visualizations, itâ€™s worth studying them.

### 3.10.1 ë´ë“œë¡œê·¸ë¨ ìˆœì„œ(Dendrogram ordering)

ê·¸ë¦¼ 3.41ì—ì„œ ì™¼ìª½ê³¼ ìƒë‹¨ì˜ ë‚˜ë¬´ ëª¨ì–‘ì€ ê³„ì¸µì  í´ëŸ¬ìŠ¤í„°ë§ ì•Œê³ ë¦¬ì¦˜ì˜ ê²°ê³¼ë¥¼ ë‚˜íƒ€ë‚´ë©° **ë´ë“œë¡œê·¸ë¨(dendrograms)**ì´ë¼ê³ ë„ ë¶ˆë¦½ë‹ˆë‹¤. í–‰ê³¼ ì—´ì˜ ìˆœì„œëŠ” ë´ë“œë¡œê·¸ë¨ì„ ê¸°ë°˜ìœ¼ë¡œ í•©ë‹ˆë‹¤. ì´ëŠ” íˆìŠ¤í† ê·¸ë¨ì˜ ì‹œê°ì  íš¨ê³¼ì— ì—„ì²­ë‚œ ì˜í–¥ì„ ë¯¸ì¹©ë‹ˆë‹¤. ê·¸ëŸ¬ë‚˜ ë‚˜íƒ€ë‚˜ëŠ” íŒ¨í„´ ì¤‘ ì–´ëŠ ê²ƒì´ ì‹¤ì œì´ê³  ì–´ëŠ ê²ƒì´ ì„ì˜ì ì¸ íŠ¸ë¦¬ ë ˆì´ì•„ì›ƒ ê²°ì •ì˜ ê²°ê³¼ì¸ì§€ ê²°ì •í•˜ê¸° ì–´ë ¤ìš¸ ìˆ˜ ìˆìŠµë‹ˆë‹¤17. ë‹¤ìŒ ì‚¬í•­ì„ ì—¼ë‘ì— ë‘¡ì‹œë‹¤:

17 ìš°ë¦¬ëŠ” [5ì¥](05-chap.html)ì—ì„œ í´ëŸ¬ìŠ¤í„°ë§ê³¼ í´ëŸ¬ìŠ¤í„°ì˜ ìœ ì˜ì„±ì„ í‰ê°€í•˜ëŠ” ë°©ë²•ì— ëŒ€í•´ ë°°ìš¸ ê²ƒì…ë‹ˆë‹¤.

  * í´ëŸ¬ìŠ¤í„° ë´ë“œë¡œê·¸ë¨ì— ë”°ë¼ í–‰ê³¼ ì—´ì„ ì •ë ¬í•˜ëŠ” ê²ƒ(ê·¸ë¦¼ 3.41ì—ì„œì²˜ëŸ¼)ì€ ì„ì˜ì ì¸ ì„ íƒì´ë©°, ë‹¤ë¥¸ ë°©ì‹ìœ¼ë¡œ ì •ë ¬í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤.

  * ë´ë“œë¡œê·¸ë¨ ìˆœì„œë¡œ ê²°ì •í•˜ë”ë¼ë„, ê° ë‚´ë¶€ ë¶„ê¸°(internal branch)ì—ì„œ ì–´ëŠ ìª½ì„ ì™¼ìª½ ë˜ëŠ” ì˜¤ë¥¸ìª½ìœ¼ë¡œ ë‘˜ì§€ëŠ” ë³¸ì§ˆì ìœ¼ë¡œ ì„ì˜ì ì¸ ì„ íƒì…ë‹ˆë‹¤. íŠ¸ë¦¬ì˜ ìœ„ìƒ(topology)ì„ ë°”ê¾¸ì§€ ì•Šê³ ë„ ê° ë¶„ê¸°ë¥¼ ë’¤ì§‘ì„ ìˆ˜ ìˆê¸° ë•Œë¬¸ì…ë‹ˆë‹¤([ê·¸ë¦¼ 5.21](05-chap.html#fig-mobile) ì°¸ì¡°).

__

ì§ˆë¬¸ 3.6

`pheatmap` í•¨ìˆ˜ëŠ” ì„œë¸ŒíŠ¸ë¦¬ì˜ ì–´ëŠ ë¶„ê¸°ë¥¼ ì™¼ìª½ê³¼ ì˜¤ë¥¸ìª½ìœ¼ë¡œ ë³´ë‚¼ì§€ ê²°ì •í•˜ëŠ” ë¬¸ì œë¥¼ ì–´ë–»ê²Œ ì²˜ë¦¬í•˜ë‚˜ìš”?

__

í•´ê²°ì±…

__

ì´ëŠ” ê¸°ë³¸ì ìœ¼ë¡œ `pheatmap`ì—ì„œ ì‚¬ìš©ë˜ëŠ” **[stats](https://cran.r-project.org/web/packages/stats/)** íŒ¨í‚¤ì§€ì˜ `hclust` í•¨ìˆ˜ ë§¤ë‰´ì–¼ í˜ì´ì§€ì— ì„¤ëª…ë˜ì–´ ìˆìŠµë‹ˆë‹¤.

__

ì§ˆë¬¸ 3.7

ë‹¤ë¥¸ ì–´ë–¤ ì •ë ¬ ë°©ë²•ë“¤ì„ ìƒê°í•  ìˆ˜ ìˆë‚˜ìš”?

__

í•´ê²°ì±…

__

ì œì•ˆëœ ë°©ë²•ë“¤ ì¤‘ì—ëŠ” ì™¸íŒì› ë¬¸ì œ(travelling salesman problem, [McCormick Jr, Schweitzer, and White 1972](16-chap.html#ref-McCormick:1972:MatrixClustering))ë‚˜ ì²« ë²ˆì§¸ ì£¼ì„±ë¶„ìœ¼ë¡œì˜ íˆ¬ì˜(ì˜ˆë¥¼ ë“¤ì–´ `pheatmap` ë§¤ë‰´ì–¼ í˜ì´ì§€ì˜ ì˜ˆì œ ì°¸ì¡°) ë“±ì´ ìˆìŠµë‹ˆë‹¤.

__

ì§ˆë¬¸ 3.8

`pheatmap` í•¨ìˆ˜ì˜ `clustering_callback` ì¸ìˆ˜ë¥¼ í™•ì¸í•´ ë³´ì„¸ìš”.

### 3.10.2 ìƒ‰ ê³µê°„(Color spaces)

ì¸ê°„ì˜ ìƒ‰ ì§€ê°([Helmholtz 1867](16-chap.html#ref-Helmholtz:1867))ì€ 3ì°¨ì›ì ì…ë‹ˆë‹¤18. ì´ ê³µê°„ì„ ë§¤ê°œë³€ìˆ˜í™”í•˜ëŠ” ë°©ë²•ì€ ì—¬ëŸ¬ ê°€ì§€ê°€ ìˆìŠµë‹ˆë‹¤. ìœ„ì—ì„œ ìš°ë¦¬ëŠ” ì´ë¯¸ 3.4ì ˆ ì‹œì‘ ë¶€ë¶„ì—ì„œ `groupColor`ì˜ ë‚´ìš©ì„ ì¶œë ¥í•  ë•Œ [0,1] ë²”ìœ„ì˜ ì„¸ ê°’ì„ ì‚¬ìš©í•˜ëŠ” RGB ìƒ‰ìƒ ëª¨ë¸ì„ ì ‘í–ˆìŠµë‹ˆë‹¤.

18 ë¬¼ë¦¬ì ìœ¼ë¡œëŠ” ë¹›ì˜ íŒŒì¥ì´ ë¬´ìˆ˜íˆ ë§ê³  ì´ë¥¼ í˜¼í•©í•˜ëŠ” ë°©ë²•ë„ ë¬´ìˆ˜íˆ ë§ê¸° ë•Œë¬¸ì—, ë‹¤ë¥¸ ì¢…ì´ë‚˜ ë¡œë´‡ì˜ ê²½ìš° ìƒ‰ ê³µê°„ì´ ë” ì ê±°ë‚˜ ë” ë§ì€ ì°¨ì›ì„ ê°€ì§ˆ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

    
    
    groupColor[1]__
    
    
        E3.25 
    "#CAB2D6" 

ì—¬ê¸°ì„œ CAëŠ” ë¹¨ê°„ìƒ‰ ì±„ë„ì˜ ê°•ë„ë¥¼ ë‚˜íƒ€ë‚´ëŠ” 16ì§„ìˆ˜ í‘œí˜„ì´ë©°, B2ëŠ” ë…¹ìƒ‰, D6ì€ íŒŒë€ìƒ‰ ì±„ë„ì˜ ê°•ë„ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤. ì‹­ì§„ìˆ˜ë¡œ ë³€í™˜í•˜ë©´ ì´ ìˆ«ìë“¤ì€ ê°ê° 202, 178, 214ì…ë‹ˆë‹¤. ì´ ê°’ë“¤ì˜ ë²”ìœ„ëŠ” 0ì—ì„œ 255ê¹Œì§€ì´ë¯€ë¡œ, ì´ë¥¼ ìµœëŒ€ê°’ìœ¼ë¡œ ë‚˜ëˆ„ë©´ RGB ì‚¼ì¤‘í•­(triplet)ì„ 3ì°¨ì› ë‹¨ìœ„ ì…ë°©ì²´ ë‚´ì˜ í•œ ì ìœ¼ë¡œ ìƒê°í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤.

[![](03-chap_files/figure-html/fig-graphics-hcl-1.png)](03-chap_files/figure-
html/fig-graphics-hcl-1.png "ê·¸ë¦¼ 3.42 (a): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-hcl-2.png)](03-chap_files/figure-
html/fig-graphics-hcl-2.png "ê·¸ë¦¼ 3.42 (b): ")

(b)

ê·¸ë¦¼ 3.42: HCL ìƒ‰ ê³µê°„ì˜ ì›. (a) íœ˜ë„(Luminance) \\(L\\)ì€ 75ë¡œ ê³ ì •ëœ ìƒíƒœì—ì„œ ê°ë„ ì¢Œí‘œ \\(H\\)(ìƒ‰ìƒ, hue)ëŠ” 0ì—ì„œ 360ê¹Œì§€ ë³€í•˜ê³  ë°˜ì§€ë¦„ ì¢Œí‘œ \\(C\\)ëŠ” 0, 10, ..., 60ì˜ ê°’ì„ ê°€ì§‘ë‹ˆë‹¤. (b) í¬ë¡œë§ˆ(Chroma) \\(C\\)ëŠ” 50ìœ¼ë¡œ ê³ ì •ë˜ê³  \\(H\\)ëŠ” ìœ„ì™€ ê°™ìœ¼ë©°, ë°˜ì§€ë¦„ ì¢Œí‘œëŠ” íœ˜ë„ \\(L\\)ë¡œ 10, 20, ..., 90ì˜ ê°’ì„ ê°€ì§‘ë‹ˆë‹¤.

`hcl` í•¨ìˆ˜ëŠ” ë‹¤ë¥¸ ì¢Œí‘œê³„ë¥¼ ì‚¬ìš©í•©ë‹ˆë‹¤. ì´ ì—­ì‹œ ìƒ‰ìƒ(hue) \\(H\\)(\\([0, 360]\\) ë²”ìœ„ì˜ ê°ë„), í¬ë¡œë§ˆ(chroma) \\(C\\)(ì–‘ìˆ˜), íœ˜ë„(luminance) \\(L\\)(\\([0, 100]\\) ë²”ìœ„ì˜ ê°’)ë¼ëŠ” ì„¸ ê°€ì§€ ì¢Œí‘œë¡œ êµ¬ì„±ë©ë‹ˆë‹¤. \\(C\\)ì˜ ìƒí•œì„ ì€ ìƒ‰ìƒê³¼ íœ˜ë„ì— ë”°ë¼ ë‹¬ë¼ì§‘ë‹ˆë‹¤.

`hcl` í•¨ìˆ˜ëŠ” CIE-LUV19ì˜ ê·¹ì¢Œí‘œ(polar coordinates)ì— ëŒ€ì‘í•˜ë©° ì˜ì—­ ì±„ìš°ê¸°(area fill)ë¥¼ ìœ„í•´ ì„¤ê³„ë˜ì—ˆìŠµë‹ˆë‹¤. í¬ë¡œë§ˆì™€ íœ˜ë„ ì¢Œí‘œë¥¼ ì¼ì •í•˜ê²Œ ìœ ì§€í•˜ê³  ìƒ‰ìƒë§Œ ë³€í™”ì‹œí‚´ìœ¼ë¡œì¨, ì¡°í™”ë¡œìš´ ìƒ‰ìƒ íŒ”ë ˆíŠ¸ë¥¼ ì‰½ê²Œ ë§Œë“¤ ìˆ˜ ìˆìœ¼ë©° ë°ì€ ìƒ‰ ì˜ì—­ì´ ì–´ë‘ìš´ ì˜ì—­ë³´ë‹¤ ë” ì»¤ ë³´ì´ëŠ” ì¡°ì‚¬(irradiation) ì°©ì‹œë¥¼ í”¼í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë˜í•œ ìš°ë¦¬ì˜ ì‹œì„ ì€ ê°•ë ¬í•œ ìƒ‰ìƒì— ëŒë¦¬ëŠ” ê²½í–¥ì´ ìˆëŠ”ë°, í¬ë¡œë§ˆ ê°’ì„ ê³ ì •í•˜ë©´ ëª¨ë“  ìƒ‰ìƒì´ ìš°ë¦¬ ëˆˆì— ë™ì¼í•˜ê²Œ ë§¤ë ¥ì ìœ¼ë¡œ ë³´ì´ê²Œ ë©ë‹ˆë‹¤.

19 CIE: êµ­ì œ ì¡°ëª… ìœ„ì›íšŒ(Commission Internationale de lâ€™Ã‰clairage) â€“ ìì„¸í•œ ë‚´ìš©ì€ ìœ„í‚¤ë°±ê³¼ ë“±ì„ ì°¸ì¡°í•˜ì„¸ìš”.

ìƒ‰ìƒí™˜(color wheel)ì—ì„œ ìƒ‰ì„ ì„ íƒí•˜ëŠ” ë°©ë²•ì€ ì—¬ëŸ¬ ê°€ì§€ê°€ ìˆìŠµë‹ˆë‹¤. _ì‚¼í•©ìƒ‰(Triads)_ì€ ìƒ‰ìƒí™˜ì—ì„œ ë“±ê°„ê²©ìœ¼ë¡œ ë–¨ì–´ì§„ ì„¸ ê°€ì§€ ìƒ‰ìƒì„ ì„ íƒí•˜ëŠ” ê²ƒì…ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ \\(H=0,120,240\\)ì€ ë¹¨ê°•, ì´ˆë¡, íŒŒë‘ì„ ì¤ë‹ˆë‹¤. _ì‚¬í•©ìƒ‰(Tetrads)_ì€ ìƒ‰ìƒí™˜ì—ì„œ ë“±ê°„ê²©ì¸ ë„¤ ê°€ì§€ ìƒ‰ìƒì´ë©°, ì¼ë¶€ ê·¸ë˜í”½ ì•„í‹°ìŠ¤íŠ¸ë“¤ì€ ê·¸ íš¨ê³¼ë¥¼ "ì—­ë™ì "ì´ë¼ê³  ì„¤ëª…í•©ë‹ˆë‹¤. _ë”°ëœ»í•œ ìƒ‰(Warm colors)_ì€ ë…¸ë€ìƒ‰ì— ê°€ê¹Œìš´ ë“±ê°„ê²© ìƒ‰ìƒ ì„¸íŠ¸ì´ê³ , _ì°¨ê°€ìš´ ìƒ‰(Cool colors)_ì€ íŒŒë€ìƒ‰ì— ê°€ê¹Œìš´ ë“±ê°„ê²© ìƒ‰ìƒ ì„¸íŠ¸ì…ë‹ˆë‹¤. _ìœ ì‚¬ ìƒ‰(Analogous color)_ ì„¸íŠ¸ëŠ” ë…¸ë‘, ì£¼í™©, ë¹¨ê°• ë˜ëŠ” ì´ˆë¡, ì²­ë¡, íŒŒë‘ê³¼ ê°™ì´ ìƒ‰ìƒí™˜ì˜ ì‘ì€ ì„¸ê·¸ë¨¼íŠ¸ ë‚´ì˜ ìƒ‰ìƒë“¤ì„ í¬í•¨í•©ë‹ˆë‹¤. _ë³´ìƒ‰(Complementary colors)_ì€ ìƒ‰ìƒí™˜ì—ì„œ ì„œë¡œ ì •ë°˜ëŒ€í¸ì— ìœ„ì¹˜í•œ ìƒ‰ìƒë“¤ì…ë‹ˆë‹¤. ì‚¬í•©ìƒ‰ì€ ë‘ ìŒì˜ ë³´ìƒ‰ìœ¼ë¡œ ì´ë£¨ì–´ì§‘ë‹ˆë‹¤. _ë¶„í•  ë³´ìƒ‰(Split complementaries)_ì€ í•œ ìŒì˜ ë³´ìƒ‰ ì¤‘ í•˜ë‚˜ë¥¼ ì–‘ì˜†ìœ¼ë¡œ ë“±ê°„ê²©ì¸ ë‘ ìƒ‰ìœ¼ë¡œ ë‚˜ëˆˆ ì„¸ ê°€ì§€ ìƒ‰ìƒì…ë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ \\(H=60,\,240-30,\,240+30\\)ì…ë‹ˆë‹¤. ì´ëŠ” ë‘ ê°œì˜ ìœ ì‚¬í•œ ë²”ì£¼ì™€ ì„¸ ë²ˆì§¸ ë‹¤ë¥¸ ë²”ì£¼ ì‚¬ì´ì˜ ì°¨ì´ë¥¼ ê°•ì¡°í•  ë•Œ ìœ ìš©í•©ë‹ˆë‹¤. ì°¸ê³  ë¬¸í—Œ([Mollon 1995](16-chap.html#ref-Mollon:1995); [Ihaka 2003](16-chap.html#ref-IhakaColorPres))ì—ì„œ ë” ìì„¸í•œ ë…¼ì˜ë¥¼ í™•ì¸í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

#### ì„  ëŒ€ ì˜ì—­(Lines versus areas)

ì„ ê³¼ ì ì˜ ê²½ìš° ë°°ê²½ê³¼ ê°•í•œ ëŒ€ë¹„ë¥¼ ì›í•˜ë¯€ë¡œ, í°ìƒ‰ ë°°ê²½ì—ì„œëŠ” ìƒëŒ€ì ìœ¼ë¡œ ì–´ë‘¡ê²Œ(ë‚®ì€ íœ˜ë„ \\(L\\)) ë§Œë“œëŠ” ê²ƒì´ ì¢‹ìŠµë‹ˆë‹¤. ì˜ì—­ ì±„ìš°ê¸°ì˜ ê²½ìš°, ì±„ë„ê°€ ë‚®ê±°ë‚˜ ì¤‘ê°„ ì •ë„ì¸ ë” ë°ì€ íŒŒìŠ¤í…” í†¤ì˜ ìƒ‰ìƒì´ ë³´í†µ ë” ë³´ê¸° ì¢‹ìŠµë‹ˆë‹¤.

## 3.11 ë°ì´í„° ë³€í™˜(Data transformations)

ëŒ€ë¶€ë¶„ì˜ ì ì´ í•œ ì˜ì—­ì— ë­‰ì³ ìˆê³  ê°€ìš© ê³µê°„ì˜ ìƒë‹¹ ë¶€ë¶„ì´ ë¹„ì–´ ìˆëŠ” í”Œë¡¯ì€ ì½ê¸° ì–´ë µìŠµë‹ˆë‹¤. ë³€ìˆ˜ì˜ ì£¼ë³€ ë¶„í¬(marginal distribution) íˆìŠ¤í† ê·¸ë¨ì´ ë‚ ì¹´ë¡œìš´ ì •ì ì„ ê°–ê³  í•œìª½ ë˜ëŠ” ì–‘ìª½ìœ¼ë¡œ ê¸´ ê¼¬ë¦¬ë¥¼ ê°€ì§„ë‹¤ë©´, ë°ì´í„°ë¥¼ ë³€í™˜í•˜ëŠ” ê²ƒì´ ë„ì›€ì´ ë  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì´ëŸ¬í•œ ê³ ë ¤ ì‚¬í•­ì€ \(x\) ë° \(y\) ë¯¸í•™ë¿ë§Œ ì•„ë‹ˆë¼ ìƒ‰ìƒ ì²™ë„ì—ë„ ì ìš©ë©ë‹ˆë‹¤. ë§ˆì´í¬ë¡œì–´ë ˆì´ ë°ì´í„°ë¥¼ í¬í•¨í•˜ëŠ” ì´ ì¥ì˜ í”Œë¡¯ë“¤ì—ì„œ ìš°ë¦¬ëŠ” ë¡œê·¸ ë³€í™˜20ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤. ì´ëŠ” ê·¸ë¦¼ 3.27ê³¼ ê°™ì€ ì‚°ì ë„ì˜ \(x\) ë° \(y\) ì¢Œí‘œë¿ë§Œ ì•„ë‹ˆë¼, ë°œí˜„ ë°°ìˆ˜ ë³€í™”(fold changes)ë¥¼ ë‚˜íƒ€ë‚´ëŠ” ê·¸ë¦¼ 3.41ì˜ ìƒ‰ìƒ ì²™ë„ì—ì„œë„ ë§ˆì°¬ê°€ì§€ì…ë‹ˆë‹¤. ë¡œê·¸ ë³€í™˜ì€ í™•ì‹¤í•œ ì˜ë¯¸ë¥¼ ê°–ê¸° ë•Œë¬¸ì— ë§¤ë ¥ì ì…ë‹ˆë‹¤. ë¡œê·¸ ë³€í™˜ëœ ì²™ë„ì—ì„œ ë™ì¼í•œ ì–‘ë§Œí¼ ìœ„ì•„ë˜ë¡œ ì›€ì§ì´ëŠ” ê²ƒì€ ì›ë˜ ì²™ë„ì—ì„œ ë™ì¼í•œ ë°°ìˆ˜ ë³€í™”ì— ëŒ€ì‘í•©ë‹ˆë‹¤: \\(\log(ax)=\log a+\log x\\).

20 `ExpressionSet` ê°ì²´ `x`ì˜ ë°ì´í„°ê°€ ì´ë¯¸ ë¡œê·¸ ë³€í™˜ëœ ìƒíƒœë¡œ ì œê³µë˜ë¯€ë¡œ ìš°ë¦¬ëŠ” ì´ë¥¼ ì•”ì‹œì ìœ¼ë¡œ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.

í•˜ì§€ë§Œ ë•Œë¡œëŠ” ë¡œê·¸ ë³€í™˜ë§Œìœ¼ë¡œëŠ” ì¶©ë¶„í•˜ì§€ ì•Šì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´ ë°ì´í„°ì— 0ì´ë‚˜ ìŒìˆ˜ ê°’ì´ í¬í•¨ë˜ì–´ ìˆê±°ë‚˜, ë¡œê·¸ ì²™ë„ì—ì„œë„ ë°ì´í„° ë¶„í¬ê°€ ë§¤ìš° ë¶ˆê· ë“±í•œ ê²½ìš°ì…ë‹ˆë‹¤. ê·¸ë¦¼ 3.43ì˜ ìœ„ìª½ íŒ¨ë„ì„ ë³´ë©´, `A`ê°€ ë‚®ì„ìˆ˜ë¡ ë¶„ì‚°ì´ ë” ì»¤ ë³´ì´ëŠ” ë“± `M`ì˜ ë¶„í¬ê°€ `A`ì— ì˜ì¡´í•œë‹¤ëŠ” ì¸ìƒì„ ë°›ê¸° ì‰½ìŠµë‹ˆë‹¤. ê·¸ëŸ¬ë‚˜ ì´ëŠ” ì „ì ìœ¼ë¡œ ì‹œê°ì  ì°©ì‹œì´ë©°, ì•„ë˜ìª½ íŒ¨ë„ì´ ì´ë¥¼ í™•ì¸í•´ ì¤ë‹ˆë‹¤. `M`ì˜ ë¶„í¬ëŠ” `A`ì™€ ë…ë¦½ì ì´ë©°, ìœ„ìª½ íŒ¨ë„ì—ì„œ ë³´ì•˜ë˜ ê²‰ë³´ê¸° ì¶”ì„¸ëŠ” ì‘ì€ `A`ì—ì„œ ì  ë°€ë„ê°€ ë” ë†’ì•˜ê¸° ë•Œë¬¸ì— ë°œìƒí•œ ê²ƒì…ë‹ˆë‹¤.

    
    
    gg = ggplot(tibble(A = Biobase::exprs(x)[, 1], M = rnorm(length(A))),
                aes(y = M))
    gg + geom_point(aes(x = A), size = 0.2)
    gg + geom_point(aes(x = rank(A)), size = 0.2)__

[![](03-chap_files/figure-html/fig-graphics-MA-1.png)](03-chap_files/figure-
html/fig-graphics-MA-1.png "ê·¸ë¦¼ 3.43 (a): ")

(a)

[![](03-chap_files/figure-html/fig-graphics-MA-2.png)](03-chap_files/figure-
html/fig-graphics-MA-2.png "ê·¸ë¦¼ 3.43 (b): ")

(b)

ê·¸ë¦¼ 3.43: ìˆœìœ„ ë³€í™˜(rank transformation)ì´ ì˜ì¡´ì„±ì˜ ì‹œê°ì  ì§€ê°ì— ë¯¸ì¹˜ëŠ” ì˜í–¥.

__

ì§ˆë¬¸ 3.9

ê·¸ë¦¼ 3.31ì—ì„œì²˜ëŸ¼ ë°€ë„ ê¸°ë°˜ ë˜ëŠ” êµ¬ê°„í™” ê¸°ë°˜ í”Œë¡œíŒ… ë°©ë²•ì„ ì‚¬ìš©í•˜ì—¬ ì‹œê°ì  ì°©ì‹œë¥¼ í”¼í•  ìˆ˜ ìˆì„ê¹Œìš”?

__

ì§ˆë¬¸ 3.10

ìˆœìœ„ ë³€í™˜ì„ íˆíŠ¸ë§µ ë“±ì˜ ìƒ‰ìƒ ì²™ë„ë¥¼ ì„ íƒí•  ë•Œë„ ì ìš©í•  ìˆ˜ ìˆì„ê¹Œìš”? ì´ë¯¸ì§€ ì²˜ë¦¬ì—ì„œ _íˆìŠ¤í† ê·¸ë¨ í‰í™œí™”(histogram equalization)_ëŠ” ë¬´ì—‡ì„ í•˜ë‚˜ìš”?

## 3.12 ìˆ˜í•™ ê¸°í˜¸ ë° ê¸°íƒ€ ê¸€ê¼´

ìš°ë¦¬ëŠ” R êµ¬ë¬¸ê³¼ LaTeX ìŠ¤íƒ€ì¼ í‘œê¸°ë²•ì´ í˜¼í•©ëœ ë°©ì‹ì„ ì‚¬ìš©í•˜ì—¬ í”Œë¡¯ ë ˆì´ë¸”ì— ìˆ˜í•™ì  í‘œê¸°ë²•ì„ ì‚¬ìš©í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤(ìì„¸í•œ ë‚´ìš©ì€ `help("plotmath")`ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”):

[![](imgs/chap3-ecdfs-bauhaus93font.png)](imgs/chap3-ecdfs-bauhaus93font.png
"ê·¸ë¦¼ 3.44: ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ, â€œBauhaus 93â€ ê¸€ê¼´ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.44: ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ, â€œBauhaus 93â€ ê¸€ê¼´ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.

    
    
    volume = function(rho, nu)
                pi^(nu/2) * rho^nu / gamma(nu/2+1)
    
    ggplot(tibble(nu    = 1:15,
      Omega = volume(1, nu)), aes(x = nu, y = Omega)) +
    geom_line() +
    xlab(expression(nu)) + ylab(expression(Omega)) +
    geom_text(label =
    "Omega(rho,nu)==frac(pi^frac(nu,2)~rho^nu, Gamma(frac(nu,2)+1))",
      parse = TRUE, x = 6, y = 1.5)__

[![](03-chap_files/figure-html/fig-graphics-
mathnot-1.png)](03-chap_files/figure-html/fig-graphics-mathnot-1.png
"ê·¸ë¦¼ 3.45: ë°˜ì§€ë¦„ \\rho=1ì¸ \\nuì°¨ì› êµ¬ì˜ ë¶€í”¼ \\Omega (\\nu=1,...,15).")

ê·¸ë¦¼ 3.45: ë°˜ì§€ë¦„ \\(\rho=1\\)ì¸ \\(\nu\\)ì°¨ì› êµ¬ì˜ ë¶€í”¼ \\(\Omega\\) (\\(\nu=1,...,15\\)).

ê²°ê³¼ëŠ” ê·¸ë¦¼ 3.45ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤. ë˜í•œ ì„¸ë¦¬í”„(serif) ê¸€ê¼´ì¸ Timesì™€ ê°™ì€ ë‹¤ë¥¸ ê¸€ê¼´ë¡œ ì „í™˜í•˜ëŠ” ê²ƒë„ ì‰½ìŠµë‹ˆë‹¤(ê·¸ë¦¼ 3.46).

    
    
    ggplot(genes, aes( x = value, color = gene)) + stat_ecdf() +
      theme(text = element_text(family = "Times"))__

[![](03-chap_files/figure-html/fig-graphics-
timesfont-1.png)](03-chap_files/figure-html/fig-graphics-timesfont-1.png
"ê·¸ë¦¼ 3.46: ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ, ë‹¤ë¥¸ ê¸€ê¼´ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.")

ê·¸ë¦¼ 3.46: ê·¸ë¦¼ 3.24ì™€ ìœ ì‚¬í•˜ì§€ë§Œ, ë‹¤ë¥¸ ê¸€ê¼´ì„ ì‚¬ìš©í–ˆìŠµë‹ˆë‹¤.

ì‚¬ì‹¤ í‘œì¤€ R ì„¤ì¹˜ì—ì„œ ì‚¬ìš©í•  ìˆ˜ ìˆëŠ” ê¸€ê¼´ ì„¸íŠ¸ëŠ” ì œí•œì ì´ì§€ë§Œ, ë‹¤í–‰íˆ Rì˜ í‘œì¤€ PostScript ê¸€ê¼´ ì´ì™¸ì˜ ê¸€ê¼´ì„ ì‰½ê²Œ ì‚¬ìš©í•  ìˆ˜ ìˆë„ë¡ ë„ì™€ì£¼ëŠ” **[extrafont](https://cran.r-project.org/web/packages/extrafont/)** íŒ¨í‚¤ì§€ê°€ ìˆìŠµë‹ˆë‹¤. R ì™¸ë¶€ì˜ ê¸€ê¼´ì„ ë¨¼ì € Rì— ì¸ì‹ì‹œì¼œì•¼ í•˜ë¯€ë¡œ ì‚¬ìš©í•˜ê¸° ì „ì— ì¶”ê°€ ì‘ì—…ì´ í•„ìš”í•©ë‹ˆë‹¤. ê¸€ê¼´ì€ ìš´ì˜ ì²´ì œ, ì›Œë“œ í”„ë¡œì„¸ì„œ ë˜ëŠ” ë‹¤ë¥¸ ê·¸ë˜í”½ ì• í”Œë¦¬ì¼€ì´ì…˜ê³¼ í•¨ê»˜ ì„¤ì¹˜ë˜ì—ˆì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤. ë”°ë¼ì„œ ì‚¬ìš© ê°€ëŠ¥í•œ ê¸€ê¼´ ì„¸íŠ¸ì™€ ë¬¼ë¦¬ì  ìœ„ì¹˜ëŠ” í‘œì¤€í™”ë˜ì–´ ìˆì§€ ì•Šìœ¼ë©°, ìš´ì˜ ì²´ì œ ë° ì¶”ê°€ êµ¬ì„±ì— ë”°ë¼ ë‹¬ë¼ì§‘ë‹ˆë‹¤. **[extrafont](https://cran.r-project.org/web/packages/extrafont/)** íŒ¨í‚¤ì§€ë¥¼ ë¶ˆëŸ¬ì˜¨ í›„ ì²« ë²ˆì§¸ ì„¸ì…˜ì—ì„œëŠ” ê¸€ê¼´ì„ ê°€ì ¸ì™€ íŒ¨í‚¤ì§€ì— ì•Œë¦¬ê¸° ìœ„í•´ `font_import` í•¨ìˆ˜ë¥¼ ì‹¤í–‰í•´ì•¼ í•©ë‹ˆë‹¤. ê·¸ëŸ° ë‹¤ìŒ ê¸€ê¼´ì„ ì‚¬ìš©í•˜ë ¤ëŠ” ê° ì„¸ì…˜ì—ì„œ í•˜ë‚˜ ì´ìƒì˜ R ê·¸ë˜í”½ ì¥ì¹˜ì— ë“±ë¡í•˜ê¸° ìœ„í•´ `loadfonts` í•¨ìˆ˜ë¥¼ í˜¸ì¶œí•´ì•¼ í•©ë‹ˆë‹¤. ë§ˆì§€ë§‰ìœ¼ë¡œ `fonttable` í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ì‚¬ìš© ê°€ëŠ¥í•œ ê¸€ê¼´ ëª©ë¡ì„ í™•ì¸í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ìì‹ ì˜ ë¨¸ì‹ ì—ì„œ ì´ ì‘ì—…ì„ ìˆ˜í–‰í•˜ëŠ” ë°©ë²•ì€ **[extrafonts](https://cran.r-project.org/web/packages/extrafonts/)** íŒ¨í‚¤ì§€ì˜ ë¬¸ì„œë¥¼ ì°¸ì¡°í•´ì•¼ í•©ë‹ˆë‹¤.

__

íƒœìŠ¤í¬

**[extrafont](https://cran.r-project.org/web/packages/extrafont/)** íŒ¨í‚¤ì§€ë¥¼ ì‚¬ìš©í•˜ì—¬ "Bauhaus 93"(ë˜ëŠ” ì‹œìŠ¤í…œì—ì„œ ì‚¬ìš© ê°€ëŠ¥í•œ ë‹¤ë¥¸ ê¸€ê¼´) ê¸€ê¼´ë¡œ ê·¸ë¦¼ 3.46ì˜ ë²„ì „ì„ ë§Œë“¤ì–´ ë³´ì„¸ìš”.

## 3.13 Genomic data

[![](imgs/EBI-genomebrowser-rnaseq.png)](imgs/EBI-genomebrowser-rnaseq.png
"FigureÂ 3.47: Screenshot from Ensembl genome browser, showing gene annotation
of a genomic region as well as a read pile-up visualization of an RNA-Seq
experiment.")

Figure 3.47: Screenshot from Ensembl genome browser, showing gene annotation
of a genomic region as well as a read pile-up visualization of an RNA-Seq
experiment.

To visualize genomic data, in addition to the general principles we have
discussed in this chapter, there are some specific considerations. The data
are usually associated with genomic coordinates. In fact genomic coordinates
offer a great organizing principle for the integration of genomic data. You
will probably have seen genome browser displays such as in Figure 3.47. Here
we will briefly show how to produce such plots programmatically, using your
data as well as public annotation. It will be a short glimpse, and we refer to
resources such as [Bioconductor](https://www.bioconductor.org) for a fuller
picture.

The main challenge of genomic data visualization is the size of genomes. We
need visualizations at multiple scales, from whole genome down to the
nucleotide level. It should be easy to zoom in and and out, and we may need
different visualization strategies for the different size scales. It can be
convenient to visualize biological molecules (genomes, genes, transcripts,
proteins) in a linear manner, although their embedding in the 3D physical
world can matter (a great deal).

Letâ€™s start with some fun examples, an ideogram plot of human chromosome 1
(Figure 3.48) and a plot of the genome-wide distribution of RNA editing sites
(Figure 3.49).

    
    
    library("ggbio")
    data("hg19IdeogramCyto", package = "biovizBase")
## 3.13 ê²Œë†ˆ ë°ì´í„°(Genomic data)

[![](imgs/EBI-genomebrowser-rnaseq.png)](imgs/EBI-genomebrowser-rnaseq.png
"ê·¸ë¦¼ 3.47: Ensembl ê²Œë†ˆ ë¸Œë¼ìš°ì €ì˜ ìŠ¤í¬ë¦°ìƒ·ìœ¼ë¡œ, ê²Œë†ˆ ì§€ì—­ì˜ ìœ ì „ì ì–´ë…¸í…Œì´ì…˜ê³¼ RNA-Seq ì‹¤í—˜ì˜ ë¦¬ë“œ íŒŒì¼ì—…(read pile-up) ì‹œê°í™”ë¥¼ ë³´ì—¬ì¤ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.47: Ensembl ê²Œë†ˆ ë¸Œë¼ìš°ì €ì˜ ìŠ¤í¬ë¦°ìƒ·. ê²Œë†ˆ ì§€ì—­ì˜ ìœ ì „ì ì–´ë…¸í…Œì´ì…˜ê³¼ RNA-Seq ì‹¤í—˜ì˜ ë¦¬ë“œ íŒŒì¼ì—…(read pile-up) ì‹œê°í™”ë¥¼ ë³´ì—¬ì¤ë‹ˆë‹¤.

ê²Œë†ˆ ë°ì´í„°ë¥¼ ì‹œê°í™”í•˜ë ¤ë©´ ì´ ì¥ì—ì„œ ë…¼ì˜í•œ ì¼ë°˜ì ì¸ ì›ì¹™ ì™¸ì—ë„ ëª‡ ê°€ì§€ êµ¬ì²´ì ì¸ ê³ ë ¤ ì‚¬í•­ì´ ìˆìŠµë‹ˆë‹¤. ë°ì´í„°ëŠ” ëŒ€ê°œ ê²Œë†ˆ ì¢Œí‘œì™€ ì—°ê´€ë˜ì–´ ìˆìŠµë‹ˆë‹¤. ì‚¬ì‹¤ ê²Œë†ˆ ì¢Œí‘œëŠ” ê²Œë†ˆ ë°ì´í„° í†µí•©ì„ ìœ„í•œ í›Œë¥­í•œ ì¡°ì§í™” ì›ì¹™ì„ ì œê³µí•©ë‹ˆë‹¤. ê·¸ë¦¼ 3.47ê³¼ ê°™ì€ ê²Œë†ˆ ë¸Œë¼ìš°ì € ë””ìŠ¤í”Œë ˆì´ë¥¼ ë³¸ ì ì´ ìˆì„ ê²ƒì…ë‹ˆë‹¤. ì—¬ê¸°ì„œëŠ” ê³µê³µ ì–´ë…¸í…Œì´ì…˜ë¿ë§Œ ì•„ë‹ˆë¼ ì—¬ëŸ¬ë¶„ì˜ ë°ì´í„°ë¥¼ ì‚¬ìš©í•˜ì—¬ ì´ëŸ¬í•œ í”Œë¡¯ì„ í”„ë¡œê·¸ë˜ë° ë°©ì‹ìœ¼ë¡œ ìƒì„±í•˜ëŠ” ë°©ë²•ì„ ê°„ë‹¨íˆ ë³´ì—¬ë“œë¦¬ê² ìŠµë‹ˆë‹¤. ì´ëŠ” ì§§ì€ ë§›ë³´ê¸°ì— ë¶ˆê³¼í•˜ë©°, ë” ì™„ì „í•œ ê·¸ë¦¼ì„ ë³´ë ¤ë©´ [Bioconductor](https://www.bioconductor.org)ì™€ ê°™ì€ ìë£Œë¥¼ ì°¸ì¡°í•˜ì‹œê¸° ë°”ëë‹ˆë‹¤.

ê²Œë†ˆ ë°ì´í„° ì‹œê°í™”ì˜ ì£¼ìš” ê³¼ì œëŠ” ê²Œë†ˆì˜ í¬ê¸°ì…ë‹ˆë‹¤. ì „ì²´ ê²Œë†ˆì—ì„œ ë‰´í´ë ˆì˜¤íƒ€ì´ë“œ ìˆ˜ì¤€ê¹Œì§€ ë‹¤ì¤‘ ì²™ë„ì—ì„œì˜ ì‹œê°í™”ê°€ í•„ìš”í•©ë‹ˆë‹¤. í™•ëŒ€ ë° ì¶•ì†Œê°€ ì‰¬ì›Œì•¼ í•˜ë©°, í¬ê¸° ì²™ë„ì— ë”°ë¼ ì„œë¡œ ë‹¤ë¥¸ ì‹œê°í™” ì „ëµì´ í•„ìš”í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. ìƒë¬¼í•™ì  ë¶„ì(ê²Œë†ˆ, ìœ ì „ì, ì „ì‚¬ì²´, ë‹¨ë°±ì§ˆ)ë¥¼ ì„ í˜•ì ìœ¼ë¡œ ì‹œê°í™”í•˜ëŠ” ê²ƒì´ í¸ë¦¬í•  ìˆ˜ ìˆì§€ë§Œ, 3D ë¬¼ë¦¬ì  ì„¸ê³„ì—ì„œì˜ ë°°ì¹˜ê°€ (ë§¤ìš°) ì¤‘ìš”í•  ìˆ˜ë„ ìˆìŠµë‹ˆë‹¤.

ëª‡ ê°€ì§€ ì¬ë¯¸ìˆëŠ” ì˜ˆì‹œë¶€í„° ì‹œì‘í•´ ë³´ê² ìŠµë‹ˆë‹¤. ì¸ê°„ 1ë²ˆ ì—¼ìƒ‰ì²´ì˜ ì´ë°ì˜¤ê·¸ë¨(ideogram) í”Œë¡¯(ê·¸ë¦¼ 3.48)ê³¼ RNA í¸ì§‘ ë¶€ìœ„ì˜ ì „ ê²Œë†ˆ ë¶„í¬ í”Œë¡¯(ê·¸ë¦¼ 3.49)ì…ë‹ˆë‹¤.

    
    
    library("ggbio")
    data("hg19IdeogramCyto", package = "biovizBase")
    plotIdeogram(hg19IdeogramCyto, subchr = "chr1")__

[![](03-chap_files/figure-html/fig-graphics-
ideogram-1-1.png)](03-chap_files/figure-html/fig-graphics-ideogram-1-1.png
"ê·¸ë¦¼ 3.48: ì¸ê°„ ê²Œë†ˆì˜ 1ë²ˆ ì—¼ìƒ‰ì²´: ì´ë°ì˜¤ê·¸ë¨ í”Œë¡¯.")

ê·¸ë¦¼ 3.48: ì¸ê°„ ê²Œë†ˆì˜ 1ë²ˆ ì—¼ìƒ‰ì²´: ì´ë°ì˜¤ê·¸ë¨ í”Œë¡¯.

`darned_hg19_subset500`ì€ ì¸ê°„ ê²Œë†ˆì—ì„œ ì„ íƒëœ 500ê°œì˜ RNA í¸ì§‘ ë¶€ìœ„ë¥¼ ë‚˜ì—´í•©ë‹ˆë‹¤. ì´ëŠ” íŒŒë¦¬, ìƒì¥ ë° ì¸ê°„ì˜ RNA í¸ì§‘ ë°ì´í„°ë² ì´ìŠ¤(DARNED, <https://darned.ucc.ie/>)ì—ì„œ ê°€ì ¸ì˜¨ ê²ƒì…ë‹ˆë‹¤. ê²°ê³¼ëŠ” ê·¸ë¦¼ 3.49ì— ë‚˜ì™€ ìˆìŠµë‹ˆë‹¤.

    
    
    library("GenomicRanges")
    data("darned_hg19_subset500", package = "biovizBase")
    autoplot(darned_hg19_subset500, layout = "karyogram",
             aes(color = exReg, fill = exReg))__

[![](03-chap_files/figure-html/fig-graphics-
darned1-1.png)](03-chap_files/figure-html/fig-graphics-darned1-1.png
"ê·¸ë¦¼ 3.49: RNA í¸ì§‘ ë¶€ìœ„ê°€ í¬í•¨ëœ ì¹´ë¦¬ì˜¤ê·¸ë¨(Karyogram). exRegëŠ” í•´ë‹¹ ë¶€ìœ„ê°€ ì½”ë”© ì˜ì—­(C), 3â€™- ë˜ëŠ” 5â€™-UTRì— ìˆëŠ”ì§€ ì—¬ë¶€ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.49: RNA í¸ì§‘ ë¶€ìœ„ê°€ í¬í•¨ëœ ì¹´ë¦¬ì˜¤ê·¸ë¨(Karyogram). `exReg`ëŠ” í•´ë‹¹ ë¶€ìœ„ê°€ ì½”ë”© ì˜ì—­(C), 3â€™- ë˜ëŠ” 5â€™-UTRì— ìˆëŠ”ì§€ ì—¬ë¶€ë¥¼ ë‚˜íƒ€ëƒ…ë‹ˆë‹¤.

__

ì§ˆë¬¸ 3.11

ê·¸ë¦¼ 3.49ì—ì„œ ì—¼ìƒ‰ì²´ì˜ ìˆœì„œë¥¼ ì–´ë–»ê²Œ ê³ ì¹˜ê³  ì—¼ìƒ‰ì²´ ê¸¸ì´ì— ëŒ€í•œ ê²½ê³ ë¥¼ ì–´ë–»ê²Œ ì—†ì•¨ ìˆ˜ ìˆì„ê¹Œìš”?

__

í•´ê²°ì±…

__

ì¸ê°„ ê²Œë†ˆì˜ hg19 ì–´ì…ˆë¸”ë¦¬ì—ì„œì˜ ì—¼ìƒ‰ì²´ ê¸¸ì´ì— ëŒ€í•œ ì •ë³´ëŠ” (ì˜ˆë¥¼ ë“¤ì–´) `ideoCyto` ë°ì´í„° ì„¸íŠ¸ì— ì €ì¥ë˜ì–´ ìˆìŠµë‹ˆë‹¤. ì—¼ìƒ‰ì²´ì˜ ìˆœì„œë¥¼ ë‹¤ì‹œ ì •í•˜ê¸° ìœ„í•´ `keepSeqlevels` í•¨ìˆ˜ë¥¼ ì‚¬ìš©í•©ë‹ˆë‹¤. ê·¸ë¦¼ 3.50ì„ ì°¸ì¡°í•˜ì„¸ìš”.

    
    
    data("ideoCyto", package = "biovizBase")
    dn = darned_hg19_subset500
    seqlengths(dn) = seqlengths(ideoCyto$hg19)[names(seqlengths(dn))]
    dn = keepSeqlevels(dn, paste0("chr", c(1:22, "X")))
    autoplot(dn, layout = "karyogram", aes(color = exReg, fill = exReg))__

[![](03-chap_files/figure-html/fig-graphics-
darned2-1.png)](03-chap_files/figure-html/fig-graphics-darned2-1.png
"ê·¸ë¦¼ 3.50: ê·¸ë¦¼ 3.49ì˜ ê°œì„ ëœ ë²„ì „ì…ë‹ˆë‹¤.")

ê·¸ë¦¼ 3.50: ê·¸ë¦¼ 3.49ì˜ ê°œì„ ëœ ë²„ì „.

__

ì§ˆë¬¸ 3.12

`darned_hg19_subset500`ì€ ì–´ë–¤ ìœ í˜•ì˜ ê°ì²´ì¸ê°€ìš”?

__

í•´ê²°ì±…

__

    
    
    darned_hg19_subset500[1:2,]__
    
    
    GRanges object with 2 ranges and 10 metadata columns:
          seqnames    ranges strand |       inchr       inrna         snp
             <Rle> <IRanges>  <Rle> | <character> <character> <character>
      [1]     chr5  86618225      - |           A           I        <NA>
      [2]     chr7  99792382      - |           A           I        <NA>
                 gene      seqReg       exReg      source      ests      esta
          <character> <character> <character> <character> <integer> <integer>
      [1]        <NA>           O        <NA>    amygdala         0         0
      [2]        <NA>           O        <NA>        <NA>         0         0
               author
          <character>
      [1]    15342557
      [2]    15342557
      -------
      seqinfo: 23 sequences from an unspecified genome; no seqlengths

ì´ëŠ” ê²Œë†ˆ ì¢Œí‘œì™€ ì—°ê´€ëœ ë°ì´í„°ë¥¼ ì €ì¥í•˜ê¸° ìœ„í•œ Bioconductor í”„ë¡œì íŠ¸ì˜ íŠ¹ìˆ˜ í´ë˜ìŠ¤ì¸ _GRanges_ ê°ì²´ì…ë‹ˆë‹¤. ì²˜ìŒ ì„¸ ê°œì˜ ì—´ì€ í•„ìˆ˜ ì‚¬í•­ì…ë‹ˆë‹¤: `seqnames`ëŠ” í¬í•¨ëœ ìƒì²´ ê³ ë¶„ìì˜ ì´ë¦„(ìš°ë¦¬ì˜ ê²½ìš° ì¸ê°„ ì—¼ìƒ‰ì²´ ì´ë¦„), `ranges`ëŠ” êµ¬ê°„ì˜ ê²Œë†ˆ ì¢Œí‘œ(ì´ ê²½ìš° ê° êµ¬ê°„ì€ ë‹¨ì¼ ë‰´í´ë ˆì˜¤íƒ€ì´ë“œë¥¼ ë‚˜íƒ€ë‚´ë¯€ë¡œ ê¸¸ì´ëŠ” ëª¨ë‘ 1ì„), ê·¸ë¦¬ê³  RNAê°€ ì „ì‚¬ë˜ëŠ” DNA `strand`ì…ë‹ˆë‹¤. ì´ í´ë˜ìŠ¤ì™€ ê´€ë ¨ ì¸í”„ë¼ë¥¼ ì‚¬ìš©í•˜ëŠ” ë°©ë²•ì— ëŒ€í•œ ìì„¸í•œ ë‚´ìš©ì€ ë¬¸ì„œ(ì˜ˆ: **[GenomicRanges](https://bioconductor.org/packages/GenomicRanges/)** íŒ¨í‚¤ì§€ì˜ ë¹„ë„¤íŠ¸)ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”. ê²Œë†ˆ ì—°ê´€ ë°ì´í„° ì„¸íŠ¸ë¡œ ì‘ì—…í•˜ë ¤ë©´ ì´ë¥¼ ë°°ìš¸ ê°€ì¹˜ê°€ ìˆìŠµë‹ˆë‹¤. ì™œëƒí•˜ë©´ ì´ëŸ¬í•œ ë°ì´í„°ë¥¼ í¸ë¦¬í•˜ê³  íš¨ìœ¨ì ì´ë©° ì•ˆì „í•˜ê²Œ ì¡°ì‘í•  ìˆ˜ ìˆê²Œ í•´ì£¼ê³  ë§ì€ ê°•ë ¥í•œ ìœ í‹¸ë¦¬í‹°ë¥¼ ì œê³µí•˜ê¸° ë•Œë¬¸ì…ë‹ˆë‹¤.

## 3.14 ì´ ì¥ì˜ ìš”ì•½

ë°ì´í„°ë¥¼ "ê°€ê³µë˜ì§€ ì•Šì€" ìƒíƒœë¡œ ë³´ê±°ë‚˜ ì²˜ë¦¬, ìš”ì•½ ë° ì¶”ë¡ ì˜ ë‹¤ì–‘í•œ ë‹¨ê³„ì—ì„œ ì‹œê°í™”í•˜ëŠ” ê²ƒì€ ì‘ìš© í†µê³„í•™ë¿ë§Œ ì•„ë‹ˆë¼ ê³¼í•™ ì „ë°˜ì—ì„œ ê°€ì¥ ì¤‘ìš”í•œ í™œë™ ì¤‘ í•˜ë‚˜ì…ë‹ˆë‹¤. ì—°ì—­ì  ì´ë¡ ì´ ë§ì§€ ì•Šê¸° ë•Œë¬¸ì— êµê³¼ì„œì—ì„œëŠ” ë•Œë•Œë¡œ ì†Œí™€íˆ ë‹¤ë£¨ì–´ì§€ê¸°ë„ í•©ë‹ˆë‹¤. í•˜ì§€ë§Œ ìˆ˜ë§ì€ ì¢‹ì€(ê·¸ë¦¬ê³  ë‚˜ìœ) ê´€í–‰ë“¤ì´ ìˆìœ¼ë©°, ì¼ë‹¨ ì£¼ì˜ë¥¼ ê¸°ìš¸ì´ê¸° ì‹œì‘í•˜ë©´ íŠ¹ì • ê·¸ë˜í”½ì´ ë©”ì‹œì§€ë¥¼ ì „ë‹¬í•˜ëŠ” ë° íš¨ê³¼ì ì¸ì§€, ë˜ëŠ” ê°•ë ¥í•˜ê³  ë¯¸ì ìœ¼ë¡œ ë§¤ë ¥ì ì¸ ë°ì´í„° ì‹œê°í™”ë¥¼ ë§Œë“¤ê¸° ìœ„í•´ ì–´ë–¤ ì„ íƒì„ í•  ìˆ˜ ìˆëŠ”ì§€ ê¸ˆë°© ì•Œê²Œ ë  ê²ƒì…ë‹ˆë‹¤. ì¤‘ìš”í•œ ì˜µì…˜ ì¤‘ì—ëŠ” í”Œë¡¯ ìœ í˜•(**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì—ì„œ `geom`ì´ë¼ ë¶€ë¥´ëŠ” ê²ƒ), ë¹„ìœ¨(ì¢…íš¡ë¹„ í¬í•¨) ë° ìƒ‰ìƒì´ ìˆìŠµë‹ˆë‹¤. _ê·¸ë˜í”½ì˜ ë¬¸ë²•(grammar of graphics)_ì€ ê·¸ë˜í”½ì— ëŒ€í•´ ì¶”ë¡ í•˜ê³  ë°ì´í„° ì‹œê°í™”ì— ëŒ€í•œ ìš°ë¦¬ì˜ ì˜ë„ë¥¼ ì»´í“¨í„°ì— ì „ë‹¬í•˜ê¸° ìœ„í•œ ê°•ë ¥í•œ ê°œë… ì„¸íŠ¸ì…ë‹ˆë‹¤.

ì˜µì…˜ì— ëŒ€í•´ ìƒê°í•˜ì§€ ì•Šê³  ì†Œí”„íŠ¸ì›¨ì–´ì˜ ê¸°ë³¸ê°’ë§Œ ì‚¬ìš©í•˜ëŠ” _ê²Œìœ¼ë¦„_ì„ í”¼í•˜ê³ , í”Œë¡¯ì„ ë³µì¡í•˜ê²Œ ë§Œë“¤ ë¿ ì‹¤ì œ ë©”ì‹œì§€ëŠ” ì—†ëŠ” ì‹œê°ì  ì¥ì‹ë“¤ì„ ì”ëœ© ì¶”ê°€í•˜ëŠ” _ì§€ë‚˜ì¹¨_ë„ í”¼í•˜ì‹­ì‹œì˜¤. ìì‹ ë§Œì˜ ì‹œê°í™”ë¥¼ ë§Œë“œëŠ” ê²ƒì€ ì—¬ëŸ¬ ë©´ì—ì„œ ì¢‹ì€ ê¸€ì„ ì“°ëŠ” ê²ƒê³¼ ë¹„ìŠ·í•©ë‹ˆë‹¤. (ë°œí‘œë‚˜ ë…¼ë¬¸ì—ì„œ) ì—¬ëŸ¬ë¶„ì˜ ë©”ì‹œì§€ë¥¼ ì „ë‹¬í•˜ëŠ” ê²ƒì€ ë§¤ìš° ì¤‘ìš”í•˜ì§€ë§Œ, ì´ë¥¼ ìœ„í•œ ê°„ë‹¨í•œ ë¹„ë²•ì€ ì—†ìŠµë‹ˆë‹¤. ë‹¤ë¥¸ ì‚¬ëŒë“¤ì´ ë§Œë“  ìˆ˜ë§ì€ ì‹œê°í™” ìë£Œë¥¼ ì£¼ì˜ ê¹Šê²Œ ì‚´í´ë³´ê³ , ìš”ë ¹ì„ í„°ë“í•˜ê¸° ìœ„í•´ ìì‹ ë§Œì˜ ì‹œê°í™”ë¥¼ ì§ì ‘ ë§Œë“œëŠ” ì‹¤í—˜ì„ í•´ë³¸ ë‹¤ìŒ, ì—¬ëŸ¬ë¶„ë§Œì˜ ìŠ¤íƒ€ì¼ì„ ê²°ì •í•˜ì‹­ì‹œì˜¤.

## 3.15 ë” ì½ì„ê±°ë¦¬

**[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)**ì— ê´€í•œ ê°€ì¥ ìœ ìš©í•œ ìë£ŒëŠ” Wickham ([2016](16-chap.html#ref-ggplot2))ì˜ 2íŒê³¼ [ggplot2](https://ggplot2.tidyverse.org) ì›¹ì‚¬ì´íŠ¸ì…ë‹ˆë‹¤. ì˜¨ë¼ì¸ì—ëŠ” ìˆ˜ë§ì€ **[ggplot2](https://cran.r-project.org/web/packages/ggplot2/)** ì½”ë“œ ìŠ¤ë‹ˆí«ì´ ìˆìœ¼ë©°, ì•½ê°„ì˜ ì—°ìŠµì„ ê±°ì¹˜ë©´ ê²€ìƒ‰ ì—”ì§„ì„ í†µí•´ ì°¾ì„ ìˆ˜ ìˆì„ ê²ƒì…ë‹ˆë‹¤. í•˜ì§€ë§Œ ë¹„íŒì ìœ¼ë¡œ ì •ë³´ë¥¼ í™•ì¸í•˜ì‹­ì‹œì˜¤. ì˜¨ë¼ì¸ì—ì„œ ì°¾ëŠ” ì½”ë“œ ì˜ˆì œëŠ” ë•Œë•Œë¡œ íŒ¨í‚¤ì§€ì˜ êµ¬ë²„ì „ì„ ì°¸ì¡°í•˜ê±°ë‚˜ í’ˆì§ˆì´ ì¢‹ì§€ ì•Šì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤.

ì´ ì‹œìŠ¤í…œì˜ ê¸°ì´ˆëŠ” Wilkinson ([2005](16-chap.html#ref-GrofGrbook))ê³¼ Tukey ([1977](16-chap.html#ref-Tukey77); [William S. Cleveland 1988](16-chap.html#ref-Tukey:1988))ì˜ ì•„ì´ë””ì–´ì— ê¸°ë°˜í•˜ê³  ìˆìŠµë‹ˆë‹¤.

## 3.16 ì—°ìŠµ ë¬¸ì œ

__

ì—°ìŠµ ë¬¸ì œ 3.1

_í…Œë§ˆ(themes)_ë¥¼ ì‚¬ìš©í•˜ì—¬ í”Œë¡¯ì˜ ì‹œê°ì  ì™¸ê´€ì„ ë³€ê²½í•´ ë³´ì„¸ìš”. ë‹¤ìŒ ì˜ˆì œë¥¼ ì‹¤í–‰í•˜ëŠ” ê²ƒë¶€í„° ì‹œì‘í•˜ì‹­ì‹œì˜¤:

    
    
    ggcars = ggplot(mtcars, aes(x = hp, y = mpg)) + geom_point()
    ggcars
    ggcars + theme_bw()
    ggcars + theme_minimal()__

ì–´ë–¤ ë‹¤ë¥¸ í…Œë§ˆë“¤ì´ ìˆë‚˜ìš”? (íŒíŠ¸: [ggplot2 ì˜¨ë¼ì¸ ë¬¸ì„œ](https://ggplot2.tidyverse.org/reference)ì˜ themes ì„¹ì…˜ì„ ì‚´í´ë³´ì„¸ìš”.) ì´ê²ƒë“¤ì´ í…Œë§ˆì˜ ì „ë¶€ì¸ê°€ìš”? (íŒíŠ¸: **[ggthemes](https://cran.r-project.org/web/packages/ggthemes/)** íŒ¨í‚¤ì§€ë¥¼ ì‚´í´ë³´ì„¸ìš”.) ìœ„ì˜ í”Œë¡¯ì„ _Economist_ ì¡ì§€ ìŠ¤íƒ€ì¼ë¡œ ë§Œë“¤ì–´ ë³´ì„¸ìš”. í‰í™œ ì„ (smoothing line)ì„ ì¶”ê°€í•´ ë³´ì„¸ìš”.

__

ì—°ìŠµ ë¬¸ì œ 3.2

ê¸°ë³¸ Rì—ì„œ í—ˆìš©ë˜ëŠ” _ìƒ‰ìƒ ì´ë¦„(color names)_ì€ ë¬´ì—‡ì¸ê°€ìš”? `colors` í•¨ìˆ˜ì˜ ë§¤ë‰´ì–¼ í˜ì´ì§€ë¥¼ ì‚´í´ë³´ì„¸ìš”. ì˜ˆì œì™€ ë°ëª¨ë¥¼ ì‹¤í–‰í•´ ë³´ì„¸ìš”.

ì¸í„°ë„· ê²€ìƒ‰ ì—”ì§„ì„ ì‚¬ìš©í•˜ì—¬ "R color names"ë¥¼ ê²€ìƒ‰í•˜ê³  ê²€ìƒ‰ ê²°ê³¼ë¡œ ë‚˜ì˜¤ëŠ” ìë£Œë“¤ì„ íƒìƒ‰í•´ ë³´ì„¸ìš”. ì˜ˆë¥¼ ë“¤ì–´, ëª¨ë“  ìƒ‰ìƒì´ í‘œì‹œëœ ì¹˜íŠ¸ ì‹œíŠ¸(cheat sheets) ë“±ì´ ìˆìŠµë‹ˆë‹¤.

ê·¸ë¦¼ 3.41ì˜ íˆíŠ¸ë§µì„ [R íŒ¨í‚¤ì§€ **beyonce**](https://github.com/dill/beyonce)ì˜ ìƒ‰ìƒ íŒ”ë ˆíŠ¸ë¥¼ ì‚¬ìš©í•˜ëŠ” ë²„ì „ìœ¼ë¡œ ë§Œë“¤ì–´ ë³´ì„¸ìš”.

__

ì—°ìŠµ ë¬¸ì œ 3.3

[xkcd](https://xkcd.com) ì›¹íˆ° ìŠ¤íƒ€ì¼ì˜ í”Œë¡¯ì„ ë§Œë“¤ì–´ ë³´ì„¸ìš”. ë‹¤ìŒ ìë£Œë“¤ì„ ì°¸ê³ í•˜ì‹­ì‹œì˜¤:

  * Stackoverflowì˜ [How can we make xkcd style graphs?](https://stackoverflow.com/questions/12675147/how-can-we-make-xkcd-style-graphs)ë¼ëŠ” ì œëª©ì˜ ìŠ¤ë ˆë“œ

  * ì´ ì¥ì˜ ì‹œì‘ ë¶€ë¶„ì„ ìƒì„±í•˜ëŠ” ì½”ë“œê°€ í¬í•¨ëœ ì´ ì±…ì˜ R ì†ŒìŠ¤ ì½”ë“œ

  * **[xkcd](https://cran.r-project.org/web/packages/xkcd/)** íŒ¨í‚¤ì§€ì˜ ë¹„ë„¤íŠ¸(vignette)

![](imgs/xkcdgraph.png)

__

ì—°ìŠµ ë¬¸ì œ 3.4

RStudio ì›¹ì‚¬ì´íŠ¸ì˜ **[shiny](https://cran.r-project.org/web/packages/shiny/)** íŠœí† ë¦¬ì–¼ì„ í™•ì¸í•´ ë³´ì„¸ìš”. ì´ ì¥ì˜ í”Œë¡¯ ì¤‘ í•˜ë‚˜ë¥¼ í‘œì‹œí•˜ì§€ë§Œ, í‘œì‹œë˜ëŠ” ìœ ì „ì ë“±ì„ ì œì–´í•  ìˆ˜ ìˆëŠ” ëŒ€í™”í˜• ìš”ì†Œê°€ í¬í•¨ëœ _shiny ì•±_ì„ ì‘ì„±í•´ ë³´ì„¸ìš”(ê·¸ë¦¼ 3.33â€”3.35).

__

ì—°ìŠµ ë¬¸ì œ 3.5

ê·¸ë˜í”½ì„ _ì§ë ¬í™”(serializing)_í•˜ëŠ” ë°©ë²•, ì¦‰ ë‚˜ì¤‘ì— ì‚¬ìš©í•˜ê¸° ìœ„í•´ íŒŒì¼ë¡œ ì €ì¥í•˜ê±°ë‚˜ ë‹¤ë¥¸ ì†Œí”„íŠ¸ì›¨ì–´ì—ì„œ ë¶ˆëŸ¬ì˜¤ê¸° ìœ„í•œ ë°©ë²•ì—ëŠ” ì–´ë–¤ ê²ƒë“¤ì´ ìˆë‚˜ìš”? ëŒ€í™”í˜• ê·¸ë˜í”½ì€ ì–´ë–»ê²Œ ì§ë ¬í™”í•  ìˆ˜ ìˆë‚˜ìš”?

__

ì—°ìŠµ ë¬¸ì œ 3.6

ì¤‘ìš”í•˜ê³  ì˜ ë§Œë“¤ì–´ì§„ ê·¸ë˜í”½ì´ ë°˜ë“œì‹œ ë³µì¡í•  í•„ìš”ëŠ” ì—†ìŠµë‹ˆë‹¤. ì˜ˆë¥¼ ë“¤ì–´, [Comirnatyì— ëŒ€í•œ ìœ ëŸ½ ì˜ì•½í’ˆì²­(EMA)ì˜ ê³µê³µ í‰ê°€ ë³´ê³ ì„œ](extras/comirnaty-epar-public-assessment-report_en.pdf)21ì˜ ê·¸ë¦¼ 9(82í˜ì´ì§€)ì™€ [XKCD 2400](https://xkcd.com/2400/)ì„ í™•ì¸í•´ ë³´ì„¸ìš”.

21 <https://www.ema.europa.eu/en/medicines/human/EPAR/comirnaty>ì—ì„œë„ í™•ì¸í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤.

Carr, Daniel B, Richard J Littlefield, WL Nicholson, and JS Littlefield. 1987.
â€œScatterplot Matrix Techniques for Large N.â€ _Journal of the American
Statistical Association_ 82 (398): 424â€“36.

Cleveland, William S. 1988. _The Collected Works of John w. Tukey: Graphics
1965-1985_. Vol. 5. CRC Press.

Cleveland, William S., Marylyn E. McGill, and Robert McGill. 1988. â€œThe Shape
Parameter of a Two-Variable Graph.â€ _Journal of the American Statistical
Association_ 83: 289â€“300.

Helmholtz, H. von. 1867. _Handbuch Der Physiologischen Optik_. Leipzig:
Leopold Voss.

Ihaka, Ross. 2003. â€œColor for Presentation Graphics.â€ In _Proceedings of the
3rd International Workshop on Distributed Statistical Computing_ , edited by
Kurt Hornik and Friedrich Leisch. Vienna, Austria:
http://www.r-project.org/conferences/DSC-2003/Proceedings/; ISSN 1609-395X.

Irizarry, R. A., B. Hobbs, F. Collin, Y. D. Beazer-Barclay, K. J. Antonellis,
U. Scherf, and T. P. Speed. 2003. â€œExploration, Normalization, and Summaries
of High Density Oligonucleotide Array Probe Level Data.â€ _Biostatistics_ 4
(2): 249â€“64.

Lawrence, Michael S., Petar Stojanov, Paz Polak, Gregory V. Kryukov, Kristian
Cibulskis, Andrey Sivachenko, Scott L. Carter, et al. 2013. â€œMutational
Heterogeneity in Cancer and the Search for New Cancer-Associated Genes.â€
_Nature_ 499 (7457): 214â€“18. <https://doi.org/10.1038/nature12213>.

McCormick Jr, William T, Paul J Schweitzer, and Thomas W White. 1972. â€œProblem
Decomposition and Data Reorganization by a Clustering Technique.â€ _Operations
Research_ 20 (5): 993â€“1009.

Mollon, John. 1995. â€œSeeing Colour.â€ In _Colour: Art and Science_ , edited by
T. Lamb and J. Bourriau. Cambridge Unversity Press.

Ohnishi, Y., W. Huber, A. Tsumura, M. Kang, P. Xenopoulos, K. Kurimoto, A. K.
Oles, et al. 2014. â€œCell-to-Cell Expression Variability Followed by Signal
Reinforcement Progressively Segregates Early Mouse Lineages.â€ _Nature Cell
Biology_ 16 (1): 27â€“37.

Tukey, John W. 1977. â€œExploratory Data Analysis.â€ _Massachusetts: Addison-
Wesley_.

Wickham, Hadley. 2010. â€œA Layered Grammar of Graphics.â€ _Journal of
Computational and Graphical Statistics_ 19 (1): 3â€“28.

â€”â€”â€”. 2016. _Ggplot2: Elegant Graphics for Data Analysis_. Springer New York.
<http://had.co.nz/ggplot2/book>.

Wilkinson, Leland. 1999. â€œDot Plots.â€ _The American Statistician_ 53 (3): 276.

â€”â€”â€”. 2005. _The Grammar of Graphics_. Springer.

Page built at 01:33 on 2025-09-01 using R version 4.5.1 (2025-06-13)

