<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>10&nbsp; Networks and Trees – Modern Statistics for Modern Biology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./imgs/msmb-favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-af5ec82acda093b5ee751184164e9432.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d63985d606c7f07261ad561a6cfde940.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="msmb.css">
<meta property="og:title" content="10&nbsp; Networks and Trees – Modern Statistics for Modern Biology">
<meta property="og:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta property="og:image" content="https://www.huber.embl.de/msmb/imgs/MSFMB-Cover2-cropped.jpg">
<meta property="og:site_name" content="Modern Statistics for Modern Biology">
<meta name="twitter:title" content="10&nbsp; Networks and Trees – Modern Statistics for Modern Biology">
<meta name="twitter:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta name="twitter:image" content="https://www.huber.embl.de/msmb/imgs/MSFMB-Cover2-cropped.jpg">
<meta name="twitter:card" content="summary">
<script type="text/javascript">
  var _paq = _paq || [];
  
  // Call disableCookies before calling trackPageView 
  _paq.push(['disableCookies']);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u='//tr-denbi.embl.de/heimdall/';
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '26']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>

</head>

<body class="floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Modern Statistics for Modern Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-chapters" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Chapters</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-chapters">    
        <li>
    <a class="dropdown-item" href="./index.html">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./00-chap.html">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./01-chap.html">
 <span class="dropdown-text">Generative Models for Discrete Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-chap.html">
 <span class="dropdown-text">Statistical Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03-chap.html">
 <span class="dropdown-text">Data visualization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04-chap.html">
 <span class="dropdown-text">Mixture Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./05-chap.html">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06-chap.html">
 <span class="dropdown-text">Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./07-chap.html">
 <span class="dropdown-text">Multivariate Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./08-chap.html">
 <span class="dropdown-text">High-Throughput Count Data &amp; Generalized Linear Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./09-chap.html">
 <span class="dropdown-text">Multivariate methods for heterogeneous data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./10-chap.html">
 <span class="dropdown-text">Networks and Trees</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./11-chap.html">
 <span class="dropdown-text">Image data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./12-chap.html">
 <span class="dropdown-text">Supervised Learning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./13-chap.html">
 <span class="dropdown-text">Design of High Throughput Experiments and their Analyses</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./14-chap.html">
 <span class="dropdown-text">Statistical Concordance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./15-chap.html">
 <span class="dropdown-text">Acknowledgements</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./16-chap.html">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Sections</h2>
   
  <ul class="collapse">
  <li><a href="#goals-for-this-chapter" id="toc-goals-for-this-chapter" class="nav-link active" data-scroll-target="#goals-for-this-chapter"><span class="header-section-number">10.1</span> Goals for this chapter</a></li>
  <li><a href="#graphs" id="toc-graphs" class="nav-link" data-scroll-target="#graphs"><span class="header-section-number">10.2</span> Graphs</a></li>
  <li><a href="#from-gene-set-enrichment-to-networks" id="toc-from-gene-set-enrichment-to-networks" class="nav-link" data-scroll-target="#from-gene-set-enrichment-to-networks"><span class="header-section-number">10.3</span> From gene set enrichment to networks</a></li>
  <li><a href="#phylogenetic-trees" id="toc-phylogenetic-trees" class="nav-link" data-scroll-target="#phylogenetic-trees"><span class="header-section-number">10.4</span> Phylogenetic Trees</a></li>
  <li><a href="#combining-phylogenetic-trees-into-a-data-analysis" id="toc-combining-phylogenetic-trees-into-a-data-analysis" class="nav-link" data-scroll-target="#combining-phylogenetic-trees-into-a-data-analysis"><span class="header-section-number">10.5</span> Combining phylogenetic trees into a data analysis</a></li>
  <li><a href="#minimum-spanning-trees" id="toc-minimum-spanning-trees" class="nav-link" data-scroll-target="#minimum-spanning-trees"><span class="header-section-number">10.6</span> Minimum spanning trees</a></li>
  <li><a href="#summary-of-this-chapter" id="toc-summary-of-this-chapter" class="nav-link" data-scroll-target="#summary-of-this-chapter"><span class="header-section-number">10.7</span> Summary of this chapter</a></li>
  <li><a href="#further-reading" id="toc-further-reading" class="nav-link" data-scroll-target="#further-reading"><span class="header-section-number">10.8</span> Further reading</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">10.9</span> Exercises</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-graphs" class="quarto-section-identifier"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Networks and Trees</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/Darwin_Tree_1837_web.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="imgs/Darwin_Tree_1837_web.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a></p>
</figure>
</div>
</div></div></div>
<p>Networks and trees are often used to represent knowledge about a biological system. They can also be used to directly encode observations from an experiment or study. Phylogenetic trees were drawn to represent family and similarity relationships between species, even before Darwin’s famous notebook sketch that gave these trees a mechanistic, causal interpretation. The meaning of the nodes and edges in a network can differ, and needs to be specified. For instance, a network might schematize relationships between proteins, such as in <a href="#fig-chemo" class="quarto-xref">Figure&nbsp;<span>10.1</span></a>, where the nodes could stand for genes or their encoded proteins, and edges could be direct physical interactions or more abstract “functional” or “genetic” interactions representing outcomes from an experiment. In this book, we use the terms <em>graph</em> and <em>network</em> largely exchangeably. The former term is a bit more evocative of the mathematical structure, the latter of the biological interpretation.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-chemo" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-chemo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-chemo-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2" title="Figure&nbsp;10.1: A small protein-protein network that represents pairwise relationships between proteins."><img src="10-chap_files/figure-html/fig-chemo-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-chemo-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.1: A small protein-protein network that represents pairwise relationships between proteins.
</figcaption>
</figure>
</div>
</div></div></div>
<p>We saw in <a href="02-chap.html" class="quarto-xref"><span>Chapter 2</span></a> that we could model sequences of state transitions as a Markov chain, which can be represented as directed graphs with weights on the edges. Metabolic pathways in which nodes are chemical metabolites and the edges represent chemical reactions. Mutation history trees are used in cancer genomics to represent lineages of mutations.</p>
<p>Transmission networks are important in studying the epidemiology of infectious diseases. As real networks can be very large, we will need special methods for representing and visualizing them. This chapter will be focused on ways of integrating graphs into a data analytic workflow.</p>
<section id="goals-for-this-chapter" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="goals-for-this-chapter"><span class="header-section-number">10.1</span> Goals for this chapter</h2>
<p>In this chapter we will</p>
<ul>
<li><p>Use the formal definition of a graph’s components: edges, vertices, layout to see how we can manipulate them in R using both adjacency matrices and lists of edges.</p></li>
<li><p>We will transform a graph object from <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> into an object that can be visualized according to the layers approach in <strong><a href="https://bioconductor.org/packages/ggplot2/">ggplot2</a></strong> using <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong>. We will experiment with covariates that we attach to graph edges and nodes.</p></li>
<li><p>Graphs are useful ways of encoding prior knowledge about a system, and we will see how they enable us to go from simple gene set analyses to meaningful biological recommendations by mapping significance scores onto the network to detect perturbation <em>hotspots</em>.</p></li>
<li><p>We will build phylogenetic trees starting with from DNA sequences and then visualize these trees with the specifically designed R packages <strong><a href="https://cran.r-project.org/web/packages/ape/">ape</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/ggtree/">ggtree</a></strong>.</p></li>
<li><p>We will combine a phylogenetic tree built from microbiome 16S rRNA data with covariates to show how the hierarchical relationship between taxa can increase the power in multiple hypothesis testing.</p></li>
<li><p>A special tree called a minimum spanning tree (MST) is very useful for testing the relations between a graph and other covariates. We’ll see how to implement different versions of what is known as the Friedman-Rafsky test. We’ll study both co-occurrence of bacteria in mice litters and strain similarities in HIV contagion networks.</p></li>
</ul>
</section>
<section id="graphs" class="level2 page-columns page-full" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="graphs"><span class="header-section-number">10.2</span> Graphs</h2>
<section id="what-is-a-graph-and-how-can-it-be-encoded" class="level3 page-columns page-full" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="what-is-a-graph-and-how-can-it-be-encoded"><span class="header-section-number">10.2.1</span> What is a graph and how can it be encoded?</h3>
<p>A <strong>graph</strong> is defined as a combination of two sets, often denoted as <span class="math inline">\((V,E)\)</span>. Here, <span class="math inline">\(V\)</span> is a set of <strong>nodes</strong> or <strong>vertices</strong>, <span class="math inline">\(E\)</span> is a set of <strong>edges</strong> between vertices. Each element of <span class="math inline">\(E\)</span> is a pair of nodes, i.e., consists of two elements of <span class="math inline">\(V\)</span>. An intuitive way to represent a graph is by its <strong>from-to</strong> representation. If we denote the set of vertices as <span class="math inline">\(V=(\text{A},
\text{B}, \text{C}, \ldots)\)</span>, then the from-to (or <strong>edge list</strong>) representation is a table of the following form.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>  from to
1    A  B
2    B  C
3    A  E
4    C  D
5    E  F</code></pre>
</div>
</div>
<p>The ordering of the rows in the from-to table plays no role. In a <strong>directed</strong> or <strong>oriented graph</strong>, the edges are ordered pairs, i.e., the first line in the above table states that there is an edge from A to B, but does not say whether there is also an edge from B to A—this would need to be denoted in a separate row of the table.</p>
<p>In an <strong>undirected graph</strong>, the edges are unordered pairs, i.e., an edge from A to B is not distinguished from an edge from B to A. Undirected graphs encode symmetric relationships between the nodes, directed graphs represent asymmetric relationships.</p>
<p>It’s important not to confuse a graph with its visualization. It is possible to draw a graph onto a two-dimensional area like in <a href="#fig-chemo" class="quarto-xref">Figure&nbsp;<span>10.1</span></a>, but this is optional, and not unique—there are always many different ways to draw the same graph. There is also no guarantee that in such a visualization, edges do not overlap. Depending on the graph, this can happen. Graphs do not live in physical space (neither 2D nor 3D) but are literally just sets of nodes and edges.</p>
<p>An alternative representation, equivalent to the from-to table, is the <strong>adjacency matrix</strong>, a quadratic matrix with as many rows (and columns) as nodes in the graph. The matrix contains a non-zero entry in the <span class="math inline">\(i\)</span>th row and <span class="math inline">\(j\)</span>th column to encode that there is an edge between the <span class="math inline">\(i\)</span>th and <span class="math inline">\(j\)</span>th vertices.</p>
<div id="wrn-graphs-undirectedmatrix" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.1
</div>
</div>
<div class="callout-body-container callout-body">
<p>For undirected graphs, what is special about the adjacency matrix <span class="math inline">\(A\)</span>?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The adjacency matrix is symmetric, i.e., <span class="math inline">\(M = M^T\)</span>. An example is shown in Figures <a href="#fig-igraphplot" class="quarto-xref"><span>10.2</span></a> and <a href="#fig-adjmatrix" class="quarto-xref"><span>10.3</span></a>. <code>g1</code> is created from a from-to table (encoded in the two-column matrix <code>edges</code>) by the code below.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"igraph"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>edges <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>, <span class="dv">2</span>,<span class="dv">3</span>, <span class="dv">3</span>,<span class="dv">4</span>, <span class="dv">4</span>,<span class="dv">5</span>, <span class="dv">4</span>,<span class="dv">6</span>), <span class="at">byrow =</span> <span class="cn">TRUE</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">=</span> <span class="fu">graph_from_edgelist</span>(edges, <span class="at">directed =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vertex_attr</span>(g1, <span class="at">name =</span> <span class="st">"name"</span>) <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(g1, <span class="at">vertex.size =</span> <span class="dv">25</span>, <span class="at">edge.width =</span> <span class="dv">5</span>, <span class="at">vertex.color =</span> <span class="st">"coral"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-igraphplot" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-igraphplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-igraphplot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3" title="Figure&nbsp;10.2: A small undirected graph with numbered nodes."><img src="10-chap_files/figure-html/fig-igraphplot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-igraphplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.2: A small undirected graph with numbered nodes.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-adjmatrix" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-adjmatrix-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-adjmatrix-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4" title="Figure&nbsp;10.3: The adjacency matrix of the graph shown in Figure&nbsp;fig-igraphplot is is a symmetric n \times n matrix of 0s and 1s, where is n is the number of nodes."><img src="10-chap_files/figure-html/fig-adjmatrix-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-adjmatrix-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.3: The adjacency matrix of the graph shown in <a href="#fig-igraphplot" class="quarto-xref">Figure&nbsp;<span>10.2</span></a> is is a symmetric <span class="math inline">\(n \times n\)</span> matrix of <span class="math inline">\(0\)</span>s and <span class="math inline">\(1\)</span>s, where is <span class="math inline">\(n\)</span> is the number of nodes.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="wrn-graphs-alternative" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.2
</div>
</div>
<div class="callout-body-container callout-body">
<p>Can you give an alternative way to that shown in the answer to <a href="#wrn-graphs-undirectedmatrix" class="quarto-xref">Question&nbsp;<span>10.1</span></a> above for creating a graph from a list of edges dataframe?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>edges <span class="ot">=</span> <span class="st">"1,3</span><span class="sc">\n</span><span class="st">2,3</span><span class="sc">\n</span><span class="st">3,4</span><span class="sc">\n</span><span class="st">4,6</span><span class="sc">\n</span><span class="st">4,5"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="fu">textConnection</span>(edges), <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>sg <span class="ot">=</span> <span class="fu">graph_from_data_frame</span>(df, <span class="at">directed =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>sg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IGRAPH 1467322 UN-- 6 5 -- 
+ attr: name (v/c)
+ edges from 1467322 (vertex names):
[1] 1--3 2--3 3--4 4--6 4--5</code></pre>
</div>
</div>
<p>Typically, you would use the <code>read.csv</code> function to read the from-to table from a separate file. Here the authors of the book decided to create the character string <code>edges</code> on the fly and to turn into a the equivalent of a file using the function <code>textConnection</code>, to keep the example more compact. The notation <code>"\n"</code> indicates a line break.</p>
</div>
</div>
</div>
<section id="elements-of-a-simple-graph" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="elements-of-a-simple-graph">Elements of a simple graph</h4>
<ul>
<li><p>The nodes or vertices. These are the colored circles with numbers in them in <a href="#fig-igraphplot" class="quarto-xref">Figure&nbsp;<span>10.2</span></a>.</p></li>
<li><p>Edges or connections, the segments that join the nodes and which can be directed or not.</p></li>
<li><p>Edge attributes, such edge length. When not specified, we suppose the edge lengths are all the same, typically, one. For instance, to compute the distance between two nodes in the graph, we sum up the lengths of the edges along the shortest path.</p></li>
<li><p>Edge and node attributes: optionally, each edge or each node can be mapped to further continuous or categorical variables, such as type, color, weight, edge width, node size, <span class="math inline">\(...\)</span> Pretty much anything is possible here, depending on the application and the intended computations.</p></li>
</ul>
<p>We also call a directed graph with edge lengths a <strong>network</strong>. The adjacency matric of a network is an <span class="math inline">\(n\times n\)</span> matrix of positive numbers corresponding to the edge lengths.</p>
</section>
<section id="basic-concepts" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="basic-concepts">Basic concepts</h4>
<p>The <em>degree</em> of a node is the number of edges connected to it. In directed graphs we differentiate between <em>in-degree</em> and <em>out-degree</em> for incoming and outgoing edges. We may further distinguish between directed graphs that contain cycles and those that do not (termed cyclic and acyclic graphs).</p>
<p>For large graphs, on can summarize overall graph structure by looking at the distributions of vertex degrees, and we can identify particularly interesting regions or specific nodes and edge in a graph with measures such as centrality and betweenness. These measures are available in various packages (<strong><a href="https://cran.r-project.org/web/packages/network/">network</a></strong>, <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong>).</p>
<p>If the number of edges is of the same order of magnitude as the number of nodes (written <span class="math inline">\(\#E\sim O(\#V)\)</span>), we say that the graph is <strong>sparse</strong>. Some graphs have many nodes, for instance, the package <strong><a href="https://bioconductor.org/packages/ppiData/">ppiData</a></strong> contains a predicted protein interaction (<code>ppipred</code>) graph on about 2500 proteins with around 20000 edges<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. A complete adjacency matrix for such a graph requires more than 6 million memory units, of which most contain 0. This is needlessly wasteful. The edge list representation of the same graph is more compact: it only uses storage where there is an edge, in our example, this amounts to 20000 memory units. One particular choice of edge list representation is sparse matrix encodings, such as implemented in the package <strong><a href="https://cran.r-project.org/web/packages/Matrix/">Matrix</a></strong>.</p>
<div class="no-row-height column-margin column-container"><div id="fn1"><p><sup>1</sup>&nbsp;Gene and species phylogenies may even be much larger.</p></div></div><p>On the other hand, in a <strong>dense</strong> graph, the number of edges is of the same order of magnitude as the number of potential edges, i.e., the square of the number of nodes (written <span class="math inline">\(\#E\sim O(\#V^2)\)</span>. Memory space can be an issue for the storage of large dense graphs.</p>
</section>
<section id="graph-layout" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="graph-layout">Graph layout</h4>
<p>We will see several examples where the same graph is plotted in different ways, either for aesthetic or practical reasons. This is done through the choice of the <strong>graph layout</strong>.</p>
<p>When the edges have lengths representing distances the problem of a 2D representation of the graph is the same as the multidimensional scaling we saw in <a href="09-chap.html" class="quarto-xref"><span>Chapter 9</span></a> It is often solved in a similar way by spreading out the vertex-points as much as possible. In the simple case of edges without lengths, the algorithms can choose different criteria. The method of Fruchterman and Reingold is a basic choice. It is based on a physics-inspired model where similar points attract and repel each other as if under the effect of (Newtonian) physical forces.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> package to do the following</p>
<ul>
<li><p>Create a dense random graph with 12 nodes and more than 50 edges.</p></li>
<li><p>Experiment plotting the graph with different layouts: place the nodes on a circle, or represent the graph as symmetrically as possible, avoid any overlapping nodes or edges.</p></li>
</ul>
</div>
</div>
</section>
<section id="graphs-from-data" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="graphs-from-data">Graphs from data</h4>
<p>Usually data do not arrive in the form of graphs. Graphical or network representations are often the result of transforming from other data types.</p>
<p><strong>From distances or similarities:</strong> graphs can simplify distance or similarity relationships between objects by binarising them. Nodes are connected if they are similar or close, and not connected if not. Thus, the input is a similarity or distance measure between all pairs of objects of interest (genes, proteins, species, phenotypes, <span class="math inline">\(...\)</span>), to which a threshold is applied. The set of measures could be realized in a dense matrix or be computed on the fly.</p>
<p><strong>Bipartite graphs:</strong> some data arrive naturally as absence or presence relationships between two types of objects, for instance, finch species and islands in the Galapagos archipelago (<a href="#fig-bipartite" class="quarto-xref">Figure&nbsp;<span>10.4</span></a>), or transcription factors and gene regulatory regions they are considered to bind to. Such relationships can be encoded with 0/1 values in a rectangular matrix, where rows represent one object type and columns the other. The resulting graph has two types of nodes, e.g., finch-nodes and island-nodes, and edges can only exist between nodes of different types, e.g., between a taxon and an island, but not between taxa, nor between islands. Edges in <a href="#fig-bipartite" class="quarto-xref">Figure&nbsp;<span>10.4</span></a> represent <em>lives on</em> relationships.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-bipartite" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-bipartite-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/BipartiteTaxa.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5" title="Figure&nbsp;10.4: This bipartite graph connects each taxon to the sites where it was observed."><img src="imgs/BipartiteTaxa.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-bipartite-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.4: This <strong>bipartite</strong> graph connects each taxon to the sites where it was observed.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-graphs-finches" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.3
</div>
</div>
<div class="callout-body-container callout-body">
<p>Load the <code>finch.csv</code> data and experiment plotting them to highlight that they represent a bipartite network.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The output of the following code is shown in <a href="#fig-finches-1" class="quarto-xref">Figure&nbsp;<span>10.5</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>finch <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">"../data/finch.csv"</span>, <span class="at">comment =</span> <span class="st">"#"</span>, <span class="at">col_types =</span> <span class="st">"cc"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>finch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 122 × 2
   .tail .head             
   &lt;chr&gt; &lt;chr&gt;             
 1 C     Large ground finch
 2 D     Large ground finch
 3 E     Large ground finch
 4 F     Large ground finch
 5 G     Large ground finch
 6 H     Large ground finch
 7 I     Large ground finch
 8 J     Large ground finch
 9 L     Large ground finch
10 M     Large ground finch
# ℹ 112 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"network"</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>finch.nw  <span class="ot">=</span> <span class="fu">as.network</span>(finch, <span class="at">bipartite =</span> <span class="cn">TRUE</span>, <span class="at">directed =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>is.island <span class="ot">=</span> <span class="fu">nchar</span>(<span class="fu">network.vertex.names</span>(finch.nw)) <span class="sc">==</span> <span class="dv">1</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(finch.nw, <span class="at">vertex.cex =</span> <span class="fl">2.5</span>, <span class="at">displaylabels =</span> <span class="cn">TRUE</span>, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">vertex.col =</span> <span class="fu">ifelse</span>(is.island, <span class="st">"forestgreen"</span>, <span class="st">"gold3"</span>),</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">label=</span> <span class="fu">sub</span>(<span class="st">" finch"</span>, <span class="st">""</span>, <span class="fu">network.vertex.names</span>(finch.nw)))</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>finch.nw <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">t</span>() <span class="sc">|&gt;</span> (\(x) x[, <span class="fu">order</span>(<span class="fu">colnames</span>(x))])()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          A B C D E F G H I J K L M N O P Q
Large ground finch        0 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1
Medium ground finch       1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 0
Small ground finch        1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0
Sharp-beaked ground finch 0 0 1 1 1 0 0 1 0 1 0 1 1 0 1 1 1
Cactus ground finch       1 1 1 0 1 1 1 1 1 1 0 1 0 1 1 0 0
Large cactus ground finch 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0
Large tree finch          0 0 1 1 1 1 1 1 1 0 0 1 0 1 1 0 0
Medium tree finch         0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0
Small tree finch          0 0 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0
Vegetarian finch          0 0 1 1 1 1 1 1 1 1 0 1 0 1 1 0 0
Woodpecker finch          0 0 1 1 1 0 1 1 0 1 0 0 0 0 0 0 0
Mangrove finch            0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0
Warbler finch             1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-finches-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-finches-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-finches-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6" title="Figure&nbsp;10.5: The finches graph. There are many ways to improve the layout, including better taking into account the bipartite nature of the graph."><img src="10-chap_files/figure-html/fig-finches-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:85.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-finches-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.5: The <code>finches</code> graph. There are many ways to improve the layout, including better taking into account the bipartite nature of the graph.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="wrn-graphs-ggnetworkong1" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.4
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make a plot of the graph <code>g1</code> using the <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong> package, with a choice of layout and provided geoms such as <code>geom_edge_link</code>, <code>geom_node_point</code> and <code>geom_node_text</code>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The output of the following code is shown in <a href="#fig-ggnetworkong1-1" class="quarto-xref">Figure&nbsp;<span>10.6</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggraph"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(g1, <span class="at">layout =</span> <span class="st">"nicely"</span>) <span class="sc">+</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>() <span class="sc">+</span> </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="at">size=</span><span class="dv">6</span>,<span class="at">color=</span><span class="st">"#8856a7"</span>) <span class="sc">+</span> </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="at">label=</span><span class="fu">vertex_attr</span>(g1)<span class="sc">$</span>name,  <span class="at">color=</span><span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-ggnetworkong1-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ggnetworkong1-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-ggnetworkong1-1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7" title="Figure&nbsp;10.6: A ggraph example."><img src="10-chap_files/figure-html/fig-ggnetworkong1-1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:33.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ggnetworkong1-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.6: A <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong> example.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="an-example-a-four-state-markov-chain" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="an-example-a-four-state-markov-chain">An example: a four state Markov chain</h4>
<p>In <a href="02-chap.html" class="quarto-xref"><span>Chapter 2</span></a>, we saw how a Markov chain can summarize transitions between nucleotides (considered the states of the system). This is often schematized by a graph. The <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> package provides many choices for graph “decoration”:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"markovchain"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>statesNames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>,<span class="st">"T"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>T1MC <span class="ot">=</span> <span class="fu">new</span>(<span class="st">"markovchain"</span>, <span class="at">states =</span> statesNames, <span class="at">transitionMatrix =</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.2</span>,<span class="fl">0.1</span>,<span class="fl">0.4</span>,<span class="fl">0.3</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="fl">0.1</span>,<span class="fl">0.2</span>,<span class="fl">0.2</span>,<span class="fl">0.5</span>,<span class="fl">0.1</span>,<span class="fl">0.1</span>,<span class="fl">0.8</span>,<span class="fl">0.0</span>),</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">nrow =</span> <span class="dv">4</span>,<span class="at">byrow =</span> <span class="cn">TRUE</span>, <span class="at">dimnames =</span> <span class="fu">list</span>(statesNames, statesNames)))</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(T1MC, <span class="at">edge.arrow.size =</span> <span class="fl">0.4</span>, <span class="at">vertex.color =</span> <span class="st">"purple"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">edge.arrow.width =</span> <span class="fl">2.2</span>, <span class="at">edge.width =</span> <span class="dv">5</span>, <span class="at">edge.color =</span> <span class="st">"blue"</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">edge.curved =</span> <span class="cn">TRUE</span>, <span class="at">edge.label.cex =</span> <span class="fl">2.5</span>, <span class="at">vertex.size=</span> <span class="dv">32</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">vertex.label.cex =</span> <span class="fl">3.5</span>, <span class="at">edge.loop.angle =</span> <span class="dv">3</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">vertex.label.family =</span> <span class="st">"sans"</span>, <span class="at">vertex.label.color =</span> <span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-fourstateMC" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-fourstateMC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/MarkovACGT.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8" title="Figure&nbsp;10.7: A four state Markov chain with arrows representing possible transitions between states."><img src="imgs/MarkovACGT.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fourstateMC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.7: A four state Markov chain with arrows representing possible transitions between states.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Markov chains are simple models of dynamical systems, and the states are represented by the nodes in the graph. The transition matrix gives us the weights on the <strong>directed</strong> edges (arrows) between the states.</p>
<div id="wrn-graphs-MCgraph" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.5
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which state do you think this Markov chain will end up in?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>All nodes except C have outgoing edges, C only has in-going edges and a self-edge. Thus, C is an absorbing state. Sooner or later any process from this chain ends up and stays in C.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="a">
<li><p>Try changing your <code>set.seed</code> function input and see if it changes the plot.</p></li>
<li><p>Access the help for this particular <code>plot</code> function.</p></li>
<li><p>Redo the graph and label the edges with the transition probabilities in green and vertices in brown.</p></li>
</ol>
</div>
</div>
<p>We will see how to build a complete example of annotated state space Markov chain graph in <a href="#imp-graphs-PNASMarkov" class="quarto-xref">Exercise&nbsp;<span>10.3</span></a>.</p>
</section>
</section>
<section id="sec-graphs-layers" class="level3" data-number="10.2.2">
<h3 data-number="10.2.2" class="anchored" data-anchor-id="sec-graphs-layers"><span class="header-section-number">10.2.2</span> Graphs with many layers: labels on edges and nodes</h3>
<p>Here is an example of plotting a graph downloaded from the <a href="https://string-db.org">STRING</a> database with annotations at the vertices.</p>
<div class="cell" data-layout-align="center" data-warning.known="unlabeled data points">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>datf <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/string_graph.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>grs <span class="ot">=</span> <span class="fu">graph_from_data_frame</span>(datf[, <span class="fu">c</span>(<span class="st">"node1"</span>, <span class="st">"node2"</span>)], <span class="at">directed =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">E</span>(grs)<span class="sc">$</span>weight <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(grs)<span class="sc">$</span>size <span class="ot">=</span> <span class="fu">centr_degree</span>(grs)<span class="sc">$</span>res</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(grs) <span class="sc">+</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_arc</span>(<span class="at">color =</span> <span class="st">"black"</span>,  <span class="at">strength =</span> <span class="fl">0.05</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>)<span class="sc">+</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="at">size =</span> <span class="fl">2.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"orange"</span>) <span class="sc">+</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_label</span>(<span class="fu">aes</span>(<span class="at">label=</span><span class="fu">vertex_attr</span>(grs)<span class="sc">$</span>name), <span class="at">size =</span> <span class="dv">3</span>, <span class="at">alpha =</span> <span class="fl">0.9</span>, <span class="at">color =</span> <span class="st">"#8856a7"</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-completechemokine" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-completechemokine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-completechemokine-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9" title="Figure&nbsp;10.8: Perturbed chemokine subnetwork uncovered in @YuGXNA using differential gene expression patterns in sorted T-cells. Notice the clique-like structure of the genes CXCR3, CXCL13, CCL19, CSCR5 and CCR7 in the right hand corner."><img src="10-chap_files/figure-html/fig-completechemokine-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-completechemokine-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.8: Perturbed chemokine subnetwork uncovered in <span class="citation" data-cites="YuGXNA">Yu et al. (<a href="16-chap.html#ref-YuGXNA" role="doc-biblioref">2012</a>)</span> using differential gene expression patterns in sorted T-cells. Notice the clique-like structure of the genes CXCR3, CXCL13, CCL19, CSCR5 and CCR7 in the right hand corner.
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-completechemokine" class="quarto-xref">Figure&nbsp;<span>10.8</span></a> shows the full perturbed chemokine subnetwork discovered in the study of breast cancer metastasis using GXNA <span class="citation" data-cites="GXNA">(<a href="16-chap.html#ref-GXNA" role="doc-biblioref">Nacu et al. 2007</a>)</span> and reported by <span class="citation" data-cites="YuGXNA">Yu et al. (<a href="16-chap.html#ref-YuGXNA" role="doc-biblioref">2012</a>)</span>.</p>
</section>
</section>
<section id="from-gene-set-enrichment-to-networks" class="level2 page-columns page-full" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="from-gene-set-enrichment-to-networks"><span class="header-section-number">10.3</span> From gene set enrichment to networks</h2>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10" title="A long unstructured laundry list of possibly differentially expressed genes can be daunting."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="A long unstructured laundry list of possibly differentially expressed genes can be daunting."></a></p>
</figure>
</div>
<figcaption>A long unstructured laundry list of possibly differentially expressed genes can be daunting.</figcaption>
</figure>
</div>
</div></div></div>
<p>In <a href="08-chap.html" class="quarto-xref"><span>Chapter 8</span></a>, we studied methods for finding a list of differentially expressed genes. Small sample sizes, coupled with efforts to maintain low FDRs, often result in low power to detect differential expression. Therefore, obtaining a long list of genes that can be confidently declared as differentially expressed is, initially, a triumph. However, understanding the underlying biology requires more than just a laundry list of significant players in a biological system.</p>
<section id="methods-using-pre-defined-gene-sets-gsea" class="level3" data-number="10.3.1">
<h3 data-number="10.3.1" class="anchored" data-anchor-id="methods-using-pre-defined-gene-sets-gsea"><span class="header-section-number">10.3.1</span> Methods using pre-defined gene sets (GSEA)</h3>
<p>One of the earliest approaches was to look for gene attributes that are <strong>overrepresented</strong> or <strong>enriched</strong> in the laundry list of significant genes. These gene classes are often based on Gene Ontology (GO) categories (for example, genes that are involved in organ growth, or genes that are involved in feeding behavior). The Gene Ontology (GO) is a collection of three ontologies that describe genes and gene products. These ontologies are restricted vocabularies that have the structure of directed acyclic graphs (DAGS). The most specific terms are the leaves of the graph. The GO graph consists of nodes (here, Gene Ontology terms) and edges from more specific terms (children) to less specific (parents), often these edges are directed. Nodes and edges can have multiple attributes that can be visualized. The main purpose of using GO annotations for a particular set of Genes designated as significant in an experiment is to look for the <strong>enrichment</strong> of a GO term in this list, we will give this term a statistical meaning below. Many other useful lists of important gene sets exist.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Find a useful database of gene sets.</p>
</div>
</div>
<p>For instance, the <code>MsigDB</code> Molecular Signature Database <span class="citation" data-cites="MSigDB">(<a href="16-chap.html#ref-MSigDB" role="doc-biblioref">Liberzon et al. 2011</a>)</span> contains many gene sets that can be accessed from within R using the function <code>getBroadSets</code> from the Bioconductor package <strong><a href="https://bioconductor.org/packages/GSEABase/">GSEABase</a></strong> roughly as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"GSEABase"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="do">## This requires a login to the website.</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>fl   <span class="ot">=</span>  <span class="st">"/path/to/msigdb_v5.1.xml"</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>gss  <span class="ot">=</span>  <span class="fu">getBroadSets</span>(fl) </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">organism</span>(gss[[<span class="dv">1</span>]])</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">sapply</span>(gss, organism))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sec-graphs-GSEA" class="level3 page-columns page-full" data-number="10.3.2">
<h3 data-number="10.3.2" class="anchored" data-anchor-id="sec-graphs-GSEA"><span class="header-section-number">10.3.2</span> Gene set analysis with two-way table tests</h3>
<div class="cell page-columns page-full" data-layout-align="center" data-tbl-cap-location="top">
<div id="tbl-hypergeom" class="cell quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-tbl-cap-location="top" data-layout-align="center">
<figure class="quarto-float quarto-float-tbl figure page-columns page-full">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-hypergeom-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.1: Although there are the same number of each category of gene found in the <em>significant</em> set, both the simulation below and the theory of testing in two-way tables shows us that the blue category is enriched.
</figcaption>
<div aria-describedby="tbl-hypergeom-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="page-columns page-full">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<table class="do-not-create-environment cell caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Yellow</th>
<th style="text-align: right;">Blue</th>
<th style="text-align: right;">Red</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Significant</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">25</td>
</tr>
<tr class="even">
<td style="text-align: left;">Universe</td>
<td style="text-align: right;">500</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">400</td>
</tr>
</tbody>
</table>
</div></div></div>
</figure>
</div>
</div>
<p>Here, we start by explaining a basic approach often called <strong>Fisher’s “exact” test</strong> or <strong>hypergeometric test</strong>ing.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11" title="So-called ‘exact’ tests because they are nonparametric and based on exhaustive enumerations: not because we are sure of the answer – this is statistics after all."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="So-called ‘exact’ tests because they are nonparametric and based on exhaustive enumerations: not because we are sure of the answer – this is statistics after all."></a></p>
</figure>
</div>
<figcaption>So-called ‘exact’ tests because they are nonparametric and based on exhaustive enumerations: <strong>not</strong> because we are sure of the answer – this is statistics after all.</figcaption>
</figure>
</div>
</div></div></div>
<p>Define a universe of candidate genes that may potentially be significant; say this universe is of size <span class="math inline">\(N\)</span>. We also have a record of the genes that actually <strong>did</strong> come out significant, of which we suppose there were <span class="math inline">\(m\)</span>.</p>
<p>We make a toy model involving balls in boxes, with a total of <span class="math inline">\(N\)</span> balls corresponding to the genes identified in the gene universe. These genes are split into different functional categories, suppose there are <span class="math inline">\(N=1,000\)</span> genes, of which 500 are yellow, 100 are blue and 400 are red. Then a subset of <span class="math inline">\(m=75\)</span> genes are labeled as <strong>significant</strong>, suppose among these significantly interesting genes, there are 25 yellow, 25 red and 25 blue. Is the blue category enriched or overrepresented?</p>
<p>We use this hypergeometric two-way table testing to account for the fact that some categories are extremely numerous and others are rarer.</p>
<div id="wrn-graphs-simulationFishertest" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.6
</div>
</div>
<div class="callout-body-container callout-body">
<p>Run a Monte Carlo experiment with 20,000 simulations and compute the p-value of significance of having 25 blues under the null hypothesis that no category is over-represented in the significant set.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Under the null the 75 are sampled randomly from our unequal boxes as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>universe <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Yellow"</span>, <span class="dv">500</span>), <span class="fu">rep</span>(<span class="st">"Blue"</span>, <span class="dv">100</span>), <span class="fu">rep</span>(<span class="st">"Red"</span>, <span class="dv">400</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>countblue <span class="ot">=</span> <span class="fu">replicate</span>(<span class="dv">20000</span>, {</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  pick75 <span class="ot">=</span> <span class="fu">sample</span>(universe, <span class="dv">75</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sum</span>(pick75 <span class="sc">==</span> <span class="st">"Blue"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(countblue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  0.000   6.000   7.000   7.496   9.000  20.000 </code></pre>
</div>
</div>
<p>The histogram in <a href="#fig-histblue" class="quarto-xref">Figure&nbsp;<span>10.9</span></a> shows that having a value as large as 25 under the null model would be extremely rare.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-histblue" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-histblue-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-histblue-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12" title="Figure&nbsp;10.9: We can see that even in 20000 simulations, no blue count comes close to being 25. We can reject such an event as having happened by chance and conclude that the blue are enriched."><img src="10-chap_files/figure-html/fig-histblue-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:50.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-histblue-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.9: We can see that even in 20000 simulations, no blue count comes close to being 25. We can reject such an event as having happened by chance and conclude that the blue are <strong>enriched</strong>.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>In the general case, the gene universe is an urn with <span class="math inline">\(N\)</span> balls, if we pick the <span class="math inline">\(m\)</span> balls at random and there is a proportion of <span class="math inline">\(k/N\)</span> blue balls, we expect to see <span class="math inline">\(km/N\)</span> blue balls in a draw of size <span class="math inline">\(k\)</span>.</p>
<section id="plotting-gene-enrichment-networks-with-goplot" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="plotting-gene-enrichment-networks-with-goplot">Plotting gene enrichment networks with <strong><a href="https://cran.r-project.org/web/packages/GOplot/">GOplot</a></strong></h4>
<p>Here we show an attractive way of summarizing the connections between the gene functional categories and the significant gene set.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-warning.known="[&quot;for a discrete variable is not advised&quot;,&quot;rows containing missing values&quot;]">
<div class="sourceCode cell-code" id="cb15" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"GOplot"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"EC"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>circ  <span class="ot">=</span>  <span class="fu">circle_dat</span>(EC<span class="sc">$</span>david, EC<span class="sc">$</span>genelist)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>chord <span class="ot">=</span>  <span class="fu">chord_dat</span>(circ, EC<span class="sc">$</span>genes, EC<span class="sc">$</span>process)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">GOChord</span>(chord, <span class="at">limit =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-GOplotEC" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-GOplotEC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-GOplotEC-1.jpeg" class="lightbox" data-gallery="quarto-lightbox-gallery-13" title="Figure&nbsp;10.10: This graph shows the correspondence between GO terms and significantly changed genes in a study on differential expression in endothelial cells from two steady state tissues (brain and heart, see @Nolan:2013). After normalization a differential expression analysis was performed giving a list of genes. A gene-annotation enrichment analysis of the set of differentially expressed genes (adjusted p-value < 0.05) was then performed with the GOplot package."><img src="10-chap_files/figure-html/fig-GOplotEC-1.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" width="880"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-GOplotEC-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.10: This graph shows the correspondence between GO terms and significantly changed genes in a study on differential expression in endothelial cells from two steady state tissues (brain and heart, see <span class="citation" data-cites="Nolan:2013">Nolan et al. (<a href="16-chap.html#ref-Nolan:2013" role="doc-biblioref">2013</a>)</span>). After normalization a differential expression analysis was performed giving a list of genes. A gene-annotation enrichment analysis of the set of differentially expressed genes (adjusted p-value &lt; 0.05) was then performed with the <strong><a href="https://cran.r-project.org/web/packages/GOplot/">GOplot</a></strong> package.
</figcaption>
</figure>
</div>
</div>
</div>
<p>In fact, the Gene Ontology graph does not necessarily capture meaningful <em>gene interactions</em> as genes from different processes often interact productively. A large amount of information remains unused, for example, all significant genes are usually given equal weight, despite the potentially large variations in their p-values.</p>
</section>
</section>
<section id="significant-subgraphs-and-high-scoring-modules" class="level3 page-columns page-full" data-number="10.3.3">
<h3 data-number="10.3.3" class="anchored" data-anchor-id="significant-subgraphs-and-high-scoring-modules"><span class="header-section-number">10.3.3</span> Significant subgraphs and high scoring modules</h3>
<p>We have at our disposal more than just the Gene Ontology. There are many different databases of gene networks from which we can choose a <strong>known skeleton</strong> graph onto which we project significance scores such as p-values from our differential expression experiment. We will follow an idea first suggested by <span class="citation" data-cites="Ideker:2002">Ideker et al. (<a href="16-chap.html#ref-Ideker:2002" role="doc-biblioref">2002</a>)</span>. This is further developed in <span class="citation" data-cites="GXNA">Nacu et al. (<a href="16-chap.html#ref-GXNA" role="doc-biblioref">2007</a>)</span>. A careful implementation with many improvements is available as the Bioconductor package <strong><a href="https://bioconductor.org/packages/BioNet/">BioNet</a></strong> <span class="citation" data-cites="BioNet">(<a href="16-chap.html#ref-BioNet" role="doc-biblioref">Beisser et al. 2010</a>)</span>. These methods all search for the subgraphs or modules of a scored-skeleton network that seem to be particularly <strong>perturbed</strong>.</p>
<p>Each gene-node in the network is assigned a score that can either be calculated from a t-statistic or a p-value. Often pathways contain both upregulated and downregulated genes; as pointed out in <span class="citation" data-cites="Ideker:2002">Ideker et al. (<a href="16-chap.html#ref-Ideker:2002" role="doc-biblioref">2002</a>)</span>, this can be captured by taking absolute values of the test statistic or just incorporating scores computed from the p-values<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. <span class="citation" data-cites="BioNet">Beisser et al. (<a href="16-chap.html#ref-BioNet" role="doc-biblioref">2010</a>)</span> model the p-values of the genes as we did in <a href="06-chap.html" class="quarto-xref"><span>Chapter 6</span></a>: mixture of non-perturbed genes whose p-values will be uniformly distributed and non uniformly distributed p-values from the perturbed genes. We model the signal in the data using a beta distribution for the p-values following <span class="citation" data-cites="Pounds2003">Pounds and Morris (<a href="16-chap.html#ref-Pounds2003" role="doc-biblioref">2003</a>)</span>.</p>
<div class="no-row-height column-margin column-container"><div id="fn2"><p><sup>2</sup>&nbsp;We’ll want something like <span class="math inline">\(-\log p\)</span>, so that small p-values give large scores.</p></div></div><p>Given our node-scoring function, we search for connected hotspots in the graph, i.e., a subgraph of genes with high combined scores.</p>
<section id="using-a-subgraph-search-algorithm" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="using-a-subgraph-search-algorithm">Using a subgraph search algorithm</h4>
<p>Finding the maximal scoring subgraph of a generic graph is known to be intractable in general (we say it is an NP-hard problem), so various approximate algorithms have been proposed. <span class="citation" data-cites="Ideker:2002">Ideker et al. (<a href="16-chap.html#ref-Ideker:2002" role="doc-biblioref">2002</a>)</span> suggested using simulated annealing, however this is slow and tends to produce large subgraphs that are difficult to interpret. <span class="citation" data-cites="GXNA">Nacu et al. (<a href="16-chap.html#ref-GXNA" role="doc-biblioref">2007</a>)</span> started with a seed vertex and gradually expand around it. <span class="citation" data-cites="BioNet">Beisser et al. (<a href="16-chap.html#ref-BioNet" role="doc-biblioref">2010</a>)</span> started the search with a so-called minimal spanning tree (MST), a graph we we will study later in this chapter.</p>
</section>
</section>
<section id="an-example-with-the-bionet-implementation" class="level3 page-columns page-full" data-number="10.3.4">
<h3 data-number="10.3.4" class="anchored" data-anchor-id="an-example-with-the-bionet-implementation"><span class="header-section-number">10.3.4</span> An example with the BioNet implementation</h3>
<div class="page-columns page-full"><p>To illustrate the method, we show data from the <strong><a href="https://bioconductor.org/packages/BioNet/">BioNet</a></strong> package. </p><div class="no-row-height column-margin column-container"><span class="margin-aside">The <code>interactome</code> data contains a connected component of the network comprising 2034 different gene products and 8399 interactions. This constitutes the skeleton graph with which we will work, see <span class="citation" data-cites="BioNet">Beisser et al. (<a href="16-chap.html#ref-BioNet" role="doc-biblioref">2010</a>)</span>.</span></div></div>
<p>The <code>dataLym</code> contains the relevant pvalues and <span class="math inline">\(t\)</span> statistics for 3,583 genes, you can access them and do the analysis as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BioNet"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"DLBCL"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"dataLym"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"interactome"</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>interactome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A graphNEL graph with undirected edges
Number of Nodes = 9386 
Number of Edges = 36504 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pval <span class="ot">=</span> dataLym<span class="sc">$</span>t.pval</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pval)  <span class="ot">=</span>  dataLym<span class="sc">$</span>label</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>subnet <span class="ot">=</span> <span class="fu">subNetwork</span>(dataLym<span class="sc">$</span>label, interactome)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>subnet <span class="ot">=</span> <span class="fu">rmSelfLoops</span>(subnet)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>subnet</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A graphNEL graph with undirected edges
Number of Nodes = 2559 
Number of Edges = 7788 </code></pre>
</div>
</div>
<section id="fit-a-beta-uniform-model" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="fit-a-beta-uniform-model">Fit a Beta-Uniform model</h4>
<div class="page-columns page-full"><p>The p-values are fit to the type of mixture we studied in <a href="04-chap.html" class="quarto-xref"><span>Chapter 4</span></a>, with a uniform component from the null with probability <span class="math inline">\(\pi_0\)</span> and a beta distribution (proportional to <span class="math inline">\(a x^{a - 1}\)</span>) for the p-values corresponding to the alternatives <span class="citation" data-cites="Pounds2003">(<a href="16-chap.html#ref-Pounds2003" role="doc-biblioref">Pounds and Morris 2003</a>)</span>. <span class="math display">\[f(x|a,\pi_0)= \pi_0 + (1-\pi_0) a x^{a - 1}\qquad \mbox{ for } 0 &lt;x \leq 1; \; 0&lt;a&lt;1\]</span>  Running the model with an <span class="math display">\[fdr\]</span> of 0.001:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">The package actually gives a different name to <span class="math inline">\(\\pi_0\)</span>: it uses <span class="math inline">\(\\lambda\)</span> and calls it the mixing parameter.</span></div></div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>fb <span class="ot">=</span> <span class="fu">fitBumModel</span>(pval, <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>fb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Beta-Uniform-Mixture (BUM) model

3583 pvalues fitted

Mixture parameter (lambda): 0.482
shape parameter (a):        0.180
log-likelihood:         4471.8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>scores<span class="ot">=</span><span class="fu">scoreNodes</span>(subnet, fb, <span class="at">fdr =</span> <span class="fl">0.001</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center" data-warning.known="All aesthetics have length...">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-plotFITBum" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-plotFITBum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-plotFITBum-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14" title="Figure&nbsp;10.11: The qqplot shows the quality of the fit of beta-uniform mixture model to the data. The red points have the theoretical quantiles from the beta distribution as the x coordinates the observed quantiles and the y coordinates. The blue line shows that this model fits nicely."><img src="10-chap_files/figure-html/fig-plotFITBum-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-plotFITBum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.11: The qqplot shows the quality of the fit of beta-uniform mixture model to the data. The red points have the theoretical quantiles from the beta distribution as the x coordinates the observed quantiles and the y coordinates. The blue line shows that this model fits nicely.
</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-histFITBum" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-histFITBum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-histFITBum-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15" title="Figure&nbsp;10.12: A histogram of the mixture components for the p-values, the beta in red and the uniform in blue, \pi_0 is the mixing proportion assigned to the null component whose distribution should be uniform."><img src="10-chap_files/figure-html/fig-histFITBum-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-histFITBum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.12: A histogram of the mixture components for the p-values, the beta in red and the uniform in blue, <span class="math inline">\(\pi_0\)</span> is the mixing proportion assigned to the null component whose distribution should be uniform.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Then we run a heuristic search for a high scoring subgraph using:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>hotSub  <span class="ot">=</span>  <span class="fu">runFastHeinz</span>(subnet, scores)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>hotSub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A graphNEL graph with undirected edges
Number of Nodes = 144 
Number of Edges = 221 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>logFC<span class="ot">=</span>dataLym<span class="sc">$</span>diff</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(logFC)<span class="ot">=</span>dataLym<span class="sc">$</span>label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="wrn-graphs-igraph" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.7
</div>
</div>
<div class="callout-body-container callout-body">
<p>We made <a href="#fig-plotBioNet" class="quarto-xref">Figure&nbsp;<span>10.13</span></a> using the following code:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotModule</span>(hotSub, <span class="at">layout =</span> layout.davidson.harel, <span class="at">scores =</span> scores,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">diff.expr =</span> logFC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-plotBioNet" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-plotBioNet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-plotBioNet-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16" title="Figure&nbsp;10.13: The subgraph found as maximally enriched for differential expression between ABC and GCB B-cell lymphoma. The nodes are colored in red and green: green shows an upregulation in ACB and red an upregulation in GBC. The shape of the nodes depicts the score: rectangles indicate a negative score, circles a positive score."><img src="10-chap_files/figure-html/fig-plotBioNet-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-plotBioNet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.13: The subgraph found as maximally enriched for differential expression between ABC and GCB B-cell lymphoma. The nodes are colored in red and green: green shows an upregulation in ACB and red an upregulation in GBC. The shape of the nodes depicts the score: rectangles indicate a negative score, circles a positive score.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Using the function <code>igraph.from.graphNEL</code>, transform the module object and plot it using the <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong> method shown in <a href="#sec-graphs-layers" class="quarto-xref"><span>Section 10.2.2</span></a>.</p>
</div>
</div>
</section>
</section>
</section>
<section id="phylogenetic-trees" class="level2 page-columns page-full" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="phylogenetic-trees"><span class="header-section-number">10.4</span> Phylogenetic Trees</h2>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-IntroTree" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-IntroTree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/SameTreePhyl.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17" title="Figure&nbsp;10.14: As mathematical objects, the hierarchical clustering trees (studied in sec-clustering) are the same as phylogenetic trees. They are rooted binary trees with labels at the tips."><img src="imgs/SameTreePhyl.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-IntroTree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.14: As mathematical objects, the hierarchical clustering trees (studied in <a href="05-chap.html" class="quarto-xref"><span>Chapter 5</span></a>) are the same as phylogenetic trees. They are <strong>rooted binary</strong> trees with labels at the tips.
</figcaption>
</figure>
</div>
</div></div></div>
<p>One really important use of graphs in biology is the construction of phylogenetic trees. Trees are graphs with no <strong>cycles</strong> (the official word for loops, whether self loops, or ones that go through several vertices). Phylogenetic trees are usually rooted binary trees that only have labels on the leaves corresponding to contemporary<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> taxa at the tips. The inner nodes correspond to <strong>ancestral</strong> sequences which have to be inferred from the <strong>contemporaneous</strong> data on the tips. Many methods use aligned DNA sequences from the different species or populations to infer or estimate the tree. The tips of the tree are usually called <strong>OTU</strong>s (Operational Taxonomic Units). The statistical <strong>parameter</strong> of interest in these analyses is the rooted binary tree with OTU labels on its leaves (see <span class="citation" data-cites="holmesima holmes2003b">Holmes (<a href="16-chap.html#ref-holmesima" role="doc-biblioref">1999</a>, <a href="16-chap.html#ref-holmes2003b" role="doc-biblioref">2003b</a>)</span> for details).</p>
<div class="no-row-height column-margin column-container"><div id="fn3"><p><sup>3</sup>&nbsp;Because they are contemporary, the trees are often represented so that the leaves are a ll the same distance from the root.</p></div></div><section id="the-example-of-hiv" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="the-example-of-hiv">The example of HIV</h4>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-HIVtree" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-HIVtree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/hivtreeplos.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18" title="Figure&nbsp;10.15: This phylogenetic tree describes the history of different HIV/SIV strains in Africa [@Wertheim:2009], [Figure from]."><img src="imgs/hivtreeplos.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-HIVtree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.15: This phylogenetic tree describes the history of different HIV/SIV strains in Africa <span class="citation" data-cites="Wertheim:2009">(<a href="16-chap.html#ref-Wertheim:2009" role="doc-biblioref">Wertheim and Worobey 2009</a>)</span>, [Figure from].
</figcaption>
</figure>
</div>
</div></div></div>
<p>HIV is a virus that protects itself by evolving very fast (within months, several mutations can appear). Its evolution can thus be followed in real time; whereas the evolution of large organisms which has happened over millions of years. HIV trees are built for medical purposes such as the detection and understanding of drug resistance. They are estimated for individual genes. Different genes can show differences in their evolutionary histories and thus produce different <strong>gene trees</strong>. The phylogenetic tree in <a href="#fig-HIVtree" class="quarto-xref">Figure&nbsp;<span>10.15</span></a> shows times when the virus switched from monkeys to humans <span class="citation" data-cites="Wertheim:2009">(<a href="16-chap.html#ref-Wertheim:2009" role="doc-biblioref">Wertheim and Worobey 2009</a>)</span>.</p>
</section>
<section id="special-elements-in-phylogenies" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="special-elements-in-phylogenies">Special elements in phylogenies</h4>
<ul>
<li><p>Most phylogenetic trees are shown rooted, the `root’ is usually found by including an outgroup in the tree tips, as we will see later.</p></li>
<li><p>Characters that are derived from this common ancestry are called homologous (geneticists doing population studies replace the term homology by identity by descent (IBD)).</p></li>
<li><p>Sisters on the tree defined by a common ancestor are called clades or monophyletic groups, they have more than just <em>similarities</em> in common.</p></li>
</ul>
</section>
<section id="markovian-models-for-evolution." class="level3 page-columns page-full" data-number="10.4.1">
<h3 data-number="10.4.1" class="anchored" data-anchor-id="markovian-models-for-evolution."><span class="header-section-number">10.4.1</span> Markovian models for evolution.</h3>
<p>To infer what happened in the ancestral species from contemporary data collected on the tips of the tree, we have to make assumptions about how substitutions and deletions occur through time. The models we use are all Markovian and said to be time homogeneous: the mutation rate is constant across history.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19" title="This is called the molecular clock hypothesis, if we do not make this assumption we run into what is known as non-identifiability (ie we can’t tell the difference between the many possible mutational histories given the observed data)."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="This is called the molecular clock hypothesis, if we do not make this assumption we run into what is known as non-identifiability (ie we can’t tell the difference between the many possible mutational histories given the observed data)."></a></p>
</figure>
</div>
<figcaption>This is called the <strong>molecular clock</strong> hypothesis, if we do not make this assumption we run into what is known as non-identifiability (ie we can’t tell the difference between the many possible mutational histories given the observed data).</figcaption>
</figure>
</div>
</div></div></div>
<section id="continuous-markov-chain-and-generator-matrix" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="continuous-markov-chain-and-generator-matrix">Continuous Markov chain and generator matrix</h4>
<p>We are going to use the Markov chain we saw in <a href="#fig-fourstateMC" class="quarto-xref">Figure&nbsp;<span>10.7</span></a> on the states [A,C,G,T]; however now we consider that the changes of state, i.e.&nbsp;mutations, occur at random times. The gaps between these mutational events will follow an exponential distribution. These continuous time Markov chains have the following properties:</p>
<ul>
<li><p><strong>No Memory</strong>. <span class="math inline">\(P(Y(u+t)=j\;|\;Y(t)=i)\)</span> does not depend on times before <span class="math inline">\(t\)</span>.</p></li>
<li><p><strong>Time homogeneity</strong>. The probability <span class="math inline">\(P(Y(h+t)=j\,|\,Y(t)=i)\)</span> does not depend on <span class="math inline">\(t\)</span>, but on <span class="math inline">\(h\)</span>, the time between the events and on <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>.</p></li>
<li><p><strong>Linearity</strong>. The instantaneous transition rate is of an approximately linear form</p></li>
</ul>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/devil.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20" title="We use an error term written here as o(h), we read this little o of h, which just means that this error terms grows much slower (i.e., sublinear) than h."><img src="imgs/devil.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="123" alt="We use an error term written here as o(h), we read this little o of h, which just means that this error terms grows much slower (i.e., sublinear) than h."></a></p>
</figure>
</div>
<figcaption>We use an error term written here as <span class="math inline">\(o(h)\)</span>, we read this little <span class="math inline">\(o\)</span> of <span class="math inline">\(h\)</span>, which just means that this error terms grows much slower (i.e., sublinear) than <span class="math inline">\(h\)</span>.</figcaption>
</figure>
</div>
</div></div></div>
<p><span class="anchored-eq" data-anchor-id="eq-ebntiwmhzopu" id="eq-ebntiwmhzopu"><span class="math display">\[
\begin{align}
P_{ij}(h)&amp;=q_{ij}h+o(h), \quad\text{for }j\neq i\\
P_{ii}(h)&amp;=1-q_i(h)+ o(h), \qquad\text{where }q_i=\sum_{j\neq i}q_{ij}.
\end{align}
\tag{10.1}\]</span></span></p>
<p><span class="math inline">\(q_{ij}\)</span> is known as the instantaneous transition rate. These rates define matrices as in <a href="#tbl-JCKimura" class="quarto-xref">Table&nbsp;<span>10.2</span></a>.</p>
<ul>
<li><strong>Exponential distribution</strong>. Times between changes are supposed to be exponentially distributed.</li>
</ul>
<!-- -->
<div class="cell page-columns page-full" data-layout-align="center" data-tbl-cap-location="margin">
<div id="tbl-JCKimura" class="cell quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-tbl-cap-location="margin" data-layout-align="center">
<figure class="quarto-float quarto-float-tbl figure page-columns page-full">
<div aria-describedby="tbl-JCKimura-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="do-not-create-environment cell caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 48%">
<col style="width: 51%">
</colgroup>
<tbody>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(Q = \begin{array}{lcccc} &amp; A &amp; T &amp; C &amp; G \\
  A &amp; -3\alpha &amp; \alpha &amp; \alpha &amp; \alpha \\
  T &amp; \alpha &amp; -3\alpha &amp; \alpha &amp; \alpha \\
  C &amp; \alpha &amp; \alpha &amp; -3\alpha &amp; \alpha \\
  G &amp; \alpha &amp; \alpha &amp; \alpha &amp; -3\alpha \\ \end{array}\)</span></td>
<td style="text-align: right;"><span class="math inline">\(Q = \begin{array}{lcccc} &amp; A &amp; T &amp; C &amp; G \\
  A &amp; -\alpha-2 \beta &amp; \beta &amp; \beta &amp; \alpha \\
  T &amp; \beta &amp; -\alpha-2 \beta &amp; \alpha &amp; \beta \\
  C &amp; \beta &amp; \alpha &amp; -\alpha-2 \beta &amp; \beta \\
  G &amp; \alpha &amp; \beta &amp; \beta &amp; -\alpha-2 \beta \\ \end{array}\)</span></td>
</tr>
</tbody>
</table>
</div>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-tbl margin-caption" id="tbl-JCKimura-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.2: Two examples of rate matrices, on the left:the Jukes-Cantor (<code>JC69</code>) model, on the right is shown the Kimura (<code>K80</code>) two parameter model.
</figcaption>
</figure>
</div>
</div>
<p>the instantaneous change probability matrix called <strong>the generator</strong>. In the simplest possible model, called a Jukes-Cantor model; all the mutations are equally likely (see the left of <a href="#tbl-JCKimura" class="quarto-xref">Table&nbsp;<span>10.2</span></a>). A slightly more flexible model, called the Kimura model is shown on the right in <a href="#tbl-JCKimura" class="quarto-xref">Table&nbsp;<span>10.2</span></a>.</p>
<div id="wrn-graphs-KimuraJC" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.8
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why do we say the Kimura model is more flexible?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The Jukes-Cantor only has one parameter, that model supposes all transitions and transversions are equally likely. In the Kimura, there is one parameter for transitions and another for the transversions (mutations occurring from purine to pyrimidines, or vice versa).</p>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<p>![Vocabulary overload here! : Transitions in this context mean mutational changes within the purines (A&lt;-&gt;G]) or within the pyrimidines (C &lt;-&gt; T); whereas when we talked about Markov chains earlier our <strong>transition</strong> matrix contains all probabilities of any state changes.](imgs/devil.png){fig-align=‘center’ width=123}</p>
</div></div></div>
<p>The most flexible model is called the Generalized Time Reversible (GTR) model; it has 6 free parameters. We are going to show an example of data simulated according to these generative models from a known tree.</p>
</section>
</section>
<section id="simulating-data-and-plotting-a-tree" class="level3 page-columns page-full" data-number="10.4.2">
<h3 data-number="10.4.2" class="anchored" data-anchor-id="simulating-data-and-plotting-a-tree"><span class="header-section-number">10.4.2</span> Simulating data and plotting a tree</h3>
<p>Suppose we already know our phylogenetic tree and want to simulate the evolution of the nucleotides down this tree. First we visualize the tree <code>tree1</code> using <code>ggtree</code>; loading the tree and the relevant packages with:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phangorn"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggtree"</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">file.path</span>(DATA,<span class="st">"tree1.RData"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <code>ggtree</code> function to plot <code>tree1</code>; make the tips of the tree green triangles, the ancestral nodes, red circles.</p>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree1, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"darkgreen"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">right =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">size =</span> <span class="dv">7</span>, <span class="at">angle =</span> <span class="dv">90</span>, <span class="at">offset =</span> <span class="fl">0.05</span>) <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">shape =</span> isTip, <span class="at">color =</span> isTip), <span class="at">size =</span> <span class="dv">5</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-Atree1b" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-Atree1b-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-Atree1b-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21" title="Figure&nbsp;10.16: This is the tree we use as our true parameter. We generate nucleotides one at a time from the root and `dropping’ them down the tree. With some probability proportional to the edge lengths, mutations occur down the branches."><img src="10-chap_files/figure-html/fig-Atree1b-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-Atree1b-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.16: This is the tree we use as our <em>true</em> parameter. We generate nucleotides one at a time from the root and `dropping’ them down the tree. With some probability proportional to the edge lengths, mutations occur down the branches.
</figcaption>
</figure>
</div>
</div></div></div>
<p>Now we generate some sequences from our tree. Each sequence starts with a new nucleotide letter generated randomly at the root; mutations may occur as we go down the tree. You can see in <a href="#fig-ggtreeAlignment" class="quarto-xref">Figure&nbsp;<span>10.17</span></a> that the colors are not equally represented, because the frequency at the root was chosen to be different from the uniform, see the following code.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>seqs6 <span class="ot">=</span> <span class="fu">simSeq</span>(tree1, <span class="at">l =</span> <span class="dv">60</span>, <span class="at">type =</span> <span class="st">"DNA"</span>, <span class="at">bf =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">3</span>)<span class="sc">/</span><span class="dv">8</span>, <span class="at">rate =</span> <span class="fl">0.1</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>seqs6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6 sequences with 60 character and 30 different site patterns.
The states are a c g t </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>mat6df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">as.character</span>(seqs6))</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ggtree</span>(tree1, <span class="at">lwd =</span> <span class="fl">1.2</span>) <span class="sc">+</span> <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">x =</span> branch), <span class="at">size =</span> <span class="dv">5</span>, <span class="at">vjust =</span> <span class="dv">2</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">gheatmap</span>(p, mat6df[, <span class="dv">1</span><span class="sc">:</span><span class="dv">60</span>], <span class="at">offset =</span> <span class="fl">0.01</span>, <span class="at">colnames =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ggtreeAlignment" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ggtreeAlignment-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-ggtreeAlignment-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22" title="Figure&nbsp;10.17: The tree on the left was used to generate the sequences on the right according to a Jukes Cantor model. The nucleotide frequencies generated at the root were quite unequal, with A and C being generated more rarely. As the sequences percolate down the tree, mutations occur, they are more likely to occur on the longer branches."><img src="10-chap_files/figure-html/fig-ggtreeAlignment-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ggtreeAlignment-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.17: The tree on the left was used to generate the sequences on the right according to a Jukes Cantor model. The nucleotide frequencies generated at the root were quite unequal, with <code>A</code> and <code>C</code> being generated more rarely. As the sequences percolate down the tree, mutations occur, they are more likely to occur on the longer branches.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-graphs-simulateseqs" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.9
</div>
</div>
<div class="callout-body-container callout-body">
<p>Experiment with the code above. Change the <code>bf</code> and <code>rate</code> arguments in the <code>simSeq</code> function to make mutations more likely. Do you think sequences generated with a very high mutation rate would make it easier to infer the tree that generated them?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Very high mutation rates result in mutations overwriting themselves and make inference more difficult. Of course, there is a sweet spot because some mutations have to occur in order for us to resolve the tree branches. After a certain time and a certain number of mutations it may be very difficult to see what was happening at the root. See <span class="citation" data-cites="Mossel:2003">Mossel (<a href="16-chap.html#ref-Mossel:2003" role="doc-biblioref">2003</a>)</span> for details.</p>
</div>
</div>
</div>
<div id="wrn-graphs-distancebias" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.10
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Estimation bias: distance underestimation.</strong><br>
1) If we only count the number of changes between two sequences using a simple Hamming distance, but there has been much evolutionary change between the two, why do we underestimate the distance between the sequences?<br>
2) Will be the bias be larger for smaller evolutionary distances?</p>
</div>
</div>
<p>The standard Markovian models of evolution we saw above enable us to improve these estimates.</p>
</section>
<section id="estimating-a-phylogenetic-tree" class="level3 page-columns page-full" data-number="10.4.3">
<h3 data-number="10.4.3" class="anchored" data-anchor-id="estimating-a-phylogenetic-tree"><span class="header-section-number">10.4.3</span> Estimating a phylogenetic tree</h3>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="imgs/Magnify.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="imgs/Magnify.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a></p>
</figure>
</div>
</div></div></div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">“In solving a problem of this sort, the grand thing is to be able to reason backward. That is a very useful accomplishment, and a very easy one, but people do not practise it much. In the everyday affairs of life it is more useful to reason forward, and so the other comes to be neglected. There are fifty who can reason synthetically for one who can reason analytically”. <strong>Sherlock Holmes</strong></span></div></div>
<p>When the true tree-parameter is known, the above-mentioned probabilistic generative models of evolution tells us what patterns to expect in the sequences. As we have seen in earlier chapters, statistics means going back from the data to reasonable estimates of the parameters; here the tree itself and the branch lengths, even the evolutionary rates can be considered to be the parameters.</p>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-igraphsteiner" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-igraphsteiner-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-igraphsteiner-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24" title="Figure&nbsp;10.18: A Steiner tree, the inner points are represented as squares. The method for creating the shortest tree that passes through all outer 1,2,5,6 is to create two inside (“ancester”) points 3 and 4."><img src="10-chap_files/figure-html/fig-igraphsteiner-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-igraphsteiner-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.18: A Steiner tree, the inner points are represented as squares. The method for creating the shortest tree that passes through all outer 1,2,5,6 is to create two inside (“ancester”) points 3 and 4.
</figcaption>
</figure>
</div>
</div></div></div>
<p>There are several approaches to estimation: tree `building’ is no exception, here are the main ones:</p>
<p><strong>A nonparametric estimate: the parsimony tree</strong> Parsimony is a nonparametric method that minimizes the number of changes necessary to explain the data, it’s solution is the same as that of the Steiner tree problem (see <a href="#fig-igraphsteiner" class="quarto-xref">Figure&nbsp;<span>10.18</span></a>).</p>
<p><strong>A parametric estimate: the maximum likelihood tree</strong> In order to estimate the tree using a maximum likelihood or Bayesian approach one needs a model for molecular evolution that integrates mutation rates and branch edge lengths. ML estimation (e.g., <code>Phyml</code>, <code>FastML</code>, <code>RaxML</code>) use efficient optimization algorithms to maximize the likelihood of a tree under the model assumptions.</p>
<p><strong>Bayesian posterior distributions for trees</strong> Bayesian estimation, MrBayes <span class="citation" data-cites="mrbayes3">(<a href="16-chap.html#ref-mrbayes3" role="doc-biblioref">Ronquist et al. 2012</a>)</span> or BEAST <span class="citation" data-cites="beast">(<a href="16-chap.html#ref-beast" role="doc-biblioref">Bouckaert et al. 2014</a>)</span> both use MCMC to find posterior distributions of the phylogenies. Bayesian methods are not directly integrated into R and require the user to import the collections of trees generated by Monte Carlo methods in order to summarize them and make confidence statements see <span class="citation" data-cites="distory-paper">Chakerian and Holmes (<a href="16-chap.html#ref-distory-paper" role="doc-biblioref">2012</a>)</span> for simple examples.</p>
<p><strong>The semi-parametric approach: distance based methods</strong> These methods, called Neighbor Joining and UPGMA, are quite similar to the hierachical clusering algorithms we already encountered in <a href="05-chap.html" class="quarto-xref"><span>Chapter 5</span></a>. However, the distance estimation steps uses the parametric evolutionary models of <a href="#tbl-JCKimura" class="quarto-xref">Table&nbsp;<span>10.2</span></a>; the `parametric’ part of why we call the method semi-parametric.</p>
<p>The neighbor-joining algorithm itself uses Steiner points as the summary of two combined points, and proceeds iteratively as in hierarchical clustering. It can be quite fast and is often used as a good starting point for the more time-consuming methods.</p>
<p>Let’s start by estimating the tree from the data <code>seqs6</code> using the <code>nj</code> (neighbor joining) on DNA distances based on the one-parameter Jukes-Cantor model, we make <a href="#fig-njtree1" class="quarto-xref">Figure&nbsp;<span>10.19</span></a> using the <code>ggtree</code> function:</p>
<div class="cell page-columns page-full" data-layout-align="center" data-warning.known="tree contained negative edge length">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>tree.nj <span class="ot">=</span> <span class="fu">nj</span>(<span class="fu">dist.ml</span>(seqs6, <span class="st">"JC69"</span>))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(tree.nj) <span class="sc">+</span> <span class="fu">geom_tiplab</span>(<span class="at">size =</span> <span class="dv">7</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-njtree1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-njtree1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-njtree1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25" title="Figure&nbsp;10.19: Trees built with a neighbor joining algorithm are very fast to compute and are often used as initial values for more expensive estimation procedures such as the maximum likelihood or parsimony."><img src="10-chap_files/figure-html/fig-njtree1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-njtree1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.19: Trees built with a neighbor joining algorithm are very fast to compute and are often used as initial values for more expensive estimation procedures such as the maximum likelihood or parsimony.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-graphs-mlscores" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.11
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate the maximum likelihood scores of the tree1 given the <code>seqs6</code> data and compare them to those of the neighbor joining tree.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-24-contents" aria-controls="callout-24" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-24" class="callout-24-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">=</span> <span class="fu">pml</span>(tree1, seqs6, <span class="at">k =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div id="wrn-graphs-AAtree" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.12
</div>
</div>
<div class="callout-body-container callout-body">
<p>When we have aligned amino acids from which we want to infer a tree, we use (<span class="math inline">\(20 \times 20\)</span>) transition matrices. The methods for estimating the phylogenetic are very similar. Try this in phangorn with an HIV amino acid sequence downloaded from <a href="https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html" class="uri">https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html</a>.</p>
</div>
</div>
<p>The quality of the tree estimates depend on the number of sequences per taxa and the distance to the root. We can evaluate the quality of the estimates either by using parametric and nonparametric bootstraps or performing Bayesian tree estimation using MCMC. For examples of how to visualize and compare the sampling distribution of trees, see <span class="citation" data-cites="distory-paper">Chakerian and Holmes (<a href="16-chap.html#ref-distory-paper" role="doc-biblioref">2012</a>)</span>.</p>
</section>
<section id="application-to-16s-rrna-data" class="level3 page-columns page-full" data-number="10.4.4">
<h3 data-number="10.4.4" class="anchored" data-anchor-id="application-to-16s-rrna-data"><span class="header-section-number">10.4.4</span> Application to 16S rRNA data</h3>
<p>In <a href="05-chap.html" class="quarto-xref"><span>Chapter 5</span></a> we saw how to use a probabilistic clustering method to denoise 16S rRNA sequences. We can now reload these denoised sequences and preprocess them before building their phylogeny<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn4"><p><sup>4</sup>&nbsp;In order to keep all the information and be able to compare sequences from different experiments, we use the sequences themselves as their label<span class="citation" data-cites="Callahan:2017">(<a href="16-chap.html#ref-Callahan:2017" role="doc-biblioref">Callahan, McMurdie, and Holmes 2017</a>)</span>.</p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dada2"</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(DATA,<span class="st">"seqtab.rds"</span>))</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>seqs <span class="ot">=</span> <span class="fu">getSequences</span>(seqtab)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seqs) <span class="ot">=</span> seqs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One of the benefits of using well-studied marker loci such as the 16S rRNA gene is the ability to taxonomically classify the sequenced variants. <strong><a href="https://bioconductor.org/packages/dada2/">dada2</a></strong> includes a naive Bayesian classifier method for this purpose <span class="citation" data-cites="wang2007naive">(<a href="16-chap.html#ref-wang2007naive" role="doc-biblioref">Wang et al. 2007</a>)</span>. This classifier compares sequence variants to training sets of classified sequences. Here we use the RDP v16 training set <span class="citation" data-cites="cole2009rdp">(<a href="16-chap.html#ref-cole2009rdp" role="doc-biblioref">Cole et al. 2009</a>)</span><a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>. For example, code for such a classification might look like this.</p>
<div class="no-row-height column-margin column-container"><div id="fn5"><p><sup>5</sup>&nbsp;See the download link on the dada2 website: <a href="https://benjjneb.github.io/dada2/training.html" class="uri">https://benjjneb.github.io/dada2/training.html</a></p></div></div><div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>fastaRef <span class="ot">=</span> <span class="st">"../tmp/rdp_train_set_16.fa.gz"</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">=</span> <span class="fu">assignTaxonomy</span>(seqtab, <span class="at">refFasta =</span> fastaRef)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the <code>assignTaxonomy</code> function runs for a while, the above code is not live and we here load a previously computed result, a table of taxonomic information:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>taxtab <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(DATA,<span class="st">"taxtab16.rds"</span>))</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(taxtab)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 268   6</code></pre>
</div>
</div>
<div id="wrn-graphs-taxtabhead" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.13
</div>
</div>
<div class="callout-body-container callout-body">
<p>Write one line of code using R’s pipe operator <code>|&gt;</code> that shows just the first 6 rows of the taxonomic information without the row names.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(taxtab) <span class="sc">|&gt;</span> <span class="st">`</span><span class="at">rownames&lt;-</span><span class="st">`</span>(<span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Kingdom    Phylum          Class         Order          
[1,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
[2,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
[3,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
[4,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
[5,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
[6,] "Bacteria" "Bacteroidetes" "Bacteroidia" "Bacteroidales"
     Family               Genus        
[1,] "Porphyromonadaceae" NA           
[2,] "Porphyromonadaceae" NA           
[3,] "Porphyromonadaceae" NA           
[4,] "Porphyromonadaceae" "Barnesiella"
[5,] "Bacteroidaceae"     "Bacteroides"
[6,] "Porphyromonadaceae" NA           </code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="wrn-graphs-treeandtax" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.14
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the difference between taxonomic and phylogenetic information?</p>
</div>
</div>
<p>Note that as the <code>seqs</code> data are randomly generated, they are “cleaner” than the real data we will have to handle.</p>
<p>In particular naturally occurring raw sequences have to be <strong>aligned</strong>. This is necessary as there are often extra nucleotides in some sequences, a consequence of what we call <em>indel</em> events<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>. Also mutations occur and appear as substitutions of one nucleotide by another.</p>
<div class="no-row-height column-margin column-container"><div id="fn6"><p><sup>6</sup>&nbsp;A nucleotide is deleted or inserted and it is often hard to distinguish which took place.</p></div></div><p>Here is an example of what the first few characters of <strong>aligned</strong> sequences looks like:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">readLines</span>(<span class="fu">file.path</span>(DATA,<span class="st">"mal2.dna.txt"</span>)) <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">12</span>) <span class="sc">|&gt;</span> <span class="fu">cat</span>(<span class="at">sep=</span><span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    11   1620
Pre1        GTACTTGTTA GGCCTTATAA GAAAAAAGT- TATTAACTTA AGGAATTATA
Pme2        GTATCTGTTA AGCCTTATAA AAAGATAGT- T-TAAATTAA AGGAATTATA
Pma3        GTATTTGTTA AGCCTTATAA GAGAAAAGTA TATTAACTTA AGGA-TTATA
Pfa4        GTATTTGTTA GGCCTTATAA GAAAAAAGT- TATTAACTTA AGGAATTATA
Pbe5        GTATTTGTTA AGCCTTATAA GAAAAA--T- TTTTAATTAA AGGAATTATA
Plo6        GTATTTGTTA AGCCTTATAA GAAAAAAGT- TACTAACTAA AGGAATTATA
Pfr7        GTACTTGTTA AGCCTTATAA GAAAGAAGT- TATTAACTTA AGGAATTATA
Pkn8        GTACTTGTTA AGCCTTATAA GAAAAGAGT- TATTAACTTA AGGAATTATA
Pcy9        GTACTCGTTA AGCCTTTTAA GAAAAAAGT- TATTAACTTA AGGAATTATA
Pvi10       GTACTTGTTA AGCCTTTTAA GAAAAAAGT- TATTAACTTA AGGAATTATA
Pga11       GTATTTGTTA AGCCTTATAA GAAAAAAGT- TATTAATTTA AGGAATTATA</code></pre>
</div>
</div>
<p>We will perform this multiple-alignment on our <code>seqs</code> data using the <strong><a href="https://bioconductor.org/packages/DECIPHER/">DECIPHER</a></strong> package <span class="citation" data-cites="wright2015decipher">(<a href="16-chap.html#ref-wright2015decipher" role="doc-biblioref">Wright 2015</a>)</span>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"DECIPHER"</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>alignment <span class="ot">=</span> <span class="fu">AlignSeqs</span>(<span class="fu">DNAStringSet</span>(seqs), <span class="at">anchor =</span> <span class="cn">NA</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use the <strong><a href="https://cran.r-project.org/web/packages/phangorn/">phangorn</a></strong> package to build the MLE tree (under the GTR model), but will use the neighbor-joining tree as our starting point.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>phangAlign <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">phyDat</span>(<span class="fu">as</span>(alignment, <span class="st">"matrix"</span>), <span class="at">type =</span> <span class="st">"DNA"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>dm <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">dist.ml</span>(phangAlign)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>treeNJ <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">NJ</span>(dm)   <span class="co"># Note: tip order != sequence order</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">pml</span>(treeNJ, <span class="at">data =</span> phangAlign)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>fitGTR <span class="ot">=</span> <span class="fu">update</span>(fit, <span class="at">k =</span> <span class="dv">4</span>, <span class="at">inv =</span> <span class="fl">0.2</span>)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>fitGTR <span class="ot">=</span> phangorn<span class="sc">::</span><span class="fu">optim.pml</span>(fitGTR, <span class="at">model =</span> <span class="st">"GTR"</span>, <span class="at">optInv =</span> <span class="cn">TRUE</span>,</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">optGamma =</span> <span class="cn">TRUE</span>,  <span class="at">rearrangement =</span> <span class="st">"stochastic"</span>,</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">control =</span> phangorn<span class="sc">::</span><span class="fu">pml.control</span>(<span class="at">trace =</span> <span class="dv">0</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="combining-phylogenetic-trees-into-a-data-analysis" class="level2 page-columns page-full" data-number="10.5">
<h2 data-number="10.5" class="anchored" data-anchor-id="combining-phylogenetic-trees-into-a-data-analysis"><span class="header-section-number">10.5</span> Combining phylogenetic trees into a data analysis</h2>
<p>We now need to combine the phylogenetic tree and the denoised read abundances with the complementary information provided about the samples from which the reads were gathered. This information about the sample is often provided as a spreadhseet (or <code>.csv</code>), and sometimes called the <em>meta</em>-data<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>. This data combination step is facilitated by the specialized containers and accessors that <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong> provides.</p>
<div class="no-row-height column-margin column-container"><div id="fn7"><p><sup>7</sup>&nbsp;We consider the prefix <em>meta</em> unhelpful and potentially confusing here: the data about the samples is just that: data.</p></div></div><p>The following set of steps contains a few data cleanup and reorganization tasks—a dull but necessary part of applied statistics—that end in the creation of the object <code>ps1</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">"../data/MIMARKS_Data_combined.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>SampleID <span class="ot">=</span> <span class="fu">paste0</span>(<span class="fu">gsub</span>(<span class="st">"00"</span>, <span class="st">""</span>, samples<span class="sc">$</span>host_subject_id), </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"D"</span>, samples<span class="sc">$</span>age<span class="dv">-21</span>) </span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">=</span> samples[<span class="sc">!</span><span class="fu">duplicated</span>(samples<span class="sc">$</span>SampleID), ] </span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="fu">rownames</span>(seqtab) <span class="sc">%in%</span> samples<span class="sc">$</span>SampleID))</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(samples) <span class="ot">=</span> samples<span class="sc">$</span>SampleID </span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>keepCols <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"collection_date"</span>, <span class="st">"biome"</span>, <span class="st">"target_gene"</span>, <span class="st">"target_subfragment"</span>, </span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"host_common_name"</span>, <span class="st">"host_subject_id"</span>, <span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"body_product"</span>, <span class="st">"tot_mass"</span>,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"diet"</span>, <span class="st">"family_relationship"</span>, <span class="st">"genotype"</span>, <span class="st">"SampleID"</span>) </span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">=</span> samples[<span class="fu">rownames</span>(seqtab), keepCols] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The sample-by-sequence feature table, the sample (meta)data, the sequence taxonomies, and the phylogenetic tree—are combined into a single object as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phyloseq"</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>pso <span class="ot">=</span> <span class="fu">phyloseq</span>(<span class="fu">tax_table</span>(taxtab), </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>               <span class="fu">sample_data</span>(samples),</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>               <span class="fu">otu_table</span>(seqtab, <span class="at">taxa_are_rows =</span> <span class="cn">FALSE</span>), </span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>               <span class="fu">phy_tree</span>(fitGTR<span class="sc">$</span>tree))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have already encountered several cases of combining heterogeneous datasets into special data classes that automate the linking and keeping consistent the different parts of the dataset (e.g., in <a href="08-chap.html" class="quarto-xref"><span>Chapter 8</span></a>, when we studied the <code>pasilla</code> data).</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Look at the detailed <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong> <a href="https://joey711.github.io/phyloseq/tutorials-index.html">documentation here</a>. Try a few filtering operations. For instance, create a subset of the data that contains the tree, taxa abundance table, the sample and taxa information for only the samples that have more than 5000 reads.</p>
</div>
</div>
<p>This can be done in one line:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">prune_samples</span>(<span class="fu">rowSums</span>(<span class="fu">otu_table</span>(pso)) <span class="sc">&gt;</span> <span class="dv">5000</span>, pso)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 268 taxa and 10 samples ]
sample_data() Sample Data:       [ 10 samples by 14 sample variables ]
tax_table()   Taxonomy Table:    [ 268 taxa by 6 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 268 tips and 266 internal nodes ]</code></pre>
</div>
</div>
<p>We can also make other data transformations while maintaining the integrity of the links between all the data components.</p>
<div id="wrn-graphs-phyloseqsubset" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.15
</div>
</div>
<div class="callout-body-container callout-body">
<p>What do the following lines of code do?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>prevalence <span class="ot">=</span> <span class="fu">apply</span>(<span class="at">X =</span> <span class="fu">otu_table</span>(pso),</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">MARGIN =</span> <span class="fu">ifelse</span>(<span class="fu">taxa_are_rows</span>(pso), <span class="at">yes =</span> <span class="dv">1</span>, <span class="at">no =</span> <span class="dv">2</span>),</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">FUN =</span> <span class="cf">function</span>(x) {<span class="fu">sum</span>(x <span class="sc">&gt;</span> <span class="dv">0</span>)})</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>prevdf <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Prevalence =</span> prevalence,</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">TotalAbundance =</span> <span class="fu">taxa_sums</span>(pso),</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">tax_table</span>(pso))</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>tab <span class="ot">=</span> <span class="fu">table</span>(prevdf<span class="sc">$</span>Phylum)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>keepPhyla <span class="ot">=</span> <span class="fu">names</span>(tab)[tab<span class="sc">&gt;</span><span class="dv">5</span>]</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>prevdf1   <span class="ot">=</span> <span class="fu">subset</span>(prevdf,   Phylum <span class="sc">%in%</span> keepPhyla)</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>ps2v      <span class="ot">=</span> <span class="fu">subset_taxa</span>(pso, Phylum <span class="sc">%in%</span> keepPhyla)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Plotting the abundances for certain bacteria can be done using barcharts. <strong><a href="https://cran.r-project.org/web/packages/ggplot2/">ggplot2</a></strong> expressions have been hardwired into suitable one-line calls in the <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong> package. There is also an interactive <a href="http://joey711.github.io/shiny-phyloseq/">Shiny-phyloseq</a> browser based tool <span class="citation" data-cites="mcmurdie2015">(<a href="16-chap.html#ref-mcmurdie2015" role="doc-biblioref">McMurdie and Holmes 2015</a>)</span>. For more details, please see the online vignettes.</p>
<section id="hierarchical-multiple-testing" class="level3 page-columns page-full" data-number="10.5.1">
<h3 data-number="10.5.1" class="anchored" data-anchor-id="hierarchical-multiple-testing"><span class="header-section-number">10.5.1</span> Hierarchical multiple testing</h3>
<p>Hypothesis testing can identify individual bacteria whose abundance relates to sample variables of interest. A standard approach is very similar to the approach we already visited in <a href="06-chap.html" class="quarto-xref"><span>Chapter 6</span></a>. Compute a test statistic for each taxa individually; then jointly adjust p-values to ensure a false discovery rate upper bound. However, this procedure does not exploit the structure among the tested hypotheses. For example, if we observe that one Ruminococcus species is strongly associated with age, but the biological reason for this sits at the genus level, then we would expect other species to have such an association as well. To integrate such information, <span class="citation" data-cites="benjamini2003hierarchical">Benjamini and Yekutieli (<a href="16-chap.html#ref-benjamini2003hierarchical" role="doc-biblioref">2003</a>)</span> and <span class="citation" data-cites="benjamini2014selective">Benjamini and Bogomolov (<a href="16-chap.html#ref-benjamini2014selective" role="doc-biblioref">2014</a>)</span> proposed a hierarchical testing procedure, where lower level taxonomic groups are only tested if higher levels are found to be be associated. In the case where many related species have a slight signal, this pooling of information can increase power.</p>
<p>We apply this method to test the association between microbial abundance and age. We use the data object <code>ps1</code>, which is similar to <code>pso</code> from above, but has undergone some additional transformation and filtering steps. We also need to apply the normalization protocols available in the <strong><a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a></strong> package, which we discussed in <a href="08-chap.html" class="quarto-xref"><span>Chapter 8</span></a>, following <span class="citation" data-cites="LoveDESeq2">Love, Huber, and Anders (<a href="16-chap.html#ref-LoveDESeq2" role="doc-biblioref">2014</a>)</span> for RNA-Seq data and <span class="citation" data-cites="mcmurdie2014">McMurdie and Holmes (<a href="16-chap.html#ref-mcmurdie2014" role="doc-biblioref">2014</a>)</span> for 16S rRNA generated count data.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># warning: !expr c("DESeqDataSet.se, design = design, ignoreRank.: some variables in design formula are characters, converting to factors")</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>ps1 <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(DATA,<span class="st">"ps1.rds"</span>))</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>ps_dds <span class="ot">=</span> <span class="fu">phyloseq_to_deseq2</span>(ps1, <span class="at">design =</span> <span class="sc">~</span> ageBin <span class="sc">+</span> family_relationship)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>geometricmean <span class="ot">=</span> <span class="cf">function</span>(x)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>   <span class="cf">if</span> (<span class="fu">all</span>(x <span class="sc">==</span> <span class="dv">0</span>)) { <span class="dv">0</span> } <span class="cf">else</span> { <span class="fu">exp</span>(<span class="fu">mean</span>(<span class="fu">log</span>(x[x <span class="sc">!=</span> <span class="dv">0</span>]))) }</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>geoMeans <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">counts</span>(ps_dds), <span class="dv">1</span>, geometricmean)</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>ps_dds <span class="ot">=</span> <span class="fu">estimateSizeFactors</span>(ps_dds, <span class="at">geoMeans =</span> geoMeans)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>ps_dds <span class="ot">=</span> <span class="fu">estimateDispersions</span>(ps_dds)</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>abund <span class="ot">=</span> <span class="fu">getVarianceStabilizedData</span>(ps_dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use the <strong><a href="https://cran.r-project.org/web/packages/structSSI/">structSSI</a></strong> package to perform the hierarchical testing <span class="citation" data-cites="sankaran2014structssi">(<a href="16-chap.html#ref-sankaran2014structssi" role="doc-biblioref">Sankaran and Holmes 2014</a>)</span>. For more convenient printing, we first shorten the names of the taxa:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(abund) <span class="ot">=</span> <span class="fu">substr</span>(<span class="fu">rownames</span>(abund), <span class="dv">1</span>, <span class="dv">5</span>) <span class="sc">|&gt;</span> <span class="fu">make.names</span>(<span class="at">unique =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The hierarchical testing procedure we are now going to do differs from standard multiple hypothesis testing in that univariate tests are done not only for every taxon, but for each higher-level taxonomic group. A helper function, <code>treePValues</code>, is available for this: it expects an edge list encoding parent-child relationships, with the first row specifying the root node.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"structSSI"</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>el <span class="ot">=</span> <span class="fu">phy_tree</span>(ps1)<span class="sc">$</span>edge</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>el0 <span class="ot">=</span> el</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>el0 <span class="ot">=</span> el0[<span class="fu">rev</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(el))), ]</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>el_names <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">rownames</span>(abund), <span class="fu">seq_len</span>(<span class="fu">phy_tree</span>(ps1)<span class="sc">$</span>Nnode))</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>el[, <span class="dv">1</span>] <span class="ot">=</span> el_names[el0[, <span class="dv">1</span>]]</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>el[, <span class="dv">2</span>] <span class="ot">=</span> el_names[el0[, <span class="dv">2</span>]]</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>unadj_p <span class="ot">=</span> <span class="fu">treePValues</span>(el, abund, <span class="fu">sample_data</span>(ps1)<span class="sc">$</span>ageBin)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now do our FDR calculations using the hierarchical testing procedure. The test results are guaranteed to control several variants of FDR, but at different levels; we defer details to <span class="citation" data-cites="benjamini2003hierarchical benjamini2014selective sankaran2014structssi">(<a href="16-chap.html#ref-benjamini2003hierarchical" role="doc-biblioref">Benjamini and Yekutieli 2003</a>; <a href="16-chap.html#ref-benjamini2014selective" role="doc-biblioref">Benjamini and Bogomolov 2014</a>; <a href="16-chap.html#ref-sankaran2014structssi" role="doc-biblioref">Sankaran and Holmes 2014</a>)</span>.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try the following code, including the interactive plotting command that will open a browser window:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>hfdr_res <span class="ot">=</span> <span class="fu">hFDR.adjust</span>(unadj_p, el, <span class="fl">0.75</span>)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hfdr_res)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co">#plot(hfdr_res, height = 5000) # not run: opens in a browser</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-structssi-hfdr" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-structssi-hfdr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/structssi-screenshot.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26" title="Figure&nbsp;10.20: A screenshot of a subtree with many differentially abundant microbes, as determined by the hierarchical testing procedure. Currently the user is hovering over the node associated with microbe GCGAG.33; this causes the adjusted p-value (0.029) to appear."><img src="imgs/structssi-screenshot.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-structssi-hfdr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.20: A screenshot of a subtree with many differentially abundant microbes, as determined by the hierarchical testing procedure. Currently the user is hovering over the node associated with microbe GCGAG.33; this causes the adjusted p-value (0.029) to appear.
</figcaption>
</figure>
</div>
</div></div></div>
<p>The plot opens in a new browser – a static screenshot of a subtree is displayed in <a href="#fig-structssi-hfdr" class="quarto-xref">Figure&nbsp;<span>10.20</span></a>. Nodes are shaded according to p-values, from blue to orange, representing the strongest to weakest associations. Grey nodes were never tested, to focus power on more promising subtrees. Scanning the full tree; it becomes clear that the association between age group and taxonomic abundances is present in only a few isolated taxonomic groups. It is quite strong in those groups. To give context to these results, we can retrieve the taxonomic identity of the rejected hypotheses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">digits =</span> <span class="dv">3</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>tax <span class="ot">=</span> <span class="fu">tax_table</span>(ps1)[, <span class="fu">c</span>(<span class="st">"Family"</span>, <span class="st">"Genus"</span>)] <span class="sc">|&gt;</span> <span class="fu">data.frame</span>()</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>tax<span class="sc">$</span>seq <span class="ot">=</span> <span class="fu">rownames</span>(abund)</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>hfdr_res<span class="sc">@</span>p.vals<span class="sc">$</span>seq <span class="ot">=</span> <span class="fu">rownames</span>(hfdr_res<span class="sc">@</span>p.vals)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(tax, hfdr_res<span class="sc">@</span>p.vals[,<span class="sc">-</span><span class="dv">3</span>]) <span class="sc">|&gt;</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(adjp) <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">9</span>) <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              Family       Genus hypothesisName hypothesisIndex
1 Porphyromonadaceae        &lt;NA&gt;           &lt;NA&gt;              NA
2 Porphyromonadaceae        &lt;NA&gt;           &lt;NA&gt;              NA
3 Porphyromonadaceae        &lt;NA&gt;           &lt;NA&gt;              NA
4 Porphyromonadaceae Barnesiella           &lt;NA&gt;              NA
5     Bacteroidaceae Bacteroides           &lt;NA&gt;              NA
6 Porphyromonadaceae Barnesiella           &lt;NA&gt;              NA
7      Rikenellaceae   Alistipes           &lt;NA&gt;              NA
8 Porphyromonadaceae        &lt;NA&gt;           &lt;NA&gt;              NA
9 Porphyromonadaceae        &lt;NA&gt;           &lt;NA&gt;              NA</code></pre>
</div>
</div>
<p>It seems that the most strongly associated bacteria all belong to family <em>Lachnospiraceae</em>.</p>
</section>
</section>
<section id="minimum-spanning-trees" class="level2 page-columns page-full" data-number="10.6">
<h2 data-number="10.6" class="anchored" data-anchor-id="minimum-spanning-trees"><span class="header-section-number">10.6</span> Minimum spanning trees</h2>
<p>A very simple and useful graph is the so-called <strong>minimum spanning tree</strong> (<strong>MST</strong>). Given a set of vertices, a <strong>spanning tree</strong> is a tree that goes through all points at least once. Examples are shown in <a href="#fig-graphs-MST" class="quarto-xref">Figure&nbsp;<span>10.21</span></a>. Given distances between vertices, the MST is the spanning tree with the minimum total length (see <a href="#fig-graphs-MST" class="quarto-xref">Figure&nbsp;<span>10.21</span></a>).</p>
<p>Greedy algorithms work well for computing the MST and there are many implementations in R: <code>mstree</code> in <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong>, <code>mst</code> in <strong><a href="https://cran.r-project.org/web/packages/ape/">ape</a></strong>, <code>spantree</code> in <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong>, <code>mst</code> in <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong>.</p>
<div id="fig-graphs-MST" class="quarto-layout-panel page-columns page-full">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-graphs-MST-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca" class="page-columns page-full">
<div class="quarto-layout-row page-full">

<div class="no-row-height column-margin column-container"><div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-graphs-MST" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-graphs-MST-1" class="quarto-float quarto-figure quarto-figure-center anchored" width="320" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-graphs-MST-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-graphs-MST-1.png" class="lightbox" data-gallery="fig-graphs-MST" title="Figure&nbsp;10.21&nbsp;(a): "><img src="10-chap_files/figure-html/fig-graphs-MST-1.png" id="fig-graphs-MST-1" class="img-fluid quarto-figure quarto-figure-center anchored figure-img" data-ref-parent="fig-graphs-MST" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-graphs-MST-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a)
</figcaption>
</figure>
</div>
</div></div></div>
<div class="quarto-layout-row page-full">

<div class="no-row-height column-margin column-container"><div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-graphs-MST" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-graphs-MST-2" class="quarto-float quarto-figure quarto-figure-center anchored" width="320" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-graphs-MST-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-graphs-MST-2.png" class="lightbox" data-gallery="fig-graphs-MST" title="Figure&nbsp;10.21&nbsp;(b): "><img src="10-chap_files/figure-html/fig-graphs-MST-2.png" id="fig-graphs-MST-2" class="img-fluid quarto-figure quarto-figure-center anchored figure-img" data-ref-parent="fig-graphs-MST" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-graphs-MST-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b)
</figcaption>
</figure>
</div>
</div></div></div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-graphs-MST-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.21: Two spanning trees for the same set of six vertices. The blue graph is the minimum spanning tree, if the Euclidean distances between the points in the 2D plane are used.
</figcaption>
</figure>
</div>
<p>Here we are going to take the DNA sequence distances between strains of HIV from patients all over the world and construct their minimum spanning tree. The result is shown in <a href="#fig-HIVMSTi" class="quarto-xref">Figure&nbsp;<span>10.22</span></a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">file.path</span>(DATA, <span class="st">"dist2009c.RData"</span>))</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>country09 <span class="ot">=</span> <span class="fu">attr</span>(dist2009c, <span class="st">"Label"</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>mstree2009 <span class="ot">=</span> ape<span class="sc">::</span><span class="fu">mst</span>(dist2009c)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>gr09 <span class="ot">=</span> <span class="fu">graph_from_adjacency_matrix</span>(mstree2009, <span class="at">mode =</span> <span class="st">"undirected"</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(gr09, <span class="at">layout=</span><span class="st">"fr"</span>) <span class="sc">+</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">color =</span> <span class="st">"black"</span>,<span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">vertex_attr</span>(gr09)<span class="sc">$</span>name), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">vertex_attr</span>(gr09)<span class="sc">$</span>name), <span class="at">color=</span><span class="st">"black"</span>,<span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">keyheight=</span><span class="fl">0.1</span>,<span class="at">keywidth=</span><span class="fl">0.1</span>,</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">title=</span><span class="st">"Countries"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-HIVMSTi" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-HIVMSTi-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-HIVMSTi-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-29" title="Figure&nbsp;10.22: The minimum spanning tree computed from DNA distances between HIV sequences from samples taken in 2009 and whose country of origin was known, data as published in the HIVdb database [@HIVdb]."><img src="10-chap_files/figure-html/fig-HIVMSTi-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="600"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-HIVMSTi-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.22: The minimum spanning tree computed from DNA distances between HIV sequences from samples taken in 2009 and whose country of origin was known, data as published in the <code>HIVdb</code> database <span class="citation" data-cites="HIVdb">(<a href="16-chap.html#ref-HIVdb" role="doc-biblioref">Rhee et al. 2003</a>)</span>.
</figcaption>
</figure>
</div>
</div>
</div>
<div id="wrn-graphs-networklabelrepel" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.16
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make the network plot again, but replace <code>geom_node_text</code> with labels that repel each other to minimize the overlapping node labels.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-33-contents" aria-controls="callout-33" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-33" class="callout-33-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See <a href="#fig-graphs-networklabelrepel" class="quarto-xref">Figure&nbsp;<span>10.23</span></a>. Maybe a better, or additional approach would be to first cluster those vertices that are very close together and from the same country.</p>
<div class="cell" data-layout-align="center" data-warning.known="too many overlaps">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggraph"</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(gr09, <span class="at">layout=</span><span class="st">"fr"</span>) <span class="sc">+</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">color =</span> <span class="st">"black"</span>,<span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">vertex_attr</span>(gr09)<span class="sc">$</span>name), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_label</span>(<span class="fu">aes</span>(<span class="at">label =</span> <span class="fu">vertex_attr</span>(gr09)<span class="sc">$</span>name), <span class="at">color=</span><span class="st">"black"</span>,<span class="at">size=</span><span class="dv">2</span>,<span class="at">repel=</span><span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">keyheight=</span><span class="fl">0.1</span>,<span class="at">keywidth=</span><span class="fl">0.1</span>,</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">title=</span><span class="st">"Countries"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-graphs-networklabelrepel" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-graphs-networklabelrepel-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-graphs-networklabelrepel-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-30" title="Figure&nbsp;10.23: Solution to Question&nbsp;wrn-graphs-networklabelrepel."><img src="10-chap_files/figure-html/fig-graphs-networklabelrepel-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-graphs-networklabelrepel-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.23: Solution to <a href="#wrn-graphs-networklabelrepel" class="quarto-xref">Question&nbsp;<span>10.16</span></a>.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>It could be preferable to use a <strong>graph layout</strong> that incorporates the known geographic coordinates. Thus, we might be able to see how the virus jumped large distances across the world through traveller mobility. We introduce approximate country coordinates, which we then <strong>jitter</strong> slightly to reduce overlapping.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rworldmap"</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">=</span> <span class="fu">match</span>(country09, countriesLow<span class="sc">$</span>NAME)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>coords2009 <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">lat =</span> countriesLow<span class="sc">$</span>LAT[mat],</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">lon =</span> countriesLow<span class="sc">$</span>LON[mat],</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">country =</span> country09)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>layoutCoordinates <span class="ot">=</span> <span class="fu">cbind</span>(</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">jitter</span>(coords2009<span class="sc">$</span>lon, <span class="at">amount =</span> <span class="dv">15</span>),</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">jitter</span>(coords2009<span class="sc">$</span>lat, <span class="at">amount =</span> <span class="dv">8</span>))</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>labc <span class="ot">=</span> <span class="fu">names</span>(<span class="fu">table</span>(country09)[<span class="fu">which</span>(<span class="fu">table</span>(country09) <span class="sc">&gt;</span> <span class="dv">1</span>)])</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>matc <span class="ot">=</span> <span class="fu">match</span>(labc, countriesLow<span class="sc">$</span>NAME)</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>dfc <span class="ot">=</span> <span class="fu">data.frame</span>(</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">latc =</span> countriesLow<span class="sc">$</span>LAT[matc],</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">lonc =</span> countriesLow<span class="sc">$</span>LON[matc],</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a>  labc)</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>dfctrans <span class="ot">=</span> dfc</span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a>dfctrans[, <span class="dv">1</span>] <span class="ot">=</span> (dfc[,<span class="dv">1</span>] <span class="sc">+</span> <span class="dv">31</span>) <span class="sc">/</span> <span class="dv">93</span></span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a>dfctrans[, <span class="dv">2</span>] <span class="ot">=</span> (dfc[,<span class="dv">2</span>] <span class="sc">+</span> <span class="dv">105</span>) <span class="sc">/</span> <span class="dv">238</span></span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a>Countries <span class="ot">=</span> <span class="fu">vertex_attr</span>(gr09)<span class="sc">$</span>name </span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-21"><a href="#cb57-21" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(gr09, <span class="at">layout=</span>layoutCoordinates) <span class="sc">+</span></span>
<span id="cb57-22"><a href="#cb57-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">color=</span>Countries),<span class="at">size =</span> <span class="dv">3</span>, <span class="at">alpha=</span><span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_arc</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">strength=</span><span class="fl">0.15</span>) <span class="sc">+</span></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="at">data=</span>dfc,<span class="fu">aes</span>(<span class="at">x=</span>lonc,<span class="at">y=</span>latc,<span class="at">label=</span>labc,<span class="at">fill=</span>labc),<span class="at">colour=</span><span class="st">"white"</span>,<span class="at">alpha=</span><span class="fl">0.8</span>,<span class="at">size=</span><span class="dv">3</span>,<span class="at">show.legend=</span>F) <span class="sc">+</span></span>
<span id="cb57-25"><a href="#cb57-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>()  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-HIVmap" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-HIVmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-HIVmap-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-31" title="Figure&nbsp;10.24: A minimum spanning tree between HIV cases. The geographic locations of the cases were jittered to reduce overlapping. The DNA sequence distances between the HIV strains were used as input to an undirected minimum spanning tree algorithm."><img src="10-chap_files/figure-html/fig-HIVmap-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="640"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-HIVmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.24: A minimum spanning tree between HIV cases. The geographic locations of the cases were jittered to reduce overlapping. The DNA sequence distances between the HIV strains were used as input to an undirected minimum spanning tree algorithm.
</figcaption>
</figure>
</div>
</div>
</div>
<p>The input to the minimum spanning tree algorithm is a distance matrix or a graph with a length edge attribute. <a href="#fig-HIVmap" class="quarto-xref">Figure&nbsp;<span>10.24</span></a> is the minimum spanning tree between cases of HIV, for which strain information was made available through the HIVdb database@HIVdb. The DNA distances were computed using the Jukes-Cantor mutation model.</p>
<div id="wrn-graphs-MSTworlddirected" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.17
</div>
</div>
<div class="callout-body-container callout-body">
<p>The above analysis provided an <strong>undirected</strong> network of connections, in fact several implementations of the minimum spanning tree (ie for instance <code>mstree</code> in <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong>) provide a directed path through the points, which can provide meaningful information on the (apparent) spread of disases. Make a directed network version of the above maps.</p>
</div>
</div>
<p>MST is a very useful component of a simple nonparametric test for detecting differences between factors that are mapped onto its vertices.</p>
<section id="mst-based-testing-the-friedmanrafsky-test" class="level3 page-columns page-full" data-number="10.6.1">
<h3 data-number="10.6.1" class="anchored" data-anchor-id="mst-based-testing-the-friedmanrafsky-test"><span class="header-section-number">10.6.1</span> MST based testing: the Friedman–Rafsky test</h3>
<p>Graph-based two-sample tests<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a> were introduced by Friedman and Rafsky <span class="citation" data-cites="friedman1979multivariate">(<a href="16-chap.html#ref-friedman1979multivariate" role="doc-biblioref">Friedman and Rafsky 1979</a>)</span> as a generalization of the Wald-Wolfowitz runs test (see <a href="#fig-WWtest" class="quarto-xref">Figure&nbsp;<span>10.25</span></a>). Our previous examples show graph vertices associated with covariates such as country of origin. Here we test whether the covariate is significantly <strong>associated</strong> to the graph structure.</p>
<div class="no-row-height column-margin column-container"><div id="fn8"><p><sup>8</sup>&nbsp;Tests that explore whether two samples are drawn from the same distribution.</p></div></div><p>The <strong>Friedman-Rafsky</strong> tests for two/multiple sample segregation on a minimum spanning tree. It was conceived as a generalization of the univariate Wald-Wolfowitz runs test. If we are comparing two samples, say men and women, whose coordinates represent a measurement of interest. We color the two groups blue and red as in <a href="#fig-WWtest" class="quarto-xref">Figure&nbsp;<span>10.25</span></a>, the Wald-Wolfowitz test looks for long runs of the same color that would indicate that the two groups have different means.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="cell-output-display page-columns page-full">
<div id="fig-WWtest" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-WWtest-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-WWtest-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-32" title="Figure&nbsp;10.25: Seeing the number of runs in a one-dimensional, two-sample, nonparametric Wald-Wolfowitz test can indicate whether the two groups have the same distributions."><img src="10-chap_files/figure-html/fig-WWtest-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-WWtest-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.25: Seeing the number of runs in a one-dimensional, two-sample, nonparametric Wald-Wolfowitz test can indicate whether the two groups have the same distributions.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Instead of looking for consecutive values of one type (‘runs’), we count the number of connected nodes of the same type.</p>
<p>Once the minimum spanning tree has been constructed, the vertices are assigned `colors’ according to the different levels of a categorical variable. We call <strong>pure</strong> edges those whose two nodes have the same level of the factor variable. We use <span class="math inline">\(S_O\)</span>, the number of <strong>pure</strong> edges as our test statistic. To evaluate whether our observed value could have occurred by chance when the groups have the same distributions, we permute the vertix labels (colors) randomly and recount how many pure edges there are. This label swapping is repeated many times, creating our null distribution for <span class="math inline">\(S\)</span>.</p>
</section>
<section id="example-bacteria-sharing-between-mice" class="level3 page-columns page-full" data-number="10.6.2">
<h3 data-number="10.6.2" class="anchored" data-anchor-id="example-bacteria-sharing-between-mice"><span class="header-section-number">10.6.2</span> Example: Bacteria sharing between mice</h3>
<p>Here we illustrate the idea on a collection of samples from mice whose stool were analyzed for their microbial content. We read in a data set with many mice and many taxa, we compute the Jaccard distance and then use the <code>mst</code> function from the <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> package. We annotate the graph with the relevant covariates as shown in the code below:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>ps1  <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(DATA,<span class="st">"ps1.rds"</span>))</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>sampledata <span class="ot">=</span> <span class="fu">data.frame</span>( <span class="fu">sample_data</span>(ps1))</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">=</span> <span class="fu">as.matrix</span>(phyloseq<span class="sc">::</span><span class="fu">distance</span>(ps1, <span class="at">method=</span><span class="st">"jaccard"</span>))</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>gr <span class="ot">=</span> <span class="fu">graph_from_adjacency_matrix</span>(d1,  <span class="at">mode =</span> <span class="st">"undirected"</span>, <span class="at">weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>net <span class="ot">=</span> igraph<span class="sc">::</span><span class="fu">mst</span>(gr)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(net)<span class="sc">$</span>id <span class="ot">=</span> sampledata[<span class="fu">names</span>(<span class="fu">V</span>(net)), <span class="st">"host_subject_id"</span>]</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(net)<span class="sc">$</span>litter <span class="ot">=</span> sampledata[<span class="fu">names</span>(<span class="fu">V</span>(net)), <span class="st">"family_relationship"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We make a <code>ggraph</code> object from the resulting igraph generated minimum spanning tree and then plot it, as shown in <a href="#fig-mstplot" class="quarto-xref">Figure&nbsp;<span>10.26</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(net, <span class="at">layout=</span><span class="st">"fr"</span>)<span class="sc">+</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_arc</span>(<span class="at">color =</span> <span class="st">"darkgray"</span>) <span class="sc">+</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> id, <span class="at">shape =</span> litter)) <span class="sc">+</span> </span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-mstplot" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-mstplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-mstplot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-33" title="Figure&nbsp;10.26: The minimum spanning tree based on Jaccard dissimilarity and annotated with the mice ID and litter factors"><img src="10-chap_files/figure-html/fig-mstplot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mstplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.26: The minimum spanning tree based on Jaccard dissimilarity and annotated with the mice ID and litter factors
</figcaption>
</figure>
</div>
</div></div></div>
<p>Now we compute the null distribution and p-value for the test, this is implemented in the <strong>phyloseqGraphTest</strong> package:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phyloseqGraphTest"</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>gt <span class="ot">=</span> <span class="fu">graph_perm_test</span>(ps1, <span class="st">"host_subject_id"</span>, <span class="at">distance=</span><span class="st">"jaccard"</span>,</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">type=</span><span class="st">"mst"</span>,  <span class="at">nperm=</span><span class="dv">1000</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>gt<span class="sc">$</span>pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.000999</code></pre>
</div>
</div>
<p>We can take a look at the complete histogram of the null distribution generatedby permutation using:</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_permutations</span>(gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-mstJaccard" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-mstJaccard-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-mstJaccard-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-34" title="Figure&nbsp;10.27: The permutation histogram of the number of pure edges in the network obtained from the minimal spanning tree with Jaccard similarity."><img src="10-chap_files/figure-html/fig-mstJaccard-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mstJaccard-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.27: The permutation histogram of the number of pure edges in the network obtained from the minimal spanning tree with Jaccard similarity.
</figcaption>
</figure>
</div>
</div></div></div>
<section id="different-choices-for-the-skeleton-graph" class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="different-choices-for-the-skeleton-graph">Different choices for the skeleton graph</h4>
<p>It is not necessary to use an MST for the skeleton graph that defines the edges. Graphs made by linking nearest neighbors <span class="citation" data-cites="schilling1986multivariate">(<a href="16-chap.html#ref-schilling1986multivariate" role="doc-biblioref">Schilling 1986</a>)</span> or distance thresholding work as well.</p>
<p>The Bioconductor package <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong> has functionality for creating graphs based on thresholding a distance matrix through the function <code>make_network</code>. We create a network by creating an edge between samples whose Jaccard dissimilarity is less than a threshold, which we set below via the parameter <code>max.dist</code>. We can also use the <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong> package to add attributes to the vertices indicating which mouse the sample came from and which litter the mouse was in. We see that in the resulting network, shown in <a href="#fig-ggnetworkplotJ" class="quarto-xref">Figure&nbsp;<span>10.28</span></a>, there is grouping of the samples by both mouse and litter.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>net <span class="ot">=</span> <span class="fu">make_network</span>(ps1, <span class="at">max.dist =</span> <span class="fl">0.35</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>sampledata <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">sample_data</span>(ps1))</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(net)<span class="sc">$</span>id <span class="ot">=</span> sampledata[<span class="fu">names</span>(<span class="fu">V</span>(net)), <span class="st">"host_subject_id"</span>]</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(net)<span class="sc">$</span>litter <span class="ot">=</span> sampledata[<span class="fu">names</span>(<span class="fu">V</span>(net)), <span class="st">"family_relationship"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb64" data-cap-location="margin"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(net, <span class="at">layout=</span><span class="st">"fr"</span>) <span class="sc">+</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">color =</span> <span class="st">"darkgray"</span>) <span class="sc">+</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> id, <span class="at">shape =</span> litter)) <span class="sc">+</span> </span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">0</span>), <span class="st">"cm"</span>))<span class="sc">+</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">1.4</span>, <span class="fl">0.3</span>),<span class="at">legend.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="st">"cm"</span>))<span class="sc">+</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>         <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">ncol=</span><span class="dv">2</span>))<span class="sc">+</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_graph</span>(<span class="at">background =</span> <span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-ggnetworkplotJ" class="quarto-float quarto-figure quarto-figure-center anchored page-columns page-full" data-fig-align="center" data-cap-location="margin">
<figure class="quarto-float quarto-float-fig figure page-columns page-full">
<div aria-describedby="fig-ggnetworkplotJ-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-ggnetworkplotJ-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-35" title="Figure&nbsp;10.28: A co-occurrence network created by using a threshold on the Jaccard dissimilarity matrix. The colors represent which mouse the sample came from; the shape represents which litter the mouse was in."><img src="10-chap_files/figure-html/fig-ggnetworkplotJ-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-margin quarto-float-caption quarto-float-fig margin-caption" id="fig-ggnetworkplotJ-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.28: A co-occurrence network created by using a threshold on the Jaccard dissimilarity matrix. The colors represent which mouse the sample came from; the shape represents which litter the mouse was in.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Note that no matter which graph we build between the samples, we can approximate a null distribution by permuting the labels of the nodes of the graph. However, sometimes it will preferable to adjust the permutation distribution to account for known structure between the covariates.</p>
</section>
</section>
<section id="friedmanrafsky-test-with-nested-covariates" class="level3 page-columns page-full" data-number="10.6.3">
<h3 data-number="10.6.3" class="anchored" data-anchor-id="friedmanrafsky-test-with-nested-covariates"><span class="header-section-number">10.6.3</span> Friedman–Rafsky test with nested covariates</h3>
<p>In the test above, we took a rather naïve approach and showed there was a significant difference between individual mice (the <code>host_subject_id</code> variable). Here we perform a slightly different permutation test to find out if we control for the difference between mice; is there a litter (the <code>family_relationship</code> variable) effect? The setup of the test is similar, it is simply how the permutations are generated which differs. We maintain the nested structure of the two factors using the <code>grouping</code> argument. We permute the <code>family_relationship</code> labels but keep the <code>host_subject_id</code> structure intact.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>gt <span class="ot">=</span> <span class="fu">graph_perm_test</span>(ps1, <span class="st">"family_relationship"</span>,</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">grouping =</span> <span class="st">"host_subject_id"</span>,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">type =</span> <span class="st">"mst"</span>, <span class="at">nperm=</span> <span class="dv">1000</span>)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>gt<span class="sc">$</span>pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.002</code></pre>
</div>
</div>
<p>This test has a small p-value, and we reject the null hypothesis that the two samples come from the same distribution. From the plot of the minimum spanning tree in <a href="#fig-mstJaccard" class="quarto-xref">Figure&nbsp;<span>10.27</span></a>, we see by eye that the samples group by litter more than we would expect by chance.</p>
<div class="cell page-columns page-full" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_permutations</span>(gt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-mstpermplotNest" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-mstpermplotNest-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-mstpermplotNest-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-36" title="Figure&nbsp;10.29: The permutation histogram obtained from the minimal spanning tree with Jaccard similarity."><img src="10-chap_files/figure-html/fig-mstpermplotNest-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="320"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mstpermplotNest-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.29: The permutation histogram obtained from the minimal spanning tree with Jaccard similarity.
</figcaption>
</figure>
</div>
</div></div></div>
<div id="wrn-graphs-nn-graph" class="callout callout-style-default callout-warning no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question&nbsp;10.18
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <span class="math inline">\(k\)</span>-nearest neighbors graph is obtained by putting an edge between two samples whenever one of them is in the set of <span class="math inline">\(k\)</span>-nearest neighbors of the other. Redo the test, defining the graph using nearest neighbors defined with the Jaccard distance. What would you conclude?</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-36-contents" aria-controls="callout-36" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-36" class="callout-36-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>gtnn1 <span class="ot">=</span> <span class="fu">graph_perm_test</span>(ps1, <span class="st">"family_relationship"</span>,</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">grouping =</span> <span class="st">"host_subject_id"</span>,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">distance =</span> <span class="st">"jaccard"</span>, <span class="at">type =</span> <span class="st">"knn"</span>, <span class="at">knn =</span> <span class="dv">1</span>)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>gtnn1<span class="sc">$</span>pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.004</code></pre>
</div>
</div>
<p><a href="#fig-knn-1-plot" class="quarto-xref">Figure&nbsp;<span>10.30</span></a> shows that pairs of samples having edges between them in this nearest neighbor graph are much more likely to be from the same litter.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_test_network</span>(gtnn1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-knn-1-plot" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-knn-1-plot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-knn-1-plot-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-37" title="Figure&nbsp;10.30: The graph obtained from a nearest-neighbor graph with Jaccard similarity."><img src="10-chap_files/figure-html/fig-knn-1-plot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-knn-1-plot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.30: The graph obtained from a nearest-neighbor graph with Jaccard similarity.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p><strong>Note: The dual graph</strong><br>
In the examples above we sought to show relationships between samples through their shared taxa. It can also be of interest to ask the question about taxa: do some of the taxa co-occur more often than one would expect? This approach can help study microbial `communities’ as they assemble in the microbiome. The methods we developed above all apply to this use-case, all one really does is transpose the data. It is always preferable with sparse data such as the microbiome to use Jaccard and not build correlation networks that can be appropriate in other settings.</p>
</section>
</section>
<section id="summary-of-this-chapter" class="level2" data-number="10.7">
<h2 data-number="10.7" class="anchored" data-anchor-id="summary-of-this-chapter"><span class="header-section-number">10.7</span> Summary of this chapter</h2>
<section id="annotated-graphs" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="annotated-graphs">Annotated graphs</h4>
<p>In this chapter we have learnt how to store and plot data that have more structure than simple arrays: graphs have edges and nodes that can also be associated to extra annotations that can be displayed usefully.</p>
</section>
<section id="important-examples-of-graphs-and-useful-r-packages" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="important-examples-of-graphs-and-useful-r-packages">Important examples of graphs and useful R packages</h4>
<p>We started by specific examples such as Markov chain graphs, phylogenetic trees and minimum spanning trees. We saw how to use the <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> packages to visualize graphs and show as much information as possible by using specific graph layout algorithms.</p>
</section>
<section id="combining-graphs-with-statistical-data" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="combining-graphs-with-statistical-data">Combining graphs with statistical data</h4>
<p>We then approached the problem of incorporating a known `skeleton’ graph into differential expression analyses. This enables use to pinpoint perturbation hotspots in a network. We saw how evolutionary models defined along rooted binary trees serve as the basis for phylogenetic tree estimation and how we can incorporate these trees as supplementary information in a differential abundance analysis using the R packages <strong><a href="https://cran.r-project.org/web/packages/structSSI/">structSSI</a></strong> and <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong>.</p>
</section>
<section id="linking-co-occurrence-to-other-variables" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="linking-co-occurrence-to-other-variables">Linking co-occurrence to other variables</h4>
<p>Graph and network tools also enable the creation of networks from co-occurrence data and can be used to visualize and test the effect of factor covariates. We saw the Friedman-Rafsky test which provides an easy way of testing dependencies of a variable with the edge structure of a skeleton graph.</p>
</section>
<section id="context-and-intepretation-aids" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="context-and-intepretation-aids">Context and intepretation aids</h4>
<p>This chapter illustrated ways of incorporating interactions of players in a network and we saw how useful it was to combine this with statistical scores. This often provides biological insight into analyses of complex biological systems.</p>
</section>
<section id="previous-knowledge-or-outcome" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="previous-knowledge-or-outcome">Previous knowledge or outcome</h4>
<p>We saw that graphs can be both useful to encode our previous knowledge, metabolic network information, gene ontologies and phylogenetic trees of known bacteria are all available in standard databases. It is beneficial in a study to incorprate all known information and doing so by combining these skeleton networks with observed data enhances our understanding of experimental results in the context of what is already known.</p>
<p>On the other hand, the graph can be the outcome that we want to predict and we saw how to build graphs from data (phylogenetic trees, co-occurrence networks and minimum spanning trees).</p>
</section>
</section>
<section id="further-reading" class="level2" data-number="10.8">
<h2 data-number="10.8" class="anchored" data-anchor-id="further-reading"><span class="header-section-number">10.8</span> Further reading</h2>
<p>For complete developments and many important consequences of the evolutionary models used in phylogenetic trees, see the books by <span class="citation" data-cites="WHLi LiandGraur">Li (<a href="16-chap.html#ref-WHLi" role="doc-biblioref">1997</a>; <a href="16-chap.html#ref-LiandGraur" role="doc-biblioref">Li and Graur 1991</a>)</span>. The book by <span class="citation" data-cites="fels:book">Felsenstein (<a href="16-chap.html#ref-fels:book" role="doc-biblioref">2004</a>)</span> is the classic text on estimating phylogenetic trees.</p>
<p>The book written by the author of the <strong><a href="https://cran.r-project.org/web/packages/ape/">ape</a></strong> packages, <span class="citation" data-cites="paradis:2011">Paradis (<a href="16-chap.html#ref-paradis:2011" role="doc-biblioref">2011</a>)</span> contains many use-cases and details about manipulation of trees in R. A review of bootstrapping for phylogenetic trees can be found in <span class="citation" data-cites="Holmes-2003">Holmes (<a href="16-chap.html#ref-Holmes-2003" role="doc-biblioref">2003a</a>)</span>.</p>
<p>We can use a tree as well as abundances in a contingency table data through an extension of PCoA-MDS called DPCoA (Double principal coordinate analysis). For microbiome data, the phylogenetic tree provides distances between taxa; these distances serve as the basis for the first PCoA. A second PCoA enables the projection of the weighted sample points. This has proved very effective in microbial ecology applications, see <span class="citation" data-cites="Purdom2010">Purdom (<a href="16-chap.html#ref-Purdom2010" role="doc-biblioref">2010</a>)</span> or <span class="citation" data-cites="Fukuyama:2012">Fukuyama et al. (<a href="16-chap.html#ref-Fukuyama:2012" role="doc-biblioref">2012</a>)</span> for details.</p>
<p>Graphs can be used to predict vertex covariates. There is a large field of applied statistics and machine learning that considers the edges in the graph as a response variable for which one can make predictions based on covariates or partial knowledge of the graph; these include <strong>ERGM</strong>’s (Exponential Random Graph Models, <span class="citation" data-cites="robins:2007">Robins et al. (<a href="16-chap.html#ref-robins:2007" role="doc-biblioref">2007</a>)</span>) and kernel methods for graphs <span class="citation" data-cites="VertKernel">(<a href="16-chap.html#ref-VertKernel" role="doc-biblioref">Schölkopf, Tsuda, and Vert 2004</a>)</span>.</p>
<p>For theoretical properties of the Friedman-Rafsky test and more examples see <span class="citation" data-cites="Bhattacharya:2015">Bhattacharya (<a href="16-chap.html#ref-Bhattacharya:2015" role="doc-biblioref">2015</a>)</span>.</p>
<p>A full list packages that deal with graphs and networks is available at: <a href="http://www.bioconductor.org/packages/release/BiocViews.html#___GraphAndNetwork" class="uri">http://www.bioconductor.org/packages/release/BiocViews.html#___GraphAndNetwork</a>.</p>
</section>
<section id="exercises" class="level2" data-number="10.9">
<h2 data-number="10.9" class="anchored" data-anchor-id="exercises"><span class="header-section-number">10.9</span> Exercises</h2>
<div id="imp-graphs-creategraphs" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.1
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a function that plots a graph starting from an adjacency matrix. Show how it works on an example.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-38-contents" aria-controls="callout-38" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-38" class="callout-38-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<!-- TODO: missing solutions here -->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>ggplotadjacency <span class="ot">=</span> <span class="cf">function</span>(a) {</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">=</span> <span class="fu">nrow</span>(a)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">=</span> <span class="fu">ncol</span>(a)</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  fromto  <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(a)</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">nrow</span>(fromto), n<span class="sc">*</span>p))</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>  fromto<span class="sc">$</span>value <span class="ot">=</span> <span class="fu">as.factor</span>(fromto<span class="sc">$</span>value)</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  cols <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"darkblue"</span>)</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> fromto, <span class="fu">aes</span>(<span class="at">x =</span> Var1, <span class="at">y =</span> Var2, <span class="at">fill =</span> value)) <span class="sc">+</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tile</span>(<span class="at">colour =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_fixed</span>(<span class="at">ratio =</span> <span class="dv">1</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, n <span class="sc">+</span> <span class="fl">0.5</span>), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, p <span class="sc">+</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cols) <span class="sc">+</span></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">""</span> , <span class="at">breaks =</span> <span class="dv">1</span><span class="sc">:</span>p, <span class="at">labels =</span> <span class="fu">paste</span>(<span class="dv">1</span><span class="sc">:</span>p)) <span class="sc">+</span></span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_reverse</span>(  <span class="at">name =</span> <span class="st">""</span> , <span class="at">breaks =</span> n<span class="sc">:</span><span class="dv">1</span>, <span class="at">labels =</span> <span class="fu">paste</span>(n<span class="sc">:</span><span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>),</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.key =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"white"</span>),</span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"white"</span>),</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"white"</span>),</span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>() </span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div id="imp-graphs-GOgraph" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.2
</div>
</div>
<div class="callout-body-container callout-body">
<p>The relationships between gene functions is organized hierarchically into a graph called the Gene Ontology (GO) graph. The biological processes are organized at finer and finer scale. Take one of the databases providing the GO information for the organisms you are interested in. Choose a gene list and build the GO graph for that list.<br>
Hint: Some examples can be found in the packages , , .</p>
</div>
</div>
<div id="imp-graphs-PNASMarkov" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.3
</div>
</div>
<div class="callout-body-container callout-body">
<p><em>Markov chain graph of transitions between states of the vaginal microbiota</em>: In <span class="citation" data-cites="DigiulioCallahan:2015">DiGiulio et al. (<a href="16-chap.html#ref-DigiulioCallahan:2015" role="doc-biblioref">2015</a>)</span> the authors use an <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> plot to represent the transitions rates between community state types CSTs using the <strong><a href="https://cran.r-project.org/web/packages/markovchain/">markovchain</a></strong> package. Load the data and the transition rates and state names into an object of the special class <code>markovchain</code> and tailor the layout carefully to include the percentage of preterm birth as a covariate for the vertices (make the vertex size proportional to this variable). Include the size of transitions between states as the width of the arrows.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-41-contents" aria-controls="callout-41" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-41" class="callout-41-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Note: the below code is not live, a version of it used to run for the authors at one point, here it is given as a starting point for the reader to finish, it has opportunities for modernization and improvement.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-PSB-MC-s" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-PSB-MC-s-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/PSB_MC_s.png" class="lightbox" data-gallery="quarto-lightbox-gallery-38" title="Figure&nbsp;10.31: This figure was originally created for the study done in @DigiulioCallahan:2015, where the nodes designate states of the vaginal microbiome and arrows represent transitions between states with different probabilities."><img src="imgs/PSB_MC_s.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-PSB-MC-s-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.31: This figure was originally created for the study done in <span class="citation" data-cites="DigiulioCallahan:2015">DiGiulio et al. (<a href="16-chap.html#ref-DigiulioCallahan:2015" role="doc-biblioref">2015</a>)</span>, where the nodes designate states of the vaginal microbiome and arrows represent transitions between states with different probabilities.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"markovchain"</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Make Markov chain object</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>mcPreg  <span class="ot">=</span>  <span class="fu">new</span>(<span class="st">"markovchain"</span>, <span class="at">states =</span> CSTs,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">transitionMatrix =</span> trans, <span class="at">name=</span><span class="st">"PregCST"</span>)</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>mcPreg</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up igraph of the markov chain</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>netMC  <span class="ot">=</span>  markovchain<span class="sc">:::</span><span class="fu">.getNet</span>(mcPreg, <span class="at">round =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now define a number of plotting parameters, and assign node colors based on the association of that CST and preterm outcome.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>wts  <span class="ot">=</span>  <span class="fu">E</span>(netMC)<span class="sc">$</span>weight<span class="sc">/</span><span class="dv">100</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>edgel  <span class="ot">=</span>  <span class="fu">get.edgelist</span>(netMC)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>elcat  <span class="ot">=</span>  <span class="fu">paste</span>(edgel[,<span class="dv">1</span>], edgel[,<span class="dv">2</span>])</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>elrev  <span class="ot">=</span>  <span class="fu">paste</span>(edgel[,<span class="dv">2</span>], edgel[,<span class="dv">1</span>])</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>edge.curved  <span class="ot">=</span>  <span class="fu">sapply</span>(elcat, <span class="cf">function</span>(x) x <span class="sc">%in%</span> elrev)</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>samples_def  <span class="ot">=</span>  <span class="fu">data.frame</span>(<span class="fu">sample_data</span>(ps))</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>samples_def  <span class="ot">=</span>  samples_def[samples<span class="sc">$</span>Preterm <span class="sc">|</span> samples<span class="sc">$</span>Term,] <span class="co"># Only those definitely assigned, i.e. not marginal</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>premat  <span class="ot">=</span>  <span class="fu">table</span>(samples_def<span class="sc">$</span>CST, samples_def<span class="sc">$</span>Preterm)</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(premat)  <span class="ot">=</span>  markovchain<span class="sc">::</span><span class="fu">states</span>(mcPreg)</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(premat)  <span class="ot">=</span>  <span class="fu">c</span>(<span class="st">"Term"</span>, <span class="st">"Preterm"</span>)</span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>premat</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>premat  <span class="ot">=</span>  premat<span class="sc">/</span><span class="fu">rowSums</span>(premat)</span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>vert.CSTclrs  <span class="ot">=</span>  CSTColors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>default.par  <span class="ot">=</span>  <span class="fu">par</span>(<span class="at">no.readonly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Define color scale</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plotting function for markov chain</span></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>plotMC  <span class="ot">=</span>  <span class="cf">function</span>(object, ...) {</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>    netMC  <span class="ot">=</span>  markovchain<span class="sc">:::</span><span class="fu">.getNet</span>(object, <span class="at">round =</span> <span class="cn">TRUE</span>)</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot.igraph</span>(<span class="at">x =</span> netMC, ...)</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Color bar for the markov chain visualization, gradient in strength of preterm association</span></span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>color.bar  <span class="ot">=</span>  <span class="cf">function</span>(lut, min, <span class="at">max=</span><span class="sc">-</span>min, <span class="at">nticks=</span><span class="dv">11</span>, <span class="at">ticks=</span><span class="fu">seq</span>(min, max, <span class="at">len=</span>nticks), <span class="at">title=</span><span class="cn">NULL</span>) {</span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>    scale <span class="ot">=</span> (<span class="fu">length</span>(lut)<span class="sc">-</span><span class="dv">1</span>)<span class="sc">/</span>(max<span class="sc">-</span>min)</span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a>    cur.par <span class="ot">=</span> <span class="fu">par</span>(<span class="at">no.readonly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">4</span>) <span class="sc">+</span> <span class="fl">0.1</span>, <span class="at">oma =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>) <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">ps =</span> <span class="dv">10</span>, <span class="at">cex =</span> <span class="fl">0.8</span>)</span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">par</span>(<span class="at">tcl=</span><span class="sc">-</span><span class="fl">0.2</span>, <span class="at">cex.axis=</span><span class="fl">0.8</span>, <span class="at">cex.lab =</span> <span class="fl">0.8</span>)</span>
<span id="cb74-15"><a href="#cb74-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(<span class="fu">c</span>(min,max), <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>), <span class="at">type=</span><span class="st">'n'</span>, <span class="at">bty=</span><span class="st">'n'</span>, <span class="at">xaxt=</span><span class="st">'n'</span>, <span class="at">xlab=</span><span class="st">", yaxt='n', ylab="</span>, <span class="at">main=</span>title)</span>
<span id="cb74-16"><a href="#cb74-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">axis</span>(<span class="dv">1</span>, <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))</span>
<span id="cb74-17"><a href="#cb74-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">length</span>(lut)<span class="sc">-</span><span class="dv">1</span>)) {</span>
<span id="cb74-18"><a href="#cb74-18" aria-hidden="true" tabindex="-1"></a>      x <span class="ot">=</span> (i<span class="dv">-1</span>)<span class="sc">/</span>scale <span class="sc">+</span> min</span>
<span id="cb74-19"><a href="#cb74-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rect</span>(x,<span class="dv">0</span>,x<span class="sc">+</span><span class="dv">1</span><span class="sc">/</span>scale,<span class="dv">10</span>, <span class="at">col=</span>lut[i], <span class="at">border=</span><span class="cn">NA</span>)</span>
<span id="cb74-20"><a href="#cb74-20" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb74-21"><a href="#cb74-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb74-22"><a href="#cb74-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-23"><a href="#cb74-23" aria-hidden="true" tabindex="-1"></a>pal  <span class="ot">=</span>  <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"grey50"</span>, <span class="st">"maroon"</span>, <span class="st">"magenta2"</span>))(<span class="dv">101</span>)</span>
<span id="cb74-24"><a href="#cb74-24" aria-hidden="true" tabindex="-1"></a>vert.clrs  <span class="ot">=</span>  <span class="fu">sapply</span>(<span class="fu">states</span>(mcPreg), <span class="cf">function</span>(x) pal[<span class="dv">1</span><span class="sc">+</span><span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span>premat[x,<span class="st">"Preterm"</span>])])</span>
<span id="cb74-25"><a href="#cb74-25" aria-hidden="true" tabindex="-1"></a>vert.sz  <span class="ot">=</span>  <span class="dv">4</span> <span class="sc">+</span> <span class="dv">2</span><span class="sc">*</span><span class="fu">sapply</span>(<span class="fu">states</span>(mcPreg),</span>
<span id="cb74-26"><a href="#cb74-26" aria-hidden="true" tabindex="-1"></a>              <span class="cf">function</span>(x) <span class="fu">nrow</span>(<span class="fu">unique</span>(<span class="fu">sample_data</span>(ps)[<span class="fu">sample_data</span>(ps)<span class="sc">$</span>CST<span class="sc">==</span>x,<span class="st">"SubjectID"</span>])))</span>
<span id="cb74-27"><a href="#cb74-27" aria-hidden="true" tabindex="-1"></a>vert.sz  <span class="ot">=</span>  vert.sz <span class="sc">*</span> <span class="fl">0.85</span></span>
<span id="cb74-28"><a href="#cb74-28" aria-hidden="true" tabindex="-1"></a>vert.font.clrs  <span class="ot">=</span>  <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"white"</span>, <span class="st">"white"</span>, <span class="st">"white"</span>, <span class="st">"white"</span>)</span>
<span id="cb74-29"><a href="#cb74-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-30"><a href="#cb74-30" aria-hidden="true" tabindex="-1"></a><span class="co"># E(netMC) to see edge list, have to define loop angles individually by the # in edge list, not vertex</span></span>
<span id="cb74-31"><a href="#cb74-31" aria-hidden="true" tabindex="-1"></a>edge.loop.angle <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">3.14</span>, <span class="fl">3.14</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">3.14</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>)<span class="sc">-</span><span class="fl">0.45</span></span>
<span id="cb74-32"><a href="#cb74-32" aria-hidden="true" tabindex="-1"></a>layout  <span class="ot">=</span>  <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.6</span>,<span class="fl">0.95</span>, <span class="fl">0.43</span>,<span class="dv">1</span>, <span class="fl">0.3</span>,<span class="fl">0.66</span>, <span class="fl">0.55</span>,<span class="fl">0.3</span>, <span class="fl">0.75</span>,<span class="fl">0.65</span>), <span class="at">nrow =</span> <span class="dv">5</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb74-33"><a href="#cb74-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-34"><a href="#cb74-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Color by association with preterm birth</span></span>
<span id="cb74-35"><a href="#cb74-35" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>), <span class="dv">2</span>, <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>), <span class="at">heights=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">10</span>))</span>
<span id="cb74-36"><a href="#cb74-36" aria-hidden="true" tabindex="-1"></a><span class="fu">color.bar</span>(pal, <span class="at">min=</span><span class="dv">0</span>, <span class="at">max=</span><span class="dv">1</span>, <span class="at">nticks=</span><span class="dv">6</span>, <span class="at">title=</span><span class="st">"Fraction preterm"</span>)</span>
<span id="cb74-37"><a href="#cb74-37" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>)<span class="sc">+</span><span class="fl">0.1</span>)</span>
<span id="cb74-38"><a href="#cb74-38" aria-hidden="true" tabindex="-1"></a>edge.arrow.size<span class="ot">=</span><span class="fl">0.8</span></span>
<span id="cb74-39"><a href="#cb74-39" aria-hidden="true" tabindex="-1"></a>edge.arrow.width<span class="ot">=</span><span class="fl">1.4</span></span>
<span id="cb74-40"><a href="#cb74-40" aria-hidden="true" tabindex="-1"></a>edge.width <span class="ot">=</span> (<span class="dv">15</span><span class="sc">*</span>wts <span class="sc">+</span> <span class="fl">0.1</span>)<span class="sc">*</span><span class="fl">0.6</span></span>
<span id="cb74-41"><a href="#cb74-41" aria-hidden="true" tabindex="-1"></a>edge.labels  <span class="ot">=</span>  <span class="fu">as.character</span>(<span class="fu">E</span>(netMC)<span class="sc">$</span>weight<span class="sc">/</span><span class="dv">100</span>)</span>
<span id="cb74-42"><a href="#cb74-42" aria-hidden="true" tabindex="-1"></a>edge.labels[edge.labels<span class="sc">&lt;</span><span class="fl">0.4</span>]  <span class="ot">=</span>  <span class="cn">NA</span>  <span class="co"># labels only for self-loops</span></span>
<span id="cb74-43"><a href="#cb74-43" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMC</span>(mcPreg, <span class="at">edge.arrow.size=</span>edge.arrow.size, <span class="at">edge.arrow.width =</span> edge.arrow.width,</span>
<span id="cb74-44"><a href="#cb74-44" aria-hidden="true" tabindex="-1"></a>       <span class="at">edge.width=</span>edge.width, <span class="at">edge.curved=</span>edge.curved,</span>
<span id="cb74-45"><a href="#cb74-45" aria-hidden="true" tabindex="-1"></a>       <span class="at">vertex.color=</span>vert.clrs, <span class="at">vertex.size=</span>(vert.sz),</span>
<span id="cb74-46"><a href="#cb74-46" aria-hidden="true" tabindex="-1"></a>       <span class="at">vertex.label.font =</span> <span class="dv">2</span>, <span class="at">vertex.label.cex =</span> <span class="dv">1</span>,</span>
<span id="cb74-47"><a href="#cb74-47" aria-hidden="true" tabindex="-1"></a>       <span class="at">vertex.label.color =</span> vert.font.clrs, <span class="at">vertex.frame.color =</span> <span class="cn">NA</span>,</span>
<span id="cb74-48"><a href="#cb74-48" aria-hidden="true" tabindex="-1"></a>       <span class="at">layout=</span>layout, <span class="at">edge.loop.angle =</span> edge.loop.angle)</span>
<span id="cb74-49"><a href="#cb74-49" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(default.par)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div id="imp-graphs-ppinteraction" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.4
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Protein interaction networks</strong>: Read the Wikipedia article about the <a href="https://en.wikipedia.org/wiki/STRING">STRING</a> database (http://www.string-db.org).<br>
The protein Cyclin B1 is encoded by the CCNB1 gene. You can read about it on wikipedia here: <a href="https://en.wikipedia.org/wiki/Cyclin_B1" class="uri">https://en.wikipedia.org/wiki/Cyclin_B1</a>.<br>
Use STRING to generate a text file (call it <em>ccnb1datsmall.txt</em>) of edges around the CCNB1 gene. Choose nodes that are connected by evidence of co-expression with a confidence higher than 0.9. Collect no more than 50 interactions and additional nodes that are two steps away from CCNB1 in the graph.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-43-contents" aria-controls="callout-43" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-43" class="callout-43-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>Go to <a href="http://www.string-db.org">http://www.string-db.org</a>.</p></li>
<li><p>Enter CCNB1 as the protein name and Homo sapiens as the organism. Click “Continue!”</p></li>
<li><p>Select the option with protein CCNB1 (the top one).</p></li>
<li><p>Scroll down to “info and Parameters …” 4a. For Active Prediction Methods – unselect everything except “Co-Expression” 4b. For required confidence – select “highest confidence (0.900) 4c. For interactors shown – select”no more than 50 interactors” 4d. For additional (white) nodes – select “100” (these are nodes two steps away from CCNB1) 4e. Click “Update Parameters”. You should get something that looks like the image below.</p></li>
<li><p>Click “save” under the picture (showing a diskette). This will open up a new window so you can choose which format to save the data.</p></li>
<li><p>Scroll down to the “Text Summary (TXT - simple tab delimited flatfile)” file and save that document as <em>ccnb1datsmall.txt</em>.</p></li>
</ol>
</div>
</div>
</div>
<div id="imp-graphs-graphinput" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.5
</div>
</div>
<div class="callout-body-container callout-body">
<p>Read the txt file <em>ccnb1datsmall.txt</em> into R and make a plot of the graph using one of the graph visualization methods covered in this chapter.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-45-contents" aria-controls="callout-45" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-45" class="callout-45-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div id="fig-ccnb1img" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ccnb1img-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="imgs/ccnb1img.png" class="lightbox" data-gallery="quarto-lightbox-gallery-39" title="Figure&nbsp;10.32: This network was created with the STRING website by setting a 2 step neighborhood around the CCNB1 gene for co-expression levels \geq 0.900."><img src="imgs/ccnb1img.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:75.0%"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ccnb1img-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.32: This network was created with the STRING website by setting a 2 step neighborhood around the CCNB1 gene for co-expression levels <span class="math inline">\(\geq\)</span> 0.900.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">=</span> <span class="fu">read.table</span>(<span class="fu">file.path</span>(DATA,<span class="st">"ccnb1datsmall.txt"</span>), <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">comment.char =</span> <span class="st">""</span>, <span class="at">stringsAsFactors =</span> <span class="cn">TRUE</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>v <span class="ot">=</span> <span class="fu">levels</span>(<span class="fu">unlist</span>(dat[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]))        <span class="co"># vertex names</span></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">length</span>(v)                        <span class="co"># number of vertices</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>e <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">match</span>(<span class="fu">as.character</span>(<span class="fu">unlist</span>(dat[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])), v),<span class="at">ncol=</span><span class="dv">2</span>) <span class="co"># edge list</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>w <span class="ot">=</span> dat<span class="sc">$</span>coexpression                 <span class="co"># edge weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>M is our co-expression network adjacency matrix. Since the STRING data only says if proteins i and j are co-expressed and doesn’t distinguish between (i,j) and (j,i) we want to make M symmetric (undirected) by considering the weight on (i,j) is the same as from (j,i). A is our co-expression graph adjacency matrix and we make <span class="math inline">\(A_{ij} = 1\)</span> if they are coexpressed.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>M <span class="ot">=</span> <span class="fu">matrix</span>(<span class="dv">0</span>, n, n)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>M[e] <span class="ot">=</span> w</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>M <span class="ot">=</span> M <span class="sc">+</span> <span class="fu">t</span>(M)</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dimnames</span>(M) <span class="ot">=</span> <span class="fu">list</span>(v, v)</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>A <span class="ot">=</span> <span class="dv">1</span><span class="sc">*</span>(M <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use default plotting parameters and generate the graph using the package <strong><a href="https://cran.r-project.org/web/packages/igraph/">igraph</a></strong> starting with <code>e</code>, the vector of edges (an alternative is to use the adjacency matrix A).</p>
<!-- TODO: there's actually no set.seed() here? -->
<p><em>Note:</em> We use a seed to make the graph always look the same. Graph layout often contains an optimization with a random component that makes the picture look different, although the graph itself is the same.</p>
<!--```{r}
#| child: devil.qmd
#| fig-cap: "We use a seed to make the graph always look the same. Graph layout often contains an optimization with a random component that makes the picture look different, although the graph itself is the same."
```-->
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(igraph)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>net <span class="ot">=</span> <span class="fu">network</span>(e, <span class="at">directed=</span><span class="cn">FALSE</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">rep</span>(<span class="dv">0</span>,<span class="dv">4</span>))</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(net, <span class="at">label=</span>v)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You could make a graph with <strong><a href="https://cran.r-project.org/web/packages/ggraph/">ggraph</a></strong>.</p>
</div>
</div>
</div>
<div id="imp-graphs-heatmap" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.6
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make a heatmap showing the adjacency matrix of the graph created in <a href="#imp-graphs-graphinput" class="quarto-xref">Exercise&nbsp;<span>10.5</span></a>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-47-contents" aria-controls="callout-47" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-47" class="callout-47-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We use defaults in making a heatmap except for changing the colors, you can experiment and add additional parameters.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>breaks  <span class="ot">=</span>  <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">seq</span>(<span class="fl">0.9</span>, <span class="dv">1</span>, <span class="at">length=</span><span class="dv">11</span>))</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>cols  <span class="ot">=</span>  <span class="fu">grey</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fu">seq</span>(<span class="fl">0.5</span>,<span class="dv">1</span>,<span class="at">length=</span><span class="dv">10</span>)))</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>ccnb1ind  <span class="ot">=</span>  <span class="fu">which</span>(v <span class="sc">==</span> <span class="st">"CCNB1"</span>)</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>vcols  <span class="ot">=</span>  <span class="fu">rep</span>(<span class="st">"white"</span>,n)</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>vcols[ccnb1ind]  <span class="ot">=</span>  <span class="st">"blue"</span></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>vcols[<span class="fu">which</span>(M[,ccnb1ind]<span class="sc">&gt;</span><span class="dv">0</span> <span class="sc">|</span> M[ccnb1ind,])]  <span class="ot">=</span>  <span class="st">"red"</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>))</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(M, <span class="at">symm =</span> <span class="cn">TRUE</span>, <span class="at">ColSideColors =</span> vcols, <span class="at">RowSideColors =</span> vcols,</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> cols, <span class="at">breaks =</span> breaks,  <span class="at">frame =</span> <span class="cn">TRUE</span>)</span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="fu">c</span>(<span class="st">"Neighbors(CCNB1)"</span>, <span class="st">"CCNB1"</span>),</span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"red"</span>,<span class="st">"blue"</span>),</span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">inset =</span> <span class="dv">0</span>, <span class="at">xpd =</span> <span class="cn">TRUE</span>,  <span class="at">border =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-heatmapCCNB1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-heatmapCCNB1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-heatmapCCNB1-1.png" class="lightbox" data-gallery="quarto-lightbox-gallery-40" title="Figure&nbsp;10.33: This represents the adjacency of the CCNB1 network – 2 step neighborhood with co-expression levels \geq 0.900, generated from R (darker is closer to 1, we ignore values < 0.9)."><img src="10-chap_files/figure-html/fig-heatmapCCNB1-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="480"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heatmapCCNB1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.33: This represents the adjacency of the CCNB1 network – 2 step neighborhood with co-expression levels <span class="math inline">\(\geq\)</span> 0.900, generated from R (darker is closer to 1, we ignore values &lt; 0.9).
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="imp-graphs-visualizeinteractions" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.7
</div>
</div>
<div class="callout-body-container callout-body">
<p>The visualization shows the strongest interactions in the two step neighborhood of CCNB1. Both the plotted graph and the heatmap image show the same data: there seems to be a cluster of proteins which are all similar to CCNB1 and there is also another cluster in the other proteins. Many of the proteins in the CCNB1 cluster are coexpressed at the same time as each other.<br>
Why might this be the case?<br>
Conversely, proteins which are coexpressed with a protein that is coexpressed with CCNB1 (two steps away) do not tend to be coexpressed with each other.<br>
Is it easier for you to see this in one of the figures (the plot or the heatmap) than the other?</p>
</div>
</div>
<div id="imp-graphs-comparephylogeny" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.8
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compare the use of <strong><a href="https://cran.r-project.org/web/packages/ape/">ape</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/phangorn/">phangorn</a></strong> in the analysis of HIV GAG data. Compute the Jukes Cantor distances between the sequences using both packages and compare them to the Hamming distances.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ape"</span>)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phangorn"</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>GAG <span class="ot">=</span> <span class="fu">read.dna</span>(<span class="fu">file.path</span>(DATA, <span class="st">"DNA_GAG_20.txt"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="imp-graphs-CurtisThreshold" class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise&nbsp;10.9
</div>
</div>
<div class="callout-body-container callout-body">
<p>Perform the Friedman–Rafksy type test with a “two-nearest” neighbor-graph using the Bray-Curtis dissimilarity.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-51-contents" aria-controls="callout-51" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-51" class="callout-51-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>gt <span class="ot">=</span> <span class="fu">graph_perm_test</span>(ps1, <span class="st">"family_relationship"</span>, <span class="at">distance =</span> <span class="st">"bray"</span>, </span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">grouping =</span> <span class="st">"host_subject_id"</span>, <span class="at">type =</span> <span class="st">"knn"</span>, <span class="at">knn =</span> <span class="dv">2</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>gt<span class="sc">$</span>pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.004</code></pre>
</div>
</div>
<div class="cell" data-layout-nrow="1" data-layout-align="center" data-warning.known="All aesthetics have length">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_test_network</span>(gt)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>permdf <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">perm=</span>gt<span class="sc">$</span>perm)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>obs <span class="ot">=</span> gt<span class="sc">$</span>observed</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>ymax <span class="ot">=</span> <span class="fu">max</span>(gt<span class="sc">$</span>perm)</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(permdf, <span class="fu">aes</span>(<span class="at">x =</span> perm)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>) <span class="sc">+</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> obs, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> obs, <span class="at">yend =</span> ymax<span class="sc">/</span><span class="dv">10</span>), <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> obs, <span class="at">y =</span> ymax<span class="sc">/</span><span class="dv">10</span>), <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"Number of pure edges"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-knn-2-plot" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-knn-2-plot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-knn-2-plot" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-knn-2-plot-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-knn-2-plot-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-knn-2-plot-1.png" class="lightbox" data-gallery="fig-knn-2-plot" title="Figure&nbsp;10.34&nbsp;(a): \text{}"><img src="10-chap_files/figure-html/fig-knn-2-plot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" data-ref-parent="fig-knn-2-plot" width="360"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-knn-2-plot-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) <span class="math inline">\(\text{}\)</span>
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-knn-2-plot" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-knn-2-plot-2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-subfloat-fig figure">
<div aria-describedby="fig-knn-2-plot-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="10-chap_files/figure-html/fig-knn-2-plot-2.png" class="lightbox" data-gallery="fig-knn-2-plot" title="Figure&nbsp;10.34&nbsp;(b): \text{}"><img src="10-chap_files/figure-html/fig-knn-2-plot-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" data-ref-parent="fig-knn-2-plot" width="360"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-knn-2-plot-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) <span class="math inline">\(\text{}\)</span>
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-knn-2-plot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.34: The graph (a) and permutation histogram (b) obtained from a two nearest-neighbor graph with Jaccard similarity.
</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-BioNet" class="csl-entry" role="listitem">
Beisser, Daniela, Gunnar W Klau, Thomas Dandekar, Tobias Müller, and Marcus T Dittrich. 2010. <span>“<span>BioNet</span>: An <span>R</span>-Package for the Functional Analysis of Biological Networks.”</span> <em>Bioinformatics</em> 26 (8): 1129–30.
</div>
<div id="ref-benjamini2014selective" class="csl-entry" role="listitem">
Benjamini, Yoav, and Marina Bogomolov. 2014. <span>“Selective Inference on Multiple Families of Hypotheses.”</span> <em>Journal of the Royal Statistical Society: Series B</em> 76 (1): 297–318.
</div>
<div id="ref-benjamini2003hierarchical" class="csl-entry" role="listitem">
Benjamini, Yoav, and Daniel Yekutieli. 2003. <span>“Hierarchical <span>FDR</span> Testing of Trees of Hypotheses.”</span> Technical report, Department of Statistics; Operations Research, Tel Aviv University.
</div>
<div id="ref-Bhattacharya:2015" class="csl-entry" role="listitem">
Bhattacharya, Bhaswar B. 2015. <span>“Power of Graph-Based Two-Sample Tests.”</span> <em>arXiv Preprint arXiv:1508.07530</em>.
</div>
<div id="ref-beast" class="csl-entry" role="listitem">
Bouckaert, Remco, Joseph Heled, Denise Kühnert, Tim Vaughan, Chieh-Hsi Wu, Dong Xie, Marc A Suchard, Andrew Rambaut, and Alexei J Drummond. 2014. <span>“<span>BEAST 2</span>: A Software Platform for Bayesian Evolutionary Analysis.”</span> <em>PLoS Computational Biology</em> 10 (4): e1003537.
</div>
<div id="ref-Callahan:2017" class="csl-entry" role="listitem">
Callahan, Benjamin J, Paul J McMurdie, and Susan P Holmes. 2017. <span>“Exact Sequence Variants Should Replace Operational Taxonomic Units in Marker Gene Data Analysis.”</span> <em>ISME Journal</em>, 1–5.
</div>
<div id="ref-distory-paper" class="csl-entry" role="listitem">
Chakerian, John, and Susan Holmes. 2012. <span>“Computational Tools for Evaluating Phylogenetic and Hierarchical Clustering Trees.”</span> <em>Journal of Computational and Graphical Statistics</em> 21 (3): 581–99.
</div>
<div id="ref-cole2009rdp" class="csl-entry" role="listitem">
Cole, J. R., Q. Wang, E. Cardenas, J. Fish, B. Chai, R. J. Farris, A. S. Kulam-Syed-Mohideen, et al. 2009. <span>“The Ribosomal Database Project: Improved Alignments and New Tools for rRNA Analysis.”</span> <em>Nucleic Acids Research</em> 37 (Supplement 1): D141–45.
</div>
<div id="ref-DigiulioCallahan:2015" class="csl-entry" role="listitem">
DiGiulio, Daniel B., Benjamin J. Callahan, Paul J. McMurdie, Elizabeth K. Costello, Deirdre J. Lyelle, Anna Robaczewska, Christine L. Sun, et al. 2015. <span>“Temporal and Spatial Variation of the Human Microbiota During Pregnancy.”</span> <em>PNAS</em>.
</div>
<div id="ref-fels:book" class="csl-entry" role="listitem">
Felsenstein, Joseph. 2004. <em>Inferring Phylogenies</em>. Boston: Sinauer.
</div>
<div id="ref-friedman1979multivariate" class="csl-entry" role="listitem">
Friedman, Jerome H, and Lawrence C Rafsky. 1979. <span>“Multivariate Generalizations of the Wald-Wolfowitz and Smirnov Two-Sample Tests.”</span> <em>The Annals of Statistics</em>, 697–717.
</div>
<div id="ref-Fukuyama:2012" class="csl-entry" role="listitem">
Fukuyama, Julia, Paul J McMurdie, Les Dethlefsen, David A Relman, and Susan Holmes. 2012. <span>“Comparisons of Distance Methods for Combining Covariates and Abundances in Microbiome Studies.”</span> In <em>Pac Symp Biocomput</em>. World Scientific.
</div>
<div id="ref-holmesima" class="csl-entry" role="listitem">
Holmes, Susan. 1999. <span>“Phylogenetic Trees: An Overview.”</span> In <em>Statistics and Genetics</em>, 81–118. IMA 112. New York: Springer.
</div>
<div id="ref-Holmes-2003" class="csl-entry" role="listitem">
———. 2003a. <span>“Bootstrapping Phylogenetic Trees: Theory and Methods.”</span> <em>Statistical Science</em> 18 (2): 241–55.
</div>
<div id="ref-holmes2003b" class="csl-entry" role="listitem">
———. 2003b. <span>“<span class="nocase">Statistics for phylogenetic trees</span>.”</span> <em>Theoretical Population Biology</em> 63 (1): 17–32.
</div>
<div id="ref-Ideker:2002" class="csl-entry" role="listitem">
Ideker, Trey, Owen Ozier, Benno Schwikowski, and Andrew F Siegel. 2002. <span>“Discovering Regulatory and Signalling Circuits in Molecular Interaction Networks.”</span> <em>Bioinformatics</em> 18 Suppl 1 (January): S233–40. <a href="http://bioinformatics.oxfordjournals.org/cgi/reprint/18/suppl\_1/S233">http://bioinformatics.oxfordjournals.org/cgi/reprint/18/suppl\_1/S233</a>.
</div>
<div id="ref-WHLi" class="csl-entry" role="listitem">
Li, Wen-Hsiung. 1997. <em>Molecular Evolution.</em> Sinauer Associates Incorporated.
</div>
<div id="ref-LiandGraur" class="csl-entry" role="listitem">
Li, Wen-Hsiung, and Dan Graur. 1991. <em>Fundamentals of Molecular Evolution</em>. Vol. 48. Sinauer Associates Sunderland, MA.
</div>
<div id="ref-MSigDB" class="csl-entry" role="listitem">
Liberzon, Arthur, Aravind Subramanian, Reid Pinchback, Helga Thorvaldsdóttir, Pablo Tamayo, and Jill P Mesirov. 2011. <span>“Molecular Signatures Database (<span>MSigDB</span>) 3.0.”</span> <em>Bioinformatics</em> 27 (12): 1739–40.
</div>
<div id="ref-LoveDESeq2" class="csl-entry" role="listitem">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Gnome Biology</em> 15 (12): 1–21.
</div>
<div id="ref-mcmurdie2014" class="csl-entry" role="listitem">
McMurdie, Paul J, and Susan Holmes. 2014. <span>“Waste Not, Want Not: Why Rarefying Microbiome Data Is Inadmissible.”</span> <em>PLoS Computational Biology</em> 10 (4): e1003531.
</div>
<div id="ref-mcmurdie2015" class="csl-entry" role="listitem">
———. 2015. <span>“Shiny-Phyloseq: Web Application for Interactive Microbiome Analysis with Provenance Tracking.”</span> <em>Bioinformatics</em> 31 (2): 282–83.
</div>
<div id="ref-Mossel:2003" class="csl-entry" role="listitem">
Mossel, Elchanan. 2003. <span>“On the Impossibility of Reconstructing Ancestral Data and Phylogenies.”</span> <em>Journal of Computational Biology</em> 10 (5): 669–76.
</div>
<div id="ref-GXNA" class="csl-entry" role="listitem">
Nacu, Serban, Rebecca Critchley-Thorne, Peter Lee, and Susan Holmes. 2007. <span>“Gene Expression Network Analysis and Applications to Immunology.”</span> <em>Bioinformatics</em> 23 (7, 7): 850–58. <a href="https://doi.org/10.1093/bioinformatics/btm019">https://doi.org/10.1093/bioinformatics/btm019</a>.
</div>
<div id="ref-Nolan:2013" class="csl-entry" role="listitem">
Nolan, Daniel J, Michael Ginsberg, Edo Israely, Brisa Palikuqi, Michael G Poulos, Daylon James, Bi-Sen Ding, et al. 2013. <span>“Molecular Signatures of Tissue-Specific Microvascular Endothelial Cell Heterogeneity in Organ Maintenance and Regeneration.”</span> <em>Developmental Cell</em> 26 (2): 204–19.
</div>
<div id="ref-paradis:2011" class="csl-entry" role="listitem">
Paradis, Emmanuel. 2011. <em>Analysis of Phylogenetics and Evolution with r</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-Pounds2003" class="csl-entry" role="listitem">
Pounds, Stan, and Stephan W Morris. 2003. <span>“Estimating the Occurrence of False Positives and False Negatives in Microarray Studies by Approximating and Partitioning the Empirical Distribution of p-Values.”</span> <em>Bioinformatics</em> 19 (10): 1236–42.
</div>
<div id="ref-Purdom2010" class="csl-entry" role="listitem">
Purdom, Elizabeth. 2010. <span>“Analysis of a Data Matrix and a Graph: Metagenomic Data and the Phylogenetic Tree.”</span> <em>Annals of Applied Statistics</em>, July.
</div>
<div id="ref-HIVdb" class="csl-entry" role="listitem">
Rhee, Soo-Yon, Matthew J Gonzales, Rami Kantor, Bradley J Betts, Jaideep Ravela, and Robert W Shafer. 2003. <span>“Human Immunodeficiency Virus Reverse Transcriptase and Protease Sequence Database.”</span> <em>Nucleic Acids Research</em> 31 (1): 298–303.
</div>
<div id="ref-robins:2007" class="csl-entry" role="listitem">
Robins, Garry, Tom Snijders, Peng Wang, Mark Handcock, and Philippa Pattison. 2007. <span>“Recent Developments in Exponential Random Graph (p*) Models for Social Networks.”</span> <em>Social Networks</em> 29 (2): 192–215.
</div>
<div id="ref-mrbayes3" class="csl-entry" role="listitem">
Ronquist, Fredrik, Maxim Teslenko, Paul van der Mark, Daniel L Ayres, Aaron Darling, Sebastian Höhna, Bret Larget, Liang Liu, Marc A Suchard, and John P Huelsenbeck. 2012. <span>“MrBayes 3.2: Efficient Bayesian Phylogenetic Inference and Model Choice Across a Large Model Space.”</span> <em>Systematic Biology</em> 61 (3): 539–42.
</div>
<div id="ref-sankaran2014structssi" class="csl-entry" role="listitem">
Sankaran, Kris, and Susan Holmes. 2014. <span>“<span class="nocase">structSSI: Simultaneous and Selective Inference for Grouped or Hierarchically Structured Data</span>.”</span> <em>Journal of Statistical Software</em> 59 (1): 1–21.
</div>
<div id="ref-schilling1986multivariate" class="csl-entry" role="listitem">
Schilling, Mark F. 1986. <span>“Multivariate Two-Sample Tests Based on Nearest Neighbors.”</span> <em>Journal of the American Statistical Association</em> 81 (395): 799–806.
</div>
<div id="ref-VertKernel" class="csl-entry" role="listitem">
Schölkopf, Bernhard, Koji Tsuda, and Jean-Philippe Vert. 2004. <em>Kernel Methods in Computational Biology</em>. MIT press.
</div>
<div id="ref-wang2007naive" class="csl-entry" role="listitem">
Wang, Q., G. M. Garrity, J. M. Tiedje, and J. R. Cole. 2007. <span>“Naive Bayesian Classifier for Rapid Assignment of rRNA Sequences into the New Bacterial Taxonomy.”</span> <em>Applied and Environmental Microbiology</em> 73 (16): 5261.
</div>
<div id="ref-Wertheim:2009" class="csl-entry" role="listitem">
Wertheim, Joel O, and Michael Worobey. 2009. <span>“Dating the Age of the <span>SIV</span> Lineages That Gave Rise to <span>HIV</span>-1 and <span>HIV</span>-2.”</span> <em>PLoS Computational Biology</em> 5 (5): e1000377.
</div>
<div id="ref-wright2015decipher" class="csl-entry" role="listitem">
Wright, Erik S. 2015. <span>“<span>DECIPHER</span>: Harnessing Local Sequence Context to Improve Protein Multiple Sequence Alignment.”</span> <em><span>BMC Bioinformatics</span></em> 16 (1): 1.
</div>
<div id="ref-YuGXNA" class="csl-entry" role="listitem">
Yu, Hongxiang, Diana L Simons, Ilana Segall, Valeria Carcamo-Cavazos, Erich J Schwartz, Ning Yan, Neta S Zuckerman, et al. 2012. <span>“<span>PRC2</span>/<span>EED-EZH2</span> Complex Is up-Regulated in Breast Cancer Lymph Node Metastasis Compared to Primary Tumor and Correlates with Tumor Proliferation in Situ.”</span> <em>PloS One</em> 7 (12): e51239.
</div>
</div>
</section>


<p>Page built at 01:33 on 2025-09-01 using R version 4.5.1 (2025-06-13)</p></main> <!-- /main --></p><p class="build-date">
Support for maintaining the online version of this book is provided by <a href="http://www.denbi.de">de.NBI</a>
  <img src="imgs/denbi.png" style="vertical-align: text-top;height: 16px"><br>
For website support please contact msmith@embl.de
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
  const anchorJS_eq = new window.AnchorJS();
  anchorJS_eq.options = {
    placement: 'left',
    class: 'eq-link',
    icon: icon
  };
  anchorJS_eq.add('.anchored-eq');
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/www\.huber\.embl\.de\/msmb");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>
