<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>References – Modern Statistics for Modern Biology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./imgs/msmb-favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-af5ec82acda093b5ee751184164e9432.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-d63985d606c7f07261ad561a6cfde940.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="msmb.css">
<meta property="og:title" content="References – Modern Statistics for Modern Biology">
<meta property="og:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta property="og:image" content="https://www.huber.embl.de/msmb/imgs/bibliothek.jpeg">
<meta property="og:site_name" content="Modern Statistics for Modern Biology">
<meta name="twitter:title" content="References – Modern Statistics for Modern Biology">
<meta name="twitter:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta name="twitter:image" content="https://www.huber.embl.de/msmb/imgs/bibliothek.jpeg">
<meta name="twitter:card" content="summary">
<script type="text/javascript">
  var _paq = _paq || [];
  
  // Call disableCookies before calling trackPageView 
  _paq.push(['disableCookies']);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u='//tr-denbi.embl.de/heimdall/';
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '26']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>

</head>

<body class="floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Modern Statistics for Modern Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-chapters" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Chapters</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-chapters">    
        <li>
    <a class="dropdown-item" href="./index.html">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./00-chap.html">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./01-chap.html">
 <span class="dropdown-text">Generative Models for Discrete Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-chap.html">
 <span class="dropdown-text">Statistical Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03-chap.html">
 <span class="dropdown-text">Data visualization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04-chap.html">
 <span class="dropdown-text">Mixture Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./05-chap.html">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06-chap.html">
 <span class="dropdown-text">Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./07-chap.html">
 <span class="dropdown-text">Multivariate Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./08-chap.html">
 <span class="dropdown-text">High-Throughput Count Data &amp; Generalized Linear Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./09-chap.html">
 <span class="dropdown-text">Multivariate methods for heterogeneous data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./10-chap.html">
 <span class="dropdown-text">Networks and Trees</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./11-chap.html">
 <span class="dropdown-text">Image data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./12-chap.html">
 <span class="dropdown-text">Supervised Learning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./13-chap.html">
 <span class="dropdown-text">Design of High Throughput Experiments and their Analyses</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./14-chap.html">
 <span class="dropdown-text">Statistical Concordance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./15-chap.html">
 <span class="dropdown-text">Acknowledgements</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./16-chap.html">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">References</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-1000Genomes" class="csl-entry" role="listitem">
1000 Genomes Project Consortium. 2012. <span>“An Integrated Map of
Genetic Variation from 1,092 Human Genomes.”</span> <em>Nature</em> 491
(7422): 56–65.
</div>
<div id="ref-Abbott:1884" class="csl-entry" role="listitem">
Abbott, Edwin A. 1884. <em>Flatland: A Romance of Many Dimensions</em>.
OUP Oxford.
</div>
<div id="ref-Agresti:2007" class="csl-entry" role="listitem">
Agresti, Alan. 2007. <em>An Introduction to Categorical Data
Analysis</em>. John Wiley.
</div>
<div id="ref-Altman:PoS:2017" class="csl-entry" role="listitem">
Altman, Naomi, and Martin Krzywinski. 2017. <span>“Points of
Significance: Interpreting p Values.”</span> <em>Nature
<span>M</span>ethods</em> 14 (3): 213–14. <a href="https://doi.org/10.1038/nmeth.4210">https://doi.org/10.1038/nmeth.4210</a>.
</div>
<div id="ref-ambroise2002selection" class="csl-entry" role="listitem">
Ambroise, Christophe, and Geoffrey J. McLachlan. 2002.
<span>“<span>S</span>election Bias in Gene Extraction on the Basis of
Microarray Gene-Expression Data.”</span> <em>PNAS</em> 99 (10): 6562–66.
</div>
<div id="ref-Anders:2010:GB" class="csl-entry" role="listitem">
Anders, Simon, and Wolfgang Huber. 2010.
<span>“<span>D</span>ifferential Expression Analysis for Sequence Count
Data.”</span> <em>Genome Biology</em> 11: R106. <a href="http://genomebiology.com/2010/11/10/R106">http://genomebiology.com/2010/11/10/R106</a>.
</div>
<div id="ref-Reyes:GnomeResearch:2012" class="csl-entry" role="listitem">
Anders, Simon, Alejandro Reyes, and Wolfgang Huber. 2012. <span>“<span class="nocase"><span>D</span>etecting differential usage of exons from
<span>RNA</span>-Seq data</span>.”</span> <em>Genome Research</em> 22
(10): 2008–17.
</div>
<div id="ref-Anscombe1948" class="csl-entry" role="listitem">
Anscombe, Francis J. 1948. <span>“The Transformation of
<span>P</span>oisson, Binomial and Negative-Binomial Data.”</span>
<em>Biometrika</em>, 246–54.
</div>
<div id="ref-Aure2017" class="csl-entry" role="listitem">
Aure, Miriam Ragle, Valeria Vitelli, Sandra Jernström, Surendra Kumar,
Marit Krohn, Eldri U Due, Tonje Husby Haukaas, et al. 2017.
<span>“Integrative Clustering Reveals a Novel Split in the Luminal
<span>A</span> Subtype of Breast Cancer with Impact on Outcome.”</span>
<em>Breast Cancer Research</em> 19 (1): 44.
</div>
<div id="ref-Bacher:GB:2016" class="csl-entry" role="listitem">
Bacher, Rhonda, and Christina Kendziorski. 2016. <span>“Design and
Computational Analysis of Single-Cell <span>RNA</span>-Sequencing
Experiments.”</span> <em>Genome Biology</em> 17 (1): 1.
</div>
<div id="ref-Baddeley2000" class="csl-entry" role="listitem">
Baddeley, Adrain, Jesper Moller, and Rasmus Waagepetersen. 2000.
<span>“Non- and Semiparametric Estimation of Interaction in
Inhomogeneous Point Patterns.”</span> <em>Statistica Neerlandica</em>
54: 329–50.
</div>
<div id="ref-Baddeley1998" class="csl-entry" role="listitem">
Baddeley, Adrian J. 1998. <span>“Spatial Sampling and Censoring.”</span>
In <em>Stochastic Geometry: Likelihood and Computation</em>, edited by
O. E. Barndorff-Nielsen, W. S. Kendall, and M. N. M. van Lieshout,
37–78. Chapman; Hall.
</div>
<div id="ref-BioNet" class="csl-entry" role="listitem">
Beisser, Daniela, Gunnar W Klau, Thomas Dandekar, Tobias Müller, and
Marcus T Dittrich. 2010. <span>“<span>BioNet</span>: An
<span>R</span>-Package for the Functional Analysis of Biological
Networks.”</span> <em>Bioinformatics</em> 26 (8): 1129–30.
</div>
<div id="ref-Belkin2003" class="csl-entry" role="listitem">
Belkin, Mikhail, and Partha Niyogi. 2003. <span>“Laplacian Eigenmaps for
Dimensionality Reduction and Data Representation.”</span> <em>Neural
Computation</em> 15 (6): 1373–96.
</div>
<div id="ref-Bellman:1961" class="csl-entry" role="listitem">
Bellman, Richard Ernest. 1961. <em>Adaptive Control Processes: A Guided
Tour</em>. Princeton University Press.
</div>
<div id="ref-BendallCell" class="csl-entry" role="listitem">
Bendall, Sean C, Garry P Nolan, Mario Roederer, and Pratip K
Chattopadhyay. 2012. <span>“A Deep Profiler’s Guide to
Cytometry.”</span> <em>Trends in Immunology</em> 33 (7): 323–32.
</div>
<div id="ref-Bengio2004" class="csl-entry" role="listitem">
Bengio, Yoshua, Jean-François Paiement, Pascal Vincent, Olivier
Delalleau, Nicolas Le Roux, and Marie Ouimet. 2004. <span>“Out-of-Sample
Extensions for <span>LLE</span>, Isomap, <span>MDS</span>, Eigenmaps,
and Spectral Clustering.”</span> <em>Advances in Neural Information
Processing Systems</em> 16: 177–84.
</div>
<div id="ref-benjamini2014selective" class="csl-entry" role="listitem">
Benjamini, Yoav, and Marina Bogomolov. 2014. <span>“Selective Inference
on Multiple Families of Hypotheses.”</span> <em>Journal of the Royal
Statistical Society: Series B</em> 76 (1): 297–318.
</div>
<div id="ref-BH:1995" class="csl-entry" role="listitem">
Benjamini, Yoav, and Yosef Hochberg. 1995. <span>“Controlling the False
Discovery Rate: A Practical and Powerful Approach to Multiple
Testing.”</span> <em>Journal of the Royal Statistical Society B</em> 57:
289–300.
</div>
<div id="ref-benjamini2003hierarchical" class="csl-entry" role="listitem">
Benjamini, Yoav, and Daniel Yekutieli. 2003. <span>“Hierarchical
<span>FDR</span> Testing of Trees of Hypotheses.”</span> Technical
report, Department of Statistics; Operations Research, Tel Aviv
University.
</div>
<div id="ref-ilastik:NatMeth:2019" class="csl-entry" role="listitem">
Berg, Stuart, Dominik Kutra, Thorben Kroeger, Christoph N Straehle,
Bernhard X Kausler, Carsten Haubold, Martin Schiegg, et al. 2019.
<span>“Ilastik: Interactive Machine Learning for (Bio)image
Analysis.”</span> <em>Nature Methods</em> 16 (12): 1226–32.
</div>
<div id="ref-Bhattacharya:2015" class="csl-entry" role="listitem">
Bhattacharya, Bhaswar B. 2015. <span>“Power of Graph-Based Two-Sample
Tests.”</span> <em>arXiv Preprint arXiv:1508.07530</em>.
</div>
<div id="ref-Bishop:PRML" class="csl-entry" role="listitem">
Bishop, Christopher M. 2006. <em>Pattern Recognition and Machine
Learning</em>. Springer.
</div>
<div id="ref-BolandMurphy" class="csl-entry" role="listitem">
Boland, Michael V., and Robert F. Murphy. 2001. <span>“<span class="nocase">A neural network classifier capable of recognizing the
patterns of all major subcellular structures in fluorescence microscope
images of <span>HeLa</span> cells</span>.”</span>
<em>Bioinformatics</em> 17 (12): 1213–23.
</div>
<div id="ref-beast" class="csl-entry" role="listitem">
Bouckaert, Remco, Joseph Heled, Denise Kühnert, Tim Vaughan, Chieh-Hsi
Wu, Dong Xie, Marc A Suchard, Andrew Rambaut, and Alexei J Drummond.
2014. <span>“<span>BEAST 2</span>: A Software Platform for Bayesian
Evolutionary Analysis.”</span> <em>PLoS Computational Biology</em> 10
(4): e1003537.
</div>
<div id="ref-Box:1987" class="csl-entry" role="listitem">
Box, George EP, Norman Richard Draper, et al. 1987. <em>Empirical
Model-Building and Response Surfaces</em>. Vol. 424. Wiley New York.
</div>
<div id="ref-box1978statistics" class="csl-entry" role="listitem">
Box, George EP, William G Hunter, and J Stuart Hunter. 1978.
<em>Statistics for Experimenters: An Introduction to Design, Data
Analysis, and Model Building</em>. John Wiley &amp; Sons.
</div>
<div id="ref-terBraak:1985" class="csl-entry" role="listitem">
Braak, Cajo ter. 1985. <span>“Correspondence Analysis of Incidence and
Abundance Data: Properties in Terms of a Unimodal Respose.”</span>
<em>Biometrics</em> 41 (January).
</div>
<div id="ref-Brodie:2006" class="csl-entry" role="listitem">
Brodie, Eoin L, Todd Z DeSantis, Dominique C Joyner, Seung M Baek, Joern
T Larsen, Gary L Andersen, Terry C Hazen, et al. 2006.
<span>“Application of a High-Density Oligonucleotide Microarray Approach
to Study Bacterial Population Dynamics During Uranium Reduction and
Reoxidation.”</span> <em>Applied and Environmental Microbiology</em> 72
(9): 6288–98.
</div>
<div id="ref-BronsteinSemendjajew" class="csl-entry" role="listitem">
Bronštein, Il’ja N., and Konstantin A Semendjajew. 1979. <em>Taschenbuch
Der Mathematik</em>. B.G. Teubner Verlagsgesellschaft, Leipzig; Verlag
Nauka, Moscow.
</div>
<div id="ref-Brooks2010" class="csl-entry" role="listitem">
Brooks, Angela N, Li Yang, Michael O Duff, Kasper D Hansen, Jung W Park,
Sandrine Dudoit, Steven E Brenner, and Brenton R Graveley. 2011.
<span>“Conservation of an <span>RNA</span> Regulatory Map Between
<span>D</span>rosophila and Mammals.”</span> <em>Genome Research</em>,
193–202. <a href="https://doi.org/10.1101/gr.108662.110">https://doi.org/10.1101/gr.108662.110</a>.
</div>
<div id="ref-bulmer2003" class="csl-entry" role="listitem">
Bulmer, Michael George. 2003. <em>Francis Galton: Pioneer of Heredity
and Biometry</em>. JHU Press.
</div>
<div id="ref-Callahan2016Bioc" class="csl-entry" role="listitem">
Callahan, Ben J, Kris Sankaran, Julia A Fukuyama, Paul J McMurdie, and
Susan P Holmes. 2016. <span>“Bioconductor Workflow for Microbiome Data
Analysis: From Raw Reads to Community Analyses.”</span>
<em>F1000Research</em> 5.
</div>
<div id="ref-Callahan:2017" class="csl-entry" role="listitem">
Callahan, Benjamin J, Paul J McMurdie, and Susan P Holmes. 2017.
<span>“Exact Sequence Variants Should Replace Operational Taxonomic
Units in Marker Gene Data Analysis.”</span> <em>ISME Journal</em>, 1–5.
</div>
<div id="ref-dada2" class="csl-entry" role="listitem">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han,
Amy J Johnson, and Susan P Holmes. 2016. <span>“<span>DADA2</span>:
<span>H</span>igh Resolution Sample Inference from Amplicon
Data.”</span> <em>Nature Methods</em>, 1–4.
</div>
<div id="ref-Cannings1968" class="csl-entry" role="listitem">
Cannings, Chris, and Anthony WF Edwards. 1968. <span>“Natural Selection
and the <span class="nocase">de Finetti</span> Diagram.”</span>
<em>Annals of Human Genetics</em> 31 (4): 421–28.
</div>
<div id="ref-caporaso2010qiime" class="csl-entry" role="listitem">
Caporaso, J. G., J. Kuczynski, J. Stombaugh, K. Bittinger, F. D.
Bushman, E. K. Costello, N. Fierer, et al. 2010. <span>“QIIME Allows
Analysis of High-Throughput Community Sequencing Data.”</span>
<em>Nature Methods</em> 7 (5): 335–36.
</div>
<div id="ref-carpenter2006cellprofiler" class="csl-entry" role="listitem">
Carpenter, Anne E, Thouis R Jones, Michael R Lamprecht, Colin Clarke, In
Han Kang, Ola Friman, David A Guertin, Joo Han Chang, Robert A
Lindquist, and Jason Moffat. 2006. <span>“<span>CellProfiler</span>:
Image Analysis Software for Identifying and Quantifying Cell
Phenotypes.”</span> <em>Genome Biology</em> 7: R100.
</div>
<div id="ref-Carr1987hexbin" class="csl-entry" role="listitem">
Carr, Daniel B, Richard J Littlefield, WL Nicholson, and JS Littlefield.
1987. <span>“Scatterplot Matrix Techniques for Large
<span>N</span>.”</span> <em>Journal of the American Statistical
Association</em> 82 (398): 424–36.
</div>
<div id="ref-distory-paper" class="csl-entry" role="listitem">
Chakerian, John, and Susan Holmes. 2012. <span>“Computational Tools for
Evaluating Phylogenetic and Hierarchical Clustering Trees.”</span>
<em>Journal of Computational and Graphical Statistics</em> 21 (3):
581–99.
</div>
<div id="ref-chaumont2012icy" class="csl-entry" role="listitem">
Chaumont, Fabrice de, Stéphane Dallongeville, Nicolas Chenouard, Nicolas
Hervé, Sorin Pop, Thomas Provoost, Vannary Meas-Yedid, et al. 2012.
<span>“<span class="nocase"><span>I</span>cy: an open bioimage
informatics platform for extended reproducible research</span>.”</span>
<em>Nature Methods</em> 9: 690–96.
</div>
<div id="ref-Chen2011" class="csl-entry" role="listitem">
Chen, Min, Yang Xie, and Michael Story. 2011. <span>“An
Exponential-Gamma Convolution Model for Background Correction of
<span>I</span>llumina <span>BeadArray</span> Data.”</span>
<em>Communications in Statistics-Theory and Methods</em> 40 (17):
3055–69.
</div>
<div id="ref-ade4" class="csl-entry" role="listitem">
Chessel, Daniel, Anne Dufour, and Jean Thioulouse. 2004. <span>“The
<span class="nocase">ade4</span> Package - i: One-Table Methods.”</span>
<em>R News</em> 4 (1): 5–10. <a href="http://CRAN.R-project.org/doc/Rnews/">http://CRAN.R-project.org/doc/Rnews/</a>.
</div>
<div id="ref-MeckeSG2013" class="csl-entry" role="listitem">
Chiu, Sung Nok, Dietrich Stoyan, Wilfrid S. Kendall, and Joseph Mecke.
2013. <em>Stochastic Geometry and Its Applications</em>. Springer.
</div>
<div id="ref-clemmensen2012sparse" class="csl-entry" role="listitem">
Clemmensen, Line, Trevor Hastie, Daniela Witten, and Bjarne Ersbøll.
2011. <span>“Sparse Discriminant Analysis.”</span>
<em>Technometrics</em> 53: 406–13.
</div>
<div id="ref-Tukey:1988" class="csl-entry" role="listitem">
Cleveland, William S. 1988. <em>The Collected Works of John w. Tukey:
Graphics 1965-1985</em>. Vol. 5. CRC Press.
</div>
<div id="ref-Cleveland:1988:Banking" class="csl-entry" role="listitem">
Cleveland, William S., Marylyn E. McGill, and Robert McGill. 1988.
<span>“The Shape Parameter of a Two-Variable Graph.”</span> <em>Journal
of the American Statistical Association</em> 83: 289–300.
</div>
<div id="ref-cole2009rdp" class="csl-entry" role="listitem">
Cole, J. R., Q. Wang, E. Cardenas, J. Fish, B. Chai, R. J. Farris, A. S.
Kulam-Syed-Mohideen, et al. 2009. <span>“The Ribosomal Database Project:
Improved Alignments and New Tools for rRNA Analysis.”</span> <em>Nucleic
Acids Research</em> 37 (Supplement 1): D141–45.
</div>
<div id="ref-Cook1977Detection" class="csl-entry" role="listitem">
Cook, R. Dennis. 1977. <span>“Detection of Influential Observation in
Linear Regression.”</span> <em>Technometrics</em>.
</div>
<div id="ref-CressieSSD1991" class="csl-entry" role="listitem">
Cressie, Noel A. 1991. <em>Statistics for Spatial Data</em>. John Wiley;
Sons.
</div>
<div id="ref-diaconis1980finite" class="csl-entry" role="listitem">
Diaconis, Persi, and David Freedman. 1980. <span>“Finite Exchangeable
Sequences.”</span> <em>The Annals of Probability</em>, 745–64.
</div>
<div id="ref-Goel:2007aa" class="csl-entry" role="listitem">
Diaconis, Persi, Sharad Goel, and Susan Holmes. 2008. <span>“Horseshoes
in Multidimensional Scaling and Kernel Methods.”</span> <em>Annals of
Applied Statistics</em> 2: 777. <a href="https://doi.org/DOI:10.1214/08-AOAS165">https://doi.org/DOI:10.1214/08-AOAS165</a>.
</div>
<div id="ref-Diaconis1994" class="csl-entry" role="listitem">
Diaconis, Persi, and Susan Holmes. 1994. <span>“Gray Codes for
Randomization Procedures.”</span> <em>Statistics and Computing</em> 4
(4): 287–302.
</div>
<div id="ref-Diaconis-Montgomery-Holmes-2007" class="csl-entry" role="listitem">
Diaconis, Persi, Susan Holmes, and Richard Montgomery. 2007.
<span>“Dynamical Bias in the Coin Toss.”</span> <em>SIAM Review</em> 49
(2): 211–35.
</div>
<div id="ref-Diday1989" class="csl-entry" role="listitem">
Diday, Edwin, and M Paula Brito. 1989. <span>“Symbolic Cluster
Analysis.”</span> In <em>Conceptual and Numerical Analysis of Data</em>,
45–84. Springer.
</div>
<div id="ref-DiggleSPP" class="csl-entry" role="listitem">
Diggle, Peter J. 2013. <em>Statistical Analysis of Spatial and
Spatio-Temporal Point Patterns</em>. Chapman; Hall/CRCs.
</div>
<div id="ref-DigiulioCallahan:2015" class="csl-entry" role="listitem">
DiGiulio, Daniel B., Benjamin J. Callahan, Paul J. McMurdie, Elizabeth
K. Costello, Deirdre J. Lyelle, Anna Robaczewska, Christine L. Sun, et
al. 2015. <span>“Temporal and Spatial Variation of the Human Microbiota
During Pregnancy.”</span> <em>PNAS</em>.
</div>
<div id="ref-Dundar2014" class="csl-entry" role="listitem">
Dundar, Murat, Ferit Akova, Halid Z. Yerebakan, and Bartek Rajwa. 2014.
<span>“A Non-Parametric <span>B</span>ayesian Model for Joint Cell
Clustering and Cluster Matching: Identification of Anomalous Sample
Phenotypes with Random Effects.”</span> <em>BMC Bioinformatics</em> 15
(1): 1–15. <a href="https://doi.org/10.1186/1471-2105-15-314">https://doi.org/10.1186/1471-2105-15-314</a>.
</div>
<div id="ref-DEKM" class="csl-entry" role="listitem">
Durbin, Richard, Sean Eddy, Anders Krogh, and Graeme Mitchison. 1998.
<em>Biological Sequence Analysis</em>. Cambridge University Press.
</div>
<div id="ref-Efron2010" class="csl-entry" role="listitem">
Efron, Bradley. 2010. <em>Large-Scale Inference: Empirical
<span>B</span>ayes Methods for Estimation, Testing, and Prediction</em>.
Cambridge University Press.
</div>
<div id="ref-Efron:1994" class="csl-entry" role="listitem">
Efron, Bradley, and Robert J Tibshirani. 1994. <em>An Introduction to
the Bootstrap</em>. CRC press.
</div>
<div id="ref-efront" class="csl-entry" role="listitem">
Efron, B., and R. Tibshirani. 1993. <em>An Introduction to the
Bootstrap</em>. Chapman &amp; Hall/CRC.
</div>
<div id="ref-Ekman:1954" class="csl-entry" role="listitem">
Ekman, Gosta. 1954. <span>“Dimensions of Color Vision.”</span> <em>The
Journal of Psychology</em> 38 (2): 467–74.
</div>
<div id="ref-Chargaff" class="csl-entry" role="listitem">
Elson, D, and E Chargaff. 1952. <span>“On the Desoxyribonucleic Acid
Content of Sea Urchin Gametes.”</span> <em>Experientia</em> 8 (4):
143–45.
</div>
<div id="ref-fels:book" class="csl-entry" role="listitem">
Felsenstein, Joseph. 2004. <em>Inferring Phylogenies</em>. Boston:
Sinauer.
</div>
<div id="ref-definetti26" class="csl-entry" role="listitem">
Finetti, Bruno de. 1926. <span>“Considerazioni Matematiche
Sull’ereditarieta Mendeliana.”</span> <em>Metron</em> 6: 3–41.
</div>
<div id="ref-fisher1935design" class="csl-entry" role="listitem">
Fisher, Ronald Aylmer. 1935. <em>The Design of Experiments</em>. Oliver
&amp; Boyd.
</div>
<div id="ref-Flury" class="csl-entry" role="listitem">
Flury, Bernard. 1997. <em>A First Course in Multivariate
Statistics</em>. Springer.
</div>
<div id="ref-freedman1991" class="csl-entry" role="listitem">
Freedman, David A. 1991. <span>“Statistical Models and Shoe
Leather.”</span> <em>Sociological Methodology</em> 21 (2): 291–313.
</div>
<div id="ref-Freedman:1997" class="csl-entry" role="listitem">
Freedman, David, Robert Pisani, and Roger Purves. 1997.
<em>Statistics</em>. New York, NY: WW Norton.
</div>
<div id="ref-friedmanbiasvariance01" class="csl-entry" role="listitem">
Friedman, Jerome H. 1997. <span>“On Bias, Variance, 0/1—Loss, and the
Curse-of-Dimensionality.”</span> <em>Data Mining and Knowledge
Discovery</em> 1: 55–77.
</div>
<div id="ref-friedman1979multivariate" class="csl-entry" role="listitem">
Friedman, Jerome H, and Lawrence C Rafsky. 1979. <span>“Multivariate
Generalizations of the Wald-Wolfowitz and Smirnov Two-Sample
Tests.”</span> <em>The Annals of Statistics</em>, 697–717.
</div>
<div id="ref-Fukuyama:2012" class="csl-entry" role="listitem">
Fukuyama, Julia, Paul J McMurdie, Les Dethlefsen, David A Relman, and
Susan Holmes. 2012. <span>“Comparisons of Distance Methods for Combining
Covariates and Abundances in Microbiome Studies.”</span> In <em>Pac Symp
Biocomput</em>. World Scientific.
</div>
<div id="ref-Bioconductor" class="csl-entry" role="listitem">
Gentleman, Robert C, Vincent J Carey, Douglas M Bates, Ben Bolstad,
Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004.
<span>“Bioconductor: Open Software Development for Computational Biology
and Bioinformatics.”</span> <em>Genome Biology</em> 5 (10): R80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80">https://doi.org/10.1186/gb-2004-5-10-r80</a>.
</div>
<div id="ref-glass2007experimental" class="csl-entry" role="listitem">
Glass, David J. 2007. <em>Experimental Design for Biologists</em>. Cold
Spring Harbor Laboratory Press.
</div>
<div id="ref-Goslee:2007" class="csl-entry" role="listitem">
Goslee, Sarah C, Dean L Urban, et al. 2007. <span>“The Ecodist Package
for Dissimilarity-Based Analysis of Ecological Data.”</span> <em>Journal
of Statistical Software</em> 22 (7): 1–19.
</div>
<div id="ref-Grantham1981" class="csl-entry" role="listitem">
Grantham, Richard, Christian Gautier, Manolo Gouy, M Jacobzone, and R
Mercier. 1981. <span>“Codon Catalog Usage Is a Genome Strategy Modulated
for Gene Expressivity.”</span> <em>Nucleic Acids Research</em> 9 (1):
213–13.
</div>
<div id="ref-greenacre2007" class="csl-entry" role="listitem">
Greenacre, Michael J. 2007. <em>Correspondence Analysis in
Practice</em>. Chapman &amp; Hall.
</div>
<div id="ref-Wickham:r4ds" class="csl-entry" role="listitem">
Grolemund, Garrett, and Hadley Wickham. 2017. <em><span>R</span> for
Data Science</em>. O’Reilly.
</div>
<div id="ref-Grun2012" class="csl-entry" role="listitem">
Grün, Bettina, Theresa Scharl, and Friedrich Leisch. 2012.
<span>“Modelling Time Course Gene Expression Data with Finite Mixtures
of Linear Additive Models.”</span> <em>Bioinformatics</em> 28 (2):
222–28. <a href="https://doi.org/10.1093/bioinformatics/btr653">https://doi.org/10.1093/bioinformatics/btr653</a>.
</div>
<div id="ref-Guillot:2013" class="csl-entry" role="listitem">
Guillot, Gilles, and François Rousset. 2013. <span>“Dismantling the
Mantel Tests.”</span> <em>Methods in Ecology and Evolution</em> 4 (4):
336–44.
</div>
<div id="ref-Hallett2012" class="csl-entry" role="listitem">
Hallett, Robin M, Anna Dvorkin-Gheva, Anita Bane, and John A Hassell.
2012. <span>“A Gene Signature for Predicting Outcome in Patients with
Basal-Like Breast Cancer.”</span> <em>Scientific Reports</em> 2.
</div>
<div id="ref-Hastie1989" class="csl-entry" role="listitem">
Hastie, Trevor, and Werner Stuetzle. 1989. <span>“Principal
Curves.”</span> <em>Journal of the American Statistical Association</em>
84 (406): 502–16.
</div>
<div id="ref-HastieTibshiraniFriedman" class="csl-entry" role="listitem">
Hastie, Trevor, Robert Tibshirani, and Jerome Friedman. 2008. <em>The
Elements of Statistical Learning</em>. 2^{\text{nd}} ed. Springer.
</div>
<div id="ref-Head:PLoSBiol:2015:pvaluehacking" class="csl-entry" role="listitem">
Head, Megan L, Luke Holman, Rob Lanfear, Andrew T Kahn, and Michael D
Jennions. 2015. <span>“The Extent and Consequences of p-Hacking in
Science.”</span> <em>PLoS Biology</em> 13 (3): e1002106.
</div>
<div id="ref-held2010cellcognition" class="csl-entry" role="listitem">
Held, M., M. H. A. Schmitz, B. Fischer, T. Walter, B. Neumann, M. H.
Olma, M. Peter, J. Ellenberg, and D. W. Gerlich. 2010.
<span>“<span>CellCognition</span>: Time-Resolved Phenotype Annotation in
High-Throughput Live Cell Imaging.”</span> <em>Nature Methods</em> 7:
747.
</div>
<div id="ref-Helmholtz:1867" class="csl-entry" role="listitem">
Helmholtz, H. von. 1867. <em>Handbuch Der Physiologischen Optik</em>.
Leipzig: Leopold Voss.
</div>
<div id="ref-Henderson:2017:SoftwareEngineeringGoogle" class="csl-entry" role="listitem">
Henderson, Fergus. 2017. <span>“<span class="nocase">Software
Engineering at Google</span>.”</span> <em>ArXiv e-Prints</em>. <a href="https://arxiv.org/abs/1702.01715">https://arxiv.org/abs/1702.01715</a>.
</div>
<div id="ref-hoeting1999" class="csl-entry" role="listitem">
Hoeting, Jennifer A, David Madigan, Adrian E Raftery, and Chris T
Volinsky. 1999. <span>“Bayesian Model Averaging: A Tutorial.”</span>
<em>Statistical Science</em>, 382–401.
</div>
<div id="ref-Holmes1985" class="csl-entry" role="listitem">
Holmes - Junca, Susan. 1985. <span>“Outils Informatiques Pour
l’<span>é</span>valuation de La Pertinence d’un r<span>é</span>sultat En
Analyse Des Donn<span>é</span>es.”</span> PhD thesis, Université
Montpellier II, France.
</div>
<div id="ref-holmesima" class="csl-entry" role="listitem">
Holmes, Susan. 1999. <span>“Phylogenetic Trees: An Overview.”</span> In
<em>Statistics and Genetics</em>, 81–118. IMA 112. New York: Springer.
</div>
<div id="ref-Holmes-2003" class="csl-entry" role="listitem">
———. 2003a. <span>“Bootstrapping Phylogenetic Trees: Theory and
Methods.”</span> <em>Statistical Science</em> 18 (2): 241–55.
</div>
<div id="ref-holmes2003b" class="csl-entry" role="listitem">
———. 2003b. <span>“<span class="nocase">Statistics for phylogenetic
trees</span>.”</span> <em>Theoretical Population Biology</em> 63 (1):
17–32.
</div>
<div id="ref-frenchway" class="csl-entry" role="listitem">
———. 2006. <span>“Multivariate Analysis: <span class="nocase">The French
way</span>.”</span> In <em>Probability and Statistics: Essays in Honor
of David a. Freedman</em>, edited by D. Nolan and T. P. Speed. Vol. 56.
IMS Lecture Notes–Monograph Series. Beachwood, OH: IMS. <a href="http://www.imstat.org/publications/lecnotes.htm">http://www.imstat.org/publications/lecnotes.htm</a>.
</div>
<div id="ref-Holmes2017" class="csl-entry" role="listitem">
———. 2018. <span>“Statistical Proof? The Problem of
Irreproducibility.”</span> <em>Bulletin of the AMS</em> 55 (1): 31–55.
</div>
<div id="ref-holmes2011psb" class="csl-entry" role="listitem">
Holmes, Susan, Alexander V Alekseyenko, Alden Timme, Tyrrell Nelson,
Pankaj Jay Pasricha, and Alfred Spormann. 2011. <span>“Visualization and
Statistical Comparisons of Microbial Communities Using r Packages on
Phylochip Data.”</span> In <em>Pacific Symposium on Biocomputing</em>,
142–53. World Scientific.
</div>
<div id="ref-holmes2005memory" class="csl-entry" role="listitem">
Holmes, Susan, Michael He, Tong Xu, and Peter P Lee. 2005. <span>“Memory
t Cells Have Gene Expression Patterns Intermediate Between Naive and
Effector.”</span> <em>PNAS</em> 102 (15): 5519–23.
</div>
<div id="ref-Holmes2009" class="csl-entry" role="listitem">
Holmes, Susan, Adam Kapelner, and Peter P Lee. 2009. <span>“An
Interactive <span>Java</span> Statistical Image Segmentation System:
<span>GemIdent</span>.”</span> <em>Journal of Statistical Software</em>
30 (10).
</div>
<div id="ref-Hornik2005" class="csl-entry" role="listitem">
Hornik, Kurt. 2005. <span>“A <span>CLUE</span> for <span>CLU</span>ster
<span>E</span>nsembles.”</span> <em>Journal of Statistical Software</em>
14 (12).
</div>
<div id="ref-Hotelling:1933ki" class="csl-entry" role="listitem">
Hotelling, Harold. 1933. <span>“Analysis of a Complex of Statistical
Variables into Principal Components.”</span> <em>Journal of Educational
Psychology</em> 24 (6): 417–41.
</div>
<div id="ref-Hotelling1944" class="csl-entry" role="listitem">
———. 1944. <span>“Some Improvements in Weighing and Other Experimental
Techniques.”</span> <em>The Annals of Mathematical Statistics</em> 15
(3): 297–306.
</div>
<div id="ref-Huber:AMS:1964" class="csl-entry" role="listitem">
Huber, Peter J. 1964. <span>“Robust Estimation of a Location
Parameter.”</span> <em>The Annals of Mathematical Statistics</em> 35:
73–101.
</div>
<div id="ref-Huber:2015" class="csl-entry" role="listitem">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc
Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015.
<span>“Orchestrating High-Throughput Genomic Analysis with
<span>Bioconductor</span>.”</span> <em>Nature Methods</em> 12 (2):
115–21.
</div>
<div id="ref-flowsort" class="csl-entry" role="listitem">
Hulett, Henry R, William A Bonner, Janet Barrett, and Leonard A
Herzenberg. 1969. <span>“Cell Sorting: Automated Separation of Mammalian
Cells as a Function of Intracellular Fluorescence.”</span>
<em>Science</em> 166 (3906): 747–49.
</div>
<div id="ref-Ideker:2002" class="csl-entry" role="listitem">
Ideker, Trey, Owen Ozier, Benno Schwikowski, and Andrew F Siegel. 2002.
<span>“Discovering Regulatory and Signalling Circuits in Molecular
Interaction Networks.”</span> <em>Bioinformatics</em> 18 Suppl 1
(January): S233–40. <a href="http://bioinformatics.oxfordjournals.org/cgi/reprint/18/suppl\_1/S233">http://bioinformatics.oxfordjournals.org/cgi/reprint/18/suppl\_1/S233</a>.
</div>
<div id="ref-Ignatiadis:JRSSB:2021" class="csl-entry" role="listitem">
Ignatiadis, Nikolaos, and Wolfgang Huber. 2021. <span>“Covariate Powered
Cross-Weighted Multiple Testing.”</span> <em>Journal of the Royal
Statistical Society: Series B</em> 83: 720–51. <a href="https://doi.org/10.1111/rssb.12411">https://doi.org/10.1111/rssb.12411</a>.
</div>
<div id="ref-Ignatiadis:2016" class="csl-entry" role="listitem">
Ignatiadis, Nikolaos, Bernd Klaus, Judith Zaugg, and Wolfgang Huber.
2016. <span>“Data-Driven Hypothesis Weighting Increases Detection Power
in Genome-Scale Multiple Testing.”</span> <em>Nature Methods</em> 13:
577–80.
</div>
<div id="ref-IhakaColorPres" class="csl-entry" role="listitem">
Ihaka, Ross. 2003. <span>“Color for Presentation Graphics.”</span> In
<em>Proceedings of the 3rd International Workshop on Distributed
Statistical Computing</em>, edited by Kurt Hornik and Friedrich Leisch.
Vienna, Austria:
http://www.r-project.org/conferences/DSC-2003/Proceedings/; ISSN
1609-395X.
</div>
<div id="ref-RlangB" class="csl-entry" role="listitem">
Ihaka, Ross, and Robert Gentleman. 1996. <span>“R: A Language for Data
Analysis and Graphics.”</span> <em>Journal of Computational and
Graphical Statistics</em> 5 (3): 299–314.
</div>
<div id="ref-Irizarry:Biostat:2003" class="csl-entry" role="listitem">
Irizarry, R. A., B. Hobbs, F. Collin, Y. D. Beazer-Barclay, K. J.
Antonellis, U. Scherf, and T. P. Speed. 2003. <span>“Exploration,
Normalization, and Summaries of High Density Oligonucleotide Array Probe
Level Data.”</span> <em>Biostatistics</em> 4 (2): 249–64.
</div>
<div id="ref-Irizarry2009" class="csl-entry" role="listitem">
Irizarry, Rafael A, Hao Wu, and Andrew P Feinberg. 2009. <span>“A
Species-Generalized Probabilistic Model-Based Definition of CpG
Islands.”</span> <em>Mammalian Genome</em> 20 (9-10): 674–80.
</div>
<div id="ref-Izenman2008" class="csl-entry" role="listitem">
Izenman, Alan Julian. 2008. <span>“Nonlinear Dimensionality Reduction
and Manifold Learning.”</span> In <em>Modern Multivariate Statistical
Techniques: Regression, Classification, and Manifold Learning</em>,
597–632. New York, NY: Springer New York.
</div>
<div id="ref-Jacob:GroupLasso:2009" class="csl-entry" role="listitem">
Jacob, Laurent, Guillaume Obozinski, and Jean-Philippe Vert. 2009.
<span>“Group Lasso with Overlap and Graph Lasso.”</span> In
<em>Proceedings of the 26th Annual International Conference on Machine
Learning</em>, 433–40. ACM.
</div>
<div id="ref-James:2013" class="csl-entry" role="listitem">
James, Gareth, Daniela Witten, Trevor Hastie, and Robert Tibshirani.
2013. <em>An Introduction to Statistical Learning</em>. Springer.
</div>
<div id="ref-Jolicoeur1960" class="csl-entry" role="listitem">
Jolicoeur, Pierre, and James E Mosimann. 1960. <span>“Size and Shape
Variation in the Painted Turtle. A Principal Component Analysis.”</span>
<em>Growth</em> 24: 339–54.
</div>
<div id="ref-Jolliffe" class="csl-entry" role="listitem">
Jolliffe, Ian. 2002. <em>Principal Component Analysis</em>. Wiley Online
Library.
</div>
<div id="ref-jones2005voronoi" class="csl-entry" role="listitem">
Jones, T., A. Carpenter, and P. Golland. 2005. <span>“Voronoi-Based
Segmentation of Cells on Image Manifolds.”</span> <em>Computer Vision
for Biomedical Image Applications</em>, 535.
</div>
<div id="ref-Josse:2016" class="csl-entry" role="listitem">
Josse, Julie, and Susan Holmes. 2016. <span>“Measuring Multivariate
Association and Beyond.”</span> <em>Statistics Surveys</em> 10: 132–67.
</div>
<div id="ref-kahneman2011" class="csl-entry" role="listitem">
Kahneman, Daniel. 2011. <em>Thinking, Fast and Slow</em>. Macmillan.
</div>
<div id="ref-Kashyap:2013" class="csl-entry" role="listitem">
Kashyap, Purna C, Angela Marcobal, Luke K Ursell, Samuel A Smits, Erica
D Sonnenburg, Elizabeth K Costello, Steven K Higginbottom, et al. 2013.
<span>“Genetically Dictated Change in Host Mucus Carbohydrate Landscape
Exerts a Diet-Dependent Effect on the Gut Microbiota.”</span>
<em>PNAS</em> 110 (42): 17059–64.
</div>
<div id="ref-Kaufman2009" class="csl-entry" role="listitem">
Kaufman, Leonard, and Peter J Rousseeuw. 2009. <em>Finding Groups in
Data: An Introduction to Cluster Analysis</em>. Vol. 344. John Wiley
&amp; Sons.
</div>
<div id="ref-Kendall:1969" class="csl-entry" role="listitem">
Kendall, David. 1969. <span>“Incidence Matrices, Interval Graphs and
Seriation in Archeology.”</span> <em>Pacific Journal of Mathematics</em>
28 (3): 565–70.
</div>
<div id="ref-Kery2015" class="csl-entry" role="listitem">
Kéry, Marc, and J Andrew Royle. 2015. <em>Applied Hierarchical Modeling
in Ecology: Analysis of Distribution, Abundance and Species Richness in
r and BUGS: Volume 1: Prelude and Static Models</em>. Academic Press.
</div>
<div id="ref-Korthauer:GB:2019" class="csl-entry" role="listitem">
Korthauer, K., P. K. Kimes, C. Duvallet, A. Reyes, A. Subramanian, M.
Teng, C. Shukla, E. J. Alm, and S. C. Hicks. 2019. <span>“<span class="nocase"><span>A</span> practical guide to methods controlling
false discoveries in computational biology</span>.”</span> <em>Genome
Biology</em> 20 (1): 118.
</div>
<div id="ref-Kozich2013" class="csl-entry" role="listitem">
Kozich, James J, Sarah L Westcott, Nielson T Baxter, Sarah K Highlander,
and Patrick D Schloss. 2013. <span>“Development of a Dual-Index
Sequencing Strategy and Curation Pipeline for Analyzing Amplicon
Sequence Data on the MiSeq Illumina Sequencing Platform.”</span>
<em>Applied and Environmental Microbiology</em> 79 (17): 5112–20.
</div>
<div id="ref-Kristiansson2009" class="csl-entry" role="listitem">
Kristiansson, Erik, Michael Thorsen, Markus J Tamás, and Olle Nerman.
2009. <span>“Evolutionary Forces Act on Promoter Length: Identification
of Enriched Cis-Regulatory Elements.”</span> <em>Molecular Biology and
Evolution</em> 26 (6): 1299–1307.
</div>
<div id="ref-Kuan2011statistical" class="csl-entry" role="listitem">
Kuan, Pei Fen, Dongjun Chung, Guangjin Pan, James A Thomson, Ron
Stewart, and Sündüz Keleş. 2011. <span>“A Statistical Framework for the
Analysis of <span>ChIP-Seq</span> Data.”</span> <em>Journal of the
American Statistical Association</em> 106 (495): 891–903.
</div>
<div id="ref-Lai:2001" class="csl-entry" role="listitem">
Lai, Tze Leung. 2001. <em>Sequential Analysis</em>. Wiley Online
Library.
</div>
<div id="ref-lange2016mm" class="csl-entry" role="listitem">
Lange, Kenneth. 2016. <em>MM Optimization Algorithms</em>. SIAM.
</div>
<div id="ref-Laufer:NatMeth:2013" class="csl-entry" role="listitem">
Laufer, Christina, Bernd Fischer, Maximilian Billmann, Wolfgang Huber,
and Michael Boutros. 2013. <span>“<span class="nocase"><span>M</span>apping genetic interactions in human cancer
cells with <span>R</span><span>N</span><span>A</span>i and
multiparametric phenotyping</span>.”</span> <em>Nature Methods</em> 10:
427–31.
</div>
<div id="ref-Lawrence:Nature:2013" class="csl-entry" role="listitem">
Lawrence, Michael S., Petar Stojanov, Paz Polak, Gregory V. Kryukov,
Kristian Cibulskis, Andrey Sivachenko, Scott L. Carter, et al. 2013.
<span>“Mutational Heterogeneity in Cancer and the Search for New
Cancer-Associated Genes.”</span> <em>Nature</em> 499 (7457): 214–18. <a href="https://doi.org/10.1038/nature12213">https://doi.org/10.1038/nature12213</a>.
</div>
<div id="ref-Leek:2010:batch" class="csl-entry" role="listitem">
Leek, Jeffrey T, Robert B Scharpf, Héctor Corrada Bravo, David Simcha,
Benjamin Langmead, W Evan Johnson, Donald Geman, Keith Baggerly, and
Rafael A Irizarry. 2010. <span>“Tackling the Widespread and Critical
Impact of Batch Effects in High-Throughput Data.”</span> <em>Nature
Reviews Genetics</em> 11 (10): 733–39.
</div>
<div id="ref-LeekStorey:2007" class="csl-entry" role="listitem">
Leek, Jeffrey T., and John D. Storey. 2007. <span>“<span class="nocase"><span>C</span>apturing heterogeneity in gene expression
studies by surrogate variable analysis</span>.”</span> <em>PLoS
Genetics</em> 3 (9): 1724–35.
</div>
<div id="ref-WHLi" class="csl-entry" role="listitem">
Li, Wen-Hsiung. 1997. <em>Molecular Evolution.</em> Sinauer Associates
Incorporated.
</div>
<div id="ref-LiandGraur" class="csl-entry" role="listitem">
Li, Wen-Hsiung, and Dan Graur. 1991. <em>Fundamentals of Molecular
Evolution</em>. Vol. 48. Sinauer Associates Sunderland, MA.
</div>
<div id="ref-MSigDB" class="csl-entry" role="listitem">
Liberzon, Arthur, Aravind Subramanian, Reid Pinchback, Helga
Thorvaldsdóttir, Pablo Tamayo, and Jill P Mesirov. 2011.
<span>“Molecular Signatures Database (<span>MSigDB</span>) 3.0.”</span>
<em>Bioinformatics</em> 27 (12): 1739–40.
</div>
<div id="ref-BiocRNASeqWorkflow" class="csl-entry" role="listitem">
Love, Michael I., Simon Anders, Vladislav Kim, and Wolfgang Huber. 2015.
<span>“<span>RNA</span>-Seq Workflow: Gene-Level Exploratory Analysis
and Differential Expression.”</span> <em>F1000Research</em> 4 (1070). <a href="https://doi.org/10.12688/f1000research.7035.1">https://doi.org/10.12688/f1000research.7035.1</a>.
</div>
<div id="ref-LoveDESeq2" class="csl-entry" role="listitem">
Love, Michael I, Wolfgang Huber, and Simon Anders. 2014.
<span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span>
<em>Gnome Biology</em> 15 (12): 1–21.
</div>
<div id="ref-mandal2009" class="csl-entry" role="listitem">
Mandal, Rakesh, Sophie St-Hilaire, John G Kie, and DeWayne Derryberry.
2009. <span>“Spatial Trends of Breast and Prostate Cancers in the United
States Between 2000 and 2005.”</span> <em>International Journal of
Health Geographics</em> 8 (1): 53.
</div>
<div id="ref-Mardia" class="csl-entry" role="listitem">
Mardia, Kanti, John T Kent, and John M Bibby. 1979. <em>Multiariate
Analysis</em>. New York: Academic Press.
</div>
<div id="ref-Marin2007" class="csl-entry" role="listitem">
Marin, Jean-Michel, and Christian Robert. 2007. <em>Bayesian Core: A
Practical Approach to Computational <span>B</span>ayesian
Statistics</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-McCormick:1972:MatrixClustering" class="csl-entry" role="listitem">
McCormick Jr, William T, Paul J Schweitzer, and Thomas W White. 1972.
<span>“Problem Decomposition and Data Reorganization by a Clustering
Technique.”</span> <em>Operations Research</em> 20 (5): 993–1009.
</div>
<div id="ref-McElreath2015" class="csl-entry" role="listitem">
McElreath, Richard. 2015. <em>Statistical Rethinking: A
<span>B</span>ayesian Course with Examples in <span class="nocase">R and
Stan</span></em>. Chapman; Hall/CRC.
</div>
<div id="ref-mclachlan2007algorithm" class="csl-entry" role="listitem">
McLachlan, Geoffrey, and Thriyambakam Krishnan. 2007. <em>The
<span>EM</span> Algorithm and Extensions</em>. Vol. 382. John Wiley
&amp; Sons.
</div>
<div id="ref-mclachlan2004" class="csl-entry" role="listitem">
McLachlan, Geoffrey, and David Peel. 2004. <em>Finite Mixture
Models</em>. John Wiley &amp; Sons.
</div>
<div id="ref-mcmurdie2014" class="csl-entry" role="listitem">
McMurdie, Paul J, and Susan Holmes. 2014. <span>“Waste Not, Want Not:
Why Rarefying Microbiome Data Is Inadmissible.”</span> <em>PLoS
Computational Biology</em> 10 (4): e1003531.
</div>
<div id="ref-mcmurdie2015" class="csl-entry" role="listitem">
———. 2015. <span>“Shiny-Phyloseq: Web Application for Interactive
Microbiome Analysis with Provenance Tracking.”</span>
<em>Bioinformatics</em> 31 (2): 282–83.
</div>
<div id="ref-mead1990design" class="csl-entry" role="listitem">
Mead, Roger. 1990. <em>The Design of Experiments: Statistical Principles
for Practical Applications</em>. Cambridge University Press.
</div>
<div id="ref-Moignard:2015" class="csl-entry" role="listitem">
Moignard, Victoria, Steven Woodhouse, Laleh Haghverdi, Andrew J Lilly,
Yosuke Tanaka, Adam C Wilkinson, Florian Buettner, et al. 2015.
<span>“Decoding the Regulatory Network of Early Blood Development from
Single-Cell Gene Expression Measurements.”</span> <em>Nature
Biotechnology</em>.
</div>
<div id="ref-Mollon:1995" class="csl-entry" role="listitem">
Mollon, John. 1995. <span>“Seeing Colour.”</span> In <em>Colour: Art and
Science</em>, edited by T. Lamb and J. Bourriau. Cambridge Unversity
Press.
</div>
<div id="ref-Mood1946" class="csl-entry" role="listitem">
Mood, Alexander M. 1946. <span>“On Hotelling’s Weighing Problem.”</span>
<em>The Annals of Mathematical Statistics</em>, 432–46.
</div>
<div id="ref-Mossel:2003" class="csl-entry" role="listitem">
Mossel, Elchanan. 2003. <span>“On the Impossibility of Reconstructing
Ancestral Data and Phylogenies.”</span> <em>Journal of Computational
Biology</em> 10 (5): 669–76.
</div>
<div id="ref-Mourant1976" class="csl-entry" role="listitem">
Mourant, AE, Ada Kopec, and K Domaniewska-Sobczak. 1976. <span>“The
Distribution of the Human Blood Groups 2nd Edition.”</span> Oxford
University Press London.
</div>
<div id="ref-Mullner:2013" class="csl-entry" role="listitem">
Müllner, Daniel. 2013. <span>“Fastcluster: Fast Hierarchical,
Agglomerative Clustering Routines for r and Python.”</span> <em>Journal
of Statistical Software</em> 53 (9): 1–18.
</div>
<div id="ref-GXNA" class="csl-entry" role="listitem">
Nacu, Serban, Rebecca Critchley-Thorne, Peter Lee, and Susan Holmes.
2007. <span>“Gene Expression Network Analysis and Applications to
Immunology.”</span> <em>Bioinformatics</em> 23 (7, 7): 850–58. <a href="https://doi.org/10.1093/bioinformatics/btm019">https://doi.org/10.1093/bioinformatics/btm019</a>.
</div>
<div id="ref-Nelson:2011" class="csl-entry" role="listitem">
Nelson, Tyrell A, Susan Holmes, Alexander Alekseyenko, Masha Shenoy,
Todd DeSantis, Cindy Wu, Gary Andersen, et al. 2010.
<span>“<span>PhyloChip</span> Microarray Analysis Reveals Altered
Gastrointestinal Microbial Communities in a Rat Model of Colonic
Hypersensitivity.”</span> <em>Neurogastroenterology &amp; Motility</em>.
</div>
<div id="ref-Neumann:2010" class="csl-entry" role="listitem">
Neumann, B., T. Walter, J. K. Heriche, J. Bulkescher, H. Erfle, C.
Conrad, P. Rogers, et al. 2010. <span>“<span class="nocase"><span>P</span>henotypic profiling of the human genome by
time-lapse microscopy reveals cell division genes</span>.”</span>
<em>Nature</em> 464 (7289): 721–27.
</div>
<div id="ref-Neyman1936" class="csl-entry" role="listitem">
Neyman, Jerzy, and Egon S Pearson. 1936. <em>Sufficient Statistics and
Uniformly Most Powerful Tests of Statistical Hypotheses</em>. University
California Press.
</div>
<div id="ref-Nolan:2013" class="csl-entry" role="listitem">
Nolan, Daniel J, Michael Ginsberg, Edo Israely, Brisa Palikuqi, Michael
G Poulos, Daylon James, Bi-Sen Ding, et al. 2013. <span>“Molecular
Signatures of Tissue-Specific Microvascular Endothelial Cell
Heterogeneity in Organ Maintenance and Regeneration.”</span>
<em>Developmental Cell</em> 26 (2): 204–19.
</div>
<div id="ref-oneill2013flow" class="csl-entry" role="listitem">
O’Neill, Kieran, Nima Aghaeepour, Josef Špidlen, and Ryan Brinkman.
2013. <span>“Flow Cytometry Bioinformatics.”</span> <em>PLoS
Computational Biology</em> 9 (12): e1003365.
</div>
<div id="ref-Ohnishi2014" class="csl-entry" role="listitem">
Ohnishi, Y., W. Huber, A. Tsumura, M. Kang, P. Xenopoulos, K. Kurimoto,
A. K. Oles, et al. 2014. <span>“Cell-to-Cell Expression Variability
Followed by Signal Reinforcement Progressively Segregates Early Mouse
Lineages.”</span> <em>Nature Cell Biology</em> 16 (1): 27–37.
</div>
<div id="ref-OzsolakMilos" class="csl-entry" role="listitem">
Ozsolak, Fatih, and Patrice M Milos. 2011. <span>“<span class="nocase"><span>R</span><span>N</span><span>A</span> sequencing:
advances, challenges and opportunities</span>.”</span> <em>Nature
Reviews Genetics</em> 12: 87–98.
</div>
<div id="ref-Pages:2016" class="csl-entry" role="listitem">
Pagès, Jérôme. 2016. <em>Multiple Factor Analysis by Example Using
<span>R</span></em>. CRC Press.
</div>
<div id="ref-paradis:2011" class="csl-entry" role="listitem">
Paradis, Emmanuel. 2011. <em>Analysis of Phylogenetics and Evolution
with r</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-EBImage" class="csl-entry" role="listitem">
Pau, Grégoire, Florian Fuchs, Oleg Sklyar, Michael Boutros, and Wolfgang
Huber. 2010. <span>“<span>EBI</span>mage <span>R</span> Package for
Image Processing with Applications to Cellular Phenotypes.”</span>
<em>Bioinformatics</em> 26 (7): 979–81.
</div>
<div id="ref-Pearson1901" class="csl-entry" role="listitem">
Pearson, Karl. 1901. <span>“<span>LIII. O</span>n Lines and Planes of
Closest Fit to Systems of Points in Space.”</span> <em>The London,
Edinburgh, and Dublin Philosophical Magazine and Journal of Science</em>
2 (11): 559–72.
</div>
<div id="ref-Perraudeau:2017" class="csl-entry" role="listitem">
Perraudeau, Fanny, Davide Risso, Kelly Street, Elizabeth Purdom, and
Sandrine Dudoit. 2017. <span>“Bioconductor Workflow for Single-Cell
<span>RNA</span> Sequencing: Normalization, Dimensionality Reduction,
Clustering, and Lineage Inference.”</span> <em>F1000Research</em> 6.
</div>
<div id="ref-Perriere2002" class="csl-entry" role="listitem">
Perrière, Guy, and Jean Thioulouse. 2002. <span>“Use and Misuse of
Correspondence Analysis in Codon Usage Studies.”</span> <em>Nucleic
Acids Research</em> 30 (20): 4548–55.
</div>
<div id="ref-Pounds2003" class="csl-entry" role="listitem">
Pounds, Stan, and Stephan W Morris. 2003. <span>“Estimating the
Occurrence of False Positives and False Negatives in Microarray Studies
by Approximating and Partitioning the Empirical Distribution of
p-Values.”</span> <em>Bioinformatics</em> 19 (10): 1236–42.
</div>
<div id="ref-Prentice:1977" class="csl-entry" role="listitem">
Prentice, IC. 1977. <span>“Non-Metric Ordination Methods in
Ecology.”</span> <em>The Journal of Ecology</em>, 85–94.
</div>
<div id="ref-Purdom2010" class="csl-entry" role="listitem">
Purdom, Elizabeth. 2010. <span>“Analysis of a Data Matrix and a Graph:
Metagenomic Data and the Phylogenetic Tree.”</span> <em>Annals of
Applied Statistics</em>, July.
</div>
<div id="ref-Purdom2005" class="csl-entry" role="listitem">
Purdom, Elizabeth, and Susan P Holmes. 2005. <span>“Error Distribution
for Gene Expression Data.”</span> <em>Statistical Applications in
Genetics and Molecular Biology</em> 4 (1).
</div>
<div id="ref-PhenoRipper" class="csl-entry" role="listitem">
Rajaram, S., B. Pavie, L. F. Wu, and S. J. Altschuler. 2012.
<span>“<span class="nocase"><span>P</span>heno<span>R</span>ipper:
software for rapidly profiling microscopy images</span>.”</span>
<em>Nature Methods</em> 9: 635–37.
</div>
<div id="ref-diabetes" class="csl-entry" role="listitem">
Reaven, GM, and RG Miller. 1979. <span>“An Attempt to Define the Nature
of Chemical Diabetes Using a Multidimensional Analysis.”</span>
<em>Diabetologia</em> 16 (1): 17–24.
</div>
<div id="ref-Reyes:PNAS:2013" class="csl-entry" role="listitem">
Reyes, Alejandro, Simon Anders, Robert J. Weatheritt, Toby J. Gibson,
Lars M. Steinmetz, and Wolfgang Huber. 2013. <span>“Drift and
Conservation of Differential Exon Usage Across Tissues in Primate
Species.”</span> <em>Proceedings of the National Academy of
Sciences</em> 110 (38): 15377–82. <a href="https://doi.org/10.1073/pnas.1307202110">https://doi.org/10.1073/pnas.1307202110</a>.
</div>
<div id="ref-Reyes:NAR:2017" class="csl-entry" role="listitem">
Reyes, Alejandro, and Wolfgang Huber. 2017. <span>“Alternative Start and
Termination Sites of Transcription Drive Most Transcript Isoform
Differences Across Human Tissues.”</span> <em>Nucleic Acids
Research</em> 46 (2): 582–92. <a href="https://doi.org/10.1093/nar/gkx1165">https://doi.org/10.1093/nar/gkx1165</a>.
</div>
<div id="ref-HIVdb" class="csl-entry" role="listitem">
Rhee, Soo-Yon, Matthew J Gonzales, Rami Kantor, Bradley J Betts, Jaideep
Ravela, and Robert W Shafer. 2003. <span>“Human Immunodeficiency Virus
Reverse Transcriptase and Protease Sequence Database.”</span>
<em>Nucleic Acids Research</em> 31 (1): 298–303.
</div>
<div id="ref-Rice:2007" class="csl-entry" role="listitem">
Rice, John. 2006. <em>Mathematical Statistics and Data Analysis</em>.
Cengage Learning.
</div>
<div id="ref-RipleySISS1988" class="csl-entry" role="listitem">
Ripley, B. D. 1988. <em>Statistical Inference for Spatial
Processes.</em> Cambridge University Press.
</div>
<div id="ref-Casella2009" class="csl-entry" role="listitem">
Robert, Christian, and George Casella. 2009. <em>Introducing
<span>M</span>onte <span>C</span>arlo Methods with <span>R</span></em>.
Springer Science &amp; Business Media.
</div>
<div id="ref-robins:2007" class="csl-entry" role="listitem">
Robins, Garry, Tom Snijders, Peng Wang, Mark Handcock, and Philippa
Pattison. 2007. <span>“Recent Developments in Exponential Random Graph
(p*) Models for Social Networks.”</span> <em>Social Networks</em> 29
(2): 192–215.
</div>
<div id="ref-edgeR:Robinson:2009" class="csl-entry" role="listitem">
Robinson, M. D., D. J. McCarthy, and G. K. Smyth. 2009. <span>“<span class="nocase">edgeR</span>: A <span>Bioconductor</span> Package for
Differential Expression Analysis of Digital Gene Expression
Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-RockeDurbin:2001" class="csl-entry" role="listitem">
Rocke, David M, and Blythe Durbin. 2001. <span>“A Model for Measurement
Error for Gene Expression Arrays.”</span> <em>Journal of Computational
Biology</em> 8 (6): 557–69.
</div>
<div id="ref-mrbayes3" class="csl-entry" role="listitem">
Ronquist, Fredrik, Maxim Teslenko, Paul van der Mark, Daniel L Ayres,
Aaron Darling, Sebastian Höhna, Bret Larget, Liang Liu, Marc A Suchard,
and John P Huelsenbeck. 2012. <span>“MrBayes 3.2: Efficient Bayesian
Phylogenetic Inference and Model Choice Across a Large Model
Space.”</span> <em>Systematic Biology</em> 61 (3): 539–42.
</div>
<div id="ref-Rosen:2012" class="csl-entry" role="listitem">
Rosen, Michael J, Benjamin J Callahan, Daniel S Fisher, and Susan P
Holmes. 2012. <span>“Denoising <span>PCR</span>-Amplified Metagenome
Data.”</span> <em>BMC <span>B</span>ioinformatics</em> 13 (1): 283.
</div>
<div id="ref-Rousseeuw:1987" class="csl-entry" role="listitem">
Rousseeuw, Peter J. 1987. <span>“Silhouettes: A Graphical Aid to the
Interpretation and Validation of Cluster Analysis.”</span> <em>Journal
of Computational and Applied Mathematics</em> 20: 53–65.
</div>
<div id="ref-Rousseeuw:RobustBook:1987" class="csl-entry" role="listitem">
Rousseeuw, Peter J., and Annick M. Leroy. 1987. <em>Robust Regression
and Outlier Detection</em>. Wiley. <a href="https://doi.org/10.1002/0471725382">https://doi.org/10.1002/0471725382</a>.
</div>
<div id="ref-Roweis2000" class="csl-entry" role="listitem">
Roweis, Sam T, and Lawrence K Saul. 2000. <span>“Nonlinear
Dimensionality Reduction by Locally Linear Embedding.”</span>
<em>Science</em> 290 (5500): 2323–26.
</div>
<div id="ref-RussImageProcessingHandbook" class="csl-entry" role="listitem">
Russ, John C., and F. Brent Neal. 2015. <em>The Image Processing
Handbook</em>. 7th ed. CRC Press;
</div>
<div id="ref-sankaran2014structssi" class="csl-entry" role="listitem">
Sankaran, Kris, and Susan Holmes. 2014. <span>“<span class="nocase">structSSI: Simultaneous and Selective Inference for
Grouped or Hierarchically Structured Data</span>.”</span> <em>Journal of
Statistical Software</em> 59 (1): 1–21.
</div>
<div id="ref-schilling1986multivariate" class="csl-entry" role="listitem">
Schilling, Mark F. 1986. <span>“Multivariate Two-Sample Tests Based on
Nearest Neighbors.”</span> <em>Journal of the American Statistical
Association</em> 81 (395): 799–806.
</div>
<div id="ref-schindelin2012fiji" class="csl-entry" role="listitem">
Schindelin, Johannes, Ignacio Arganda-Carreras, Erwin Frise, Verena
Kaynig, Mark Longair, Tobias Pietzsch, Stephan Preibisch, et al. 2012.
<span>“<span class="nocase"><span>F</span>iji: an open-source platform
for biological-image analysis</span>.”</span> <em>Nature Methods</em> 9:
676–82.
</div>
<div id="ref-mothur" class="csl-entry" role="listitem">
Schloss, P D, S L Westcott, T Ryabin, J R Hall, M Hartmann, E B
Hollister, R A Lesniewski, et al. 2009. <span>“<span class="nocase">Introducing mothur: Open-Source, Platform-Independent,
Community-Supported Software for Describing and Comparing Microbial
Communities</span>.”</span> <em>Applied and Environmental
Microbiology</em> 75 (23): 7537–41.
</div>
<div id="ref-schloss2012stabilization" class="csl-entry" role="listitem">
Schloss, P. D., A. M. Schuber, J. P. Zackular, K. D. Iverson, Young V.
B., and Petrosino J. F. 2012. <span>“Stabilization of the Murine Gut
Microbiome Following Weaning.”</span> <em>Gut Microbes</em> 3 (4):
383–93.
</div>
<div id="ref-VertKernel" class="csl-entry" role="listitem">
Schölkopf, Bernhard, Koji Tsuda, and Jean-Philippe Vert. 2004.
<em>Kernel Methods in Computational Biology</em>. MIT press.
</div>
<div id="ref-SchwederSpjotvoll1982" class="csl-entry" role="listitem">
Schweder, T., and E. Spjøtvoll. 1982. <span>“Plots of <span class="nocase">P-values</span> to Evaluate Many Tests
Simultaneously.”</span> <em>Biometrika</em> 69: 493–502. <a href="https://doi.org/10.1093/biomet/69.3.493">https://doi.org/10.1093/biomet/69.3.493</a>.
</div>
<div id="ref-senn2004randomization" class="csl-entry" role="listitem">
Senn, Stephen. 2004. <span>“Controversies Concerning Randomization and
Additivity in Clinical Trials.”</span> <em>Statistics in Medicine</em>
23: 3729–53.
</div>
<div id="ref-MathematicalMorphology" class="csl-entry" role="listitem">
Serra, Jean. 1983. <em>Image Analysis and Mathematical Morphology</em>.
Academic Press.
</div>
<div id="ref-Setiadi2010" class="csl-entry" role="listitem">
Setiadi, A Francesca, Nelson C Ray, Holbrook E Kohrt, Adam Kapelner,
Valeria Carcamo-Cavazos, Edina B Levic, Sina Yadegarynia, et al. 2010.
<span>“Quantitative, Architectural Analysis of Immune Cell Subsets in
Tumor-Draining Lymph Nodes from Breast Cancer Patients and Healthy Lymph
Nodes.”</span> <em>PLoS One</em> 5 (8): e12420.
</div>
<div id="ref-CosmaShalizi2017" class="csl-entry" role="listitem">
Shalizi, Cosma. 2017. <em>Advanced Data Analysis from an Elementary
Point of View</em>. Cambridge University Press. <a href="https://www.stat.cmu.edu/~cshalizi/ADAfaEPoV/ADAfaEPoV.pdf">https://www.stat.cmu.edu/~cshalizi/ADAfaEPoV/ADAfaEPoV.pdf</a>.
</div>
<div id="ref-Slonim:2005" class="csl-entry" role="listitem">
Slonim, Noam, Gurinder Singh Atwal, Gašper Tkačik, and William Bialek.
2005. <span>“Information-Based Clustering.”</span> <em>PNAS</em> 102
(51): 18297–302.
</div>
<div id="ref-Stegle:2010" class="csl-entry" role="listitem">
Stegle, O., L. Parts, R. Durbin, and J. Winn. 2010. <span>“<span class="nocase"><span>A</span> <span>B</span>ayesian framework to account
for complex non-genetic factors in gene expression levels greatly
increases power in e<span>Q</span><span>T</span><span>L</span>
studies</span>.”</span> <em>PLoS Computational Biology</em> 6 (5):
e1000770.
</div>
<div id="ref-SteijgerBertone:2013" class="csl-entry" role="listitem">
Steijger, T., J. F. Abril, P. G. Engstrom, F. Kokocinski, T. J. Hubbard,
R. Guigo, J. Harrow, et al. 2013. <span>“<span class="nocase"><span>A</span>ssessment of transcript reconstruction
methods for <span class="nocase">RNA-seq</span></span>.”</span>
<em>Nature Methods</em> 10 (12): 1177–84.
</div>
<div id="ref-Stigler:sevenpillars" class="csl-entry" role="listitem">
Stigler, Stephen M. 2016. <em>The Seven Pillars of Statistical
Wisdom</em>. Harvard University Press.
</div>
<div id="ref-Storey:AnnStat:2003" class="csl-entry" role="listitem">
Storey, John D. 2003. <span>“The Positive False Discovery Rate: A
Bayesian Interpretation and the q-Value.”</span> <em>The Annals of
Statistics</em> 31 (6). <a href="https://doi.org/10.1214/aos/1074290335">https://doi.org/10.1214/aos/1074290335</a>.
</div>
<div id="ref-Strang:09" class="csl-entry" role="listitem">
Strang, Gilbert. 2009. <em>Introduction to <span>L</span>inear
<span>A</span>lgebra</em>. Fourth. Wellesley-Cambridge Press.
</div>
<div id="ref-Tenenbaum2000" class="csl-entry" role="listitem">
Tenenbaum, Joshua B, Vin De Silva, and John C Langford. 2000. <span>“A
Global Geometric Framework for Nonlinear Dimensionality
Reduction.”</span> <em>Science</em> 290 (5500): 2319–23.
</div>
<div id="ref-Tibshirani1996" class="csl-entry" role="listitem">
Tibshirani, Robert. 1996. <span>“Regression Shrinkage and Selection via
the Lasso.”</span> <em>Journal of the Royal Statistical Society. Series
B (Methodological)</em>, 267–88.
</div>
<div id="ref-gap2001" class="csl-entry" role="listitem">
Tibshirani, Robert, Guenther Walther, and Trevor Hastie. 2001.
<span>“Estimating the Number of Clusters in a Data Set via the Gap
Statistic.”</span> <em>JRSSB</em> 63 (2): 411–23.
</div>
<div id="ref-Trosset2008" class="csl-entry" role="listitem">
Trosset, Michael W, and Carey E Priebe. 2008. <span>“The Out-of-Sample
Problem for Classical Multidimensional Scaling.”</span>
<em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4635–42.
</div>
<div id="ref-Tseng:2005" class="csl-entry" role="listitem">
Tseng, George C, and Wing H Wong. 2005. <span>“Tight Clustering: A
Resampling-Based Approach for Identifying Stable and Tight Patterns in
Data.”</span> <em>Biometrics</em> 61 (1): 10–16.
</div>
<div id="ref-Tukey77" class="csl-entry" role="listitem">
Tukey, John W. 1977. <span>“Exploratory Data Analysis.”</span>
<em>Massachusetts: Addison-Wesley</em>.
</div>
<div id="ref-tversky1974heuristics" class="csl-entry" role="listitem">
Tversky, Amos, and Daniel Kahneman. 1974. <span>“Heuristics and Biases:
Judgement Under Uncertainty.”</span> <em>Science</em> 185: 1124–30.
</div>
<div id="ref-tversky1975judgment" class="csl-entry" role="listitem">
———. 1975. <span>“Judgment Under Uncertainty: Heuristics and
Biases.”</span> In <em>Utility, Probability, and Human Decision
Making</em>, 141–62. Springer.
</div>
<div id="ref-Verhulst:1845" class="csl-entry" role="listitem">
Verhulst, Pierre-François. 1845. <span>“Recherches <span class="nocase">mathématiques</span> Sur La Loi d’accroissement de La
Population.”</span> <em>Nouveaux <span>Mémoires</span> de
l’<span>Académie</span> <span>Royale</span> Des <span>Sciences</span> Et
<span>Belles</span>-<span>Lettres</span> de <span>Bruxelles</span></em>
18: 1–42.
</div>
<div id="ref-FoundationSignalProcessing" class="csl-entry" role="listitem">
Vetterli, Martin, Jelena Kovačević, and Vivek Goyal. 2014.
<em>Foundations of Signal Processing</em>. Cambridge University Press.
</div>
<div id="ref-wang2007naive" class="csl-entry" role="listitem">
Wang, Q., G. M. Garrity, J. M. Tiedje, and J. R. Cole. 2007.
<span>“Naive Bayesian Classifier for Rapid Assignment of rRNA Sequences
into the New Bacterial Taxonomy.”</span> <em>Applied and Environmental
Microbiology</em> 73 (16): 5261.
</div>
<div id="ref-Wasserstein2016:ASA" class="csl-entry" role="listitem">
Wasserstein, Ronald L, and Nicole A Lazar. 2016. <span>“The
<span>ASA</span>’s Statement on p-Values: Context, Process, and
Purpose.”</span> <em>The American Statistician</em>.
</div>
<div id="ref-Wertheim:2009" class="csl-entry" role="listitem">
Wertheim, Joel O, and Michael Worobey. 2009. <span>“Dating the Age of
the <span>SIV</span> Lineages That Gave Rise to <span>HIV</span>-1 and
<span>HIV</span>-2.”</span> <em>PLoS Computational Biology</em> 5 (5):
e1000377.
</div>
<div id="ref-Wickham:LayeredGrammar" class="csl-entry" role="listitem">
Wickham, Hadley. 2010. <span>“A Layered Grammar of Graphics.”</span>
<em>Journal of Computational and Graphical Statistics</em> 19 (1): 3–28.
</div>
<div id="ref-Wickham:TidyData" class="csl-entry" role="listitem">
———. 2014. <span>“Tidy Data.”</span> <em>Journal of Statistical
Software</em> 59 (10).
</div>
<div id="ref-ggplot2" class="csl-entry" role="listitem">
———. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>.
Springer New York. <a href="http://had.co.nz/ggplot2/book">http://had.co.nz/ggplot2/book</a>.
</div>
<div id="ref-Wiel:StatMed:2016" class="csl-entry" role="listitem">
Wiel, Mark A, Tonje G Lien, Wina Verlaat, Wessel N Wieringen, and Saskia
M Wilting. 2016. <span>“Better Prediction by Use of Co-Data: Adaptive
Group-Regularized Ridge Regression.”</span> <em>Statistics in
Medicine</em> 35 (3): 368–81.
</div>
<div id="ref-Wilkinson:DotPlots:1999" class="csl-entry" role="listitem">
Wilkinson, Leland. 1999. <span>“Dot Plots.”</span> <em>The American
Statistician</em> 53 (3): 276.
</div>
<div id="ref-GrofGrbook" class="csl-entry" role="listitem">
———. 2005. <em>The Grammar of Graphics</em>. Springer.
</div>
<div id="ref-Wills2013" class="csl-entry" role="listitem">
Wills, Quin F, Kenneth J Livak, Alex J Tipping, Tariq Enver, Andrew J
Goldson, Darren W Sexton, and Chris Holmes. 2013. <span>“Single-Cell
Gene Expression Analysis Reveals Genetic Associations Masked in
Whole-Tissue Experiments.”</span> <em>Nature Biotechnology</em> 31 (8):
748–52.
</div>
<div id="ref-Wilson:Goodenough:2017" class="csl-entry" role="listitem">
Wilson, Greg, Jennifer Bryan, Karen Cranston, Justin Kitzes, Lex
Nederbragt, and Tracy K. Teal. 2017. <span>“Good Enough Practices in
Scientific Computing.”</span> Edited by Francis Ouellette.
<em><span>PLOS</span> <span>C</span>omputational
<span>B</span>iology</em> 13 (6): e1005510. <a href="https://doi.org/10.1371/journal.pcbi.1005510">https://doi.org/10.1371/journal.pcbi.1005510</a>.
</div>
<div id="ref-witten2011penalized" class="csl-entry" role="listitem">
Witten, Daniela M, and Robert Tibshirani. 2011. <span>“Penalized
Classification Using <span>F</span>isher’s Linear Discriminant.”</span>
<em>JRSSB</em> 73 (5): 753–72.
</div>
<div id="ref-Witten2009" class="csl-entry" role="listitem">
Witten, Daniela M, Robert Tibshirani, and Trevor Hastie. 2009. <span>“A
Penalized Matrix Decomposition, with Applications to Sparse Principal
Components and Canonical Correlation Analysis.”</span>
<em>Biostatistics</em>, kxp008.
</div>
<div id="ref-wright2015decipher" class="csl-entry" role="listitem">
Wright, Erik S. 2015. <span>“<span>DECIPHER</span>: Harnessing Local
Sequence Context to Improve Protein Multiple Sequence Alignment.”</span>
<em><span>BMC Bioinformatics</span></em> 16 (1): 1.
</div>
<div id="ref-wu2011experiments" class="csl-entry" role="listitem">
Wu, CF Jeff, and Michael S Hamada. 2011. <em>Experiments: Planning,
Analysis, and Optimization</em>. Vol. 552. John Wiley &amp; Sons.
</div>
<div id="ref-YuGXNA" class="csl-entry" role="listitem">
Yu, Hongxiang, Diana L Simons, Ilana Segall, Valeria Carcamo-Cavazos,
Erich J Schwartz, Ning Yan, Neta S Zuckerman, et al. 2012.
<span>“<span>PRC2</span>/<span>EED-EZH2</span> Complex Is up-Regulated
in Breast Cancer Lymph Node Metastasis Compared to Primary Tumor and
Correlates with Tumor Proliferation in Situ.”</span> <em>PloS One</em> 7
(12): e51239.
</div>
<div id="ref-Zeileis:2008" class="csl-entry" role="listitem">
Zeileis, Achim, Christian Kleiber, and Simon Jackman. 2008.
<span>“Regression Models for Count Data in <span>R</span>.”</span>
<em>Journal of Statistical Software</em> 27 (8). <a href="http://www.jstatsoft.org/v27/i08/">http://www.jstatsoft.org/v27/i08/</a>.
</div>
<div id="ref-Zeller:MSB:2014" class="csl-entry" role="listitem">
Zeller, Georg, Julien Tap, Anita Y Voigt, Shinichi Sunagawa, Jens Roat
Kultima, Paul I Costea, Aurélien Amiot, et al. 2014.
<span>“<span>P</span>otential of Fecal Microbiota for Early-Stage
Detection of Colorectal Cancer.”</span> <em>Molecular Systems
Biology</em> 10 (11): 766. <a href="https://doi.org/10.15252/msb.20145645">https://doi.org/10.15252/msb.20145645</a>.
</div>
<div id="ref-Zou2006" class="csl-entry" role="listitem">
Zou, Hui, Trevor Hastie, and Robert Tibshirani. 2006. <span>“Sparse
Principal Component Analysis.”</span> <em>Journal of Computational and
Graphical Statistics</em> 15 (2): 265–86.
</div>
</div>



<p>Page built at 01:33 on 2025-09-01 using R version 4.5.1 (2025-06-13)</p></main> <!-- /main --></p><p class="build-date">
Support for maintaining the online version of this book is provided by <a href="http://www.denbi.de">de.NBI</a>
  <img src="imgs/denbi.png" style="vertical-align: text-top;height: 16px"><br>
For website support please contact msmith@embl.de
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
  const anchorJS_eq = new window.AnchorJS();
  anchorJS_eq.options = {
    placement: 'left',
    class: 'eq-link',
    icon: icon
  };
  anchorJS_eq.add('.anchored-eq');
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/www\.huber\.embl\.de\/msmb");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>
